<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />

    

    
    <title>è¯»ä¹¦ç¬”è®°-Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç† | è‹‘é•¿çš„Index</title>
    
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    
    <meta name="keywords" content="Python,æ•°æ®å¤„ç†" />
    
    <meta name="description" content="å½“ç”Ÿç‰©ä¸Pythonäº¤ç»‡ä¹‹æ—¶ æ•…äº‹å°†å°±æ­¤å±•å¼€">
<meta property="og:type" content="article">
<meta property="og:title" content="è¯»ä¹¦ç¬”è®°-Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç†">
<meta property="og:url" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/">
<meta property="og:site_name" content="è‹‘é•¿çš„Index">
<meta property="og:description" content="å½“ç”Ÿç‰©ä¸Pythonäº¤ç»‡ä¹‹æ—¶ æ•…äº‹å°†å°±æ­¤å±•å¼€">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png">
<meta property="article:published_time" content="2020-10-04T10:00:05.000Z">
<meta property="article:modified_time" content="2021-06-21T12:47:59.553Z">
<meta property="article:author" content="è‹‘é•¿">
<meta property="article:tag" content="Python">
<meta property="article:tag" content="æ•°æ®å¤„ç†">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png">
    

    
        <link rel="alternate" href="/atom.xml" title="è‹‘é•¿çš„Index" type="application/atom+xml" />
    

    
        <link rel="icon" href="/css/images/favicon.ico" />
    

    
<link rel="stylesheet" href="/libs/font-awesome/css/all.css">

    
<link rel="stylesheet" href="/libs/titillium-web/styles.css">

    
<link rel="stylesheet" href="/libs/source-code-pro/styles.css">


    
<link rel="stylesheet" href="/css/style.css">


    
<script src="/libs/jquery/3.4.1/jquery.min.js"></script>

    
    
        
<link rel="stylesheet" href="/libs/lightgallery/css/lightgallery.min.css">

    
    
        
<link rel="stylesheet" href="/libs/justified-gallery/justifiedGallery.min.css">

    
    
    
        <script>
var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = "//hm.baidu.com/hm.js?67fc4a038906ef55e9307dac98a80fe2";
  var s = document.getElementsByTagName("script")[0];
  s.parentNode.insertBefore(hm, s);
})();
</script>

    


<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
</head>

<body>
    <div id="wrap">
        <header id="header">
    <div id="header-outer" class="outer">
        <div class="container">
            <div class="container-inner">
                <div id="header-title">
                    <h1 class="logo-wrap">
                        <a href="/" class="logo"></a>
                    </h1>
                    
                        <h2 class="subtitle-wrap">
                            <p class="subtitle">Life &amp; Science &amp; Technology</p>
                        </h2>
                    
                </div>
                <div id="header-inner" class="nav-container">
                    <a id="main-nav-toggle" class="nav-icon fa fa-bars"></a>
                    <div class="nav-container-inner">
                        <ul id="main-nav">
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/">ä¸»é¡µ</a>
                                </li>
                            
                                        <ul class="main-nav-list"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%94%9F%E6%B4%BB/">ç”Ÿæ´»</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%94%9F%E6%B4%BB/%E6%90%9E%E4%BA%8B%E6%83%85%E5%90%96/">æäº‹æƒ…å–</a></li></ul></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">ä¿¡æ¯æŠ€æœ¯</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E5%A4%9A%E5%AA%92%E4%BD%93/">å¤šåª’ä½“</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">ç”Ÿç‰©</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">ç©ºå¤©</a></li></ul></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/">ç´¢å¼•</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%AE%9E%E6%97%B6%E4%BF%A1%E6%81%AF/">å®æ—¶ä¿¡æ¯</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%B4%A2%E5%BC%95/">å·¥ä½œæµç´¢å¼•</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E6%94%BF%E7%AD%96%E7%B4%A2%E5%BC%95/">æ”¿ç­–ç´¢å¼•</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">èµ„æºç´¢å¼•</a></li></ul></li></ul>
                                    
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/suo-yin/zi-yuan-suo-yin/chang-yong-zi-yuan-suo-yin/">å¸¸ç”¨ç´¢å¼•</a>
                                </li>
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/sheng-huo/gao-shi-qing-a/blog-gu-gu-gu/">Blogå’•å’•å’•</a>
                                </li>
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/sheng-huo/about/">å…³äº</a>
                                </li>
                            
                        </ul>
                        <nav id="sub-nav">
                            <div id="search-form-wrap">

    <form class="search-form">
        <input type="text" class="ins-search-input search-form-input" placeholder="æœç´¢" />
        <button type="submit" class="search-form-submit"></button>
    </form>
    <div class="ins-search">
    <div class="ins-search-mask"></div>
    <div class="ins-search-container">
        <div class="ins-input-wrapper">
            <input type="text" class="ins-search-input" placeholder="æƒ³è¦æŸ¥æ‰¾ä»€ä¹ˆ..." />
            <span class="ins-close ins-selectable"><i class="fa fa-times-circle"></i></span>
        </div>
        <div class="ins-section-wrapper">
            <div class="ins-section-container"></div>
        </div>
    </div>
</div>
<script>
(function (window) {
    var INSIGHT_CONFIG = {
        TRANSLATION: {
            POSTS: 'æ–‡ç« ',
            PAGES: 'é¡µé¢',
            CATEGORIES: 'åˆ†ç±»',
            TAGS: 'æ ‡ç­¾',
            UNTITLED: '(æœªå‘½å)',
        },
        ROOT_URL: '/',
        CONTENT_URL: '/content.json',
    };
    window.INSIGHT_CONFIG = INSIGHT_CONFIG;
})(window);
</script>

<script src="/js/insight.js"></script>


</div>
                        </nav>
                    </div>
                </div>
            </div>
        </div>
    </div>
</header>
        <div class="container">
            <div class="main-body container-inner">
                <div class="main-body-inner">
                    <section id="main">
                        <div class="main-body-header">
    <h1 class="header">
    
    <a class="page-title-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><i class="icon fa fa-angle-right"></i><a class="page-title-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">ç”Ÿç‰©</a>
    </h1>
</div>

                        <div class="main-body-content">
                            <article id="post-è¯»ä¹¦ç¬”è®°-Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç†" class="article article-single article-type-post" itemscope itemprop="blogPost">
    <div class="article-inner">
        
            <header class="article-header">
                
    
        <h1 class="article-title" itemprop="name">
        è¯»ä¹¦ç¬”è®°-Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç†
        </h1>
    

            </header>
        
        
            <div class="article-meta">
                
    <div class="article-date">
      <i class="fa fa-calendar"></i>
      <a href="/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/" class="article-date">
         <time datetime="2020-10-04T10:00:05.000Z" itemprop="datePublished">2020-10-04</time>
      </a>
    </div>


                

                
    <div class="article-tag">
        <i class="fa fa-tag"></i>
        <a class="tag-link-link" href="/tags/Python/" rel="tag">Python</a>, <a class="tag-link-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" rel="tag">æ•°æ®å¤„ç†</a>
    </div>

                

                
   <div class="article-counter">
      <i class="fa fa-file-word"></i>
      <span class="post-count-item-text">æœ¬æ–‡å­—æ•°: </span>
      <span class="post-count">18.3k</span>
      <i class="fa fa-clock"></i>
      <span class="post-time-item-text">é˜…è¯»æ—¶é•¿: </span>
      <span class="post-count">87 åˆ†é’Ÿ </span>
   </div>


            </div>
        
        
        <br>
        <div class="article-entry" itemprop="articleBody">
            <img src="/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png" class>
<blockquote><p><span style="font-size:24px"><strong>å½“ç”Ÿç‰©ä¸Pythonäº¤ç»‡ä¹‹æ—¶</strong></span></p>
<p><span style="font-size:24px"><strong>æ•…äº‹å°†å°±æ­¤å±•å¼€</strong></span></p>
</blockquote>
<span id="more"></span>
<div class="toc">
<!-- toc -->
<ul>
<li><a href="#shu-ji-jian-jie-books">ä¹¦ç±ç®€ä»‹ <span class="github-emoji"><span>ğŸ“š</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4da.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a></li>
<li><a href="#xie-zai-qian-mian-writing-hand">å†™åœ¨å‰é¢ <span class="github-emoji"><span>âœ</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/270d.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a></li>
<li><a href="#gu-shi-jiu-ci-zhan-kai-book">æ•…äº‹å°±æ­¤å±•å¼€ <span class="github-emoji"><span>ğŸ“–</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4d6.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a>
<ul>
<li><a href="#di-yi-bu-fen-ru-men">ç¬¬ä¸€éƒ¨åˆ† å…¥é—¨</a></li>
<li><a href="#di-er-bu-fen-shu-ju-guan-li">ç¬¬äºŒéƒ¨åˆ† æ•°æ®ç®¡ç†</a></li>
<li><a href="#di-san-bu-fen-mo-kuai-hua-bian-cheng">ç¬¬ä¸‰éƒ¨åˆ† æ¨¡å—åŒ–ç¼–ç¨‹</a></li>
<li><a href="#di-si-bu-fen-shu-ju-ke-shi-hua">ç¬¬å››éƒ¨åˆ† æ•°æ®å¯è§†åŒ–</a></li>
<li><a href="#di-wu-bu-fen-biopython">ç¬¬äº”éƒ¨åˆ† Biopython</a></li>
<li><a href="#di-liu-bu-fen-bian-cheng-mi-ji">ç¬¬å…­éƒ¨åˆ† ç¼–ç¨‹ç§˜ç¬ˆ</a></li>
<li><a href="#fu-lu">é™„å½•</a></li>
</ul>
</li>
</ul>
<!-- tocstop -->
</div>
<p>æ¥è§¦ <code>è®¡ç®—æœºæŠ€æœ¯(IT)</code> ä¹Ÿå·²ç»æœ‰äº›å¹´å¤´äº†</p>
<p>è®¡ç®—æœºæŠ€æœ¯è¡¨ç°å‡ºæ¥çš„ <code>é«˜æ•ˆ</code> çš„ç”Ÿäº§åŠ›ç€å®è®©äººèµå¹, è®©äººæœ‰æ— ç©·çš„åŠ¨åŠ›å‚ä¸å…¶ä¸­</p>
<p>å¦‚ä»Šå…¥äº†ç”Ÿç‰©çš„å‘, è‡ªç„¶ä¸€ç›´æœ‰æ¢ç´¢å„ç§ç”Ÿç‰©å’Œä¿¡æ¯æŠ€æœ¯äº¤èçš„çŸ¥è¯†é¢†åŸŸ</p>
<p>ä½†æ˜¯æµ…å°è¾„æ­¢çš„å…´è¶£å¼çš„æ¸¸ç©, ç»ˆç©¶æ˜¯ä¸å¤ªå¤Ÿç”¨çš„</p>
<p>å€’ä¸å¦‚é™ä¸‹å¿ƒæ¥, å®é™…åœ°ç³»ç»Ÿåœ°å­¦ä¹ ä¸€äº›å†…å®¹</p>
<h1><a href="#shu-ji-jian-jie" class="header-anchor"> âš“ï¸</a><span id="shu-ji-jian-jie">ä¹¦ç±ç®€ä»‹ <span class="github-emoji"><span>ğŸ“š</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4da.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<p>æŒ‘é€‰çš„ä¹¦æ˜¯ä» <code>æ±Ÿç§‘å¤§é•¿å±±æ ¡åŒºå›¾ä¹¦é¦†</code> é‡Œç¿»æ‰¾åˆ°çš„</p>
<p>ä½äº <code>è‡ªç„¶ç§‘å­¦å›¾ä¹¦ç¬¬2é˜…è§ˆå®¤(5æ¥¼)</code>, 5å·æ¶Bé¢5å±‚, ç´¢ä¹¦å·: <code>Q811.4/51</code></p>
<blockquote>
<p>@ä¿¡æ¯æ¥è‡ª<a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/search_adv.php#/index">æ±Ÿç§‘å¤§å›¾ä¹¦é¦†v5.6ä¹¦ç›®æ£€ç´¢ç³»ç»Ÿ</a></p>
<ul>
<li>
<p>é¢˜å/è´£ä»»è€…: <a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/item.php?marc_no=4b33443333726c7851522b2b4e57577445694d3150513d3d&amp;list=1">Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç†</a>/(æ„) Allegra Via, (å¾·) Kristian Rother, (æ„) Anna Tramontanoè‘— å¢å®è¶…, é™ˆä¸€æƒ…, æç»å¨Ÿè¯‘</p>
</li>
<li>
<p>å‡ºç‰ˆå‘è¡Œé¡¹: åŒ—äº¬:ç”µå­å·¥ä¸šå‡ºç‰ˆç¤¾,2017.01</p>
</li>
<li>
<p>ISBNåŠå®šä»·: 978-7-121-30382-1/CNY69.00</p>
</li>
<li>
<p>è½½ä½“å½¢æ€é¡¹: 17, 318é¡µ:å›¾;26cm</p>
</li>
<li>
<p>ç»Ÿä¸€é¢˜å: <a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/openlink.php?title=Managing+your+biological+data+with+Python">Managing your biological data with Python</a></p>
</li>
<li>
<p>ä¸­å›¾æ³•åˆ†ç±»å·:<a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/openlink.php?coden=Q811.4-39">Q811.4-39</a></p>
</li>
<li>
<p>ç‰ˆæœ¬é™„æ³¨:ç”±Taylor &amp; Franciså‡ºç‰ˆé›†å›¢æ——ä¸‹ï¼ŒCRCå‡ºç‰ˆå…¬å¸æˆæƒç¿»è¯‘å‡ºç‰ˆ</p>
</li>
<li>
<p>æè¦æ–‡æ‘˜é™„æ³¨:</p>
</li>
</ul>
<p>æœ¬ä¹¦ç”±å…­éƒ¨åˆ†ç»„æˆï¼šPythonè¯­è¨€åŸºæœ¬ä»‹ç»ï¼Œè¯­è¨€æ‰€æœ‰æˆåˆ†ä»‹ç»ï¼Œé«˜çº§ç¼–ç¨‹ï¼Œæ•°æ®å¯è§†åŒ–ï¼Œç”Ÿç‰©ä¿¡æ¯é€šç”¨åŒ…Biopythonï¼Œæœ€åç»™å‡º20ä¸ª"ç¼–ç¨‹ç§˜ç¬ˆâ€ï¼ŒèŒƒå›´æ¶µç›–äº†ä»äºŒçº§ç»“æ„é¢„æµ‹ã€å¤šåºåˆ—æ¯”å¯¹åˆ°è›‹ç™½è´¨ä¸‰ç»´ç»“æ„çš„å¹¿æ³›è¯é¢˜ã€‚æ­¤å¤–ï¼Œé™„å½•è¿˜åŒ…æ‹¬äº†å¤§é‡çš„ç”Ÿç‰©ä¿¡æ¯å¸¸ç”¨èµ„æºçš„ä¿¡æ¯ã€‚</p>
</blockquote>
<p>ä¸è¿‡æ‰¾ä¹Ÿè´¹äº†ç•ªåŠŸå¤«, è€Œä¸”å‘ç°å­¦æ ¡å…³äº<code>ç”Ÿç‰©ä¿¡æ¯å­¦</code>æ–¹é¢çš„ä¹¦è¿˜çœŸçš„ä¸æ˜¯éå¸¸å¤š</p>
<p>å¯è§<code>ç”Ÿç‰©ä¿¡æ¯å­¦</code>çš„ç¡®è¿˜æ˜¯æ¯”è¾ƒ<strong>æ–°</strong>çš„ä¸€é—¨å­¦ç§‘</p>
<p><a target="_blank" rel="noopener" href="http://libgen.rs/book/index.php?md5=B980D099C8E284CBEE2241C4A17319BC">Library Genesis: Via, Allegra; Rother, Kristian; Tramontano, Anna - Managing Your Biological Data with Python (libgen.rs)</a></p>
<h1><a href="#xie-zai-qian-mian" class="header-anchor"> âš“ï¸</a><span id="xie-zai-qian-mian">å†™åœ¨å‰é¢ <span class="github-emoji"><span>âœ</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/270d.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<p>æ­¤è¯»ä¹¦ç¬”è®°ä¼šå¿½ç•¥äº¦æˆ–è€…å¢åŠ ä¸€äº›å†…å®¹, æ¯”å¦‚</p>
<ul>
<li>å¿½ç•¥åŸºç¡€çš„<code>Pythonè¯­æ³•</code>: æœ¬ä¹¦è¿˜æ˜¯ä½¿ç”¨äº†å¤§é‡çš„ä¾‹å­æ¥è®²è§£Pythonçš„åŸºç¡€è¯­æ³•çš„, ä½†æ˜¯å¯¹äºç¬”è€…æ¥è¯´å·²æ— å¿…è¦. éœ€è¦è¡¥å……åŸºç¡€è¯­æ³•çš„å°ä¼™ä¼´å¯ä»¥çœ‹è¿™é‡Œ =&gt; <a target="_blank" rel="noopener" href="https://www.runoob.com/python3/python3-tutorial.html">Python3 æ•™ç¨‹ | èœé¸Ÿæ•™ç¨‹ (runoob.com)</a></li>
<li>å¢åŠ ä¸€äº›æ¶‰åŠåˆ°çš„ç”Ÿç‰©å­¦ç†è®ºç»†èŠ‚: è¿™ä¸»è¦æ˜¯è¡¥å……ä¹¦ä¸­çš„å†…å®¹, ä½¿å¾—æ˜“äºç†è§£ç›¸å…³çš„ä¾‹å­</li>
</ul>
<p>å½“ç„¶, å…¶å®ä¸»è¦æ˜¯ä»€ä¹ˆå€¼å¾—è¡¥å……å°±ä¼šè¡¥å……çš„ :happy:</p>
<p>æœ¬ç¬”è®°æŒ‰ç…§åŸä¹¦çš„ç›®å½•é¡ºåºå±•å¼€, ä½†æ˜¯ä»…ä¿ç•™å°‘éƒ¨åˆ†çš„æ ‡é¢˜, ä¸»è¦ä»¥<code>æŠ€æœ¯æµç¨‹</code>æ¥æ•´åˆç¬”è®°</p>
<p>æœ¬ä¹¦çš„ä»£ç ä½¿ç”¨çš„Pythonçš„ç‰ˆæœ¬ä¸º <code>Python 2.6.6</code></p>
<p>ç¬”è€…ä½¿ç”¨ç”± <code>Anaconda</code> æ„å»ºçš„Pythonç‰ˆæœ¬ä¸º <code>python-2.7.13</code> çš„è™šæ‹Ÿç¯å¢ƒæ¥å¤ç°ä¹¦ä¸­çš„ä»£ç </p>
<p>æœŸé—´å¯èƒ½ä¼šå°è¯•ä½¿ç”¨ <code>Python3.8</code> çš„ç‰ˆæœ¬æ¥é‡æ„é‡Œé¢çš„ä»£ç </p>
<p>æ²¡æœ‰ç‰¹åˆ«è¯´æ˜, ç¬”è®°é‡Œçš„Pythonç‰ˆæœ¬åˆ™ä¸º <code>python-2.7.13</code></p>
<p>äº‹å®ä¸Š, æŒæ¡å……è¶³çš„ç”Ÿç‰©å­¦çŸ¥è¯†æ˜¯éå¸¸é‡è¦çš„</p>
<p>è®¡ç®—æœºæŠ€æœ¯åªæ˜¯å¸®åŠ©æˆ‘ä»¬è‡ªåŠ¨åŒ–åœ°å®Œæˆäº†åŸºäºç”Ÿç‰©å­¦è®¤è¯†çš„å·¥ä½œ</p>
<p>æœ¬ä¹¦ä¹Ÿåªæ˜¯è®²è§£äº†å¦‚ä½•å°†Pythonå¼•å…¥åˆ†æä»¥åŠæ•°æ®ç®¡ç†çš„å·¥ä½œæµ</p>
<p>å½“ä½ æŒæ¡çœŸæ­£çš„ç®—æ³•å’Œç†è®ºå®éªŒèƒŒæ™¯, ä½ å°†å¯ä»¥æ›´å¥½åœ°åˆ©ç”¨Pythonè¿™ä¸ªå¼ºå¤§çš„å·¥å…·</p>
<p><strong>å¤šå¤šåšå®éªŒ, çœ‹çœ‹ç†è®ºä¹¦å§</strong></p>
<h1><a href="#gu-shi-jiu-ci-zhan-kai" class="header-anchor"> âš“ï¸</a><span id="gu-shi-jiu-ci-zhan-kai">æ•…äº‹å°±æ­¤å±•å¼€ <span class="github-emoji"><span>ğŸ“–</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4d6.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<h2><a href="#di-yi-bu-fen-ru-men" class="header-anchor"> âš“ï¸</a><span id="di-yi-bu-fen-ru-men">ç¬¬ä¸€éƒ¨åˆ† å…¥é—¨</span></h2>
<p>æœ¬ä¹¦çš„ä½œè€…æè¿°äº†ä»–å¸Œæœ›è¿™æœ¬ä¹¦å¸¦ç»™è¯»è€…ä»€ä¹ˆ</p>
<blockquote>
<p>æˆ‘ä»¬å¸Œæœ›æœ¬ä¹¦èƒ½æˆä¸ºå¸®åŠ©è¯»è€…æ¯å¤©æ”€ç™»æ•°æ®ä¹‹å±±çš„åœ°å›¾, å¸Œæœ›è¯»è€…èƒ½è®©ç¼–ç¨‹ä¸ºå·¥ä½œæä¾›ä¾¿åˆ©, è€Œä¸å¿…å…ˆæˆä¸ºä¸€ä¸ªä¸“ä¸šçš„è½¯ä»¶å¼€å‘è€…</p>
</blockquote>
<p>ä½œè€…å»ºè®®åƒå­¦ä¹ å¾·è¯­ä¸€æ ·, å­¦ä¹ Pythonè¯­æ³•, æŠŠ<code>Pythonæ ‡å‡†åº“</code> å½“ä½œå­—å…¸æ¥æŸ¥</p>
<p><code>Pythonæ ‡å‡†åº“</code> :</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/3.7/library/index.html">Python æ ‡å‡†åº“ â€” Python 3.7.8 æ–‡æ¡£</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/index.html">Python æ ‡å‡†åº“ â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
</ul>
<p><code>ä¹¦ä¸­çš„ä»£ç </code>å’Œ<code>ä¸€äº›æ€»ç»“</code>å‚è€ƒ:</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://ugoproto.github.io/ugo_py_doc/managing_your_biological_data_with_python/">Managing Your Biological Data with Python - ugo_py_doc (ugoproto.github.io)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://github.com/wqhf/Managing-Your-Biological-Data-with-Python">Managing-Your-Biological-Data-with-Python (github.com)</a></p>
</li>
</ul>
<h3><a href="#di-1-zhang-python-shell" class="header-anchor"> âš“ï¸</a><span id="di-1-zhang-python-shell">ç¬¬1ç«  Python shell</span></h3>
<p>å…³äº<code>Python</code>çš„å®‰è£…å¯ä»¥å‚è€ƒè¿™ç¯‡æ–‡ç«  <a href="/anaconda-zhi-lu/" title="Anacondaä¹‹è·¯">Anacondaä¹‹è·¯</a></p>
<p>ä¸»è¦ä»‹ç»äº†å¦‚ä½•ä½¿ç”¨<code>Anaconda</code>æ¥æ„å»º<code>Python</code>è™šæ‹Ÿç¯å¢ƒ</p>
<blockquote>
<p><code>Anaconda</code>æ˜¯ä¸€ä¸ªå¯ä»¥ç®¡ç†<code>PythonåŒ…</code>å’Œ<code>Pythonç¯å¢ƒ</code>çš„è½¯ä»¶</p>
<p>æ‰€è°“<code>PythonåŒ…</code>æŒ‡çš„æ˜¯<strong>åˆ«äººå†™</strong>(ç¬¬ä¸‰æ–¹)çš„æœ‰ä¸€å®šåŠŸèƒ½çš„<code>Pythonä»£ç </code></p>
<p>æ‰€è°“<code>Pythonç¯å¢ƒ</code>å…¶å®å°±æ˜¯<code>Pythonä»£ç </code>è¿è¡Œæ—¶è°ƒç”¨çš„å…·ä½“ç‰ˆæœ¬çš„<code>Pythonè§£é‡Šå™¨</code></p>
<p>æ‰€è°“<code>ç®¡ç†</code>æŒ‡çš„å°±æ˜¯æˆ‘ä»¬å¯ä»¥é€šè¿‡<code>Anaconda</code>çš„<code>conda</code>æŒ‡ä»¤, éšæ„åœ°åˆ›å»ºåŒ…å«æœ‰æˆ‘ä»¬æŒ‡å®šçš„<code>PythonåŒ…</code>çš„æŒ‡å®šç‰ˆæœ¬çš„<code>Pythonç¯å¢ƒ</code>, è¿™æ ·åšçš„ç›®çš„æ˜¯è§£å†³å…¼å®¹æ€§é—®é¢˜(ç›¸äº’ä¾èµ–é—®é¢˜)</p>
<p>æ¯”å¦‚è¿™æœ¬ä¹¦çš„ä»£ç æ˜¯<code>Python2</code>æ—¶ä»£ç¼–å†™çš„, æˆ‘è‡ªç„¶æ˜¯ä¸å¯ä»¥ç”¨<code>Python3</code>çš„è§£é‡Šå™¨æ¥è¿è¡Œçš„.</p>
<p>é‚£å°±å¯ä»¥åˆ›å»ºä¸€ä¸ª<code>Python2</code>çš„ç¯å¢ƒæ¥å¤ç°ä¹¦ä¸­çš„ä»£ç , ç„¶åå†åˆ›å»ºä¸€ä¸ª<code>Python3</code>çš„ç¯å¢ƒæ¥é‡æ–°ç¼–å†™ä»–ä»¬(é‡æ„)</p>
</blockquote>
<p>è¿™æ ·æˆ‘ä»¬å°±å¯ä»¥å¤ç°ä¹¦ä¸­çš„ä»£ç äº†, è€Œä¸”è¿˜å¯ä»¥è½»æ¾çš„é‡æ„è€æ—§çš„ä»£ç </p>
<h4><a href="#1-2-an-li-ji-suan-atp-shui-jie-de-dg" class="header-anchor"> âš“ï¸</a><span id="1-2-an-li-ji-suan-atp-shui-jie-de-dg">1.2 æ¡ˆä¾‹ï¼š è®¡ç®—ATPæ°´è§£çš„Î”G</span></h4>
<ul>
<li><code>å‰å¸ƒæ–¯è‡ªç”±èƒ½</code>:</li>
</ul>
<blockquote>
<p>åœ¨<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%83%AD%E5%8A%9B%E5%AD%A6">çƒ­åŠ›å­¦</a>é‡Œï¼Œ<strong>å‰å¸ƒæ–¯èƒ½</strong>ï¼ˆè‹±è¯­ï¼š<strong>Gibbs Free Energy</strong>ï¼‰ï¼Œåˆç§°<strong>å‰å¸ƒæ–¯è‡ªç”±èƒ½</strong>ã€<strong>å‰å¸ƒæ–¯å‡½æ•°</strong>ã€<strong>è‡ªç”±ç„“</strong>ï¼Œå¸¸ç”¨è‹±æ–‡å­—æ¯<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/f5f3c8921a3b352de45446a6789b104458c9f90b" alt>æ ‡è®°ã€‚å‰å¸ƒæ–¯èƒ½æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%9C%8B%E9%9A%9B%E7%B4%94%E5%8C%96%E5%AD%B8%E5%92%8C%E6%87%89%E7%94%A8%E5%8C%96%E5%AD%B8%E8%81%AF%E5%90%88%E6%9C%83">å›½é™…åŒ–å­¦è”ä¼š</a>å»ºè®®é‡‡ç”¨çš„åç§°ã€‚å‰å¸ƒæ–¯èƒ½æ˜¯æè¿°ç³»ç»Ÿçš„çƒ­åŠ›æ€§è´¨çš„ä¸€ç§<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%86%B1%E5%8A%9B%E5%8B%A2">çƒ­åŠ›åŠ¿</a>ï¼Œå®šä¹‰ä¸º[<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-cn/%E5%90%89%E5%B8%83%E6%96%AF%E8%83%BD#cite_note-Greiner1995-1">1]</a>:101[<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-cn/%E5%90%89%E5%B8%83%E6%96%AF%E8%83%BD#cite_note-Perrot-2">2]</a>:128-129</p>
<p><img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/6212e95f1f023c632b04fa0600e5a37259383a6c" alt></p>
<p>å…¶ä¸­ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/458a728f53b9a0274f059cd695e067c430956025" alt>æ˜¯ç³»ç»Ÿçš„<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%86%85%E8%83%BD">å†…èƒ½</a>ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/ec7200acd984a1d3a3d7dc455e262fbe54f7f6e0" alt>æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%B5%95%E5%B0%8D%E6%BA%AB%E5%BA%A6">ç»å¯¹æ¸©åº¦</a>ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/4611d85173cd3b508e67077d4a1252c9c05abca2" alt>æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%86%B5">ç†µ</a>ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/81eac1e205430d1f40810df36a0edffdc367af36" alt>æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%8E%8B%E5%BC%BA">å‹å¼º</a>ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/af0f6064540e84211d0ffe4dac72098adfa52845" alt>æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E4%BD%93%E7%A7%AF">ä½“ç§¯</a>ï¼Œ<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/75a9edddcca2f782014371f75dca39d7e13a9c1b" alt>æ˜¯<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%84%93">ç„“</a>ã€‚</p>
</blockquote>
<p>$$<br>
\Delta G=\Delta G^{0}+R T * \ln \left([A D P] *\left[P_{i}\right] /[A T P]\right)<br>
$$</p>
<p>$$<br>
[?]: æŒ‡ ? çš„æ‘©å°”æµ“åº¦   \\  Î”G : å‰å¸ƒæ–¯è‡ªç”±èƒ½  \\  Î”G^0 : æ ‡å‡†ååº”å‰å¸ƒæ–¯èƒ½   \\  R : æ‘©å°”æ°”ä½“å¸¸æ•°  \\  T : æ¸©åº¦  \\<br>
$$</p>
<ul>
<li>pythonçš„ <code>math</code> æ¨¡å—:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">import</span> math  <span class="comment"># å¯¼å…¥æ•°å­¦è®¡ç®—æ¨¡å—</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">dir</span>(math)    <span class="comment"># dir()å‡½æ•°å¯ä»¥è·å–ä¼ å…¥çš„æ¨¡å—çš„å†…éƒ¨çš„å‡½æ•°</span></span><br><span class="line">[<span class="string">'__doc__'</span>, <span class="string">'__name__'</span>, <span class="string">'__package__'</span>, <span class="string">'acos'</span>, <span class="string">'acosh'</span>, <span class="string">'asin'</span>, <span class="string">'asinh'</span>, <span class="string">'atan'</span>, <span class="string">'atan2'</span>, <span class="string">'atanh'</span>, <span class="string">'ceil'</span>, <span class="string">'copysign'</span>, <span class="string">'cos'</span>, <span class="string">'cosh'</span>, <span class="string">'degrees'</span>, <span class="string">'e'</span>, <span class="string">'erf'</span>, <span class="string">'erfc'</span>, <span class="string">'exp'</span>, <span class="string">'expm1'</span>, <span class="string">'fabs'</span>, <span class="string">'factorial'</span>, <span class="string">'floor'</span>, <span class="string">'fmod'</span>, <span class="string">'frexp'</span>, <span class="string">'fsum'</span>, <span class="string">'gamma'</span>, <span class="string">'hypot'</span>, <span class="string">'isinf'</span>, <span class="string">'isnan'</span>, <span class="string">'ldexp'</span>, <span class="string">'lgamma'</span>, <span class="string">'log'</span>, <span class="string">'log10'</span>, <span class="string">'log1p'</span>, <span class="string">'modf'</span>, <span class="string">'pi'</span>, <span class="string">'pow'</span>, <span class="string">'radians'</span>, <span class="string">'sin'</span>, <span class="string">'sinh'</span>, <span class="string">'sqrt'</span>, <span class="string">'tan'</span>, <span class="string">'tanh'</span>, <span class="string">'trunc'</span>]</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>math.log(...)  <span class="comment"># ä¼ å…¥å‚æ•°, è¿”å›å®ƒçš„è‡ªç„¶å¯¹æ•°</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">help</span>(math.log) <span class="comment"># ä½¿ç”¨`å¸®åŠ©`è¿™ä¸ªå†…å»ºå‡½æ•°æŸ¥çœ‹æŸä¸ªå‡½æ•°çš„ä½¿ç”¨è¯´æ˜</span></span><br><span class="line">Help on built-<span class="keyword">in</span> function log <span class="keyword">in</span> module math:</span><br><span class="line">log(...)</span><br><span class="line">  log(x[, base])</span><br><span class="line">    Return the logarithm of x to the given base.</span><br><span class="line">    If the base <span class="keyword">not</span> specified, returns the natural logarithm (base e) of x.</span><br></pre></td></tr></tbody></table></figure>
<h4><a href="#1-3-4-ji-suan" class="header-anchor"> âš“ï¸</a><span id="1-3-4-ji-suan">1.3.4 è®¡ç®—</span></h4>
<p>è¿™é‡Œå¼•å…¥ä¸€ç¯‡åšæ–‡æ¥è¡¥å……å„ç§<code>ç¬¦å·</code>çš„æ„æ€</p>
<p><a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_38981172/article/details/90176525">pythonä¸­çš„å„ç§ç¬¦å·ï¼ˆæ¬¢è¿è¡¥å……ï¼‰_æ‰æœ¨äºº-CSDNåšå®¢</a></p>
<p>ç‰¹åˆ«ä¹‹å¤„åœ¨Pythonä¸­, <code>^</code> ç¬¦å·</p>
<table>
<thead>
<tr>
<th style="text-align:center">ç¬¦å·</th>
<th style="text-align:center">åŠŸèƒ½</th>
<th style="text-align:center">ä¾‹å­</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center"><code>&lt;sup&gt;</code></td>
<td style="text-align:center">æŒ‰ä½å¼‚æˆ–è¿ç®—ç¬¦ï¼šå½“ä¸¤å¯¹åº”çš„äºŒè¿›ä½ç›¸å¼‚æ—¶ï¼Œç»“æœä¸º1</td>
<td style="text-align:center">(a  b) è¾“å‡ºç»“æœ 49 ï¼ŒäºŒè¿›åˆ¶è§£é‡Šï¼š 0011 0001</td>
</tr>
<tr>
<td style="text-align:center"><code>**</code></td>
<td style="text-align:center">å¹‚ - è¿”å›xçš„yæ¬¡å¹‚</td>
<td style="text-align:center">a**b ä¸º10çš„20æ¬¡æ–¹ï¼Œ è¾“å‡ºç»“æœ 100000000000000000000</td>
</tr>
</tbody>
</table>
<p>å‚è§ <code>Python2.7.18æ–‡æ¡£</code> @<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/operator.html?highlight=%E8%BF%90%E7%AE%97%E7%AC%A6#mapping-operators-to-functions">9.9.1. å°†è¿ç®—ç¬¦æ˜ å°„åˆ°å‡½æ•°</a></p>
<ul>
<li>è®¡ç®—æµç¨‹:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>ATP = <span class="number">3.5</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>ADP = <span class="number">1.8</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>Pi = <span class="number">5.0</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>R = <span class="number">0.00831</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>T = <span class="number">298</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>deltaG0 = -<span class="number">30.5</span></span><br><span class="line">&gt;&gt;&gt;</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">import</span> math</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>deltaG0 + R * T * math.log(ADP * Pi / ATP)</span><br><span class="line">-<span class="number">28.161154161098693</span></span><br></pre></td></tr></tbody></table></figure>
<p>é€šè¿‡å¯¼å…¥å„ç§å‚æ•°, å¯ä»¥ç›´æ¥è®¡ç®—å‡º<code>å‰å¸ƒæ–¯è‡ªç”±èƒ½</code></p>
<h3><a href="#di-2-zhang-di-yi-ge-python-cheng-xu" class="header-anchor"> âš“ï¸</a><span id="di-2-zhang-di-yi-ge-python-cheng-xu">ç¬¬2ç«  ç¬¬ä¸€ä¸ªPythonç¨‹åº</span></h3>
<h4><a href="#2-2-an-li-ru-he-ji-suan-yi-dao-su-xu-lie-zhong-de-an-ji-suan-pin-lu" class="header-anchor"> âš“ï¸</a><span id="2-2-an-li-ru-he-ji-suan-yi-dao-su-xu-lie-zhong-de-an-ji-suan-pin-lu">2.2æ¡ˆä¾‹ï¼š å¦‚ä½•è®¡ç®—èƒ°å²›ç´ åºåˆ—ä¸­çš„æ°¨åŸºé…¸é¢‘ç‡ *</span></h4>
<ul>
<li>è®¡ç®—æµç¨‹:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># insulin [Homo sapiens] GI:386828</span></span><br><span class="line"><span class="comment"># extracted 51 amino acids of A+B chain</span></span><br><span class="line">insulin = <span class="string">"GIVEQCCTSICSLYQLENYCNFVNQHLCGSHLVEALYLVCGERGFFYTPKT"</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> amino_acid <span class="keyword">in</span> <span class="string">"ACDEFGHIKLMNPQRSTVWY"</span>:</span><br><span class="line">    number = insulin.count(amino_acid)</span><br><span class="line">    <span class="built_in">print</span> amino_acid,number</span><br><span class="line"></span><br><span class="line">A <span class="number">1</span></span><br><span class="line">C <span class="number">6</span></span><br><span class="line">D <span class="number">0</span></span><br><span class="line">E <span class="number">4</span></span><br><span class="line">F <span class="number">3</span></span><br><span class="line">G <span class="number">4</span></span><br><span class="line">H <span class="number">2</span></span><br><span class="line">I <span class="number">2</span></span><br><span class="line">K <span class="number">1</span></span><br><span class="line">L <span class="number">6</span></span><br><span class="line">M <span class="number">0</span></span><br><span class="line">N <span class="number">3</span></span><br><span class="line">P <span class="number">1</span></span><br><span class="line">Q <span class="number">3</span></span><br><span class="line">R <span class="number">1</span></span><br><span class="line">S <span class="number">3</span></span><br><span class="line">T <span class="number">3</span></span><br><span class="line">V <span class="number">4</span></span><br><span class="line">W <span class="number">0</span></span><br><span class="line">Y <span class="number">4</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>â€œå­—ç¬¦ä¸²1â€.count(â€œå­—ç¬¦ä¸²2â€) : è¿”å›<code>å­—ç¬¦ä¸²1</code>ä¸­å«æœ‰å¤šå°‘<code>å­—ç¬¦ä¸²2</code>çš„æ•°é‡</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/string.html?highlight=count#string.count">7.1. string â€” å¸¸è§çš„å­—ç¬¦ä¸²æ“ä½œ â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p><code>string.``count</code>(<em>s</em>, <em>sub</em>[, <em>start</em>[, <em>end</em>]])</p>
<p>Return the number of (non-overlapping) occurrences of substring <em>sub</em> in string <code>s[start:end]</code>. Defaults for <em>start</em> and <em>end</em> and interpretation of negative values are the same as for slices.</p>
</li>
</ul>
</li>
<li>
<p>æ³¨æ„ç¼©è¿›é”™è¯¯ï¼šIndentationError: unindent does not match any outer indentation level</p>
</li>
</ul>
<p>åˆ‡ä¸å¯æ··ç”¨ <code>ç©ºæ ¼ç¼©è¿›</code>å’Œ<code>åˆ¶è¡¨ç¬¦ç¼©è¿›</code>, å¦åˆ™ä½ ä¼šç–¯çš„</p>
<ul>
<li>Python çš„åæ–œæ æ¢è¡Œ:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="number">3</span> + <span class="number">5</span> + \</span><br><span class="line"><span class="meta">... </span><span class="number">7</span></span><br><span class="line"><span class="number">15</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>å†…å»ºå‡½æ•°(build-in): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#zip">2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<p>æ¯”å¦‚ <code>len()</code> å¯ä»¥ç›´æ¥è°ƒç”¨, å¹¶è¿”å›ç»“æœ</p>
</li>
<li>
<p>å¯¹è±¡æ–¹æ³•: æŸä¸ªå¯¹è±¡(ç±»çš„å®ä¾‹)çš„å†…éƒ¨çš„æ–¹æ³•(å‡½æ•°), é€šè¿‡<code>.</code>ç¬¦å·è°ƒç”¨, å¦‚ <code>" ".count(' ')</code></p>
</li>
<li>
<p><code>in</code>è¿­ä»£:</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> ... <span class="keyword">in</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="comment"># `in` å¯ä»¥è¿­ä»£å­—ç¬¦ä¸²ç­‰</span></span><br></pre></td></tr></tbody></table></figure>
<p>è¡¥å…… in</p>
<ul>
<li><code>,</code> å¯¹ <code>print</code> è¾“å‡ºçš„å½±å“:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span> <span class="number">1</span>,<span class="number">2</span>,   <span class="comment"># æ³¨æ„æœ«å°¾çš„é€—å·å¯ä»¥æŠ‘åˆ¶æ¢è¡Œç¬¦</span></span><br><span class="line"><span class="built_in">print</span> <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="number">1</span> <span class="number">2</span> <span class="number">3</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>è½¬ä¹‰å­—ç¬¦<code>\</code>:</p>
</li>
<li>
<p>å¼•å·ä¹‹é—´çš„åµŒå¥—:</p>
</li>
<li>
<p><code>randomæ¨¡å—</code>:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/random.html#module-random">9.6. random â€” ç”Ÿæˆä¼ªéšæœºæ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
</ul>
<h4><a href="#2-5-zi-ce-ti" class="header-anchor"> âš“ï¸</a><span id="2-5-zi-ce-ti">2.5è‡ªæµ‹é¢˜</span></h4>
<blockquote>
<p>2009å¹´, Elizabeth H. Blackburn, Carol W. Greide å’Œ Jack W. Szostak å› å‘ç°ç«¯ç²’é…¶åŠŸèƒ½è€Œè·å¾—è¯ºè´å°”å¥–, ç«¯ç²’é…¶è´Ÿè´£å»¶ä¼¸æŸ“è‰²ä½“çš„æœ«ç«¯. é€šè¿‡NCBIçš„è›‹ç™½è´¨æ•°æ®åº“æ£€ç´¢äººç±»äºšå‹Içš„ç«¯ç²’çš„11132ä¸ªæ®‹åŸºåºåˆ—, è¯•é—®å…¶ä¸­å“ªç§æ°¨åŸºé…¸å‡ºç°çš„æœ€ä¸ºé¢‘ç¹?</p>
</blockquote>
<h2><a href="#di-er-bu-fen-shu-ju-guan-li" class="header-anchor"> âš“ï¸</a><span id="di-er-bu-fen-shu-ju-guan-li">ç¬¬äºŒéƒ¨åˆ† æ•°æ®ç®¡ç†</span></h2>
<h3><a href="#di-3-zhang-fen-xi-shu-ju-lie" class="header-anchor"> âš“ï¸</a><span id="di-3-zhang-fen-xi-shu-ju-lie">ç¬¬3ç«  åˆ†ææ•°æ®åˆ—</span></h3>
<p>è¯‘è€…æ¨èæ ‡å‡†åº“æ¨¡å— <code>collections</code> :</p>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/collections.html#module-collections">8.3. collections â€” High-performance container datatypes â€” Python 2.7.18 æ–‡æ¡£</a></p>
<h4><a href="#3-2-an-li-shu-tu-chang-du" class="header-anchor"> âš“ï¸</a><span id="3-2-an-li-shu-tu-chang-du">3.2æ¡ˆä¾‹ï¼š æ ‘çªé•¿åº¦</span></h4>
<p>ä½œè€…ä½¿ç”¨äº† <code>ImageJ</code> æ¥è®¡ç®—æ ‘çªçš„é•¿åº¦å¹¶æŠŠå€¼ä¼ å…¥æ–‡æœ¬æ–‡ä»¶</p>
<p>è½¯ä»¶æŒ‡è·¯: <a target="_blank" rel="noopener" href="https://imagej.net/Welcome">ImageJ</a></p>
<ul>
<li><code>in</code> è¿­ä»£:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># forå¾ªç¯ä¸­æŒ‰è¡Œè¿­ä»£æ–‡ä»¶çš„æ–¹æ³•</span></span><br><span class="line"><span class="keyword">for</span> ... <span class="keyword">in</span> <span class="built_in">open</span>(...):</span><br><span class="line">    ...</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>å†…å»ºå‡½æ•°(å†…ç½®å‡½æ•°): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=max#max">2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># å†…å»ºå‡½æ•° </span></span><br><span class="line"><span class="built_in">sum</span>(); <span class="built_in">min</span>(); <span class="built_in">max</span>(); <span class="built_in">len</span>()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>%</code> æ ¼å¼åŒ–è¾“å‡º::</p>
</li>
<li>
<p>æ–‡ä»¶çš„<code>readlines</code> æ–¹æ³•: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=readlines#file.readlines">file.readlines([sizehint]) â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
</ul>
<p>æ­¤æ–¹æ³•å¯ä»¥å°†æ–‡ä»¶çš„å†…å®¹<code>æŒ‰è¡Œ</code>ç»„åˆæˆ<code>åˆ—è¡¨</code>è¿”å›</p>
<ul>
<li>
<p><code>ç¡¬ç¼–ç (hard-coding)</code>:</p>
</li>
<li>
<p><strong>ä¸å…³é—­æ–‡ä»¶</strong>çš„éšæ‚£:</p>
</li>
</ul>
<p>å½“åœ¨å…¶ä»–æ—¶å€™è¯»å–æ–‡ä»¶çš„æ—¶å€™, ä¹‹å‰å†™å…¥çš„éƒ¨åˆ†å†…å®¹å› ä¸ºæ²¡æœ‰è¶…è¿‡ç¼“å†²åŒºå¤§å°, è¿˜å¤„äºå†…å­˜å†….</p>
<p>è¿™å°†å¯¼è‡´æœ‰æ—¶å€™ä½ å°†æ— æ³•è·å–æƒ³è¦çš„å†…å®¹, åˆ‡è®°åŠæ—¶å…³é—­æ–‡ä»¶.</p>
<ul>
<li>
<p><code>ç¼“å†²åŒº</code>:</p>
</li>
<li>
<p>æ–‡ä»¶çš„ <code>writelines</code> æ–¹æ³•:</p>
</li>
</ul>
<p>å°†<code>åˆ—è¡¨</code>æŒ‰<code>é¡¹</code>æ— ç©ºæ ¼è¿ç»­å†™å…¥<code>æ–‡ä»¶</code>, åˆ—è¡¨å†…åªæ”¯æŒ <code>å­—ç¬¦ä¸²</code> .</p>
<blockquote>
<p><code>file.writelines</code>(<em>sequence</em>)<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.writelines">Â¶</a></p>
<p>Write a sequence of strings to the file. The sequence can be any iterable object producing strings, typically a list of strings. There is no return value. (The name is intended to match <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.readlines"><code>readlines()</code></a>; <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.writelines"><code>writelines()</code></a> does not add line separators.)</p>
</blockquote>
<ul>
<li><code>" ".join()</code>æ–¹æ³•:</li>
</ul>
<p>è¿”å›ä¸€ä¸ªç”±<code>å¯è¿­ä»£å¯¹è±¡</code>ä¸­çš„å­—ç¬¦ä¸²<code>ä¸²è”" "</code>ç»„æˆçš„å­—ç¬¦ä¸²ã€‚</p>
<blockquote>
<p>str.join(<em>iterable</em>)<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=join#str.join">Â¶</a></p>
<p>Return a string which is the concatenation of the strings in <em>iterable</em>. If there is any Unicode object in <em>iterable</em>, return a Unicode instead. A <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/exceptions.html#exceptions.TypeError"><code>TypeError</code></a> will be raised if there are any non-string or non Unicode object values in <em>iterable</em>. The separator between elements is the string providing this method.</p>
</blockquote>
<h3><a href="#di-4-zhang-jie-xi-shu-ju-ji-lu" class="header-anchor"> âš“ï¸</a><span id="di-4-zhang-jie-xi-shu-ju-ji-lu">ç¬¬4ç«  è§£ææ•°æ®è®°å½•</span></h3>
<h4><a href="#4-2-an-li-zheng-he-zhi-pu-shu-ju-zhuan-hua-dao-dai-xie-tong-lu-zhong" class="header-anchor"> âš“ï¸</a><span id="4-2-an-li-zheng-he-zhi-pu-shu-ju-zhuan-hua-dao-dai-xie-tong-lu-zhong">4.2 æ¡ˆä¾‹ï¼š æ•´åˆè´¨è°±æ•°æ®ï¼Œ è½¬åŒ–åˆ°ä»£è°¢é€šè·¯ä¸­</span></h4>
<ul>
<li>
<p><code>UniProt</code> å½¢å¼:  <a target="_blank" rel="noopener" href="https://www.uniprot.org/">UniProt</a></p>
<blockquote>
<p>The mission of <a target="_blank" rel="noopener" href="https://www.uniprot.org/help/about">UniProt</a> is to provide the scientific community with a comprehensive, high-quality and freely accessible resource of protein sequence and functional information.</p>
</blockquote>
<ul>
<li>~ AC:</li>
</ul>
</li>
<li>
<p><code>CSV</code> é€—å·åˆ†éš”æ–‡ä»¶:</p>
</li>
<li>
<p><code>TSV</code> åˆ¶è¡¨ç¬¦åˆ†éš”æ–‡ä»¶:</p>
</li>
<li>
<p><code>MS</code> è´¨è°±:</p>
</li>
<li>
<p>ä½œè€…æåˆ°çš„è½¯ä»¶ä¸ç½‘ç«™:</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="http://www.matrixscience.com/">Mascot search engine</a>: Protein identification software for <strong>mass spec</strong>(è´¨è°±) data</p>
<blockquote>
<p>Welcome to the home of Mascot software, the benchmark for identification, characterisation and quantitation of proteins using mass spectrometry data. Here, you can learn more about the tools developed by Matrix Science to get the best out of your data, whatever your chosen instrument.</p>
</blockquote>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://reactome.org/">Reactome Pathway Database</a>:  <code>ä»£è°¢é€šè·¯</code>èµ„æºè·å–</p>
<blockquote>
<p>Reactome is a free, open-source, curated and peer-reviewed pathway database. Our goal is to provide intuitive bioinformatics tools for the visualization, interpretation and analysis of pathway knowledge to support basic research, genome analysis, modeling, systems biology and education.</p>
<p>If you use Reactome in Asia, we suggest using our Chinese mirror site at <a target="_blank" rel="noopener" href="http://reactome.ncpsb.org/">reactome.ncpsb.org</a>.</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># proteins participating in cell cycle å‚ä¸ç»†èƒé€šè·¯çš„è›‹ç™½</span></span><br><span class="line">list_a =[]</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">"cell_cycle_proteins.txt"</span>) :</span><br><span class="line">    list_a.append(line.strip())</span><br><span class="line"><span class="built_in">print</span> list_a</span><br><span class="line"></span><br><span class="line"><span class="comment"># proteins expressed in a given cancer cell åœ¨ç»™å®šçš„ç»†èƒä¸­è¡¨è¾¾çš„è›‹ç™½</span></span><br><span class="line">list_b =[]</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">"cancer_cell_proteins.txt"</span>) :</span><br><span class="line">    list_b.append(line.strip())</span><br><span class="line"><span class="built_in">print</span> list_b</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> protein <span class="keyword">in</span> list_a:  <span class="comment">#è¿­ä»£åˆ—è¡¨a</span></span><br><span class="line">    <span class="keyword">if</span> protein <span class="keyword">in</span> list_b:  <span class="comment">#åˆ¤æ–­æ˜¯ä¸æ˜¯åœ¨åˆ—è¡¨bä¸­</span></span><br><span class="line">        <span class="built_in">print</span> protein, <span class="string">'detected in the cancer cell'</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">print</span> protein, <span class="string">'not observed'</span></span><br></pre></td></tr></tbody></table></figure>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#if">7.1. The <code>if</code> statement</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># if è¯­å¥</span></span><br><span class="line"><span class="keyword">if</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">elif</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">elif</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    ....</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>åˆ—è¡¨æ¨å¯¼å¼: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/datastructures.html?highlight=%E5%88%97%E8%A1%A8#list-comprehensions">5.1.4. åˆ—è¡¨æ¨å¯¼å¼</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">bases = [<span class="string">'A'</span>,<span class="string">'C'</span>,<span class="string">'T'</span>,<span class="string">'G'</span>]</span><br><span class="line">seq = <span class="string">'GGACXCAGXXGATT'</span></span><br><span class="line">seqlist = [base <span class="keyword">for</span> base <span class="keyword">in</span> seq <span class="keyword">if</span> base <span class="keyword">in</span> bases]</span><br><span class="line"><span class="built_in">print</span> seqlist</span><br><span class="line"></span><br><span class="line">[<span class="string">'G'</span>, <span class="string">'G'</span>, <span class="string">'A'</span>, <span class="string">'C'</span>, <span class="string">'C'</span>, <span class="string">'A'</span>, <span class="string">'G'</span>, <span class="string">'G'</span>, <span class="string">'A'</span>, <span class="string">'T'</span>, <span class="string">'T'</span>]</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>ESR1</code>è›‹ç™½:</p>
<blockquote>
<p>Protein       : Estrogen receptor  é›Œæ¿€ç´ å—ä½“</p>
<p>Gene          : ESR1</p>
<p>Organism  : <em>Homo sapiens (Human)</em></p>
</blockquote>
<ul>
<li>
<p>è›‹ç™½è¯¦æƒ…<a target="_blank" rel="noopener" href="https://www.uniprot.org/uniprot/P03372">ESR1 - Estrogen receptor - Homo sapiens (Human) - ESR1 gene &amp; protein (uniprot.org)</a></p>
</li>
<li>
<p><code>ESR1</code>è›‹ç™½åºåˆ—: <a target="_blank" rel="noopener" href="https://www.uniprot.org/uniprot/P03372.fasta">https://www.uniprot.org/uniprot/P03372.fasta</a></p>
<blockquote>
<p>&gt;sp|P03372|ESR1_HUMAN Estrogen receptor OS=Homo sapiens OX=9606 GN=ESR1 PE=1 SV=2</p>
<p>MTMTLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY<br>
EFNAAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF<br>
LQPHGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK<br>
ETRYCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC<br>
RLRKCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR<br>
SKKNSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW<br>
AKRVPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG<br>
MVEIFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD<br>
KITDTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL<br>
LEMLDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># è¯»å–æ•´ä¸ªFASTAè®°å½•, æ‰¾åˆ°é‡Œé¢æ‰€å±ä¸º'Homo sapiens'çš„åºåˆ—</span></span><br><span class="line"></span><br><span class="line">fasta_file = <span class="built_in">open</span>(<span class="string">'SwissProt.fasta'</span>,<span class="string">'r'</span>)  <span class="comment"># è¯»å–å¤šåºåˆ—Fastaæ–‡ä»¶</span></span><br><span class="line">out_file = <span class="built_in">open</span>(<span class="string">'SwissProtHuman.fasta'</span>,<span class="string">'w'</span>) <span class="comment"># è¾“å‡ºæ–‡ä»¶å‡†å¤‡</span></span><br><span class="line"></span><br><span class="line">seq = <span class="string">''</span>  <span class="comment"># åºåˆ—å­˜è´®å˜é‡</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> fasta_file:  <span class="comment">#æŒ‰è¡Œè¿­ä»£</span></span><br><span class="line">    <span class="keyword">if</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span> <span class="keyword">and</span> seq == <span class="string">''</span>:  <span class="comment"># åˆ¤æ–­æ˜¯ä¸æ˜¯å½“å‰åœ¨è¯»è®°å½•çš„ç¬¬ä¸€è¡Œ</span></span><br><span class="line">        header = line</span><br><span class="line">    <span class="keyword">elif</span> line[<span class="number">0</span>] != <span class="string">'&gt;'</span>:  <span class="comment"># åˆ¤æ–­ä¸æ˜¯é¦–è¡Œ(æ²¡æœ‰'&gt;'), ç›´æ¥å°†æ­¤è¡Œæ”¾å…¥åºåˆ—</span></span><br><span class="line">        seq = seq + line</span><br><span class="line">    <span class="keyword">elif</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span> <span class="keyword">and</span> seq != <span class="string">''</span>: <span class="comment"># å¦‚æœæ˜¯é¦–è¡Œä¸”åºåˆ—æ˜¯ç©ºçš„è¯, è¯´æ˜ä¸Šä¸€æ¡çš„è®°å½•è¯»å®Œäº†</span></span><br><span class="line">        <span class="keyword">if</span> <span class="string">"Homo sapiens"</span> <span class="keyword">in</span> header:   <span class="comment"># è¿™é‡Œå…ˆåˆ¤æ–­ä¸Šä¸€æ¡è®°å½•çš„è®°å½•æ˜¯ä¸æ˜¯æˆ‘ä»¬è¦çš„,æ˜¯å°±å­˜èµ·æ¥</span></span><br><span class="line">            out_file.write(header + seq)</span><br><span class="line">        seq = <span class="string">''</span>                       <span class="comment"># æ¸…ç©º</span></span><br><span class="line">        header = line                  <span class="comment"># ä¸Šä¸€æ¡çºªå½•æå®Œå•¦, æŠŠç°åœ¨åœ¨è¯»çš„è¿™ä¸€è¡Œä¿å­˜è‡³haeder</span></span><br><span class="line">        </span><br><span class="line"><span class="keyword">if</span> <span class="string">"Homo sapiens"</span> <span class="keyword">in</span> header:           <span class="comment"># å…¨éƒ¨è¯»å®Œäº†, ä½†æ˜¯æœ€åä¸€æ¡è®°å½•è¿˜æ²¡å¤„ç†, è¿™é‡Œå¤„ç†æ‰</span></span><br><span class="line">    out_file.write(header + seq)</span><br><span class="line">out_file.close()</span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#di-5-zhang-sou-suo-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-5-zhang-sou-suo-shu-ju">ç¬¬5ç«  æœç´¢æ•°æ®</span></h3>
<h4><a href="#5-2-an-li-jiang-rna-xu-lie-fan-yi-wei-xiang-ying-de-dan-bai-zhi-xu-lie" class="header-anchor"> âš“ï¸</a><span id="5-2-an-li-jiang-rna-xu-lie-fan-yi-wei-xiang-ying-de-dan-bai-zhi-xu-lie">5.2æ¡ˆä¾‹ï¼š å°†RNAåºåˆ—ç¿»è¯‘ä¸ºç›¸åº”çš„è›‹ç™½è´¨åºåˆ—</span></h4>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/nuccore/A06662.1?from=470&amp;to=577&amp;report=fasta">Synthetic nucleotide sequence of the human GSH transferase pi gene - Nucleotide - NCBI (nih.gov)</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">codon_table={</span><br><span class="line"><span class="string">'GCU'</span>:<span class="string">'A'</span>,<span class="string">'GCC'</span>:<span class="string">'A'</span>,<span class="string">'GCA'</span>:<span class="string">'A'</span>,<span class="string">'GCG'</span>:<span class="string">'A'</span>,<span class="string">'CGU'</span>:<span class="string">'R'</span>,</span><br><span class="line"><span class="string">'CGC'</span>:<span class="string">'R'</span>,<span class="string">'CGA'</span>:<span class="string">'R'</span>,<span class="string">'CCS'</span>:<span class="string">'R'</span>,<span class="string">'AGA'</span>:<span class="string">'R'</span>,<span class="string">'AGG'</span>:<span class="string">'R'</span>,</span><br><span class="line"><span class="string">'UCU'</span>:<span class="string">'S'</span>,<span class="string">'UCC'</span>:<span class="string">'S'</span>,<span class="string">'UCA'</span>:<span class="string">'S'</span>,<span class="string">'UCG'</span>:<span class="string">'S'</span>,<span class="string">'AGU'</span>:<span class="string">'S'</span>,</span><br><span class="line"><span class="string">'AGC'</span>:<span class="string">'S'</span>,<span class="string">'AUU'</span>:<span class="string">'T'</span>,<span class="string">'AUC'</span>:<span class="string">'I'</span>,<span class="string">'AUA'</span>:<span class="string">'T'</span>,<span class="string">'AUU'</span>:<span class="string">'I'</span>,</span><br><span class="line"><span class="string">'AUC'</span>:<span class="string">'I'</span>,<span class="string">'AUA'</span>:<span class="string">'I'</span>,<span class="string">'UUA'</span>:<span class="string">'L'</span>,<span class="string">'UUG'</span>:<span class="string">'L'</span>,<span class="string">'CUU'</span>:<span class="string">'L'</span>,</span><br><span class="line"><span class="string">'CUC'</span>:<span class="string">'L'</span>,<span class="string">'CUA'</span>:<span class="string">'L'</span>,<span class="string">'CUG'</span>:<span class="string">'L'</span>,<span class="string">'GGU'</span>:<span class="string">'G'</span>,<span class="string">'GGC'</span>:<span class="string">'G'</span>,</span><br><span class="line"><span class="string">'GGA'</span>:<span class="string">'G'</span>,<span class="string">'GSG'</span>:<span class="string">'G'</span>,<span class="string">'GUU'</span>:<span class="string">'V'</span>,<span class="string">'GUC'</span>:<span class="string">'V'</span>,<span class="string">'GUA'</span>:<span class="string">'V'</span>,</span><br><span class="line"><span class="string">'GUG'</span>:<span class="string">'V'</span>,<span class="string">'ACU'</span>:<span class="string">'T'</span>,<span class="string">'ACC'</span>:<span class="string">'T'</span>,<span class="string">'ACA'</span>:<span class="string">'T'</span>,<span class="string">'ACG'</span>:<span class="string">'T'</span>,</span><br><span class="line"><span class="string">'CCU'</span>:<span class="string">'P'</span>,<span class="string">'CCC'</span>:<span class="string">'P'</span>,<span class="string">'CCA'</span>:<span class="string">'P'</span>,<span class="string">'CCG'</span>:<span class="string">'P'</span>,<span class="string">'AAU'</span>:<span class="string">'N'</span>,</span><br><span class="line"><span class="string">'AAC'</span>:<span class="string">'N'</span>,<span class="string">'GAU'</span>:<span class="string">'D'</span>,<span class="string">'GAC'</span>:<span class="string">'D'</span>,<span class="string">'UGU'</span>:<span class="string">'C'</span>,<span class="string">'UGC'</span>:<span class="string">'C'</span>,</span><br><span class="line"><span class="string">'CAA'</span>:<span class="string">'Q'</span>,<span class="string">'CAG'</span>:<span class="string">'Q'</span>,<span class="string">'GAA'</span>:<span class="string">'E'</span>,<span class="string">'GAG'</span>:<span class="string">'E'</span>,<span class="string">'CAU'</span>:<span class="string">'H'</span>,</span><br><span class="line"><span class="string">'CAC'</span>:<span class="string">'H'</span>,<span class="string">'AAA'</span>:<span class="string">'K'</span>,<span class="string">'AAG'</span>:<span class="string">'K'</span>,<span class="string">'UUU'</span>:<span class="string">'F'</span>,<span class="string">'UUC'</span>:<span class="string">'F'</span>,</span><br><span class="line"><span class="string">'UAU'</span>:<span class="string">'Y'</span>,<span class="string">'UAC'</span>:<span class="string">'Y'</span>,<span class="string">'AUG'</span>:<span class="string">'M'</span>,<span class="string">'UGG'</span>:<span class="string">'W'</span>,</span><br><span class="line"><span class="string">'UAG'</span>:<span class="string">'STOP'</span>,<span class="string">'UGA'</span>:<span class="string">'STOP'</span>,<span class="string">'UAA'</span>:<span class="string">'STOP'</span>}</span><br><span class="line"></span><br><span class="line">rna = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">'A06662-RNA.fasta'</span>):</span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> line.startswith(<span class="string">'&gt;'</span>):</span><br><span class="line">        rna = rna +line.strip()</span><br><span class="line"><span class="comment"># ä¸€æ¬¡ç¿»è¯‘è¿›è¡Œä¸€æ¬¡é˜…è¯»æ¡†è¯»å–</span></span><br><span class="line"><span class="keyword">for</span> frame <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">3</span>):</span><br><span class="line">    prot = <span class="string">''</span></span><br><span class="line">    <span class="built_in">print</span> <span class="string">'Reading frame '</span> + <span class="built_in">str</span>(frame + <span class="number">1</span>)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(frame, <span class="built_in">len</span>(rna), <span class="number">3</span>):</span><br><span class="line">        codon = rna[i:i+<span class="number">3</span>]</span><br><span class="line">        <span class="keyword">if</span> codon <span class="keyword">in</span> codon_table:</span><br><span class="line">            <span class="keyword">if</span> codon_table[codon] == <span class="string">'STOP'</span>:</span><br><span class="line">                prot= prot + <span class="string">'*'</span></span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                prot = prot + codon_table[codon]</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># å¦‚æœè¦è¯†åˆ«çš„çŸ­åºåˆ—ä¸åœ¨tableé‡Œ, åˆ™ç”¨'-'æ›¿ä»£</span></span><br><span class="line">            prot = prot + <span class="string">'-'</span></span><br><span class="line">        </span><br><span class="line">    <span class="comment"># æ ¼å¼åŒ–è¾“å‡ºæˆ48åˆ—çš„å—</span></span><br><span class="line">    i = <span class="number">0</span></span><br><span class="line">    <span class="keyword">while</span> i &lt; <span class="built_in">len</span>(prot):</span><br><span class="line">        <span class="built_in">print</span> prot[i:i + <span class="number">48</span>]</span><br><span class="line">        i = i + <span class="number">48</span></span><br><span class="line">        </span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>str.startswith</code>(<em>prefix</em>[, <em>start</em>[, <em>end</em>]]) <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=startswi#str.startswith">5. å†…ç½®ç±»å‹ â€” Python 2.7.18 æ–‡æ¡£</a></p>
<p>å¦‚æœå­—ç¬¦ä¸²ä»¥æŒ‡å®šçš„ <em>prefix</em> å¼€å§‹åˆ™è¿”å› <code>True</code>ï¼Œå¦åˆ™è¿”å› <code>False</code>ã€‚ <em>prefix</em> ä¹Ÿå¯ä»¥ä¸ºç”±å¤šä¸ªä¾›æŸ¥æ‰¾çš„å‰ç¼€æ„æˆçš„å…ƒç»„ã€‚ å¦‚æœæœ‰å¯é€‰é¡¹ <em>start</em>ï¼Œå°†ä»æ‰€æŒ‡å®šä½ç½®å¼€å§‹æ£€æŸ¥ã€‚ å¦‚æœæœ‰å¯é€‰é¡¹ <em>end</em>ï¼Œå°†åœ¨æ‰€æŒ‡å®šä½ç½®åœæ­¢æ¯”è¾ƒã€‚</p>
</li>
<li>
<p>æ ¸é…¸åºåˆ—<code>(å¼€æ”¾)é˜…è¯»æ¡†(Open Reading Frame)</code>:</p>
<ul>
<li>
<p>ç”±äºæ ¸ç³–ä½“ç¿»è¯‘mRNAæ—¶, <strong>ä¸‰ä¸ªç¢±åŸºç¿»è¯‘ä¸ºä¸€ä¸ªæ°¨åŸºé…¸</strong>, é‚£ä¹ˆåŒä¸€ä¸ªåºåˆ—ç†è®ºä¸Šå­˜åœ¨ä¸‰ä¸ªä¸åŒçš„é˜…è¯»æ¡†</p>
</li>
<li>
<p>å¯¹åº”äº<code>ç§»ç çªå˜</code></p>
<blockquote>
<p>å¯¹äº åºåˆ— <code>ATCGATCG</code> æœ‰ <code>ATC GAT CG</code>, <code>A TCG ATC G</code>, <code>AT CGA TCG</code> ä¸‰ç§æ–¹æ³•è§£è¯»</p>
</blockquote>
</li>
<li>
<p><code>ORF</code>æŒ‡çš„æ˜¯DNAä¸Šçš„åºåˆ—ï¼Œä»5â€™ç«¯ç¿»è¯‘èµ·å§‹å¯†ç å­ATGåˆ°ç»ˆæ­¢å¯†ç å­(TAA,TAG,TGA)çš„è›‹ç™½è´¨ç¼–ç åºåˆ—ã€‚å¯¹äºä»»æ„ç»™å®šçš„ä¸€æ®µDNAï¼Œæœ‰ä¸¤ä¸ªé—®é¢˜éœ€è¦è€ƒè™‘ï¼Œ</p>
<ul>
<li>
<p>ä¸€æ˜¯DNAåŒé“¾ä¸­çš„å“ªæ¡æ˜¯ç¼–ç é“¾</p>
</li>
<li>
<p>äºŒæ˜¯ç¼–ç åŒºç©¶ç«Ÿä»ç¬¬ä¸€ä¸ªç¢±åŸºå¼€å§‹è¿›è¡Œç¼–ç <br>
æ‰€ä»¥æ¯æ¡é“¾éƒ½æœ‰æ½œåœ¨çš„3ç§ORFï¼Œè€Œå¯¹äºåŒé“¾DNAæ¥è¯´å°±æœ‰6ç§å¯èƒ½çš„ORFã€‚ä¹Ÿå°±æ˜¯è¯´å…ˆä»ç»™å®šçš„DNAå•é“¾ä¸ºæ¨¡ç‰ˆï¼Œåˆ†åˆ«ä»5â€™-3â€™æ–¹å‘ç¬¬123ä¸ªç¢±åŸºå¼€å§‹ç¿»è¯‘ï¼Œå†ä»¥äº’è¡¥é“¾ä¸ºæ¨¡ç‰ˆï¼Œåˆ†åˆ«ä»3â€™-5â€™æ–¹å‘çš„ç¬¬123ä¸ªç¢±åŸºå¼€å§‹ç¿»è¯‘ï¼Œå¾—åˆ°å¦å¤–3ç§ç¿»è¯‘ç»“æœã€‚æ­£é“¾ä¸Šçš„æ­£å‘è¯»ç æ¡†forwardï¼Œè´Ÿé“¾çš„ä¸ºåå‘è¯»ç æ¡†reverseã€‚<strong>6ä¸ªæ½œåœ¨ORFä¸­ï¼Œä¸€èˆ¬é€‰æ‹©ä¸­é—´æ²¡æœ‰è¢«ç»ˆæ­¢å¯†ç å­éš”å¼€çš„æœ€å¤§çš„è¯»ç æ¡†ä¸ºæ­£ç¡®çš„ç»“æœ</strong>ã€‚</p>
<p>@<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/5789365faa7e">4ï¸âƒ£ æ ¸é…¸åºåˆ—ç‰¹å¾åˆ†æ(1):å¼€æ”¾é˜…è¯»æ¡†è¯†åˆ« - ç®€ä¹¦ (jianshu.com)</a></p>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>æ–‡ä»¶è¯»å–çš„<code>.next()</code>æ–¹æ³•: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next">file.next  2.7æ–‡æ¡£</a></p>
<blockquote>
<p><code>file.next</code>()</p>
<p>A file object is its own iterator, for example <code>iter(f)</code> returns <em>f</em> (unless <em>f</em> is closed). When a file is used as an iterator, typically in a <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#for"><code>for</code></a> loop (for example, <code>for line in f: print line.strip()</code>), the <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> method is called repeatedly. This method returns the next input line, or raises <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/exceptions.html#exceptions.StopIteration"><code>StopIteration</code></a> when EOF is hit when the file is open for reading (behavior is undefined when the file is open for writing). In order to make a <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#for"><code>for</code></a> loop the most efficient way of looping over the lines of a file (a very common operation), the <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> method uses a hidden read-ahead buffer. As a consequence of using a read-ahead buffer, combining <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> with other file methods (like <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.readline"><code>readline()</code></a>) does not work right. However, using <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.seek"><code>seek()</code></a> to reposition the file to an absolute position will flush the read-ahead buffer.</p>
</blockquote>
</li>
<li>
<p><code>for</code>å¾ªç¯ä¸­çš„ <code>else</code> :   <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=else#break-and-continue-statements-and-else-clauses-on-loops">break-and-continue-statements-and-else-clauses-on-loops</a></p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> n <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">2</span>, <span class="number">10</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">for</span> x <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">2</span>, n):</span><br><span class="line"><span class="meta">... </span>        <span class="keyword">if</span> n % x == <span class="number">0</span>:</span><br><span class="line"><span class="meta">... </span>            <span class="built_in">print</span> n, <span class="string">'equals'</span>, x, <span class="string">'*'</span>, n/x</span><br><span class="line"><span class="meta">... </span>            <span class="keyword">break</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">else</span>:   <span class="comment"># æ­¤elseä¸ç¬¬äºŒä¸ªforå¯¹åº”, å½“å‘ç”Ÿäº† break å, æ‰§è¡Œè¿™é‡Œ</span></span><br><span class="line"><span class="meta">... </span>        <span class="comment"># loop fell through without finding a factor</span></span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> n, <span class="string">'is a prime number'</span></span><br><span class="line">...</span><br><span class="line"><span class="number">2</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">3</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">4</span> equals <span class="number">2</span> * <span class="number">2</span></span><br><span class="line"><span class="number">5</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">6</span> equals <span class="number">2</span> * <span class="number">3</span></span><br><span class="line"><span class="number">7</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">8</span> equals <span class="number">2</span> * <span class="number">4</span></span><br><span class="line"><span class="number">9</span> equals <span class="number">3</span> * <span class="number">3</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>é¢„æµ‹è›‹ç™½è´¨åºåˆ—ä¸­çš„æ— åº(ç¯)åŒºåŸŸ: æ—¢ä¸æ˜¯<code>Î±èºæ—‹</code>ä¹Ÿä¸æ˜¯<code>Î²æŠ˜å </code>çš„åŒºåŸŸ, å°±æ˜¯æ— åºçš„.
<ul>
<li><code>æ ¸å¿ƒæ€æƒ³</code>: æ¯ä¸ªæ°¨åŸºé…¸éƒ½æœ‰ä¸€ä¸ªç‰¹å®šçš„äºŒçº§ç»“æ„å…ƒä»¶å€¾å‘, é€šè¿‡å¤§é‡å·²çŸ¥çš„<code>è›‹ç™½è´¨ç»“æ„(PDB)</code>çš„äºŒçº§ç»“æ„å…ƒä»¶ä¸­å„ç±»æ°¨åŸºé…¸ç±»å‹å‡ºç°çš„<code>é¢‘ç‡f</code>,æ¥ä¼°è®¡å€¾å‘.</li>
<li>ä½¿ç”¨<code>1-f</code>è®¡ç®—: <code>f</code>æ˜¯ä»¥æœ‰äºŒçº§ç»“æ„å…ƒä»¶çš„éƒ¨åˆ†ä¸ºåŸºæ•°çš„(å°±æ˜¯æœ‰åºçš„), é‚£ä¹ˆåœ¨è¿™æ ·çš„åŸºæ•°é‡ŒæŸç§æ°¨åŸºé…¸å‡ºç°çš„è¶Šå¤š(é¢‘æ•°æ¥è¿‘1)åˆ™å€¾å‘äºæœ‰åºçš„å¯èƒ½æ€§è¶Šå¤§.</li>
<li>å¯¹20ä¸ªæ°¨åŸºé…¸å€¾å‘çš„å€¼è¿›è¡Œæ ‡å‡†åŒ–å’Œå¼•å…¥è´Ÿå€¼, è®¾å®šä¸€ä¸ªæ˜¯å¦ä¸ºæ— åºçš„<code>é˜ˆå€¼</code>, å¤§äºé˜ˆå€¼ä¸º<code>æ— åº(ç¯)</code>, å°äºä¸º<code>æœ‰åº(äºŒçº§ç»“æ„)</code>.</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">propensities ={</span><br><span class="line"><span class="string">'N'</span>: <span class="number">0.2299</span>, <span class="string">'P'</span>:  <span class="number">0.5523</span>, <span class="string">'Q'</span>:-<span class="number">0.18770</span>, <span class="string">'A'</span>:-<span class="number">0.2615</span>,</span><br><span class="line"><span class="string">'R'</span>:-<span class="number">0.1766</span>, <span class="string">'S'</span>:  <span class="number">0.1429</span>, <span class="string">'C'</span>:-<span class="number">0.01515</span>, <span class="string">'T'</span>: <span class="number">0.0089</span>,</span><br><span class="line"><span class="string">'D'</span>: <span class="number">0.2276</span>, <span class="string">'E'</span>: -<span class="number">0.2047</span>, <span class="string">'V'</span>:-<span class="number">0.38620</span>, <span class="string">'F'</span>:-<span class="number">0.2256</span>,</span><br><span class="line"><span class="string">'W'</span>:-<span class="number">0.2434</span>, <span class="string">'G'</span>:  <span class="number">0.4332</span>, <span class="string">'H'</span>:-<span class="number">0.00120</span>, <span class="string">'Y'</span>:-<span class="number">0.2075</span>,</span><br><span class="line"><span class="string">'I'</span>:-<span class="number">0.4222</span>, <span class="string">'K'</span>: -<span class="number">0.1001</span>, <span class="string">'L'</span>: <span class="number">0.33793</span>, <span class="string">'M'</span>:-<span class="number">0.2259</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://www.rcsb.org/">RCSB PDB: </a></p>
<ul>
<li></li>
<li>
<p>æ•°æ®ç¤ºä¾‹æ¦‚è§ˆ: <a target="_blank" rel="noopener" href="https://www.rcsb.org/structure/6XM5">RCSB PDB - 6XM5: Structure of SARS-CoV-2 spike at pH 5.5, all RBDs down</a></p>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br></pre></td><td class="code"><pre><span class="line">HEADER    VIRAL PROTEIN                           <span class="number">29</span>-JUN-<span class="number">20</span>   <span class="number">6</span>XM5              </span><br><span class="line"><span class="built_in">TITLE</span>     STRUCTURE OF SARS-COV-<span class="number">2</span> SPIKE <span class="built_in">AT</span> PH <span class="number">5</span>.<span class="number">5</span>, ALL RBDS DOWN                </span><br><span class="line">COMPND    MOL_ID: <span class="number">1</span>;                                                            </span><br><span class="line">COMPND   <span class="number">2</span> MOLECULE: SPIKE GLYCOPROTEIN;                                        </span><br><span class="line">...                    </span><br><span class="line">COMPND   <span class="number">6</span> MUTATION: YES                                                        </span><br><span class="line">SOURCE    MOL_ID: <span class="number">1</span>;                                                            </span><br><span class="line">SOURCE   <span class="number">2</span> ORGANISM_SCIENTIFIC: SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS   </span><br><span class="line">...                          </span><br><span class="line">SOURCE   <span class="number">9</span> EXPRESSION_SYSTEM_CELL_LINE: FREESTYLE <span class="number">293</span>F                          </span><br><span class="line">KEYWDS    SARS-COV-<span class="number">2</span> SPIKE, COVID19, VIRAL PROTEIN                              </span><br><span class="line">EXPDTA    ELECTRON MICROSCOPY                                                   </span><br><span class="line">AUTHOR    T.ZHOU,Y.TSYBOVSKY,A.OLIA,P.D.KWONG                                   </span><br><span class="line">REVDAT   <span class="number">2</span>   <span class="number">12</span>-AUG-<span class="number">20</span> <span class="number">6</span>XM5    <span class="number">1</span>       JRNL   HETSYN                            </span><br><span class="line">REVDAT   <span class="number">1</span>   <span class="number">29</span>-JUL-<span class="number">20</span> <span class="number">6</span>XM5    <span class="number">0</span>                                                </span><br><span class="line">JRNL        AUTH   T.ZHOU,Y.TSYBOVSKY,A.S.OLIA,J.GORMAN,M.A.RAPP,G.CERUTTI,     </span><br><span class="line">...  </span><br><span class="line">JRNL        TITL   CRYO-EM STRUCTURES DELINEATE A PH-DEPENDENT SWITCH THAT         ...</span><br><span class="line">JRNL        PMID   <span class="number">32637958</span>                                                     </span><br><span class="line">JRNL        DOI    <span class="number">10</span>.<span class="number">1101</span>/<span class="number">2020</span>.<span class="number">07</span>.<span class="number">04</span>.<span class="number">187989</span>                                    </span><br><span class="line">REMARK   <span class="number">2</span>                                                                      </span><br><span class="line">...       </span><br><span class="line">REMARK   <span class="number">3</span>   SOFTWARE PACKAGES      : LATITUDE, CTFFIND, UCSF CHIMERA, COOT,    </span><br><span class="line">REMARK   <span class="number">3</span>                            PHENIX, RELION, RELION, RELION, RELION    </span><br><span class="line">...         </span><br><span class="line">REMARK   <span class="number">3</span>   OVERALL ANISOTROPIC B VALUE  : NULL                                </span><br><span class="line">...</span><br><span class="line">REMARK <span class="number">900</span> STRUCTURE OF SARS-COV-<span class="number">2</span> SPIKE <span class="built_in">AT</span> PH <span class="number">5</span>.<span class="number">5</span>, ALL RBDS DOWN               </span><br><span class="line">DBREF  <span class="number">6</span>XM5 A   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">DBREF  <span class="number">6</span>XM5 B   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">DBREF  <span class="number">6</span>XM5 C   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">SEQADV <span class="number">6</span>XM5 GLY A  <span class="number">682</span>  UNP  P0DTC2    ARG   <span class="number">682</span> ENGINEERED MUTATION            </span><br><span class="line">SEQADV <span class="number">6</span>XM5 SER A  <span class="number">683</span>  UNP  P0DTC2    ARG   <span class="number">683</span> ENGINEERED MUTATION    </span><br><span class="line">...</span><br><span class="line">SEQADV <span class="number">6</span>XM5 LYS C <span class="number">1288</span>  UNP  P0DTC2              EXPRESSION TAG                 </span><br><span class="line">SEQRES   <span class="number">1</span> A <span class="number">1275</span>  GLN CYS VAL ASN LEU THR THR ARG THR GLN LEU PRO PRO          </span><br><span class="line">SEQRES   <span class="number">2</span> A <span class="number">1275</span>  ALA TYR THR ASN SER PHE THR ARG GLY VAL TYR TYR PRO       </span><br><span class="line">...</span><br><span class="line">SEQRES  <span class="number">99</span> C <span class="number">1275</span>  LYS                                                          </span><br><span class="line">HET    NAG  D   <span class="number">1</span>      <span class="number">14</span>                                                       </span><br><span class="line">HET    NAG  D   <span class="number">2</span>      <span class="number">14</span>                                                    </span><br><span class="line">...</span><br><span class="line">HETNAM     NAG <span class="number">2</span>-ACETAMIDO-<span class="number">2</span>-DEOXY-BETA-D-GLUCOPYRANOSE                         </span><br><span class="line">...</span><br><span class="line">FORMUL   <span class="number">4</span>  NAG    <span class="number">58</span>(C8 H15 N O6)                                              </span><br><span class="line">HELIX    <span class="number">1</span> AA1 ASP A  <span class="number">294</span>  LYS A  <span class="number">304</span>  <span class="number">1</span>                                  <span class="number">11</span>    </span><br><span class="line">...      </span><br><span class="line">SHEET    <span class="number">3</span> AA1 <span class="number">8</span> TYR A <span class="number">265</span>  TYR A <span class="number">269</span> -<span class="number">1</span>  O  TYR A <span class="number">265</span>   N  PHE A  <span class="number">65</span>           </span><br><span class="line">...</span><br><span class="line">LINK         O4  NAG S   <span class="number">1</span>                 C1  NAG S   <span class="number">2</span>     <span class="number">1555</span>   <span class="number">1555</span>  <span class="number">1</span>.<span class="number">44</span>  </span><br><span class="line">CRYST1    <span class="number">1</span>.<span class="number">000</span>    <span class="number">1</span>.<span class="number">000</span>    <span class="number">1</span>.<span class="number">000</span>  <span class="number">90</span>.<span class="number">00</span>  <span class="number">90</span>.<span class="number">00</span>  <span class="number">90</span>.<span class="number">00</span> P <span class="number">1</span>                      </span><br><span class="line">ORIGX1      <span class="number">1</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>        <span class="number">0</span>.<span class="number">00000</span>                         </span><br><span class="line">...</span><br><span class="line">SCALE3      <span class="number">0</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>  <span class="number">1</span>.<span class="number">000000</span>        <span class="number">0</span>.<span class="number">00000</span>                         </span><br><span class="line">ATOM      <span class="number">1</span>  N   ALA A  <span class="number">27</span>     <span class="number">188</span>.<span class="number">584</span> <span class="number">130</span>.<span class="number">357</span> <span class="number">168</span>.<span class="number">091</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           N  </span><br><span class="line">ATOM      <span class="number">2</span>  CA  ALA A  <span class="number">27</span>     <span class="number">189</span>.<span class="number">571</span> <span class="number">130</span>.<span class="number">261</span> <span class="number">169</span>.<span class="number">159</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           C  </span><br><span class="line">ATOM      <span class="number">3</span>  C   ALA A  <span class="number">27</span>     <span class="number">190</span>.<span class="number">053</span> <span class="number">131</span>.<span class="number">641</span> <span class="number">169</span>.<span class="number">567</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           C  </span><br><span class="line">ATOM   <span class="number">8249</span>  OG  SER A1147     <span class="number">191</span>.<span class="number">132</span> <span class="number">184</span>.<span class="number">224</span> <span class="number">284</span>.<span class="number">141</span>  <span class="number">1</span>.<span class="number">00152</span>.<span class="number">29</span>           O  </span><br><span class="line">TER    <span class="number">8250</span>      SER A1147                                                      </span><br><span class="line">ATOM   <span class="number">8251</span>  N   ALA B  <span class="number">27</span>     <span class="number">132</span>.<span class="number">579</span> <span class="number">208</span>.<span class="number">265</span> <span class="number">174</span>.<span class="number">154</span>  <span class="number">1</span>.<span class="number">00140</span>.<span class="number">96</span>           N  </span><br><span class="line">...</span><br><span class="line">TER   <span class="number">24584</span>      SER C1147                                                      </span><br><span class="line">HETATM24585  C1  NAG D   <span class="number">1</span>     <span class="number">184</span>.<span class="number">739</span> <span class="number">152</span>.<span class="number">280</span> <span class="number">150</span>.<span class="number">742</span>  <span class="number">1</span>.<span class="number">00114</span>.<span class="number">03</span>           C  </span><br><span class="line">HETATM24586  C2  NAG D   <span class="number">1</span>     <span class="number">183</span>.<span class="number">632</span> <span class="number">153</span>.<span class="number">136</span> <span class="number">150</span>.<span class="number">126</span>  <span class="number">1</span>.<span class="number">00114</span>.<span class="number">03</span>           C  </span><br><span class="line">...</span><br><span class="line">CONECT  <span class="number">29225033</span>                                                                </span><br><span class="line">CONECT  <span class="number">70225047</span>                                                                </span><br><span class="line">...</span><br><span class="line">CONECT2539625389                                                                </span><br><span class="line">MASTER      <span class="number">907</span>    <span class="number">0</span>   <span class="number">58</span>   <span class="number">69</span>  <span class="number">173</span>    <span class="number">0</span>    <span class="number">0</span>    <span class="number">625393</span>    <span class="number">3</span>  <span class="number">936</span>  <span class="number">297</span>          </span><br><span class="line">END                                                                             </span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
</ul>
<h4><a href="#5-5-zi-ce-ti" class="header-anchor"> âš“ï¸</a><span id="5-5-zi-ce-ti">5.5 è‡ªæµ‹é¢˜</span></h4>
<ul>
<li>
<p><code>NCBI</code>ä¸‹è½½<code>FASTAæ ¼å¼</code>RNAåºåˆ—:</p>
</li>
<li>
<p><code>PubMed</code>æ–‡çŒ®æ£€ç´¢:</p>
</li>
<li>
<p>äºŒçº§ç»“æ„å…ƒç´ é¢„æµ‹:</p>
<ul>
<li>æ°¨åŸºé…¸äºŒçº§ç»“æ„åå¥½è¡¨: <a target="_blank" rel="noopener" href="http://www.bmrb.wisc.edu/referenc/choufas.html">Amino Acid Secondary Structure Preferences</a></li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center"><strong>A.A.</strong></th>
<th style="text-align:center"><strong>Helix</strong></th>
<th style="text-align:center"><strong>Helix</strong></th>
<th style="text-align:center"><strong>Sheet</strong></th>
<th style="text-align:center"><strong>Sheet</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center"><strong>A.A.</strong></td>
<td style="text-align:center"><strong>Designation</strong></td>
<td style="text-align:center"><strong>P</strong></td>
<td style="text-align:center"><strong>Designation</strong></td>
<td style="text-align:center"><strong>P</strong></td>
</tr>
<tr>
<td style="text-align:center"><strong>A</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.45</td>
<td style="text-align:center">I</td>
<td style="text-align:center">0.97</td>
</tr>
<tr>
<td style="text-align:center"><strong>C</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.77</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.30</td>
</tr>
<tr>
<td style="text-align:center"><strong>D</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.98</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.80</td>
</tr>
<tr>
<td style="text-align:center"><strong>E</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.53</td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.26</td>
</tr>
<tr>
<td style="text-align:center"><strong>F</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.12</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.28</td>
</tr>
<tr>
<td style="text-align:center"><strong>G</strong></td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.53</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.81</td>
</tr>
<tr>
<td style="text-align:center"><strong>H</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.24</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.71</td>
</tr>
<tr>
<td style="text-align:center"><strong>I</strong></td>
<td style="text-align:center">I</td>
<td style="text-align:center">1.00</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.60</td>
</tr>
<tr>
<td style="text-align:center"><strong>K</strong></td>
<td style="text-align:center">I</td>
<td style="text-align:center">1.07</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.74</td>
</tr>
<tr>
<td style="text-align:center"><strong>L</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.34</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.22</td>
</tr>
<tr>
<td style="text-align:center"><strong>M</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.20</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.67</td>
</tr>
<tr>
<td style="text-align:center"><strong>N</strong></td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.73</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.65</td>
</tr>
<tr>
<td style="text-align:center"><strong>P</strong></td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.59</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.62</td>
</tr>
<tr>
<td style="text-align:center"><strong>Q</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.17</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.23</td>
</tr>
<tr>
<td style="text-align:center"><strong>R</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.79</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.90</td>
</tr>
<tr>
<td style="text-align:center"><strong>S</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.79</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.72</td>
</tr>
<tr>
<td style="text-align:center"><strong>T</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.82</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.20</td>
</tr>
<tr>
<td style="text-align:center"><strong>V</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.14</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.65</td>
</tr>
<tr>
<td style="text-align:center"><strong>W</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.14</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.19</td>
</tr>
<tr>
<td style="text-align:center"><strong>Y</strong></td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.61</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.29</td>
</tr>
</tbody>
</table>
<ul>
<li>
<p><code>Designations</code>:</p>
<ul>
<li>H = Strong Former, h = Former, I = Weak Former, i = Indifferent, B = Strong Breaker, b = Breaker;</li>
</ul>
</li>
<li>
<p><code>P</code> : Conformational Parameter</p>
</li>
</ul>
<h3><a href="#di-6-zhang-guo-lu-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-6-zhang-guo-lu-shu-ju">ç¬¬6ç«  è¿‡æ»¤æ•°æ®</span></h3>
<h4><a href="#6-2-an-li-shi-yong-rnaseq-shu-chu-shu-ju" class="header-anchor"> âš“ï¸</a><span id="6-2-an-li-shi-yong-rnaseq-shu-chu-shu-ju">6.2 æ¡ˆä¾‹ï¼š ä½¿ç”¨RNAseqè¾“å‡ºæ•°æ®</span></h4>
<ul>
<li>
<p><strong>æ–°ä¸€ä»£æµ‹åº(Next-generation sequencingï¼ŒNGS)</strong>:</p>
<blockquote>
<p>æœ€æ—©å¹¿æ³›åº”ç”¨æµ‹åºæŠ€æœ¯ä¸º 70 å¹´ä»£çš„ <strong>Sanger</strong> æ³•ï¼Œè¿™ä¹Ÿæ˜¯å®Œæˆäººç±»åŸºå› ç»„è®¡åˆ’çš„åŸºç¡€ï¼Œå› å…¶æµ‹åºé€šé‡ä½ã€è´¹æ—¶è´¹åŠ›ï¼Œç§‘å­¦å®¶ä»¬ä¸€ç›´åœ¨å¯»æ±‚é€šé‡æ›´é«˜ã€é€Ÿåº¦æ›´å¿«ã€ä»·æ ¼æ›´ä¾¿å®œã€è‡ªåŠ¨åŒ–ç¨‹åº¦æ›´é«˜çš„æµ‹åºæŠ€æœ¯ã€‚è‡ª 2005 å¹´ä»¥æ¥ï¼Œä»¥Roche å…¬å¸çš„ <strong>454 æŠ€æœ¯</strong>ã€Illumina å…¬å¸çš„ <strong>Solexa</strong> æŠ€æœ¯ä»¥åŠ ABI å…¬å¸çš„ <strong>SOLiD</strong> æŠ€æœ¯ä¸ºæ ‡å¿—çš„é«˜é€šé‡æµ‹åºæŠ€æœ¯ç›¸ç»§è¯ç”Ÿ [6]ã€‚ç›¸è¾ƒäºä¼ ç»Ÿæ–¹æ³•ï¼Œè¯¥æŠ€æœ¯ä¸»è¦ç‰¹ç‚¹æ˜¯æµ‹åºé€šé‡é«˜ã€æµ‹åºæ—¶é—´å’Œæˆæœ¬æ˜¾è‘—ä¸‹é™ï¼Œå¯ä»¥ä¸€æ¬¡å¯¹å‡ åä¸‡åˆ°å‡ ç™¾ä¸‡æ¡ DNA åˆ†å­åºåˆ—æµ‹å®šï¼Œè¿™ä½¿æŸç‰©ç§å…¨åŸºå› ç»„å’Œè½¬å½•ç»„çš„å…¨è²Œç»†è‡´åˆ†ææˆä¸ºå¯èƒ½ï¼Œåˆç§°ä¸º<strong>æ·±åº¦æµ‹åº</strong>ï¼Œå¾ˆå¤šæ–‡çŒ®ä¸­ç§°å…¶ä¸ºæ–°ä¸€ä»£æµ‹åºæŠ€æœ¯ï¼Œè¶³è§å…¶åˆ’æ—¶ä»£æ„ä¹‰ [7]ã€‚</p>
<p>@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54729244">è½¬å½•ç»„æµ‹åºæŠ€æœ¯çš„åº”ç”¨åŠå‘å±•ç»¼è¿° - çŸ¥ä¹ (zhihu.com)</a> <strong>æ¨èé˜…è¯»</strong></p>
</blockquote>
</li>
</ul>
<p><img src="https://pic2.zhimg.com/v2-d4a7d909a5aea6dd5b31f4072b5eac51_r.jpg" alt="å››ç§è½¬å½•å¹³å°çš„æ¯”è¾ƒ"></p>
<ul>
<li>
<p><strong>è½¬å½•ç»„</strong>(<strong>Transcriptome</strong>)æµ‹åºæŠ€æœ¯|<strong>RNA-Seq(RNA sequencing)æŠ€æœ¯</strong> :</p>
<ul>
<li>
<p><code>è½¬å½•æœ¬</code>å°±æ˜¯åŸºå› é€šè¿‡è½¬å½•å½¢æˆçš„é€šè¿‡å‰ªåˆ‡ç”Ÿæˆçš„ä¸€ç§æˆ–å¤šç§å¯ä¾›ç¼–ç è›‹ç™½è´¨çš„æˆç†Ÿçš„mRNAã€‚å‰ªåˆ‡çš„è¿‡ç¨‹ä¸­å¯èƒ½ä¼šå‰ªåˆ‡æ‰å¤–æ˜¾å­ï¼Œä¹Ÿæœ‰å¯èƒ½ä¿ç•™éƒ¨åˆ†å†…å«å­ï¼Œè¿™æ ·å°±å½¢æˆäº†å¤šç§mRNAå³å¤šä¸ªè½¬å½•æœ¬ã€‚<br>
@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/136389445_170798">åŸºå› ç ”ç©¶ä¸çŸ¥é“è½¬å½•æœ¬ï¼Œåšä»€ä¹ˆå¼•ç‰©è®¾è®¡ï¼Ÿï¼ (sohu.com)</a></p>
</li>
<li>
<p>RNA-Seqåˆ©ç”¨é«˜é€šé‡æµ‹åºæŠ€æœ¯å¯¹ç»„ç»‡æˆ–ç»†èƒä¸­<strong>æ‰€æœ‰RNA åè½¬å½•è€ŒæˆcDNAæ–‡åº“</strong>è¿›è¡Œæµ‹åºï¼Œé€šè¿‡ç»Ÿè®¡ç›¸å…³è¯»æ®µ(reads)æ•°è®¡ç®—å‡ºä¸åŒRNAçš„è¡¨è¾¾é‡ï¼Œå‘ç°æ–°çš„è½¬å½•æœ¬ï¼›å¦‚æœæœ‰åŸºå› ç»„å‚è€ƒåºåˆ—ï¼Œå¯ä»¥<strong>æŠŠè½¬å½•æœ¬æ˜ å°„å›åŸºå› ç»„</strong>ï¼Œç¡®å®šè½¬å½•æœ¬ä½ç½®ã€å‰ªåˆ‡æƒ…å†µç­‰æ›´ä¸ºå…¨é¢çš„é—ä¼ ä¿¡æ¯ï¼Œå·²å¹¿æ³›åº”ç”¨äºç”Ÿç‰©å­¦ç ”ç©¶ã€åŒ»å­¦ç ”ç©¶ã€ä¸´åºŠç ”ç©¶å’Œè¯ç‰©ç ”å‘ç­‰ã€‚<br>
@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54729244">è½¬å½•ç»„æµ‹åºæŠ€æœ¯çš„åº”ç”¨åŠå‘å±•ç»¼è¿° - çŸ¥ä¹ (zhihu.com)</a> <strong>æ¨èé˜…è¯»</strong></p>
</li>
<li>
<p>å¤§è‡´æµç¨‹:</p>
<p>æ¯”å¯¹readåˆ°åŸºå› ç»„ï¼Œè½¬å½•æœ¬ç»„è£…ï¼Œå·®å¼‚è¡¨è¾¾åˆ†æå’Œå·®å¼‚åŸºå› æ³¨é‡Š</p>
<p>è½¬å½•æœ¬ç»„è£…ä¸»è¦å°±æ˜¯å°†æ¯”å¯¹åˆ°å‚è€ƒåŸºå› ç»„çš„readç»„è£…æˆè½¬å½•æœ¬æ¥ç€è¿›è¡Œä¸‹æ¸¸çš„æ–°è½¬å½•æœ¬é‰´å®šå’Œè½¬å½•æœ¬å®šé‡ç­‰åˆ†æï¼Œå¸¸è§è½¯ä»¶æœ‰StringTieå’Œcufflinksã€‚è€Œå¯¹äºæ— å‚è½¬å½•ç»„åˆ†ææ¥è¯´ï¼Œè½¬å½•æœ¬ç»„è£…ä¸»è¦æ˜¯å°†æµ‹åºreadsä»å¤´ç»„è£…æˆè½¬å½•æœ¬ï¼Œå¸¸è§è½¯ä»¶æœ‰Trinity,Oaseså’ŒSOAPdenovo-Transã€‚</p>
<p>@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/413528571_652735">è½¬å½•ç»„æµ‹åºï¼ˆè½¬å½•æœ¬ç»„è£…ç¯‡ï¼‰_åˆ†æ (sohu.com)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/687766131808">è½¬å½•ç»„åˆ†ææµç¨‹ - ç®€ä¹¦ (jianshu.com)</a></p>
</li>
<li>
<p><strong>æ— å‚è½¬å½•ç»„å’Œæœ‰å‚è½¬å½•ç»„çš„åŒºåˆ«</strong>: å¦‚æœæ‰€ç ”ç©¶çš„ç‰©ç§æœ‰ç»„è£…æ³¨é‡Šè´¨é‡è¾ƒå¥½åŸºå› ç»„åºåˆ—ï¼Œä¸”å’Œè¯¥åŸºå› ç»„åºåˆ—æ¯”å¯¹æ•ˆç‡è¾ƒé«˜ï¼Œé‚£ä¹ˆå¯ä»¥é‡‡ç”¨æœ‰å‚è½¬å½•ç»„çš„åˆ†æç­–ç•¥ï¼Œç›´æ¥è¿›è¡Œåˆ†æã€‚åä¹‹ï¼Œåˆ™éœ€è¦æŒ‰ç…§æ— å‚è½¬å½•ç»„çš„åˆ†æç­–ç•¥è¿›è¡Œè½¬å½•æœ¬ç»„è£…ï¼Œæ„å»ºunigene(åŸºå› ç°‡)åº“ï¼Œç„¶åè¿›è¡Œåç»­åˆ†æã€‚<br>
@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/190736664">è½¬å½•ç»„æµ‹åºï¼šå¸¸è§åŸºç¡€çŸ¥è¯†å‡†å¤‡ä¸€ - çŸ¥ä¹ (zhihu.com)</a></p>
</li>
</ul>
</li>
<li>
<p>è½¬å½•æœ¬æ–‡ä»¶ <code>.gtf</code> : è®°å½•äº†å„ä¸ªæ ·æœ¬è£…é…å¥½çš„è½¬å½•æœ¬</p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="http://cole-trapnell-lab.github.io/cufflinks/">Cufflinks (cole-trapnell-lab.github.io)</a></p>
<blockquote>
<p>Cufflinks assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</p>
</blockquote>
<ul>
<li><a target="_blank" rel="noopener" href="http://cole-trapnell-lab.github.io/cufflinks/cuffcompare/">Cufflinks <strong>cuffcompare</strong></a></li>
</ul>
<blockquote>
<p>Cufflinks includes a program that you can use to help analyze the transfrags you assemble. The program <strong>cuffcompare</strong> helps you:</p>
<ul>
<li>Compare your assembled transcripts to a reference annotation</li>
<li>Track Cufflinks transcripts across multiple experiments (e.g. across a time course)</li>
</ul>
</blockquote>
</li>
<li>
<p>è½¬å½•æœ¬æ¯”å¯¹æ–‡ä»¶ <code>.tracking</code> : å¯¼å…¥è½¬å½•æœ¬æ–‡ä»¶(<code>.gtf</code>)åˆ° <code>cuffcompare</code> ç”Ÿæˆ</p>
</li>
<li>
<p>å…ƒç»„æ–¹æ³•:</p>
<ul>
<li>
<p><code>intersection</code>(*<em>others</em>) : è¿”å›ä¸€ä¸ªæ–°é›†åˆï¼Œå…¶ä¸­åŒ…å«åŸé›†åˆä»¥åŠ others æŒ‡å®šçš„æ‰€æœ‰é›†åˆä¸­å…±æœ‰çš„å…ƒç´ ã€‚</p>
<p>@<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=intersection#frozenset.intersection">intersection 5. å†…ç½®ç±»å‹ â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p><code>difference</code>(*<em>others</em>)  : è¿”å›ä¸€ä¸ªæ–°é›†åˆï¼Œå…¶ä¸­åŒ…å«åŸé›†åˆä¸­åœ¨ others æŒ‡å®šçš„å…¶ä»–é›†åˆä¸­ä¸å­˜åœ¨çš„å…ƒç´ ã€‚</p>
<p>@<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=intersection#frozenset.difference">difference 5. å†…ç½®ç±»å‹ â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p>æ›´å¤šæ–¹æ³•å‚è§æ–‡æ¡£</p>
</li>
</ul>
</li>
<li>
<p>åˆ—è¡¨çš„<code>.count()</code>æ–¹æ³•:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/array.html?highlight=count#array.array.count">8.6. array â€” é«˜æ•ˆçš„æ•°å€¼æ•°ç»„ â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li><code>array.count</code>(<em>x</em>) : è¿”å› <em>x</em> åœ¨æ•°ç»„ä¸­çš„å‡ºç°æ¬¡æ•°ã€‚</li>
</ul>
</li>
<li>
<p>å†…ç½®å‡½æ•°ç›¸å…³: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#built-in-functions">2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li><code>pop()</code>:</li>
<li><code>reduce()</code>:</li>
<li><code>del()</code>:</li>
<li><code>remove()</code>:</li>
<li><code>enumerate()</code>:</li>
<li></li>
</ul>
</li>
<li>
<p><code>.readlines()</code> è¿”å›çš„åˆ—è¡¨å¯ä»¥ç´¢å¼•:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">lines = <span class="built_in">open</span>(<span class="string">'text.txt'</span>).readlines()</span><br><span class="line"><span class="built_in">open</span>(<span class="string">'new.txt'</span>, <span class="string">'w'</span>).writelines(lines[<span class="number">2</span>:<span class="number">4</span>]+lines[<span class="number">6</span>:])</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>ä½¿ç”¨<code>åˆ—è¡¨æ¨å¯¼å¼</code>ç­›é€‰:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">data_a = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>]</span><br><span class="line">data = [x <span class="keyword">for</span> x <span class="keyword">in</span> data_a <span class="keyword">if</span> x != <span class="number">3</span>]</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>å»é‡å¦™æ³•-<code>å…ƒç»„å»é‡</code>:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">input_file = <span class="built_in">open</span>(<span class="string">'UniprotID.txt'</span>)</span><br><span class="line">unique = <span class="built_in">set</span>(input_file)</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p><code>CD-HIT</code>(å¯å®¹é”™çš„é«˜åŒæºæ€§èšç±»æ•°æ®åº“): å¯ä»¥å¿«é€Ÿåœ°çš„åŸºäºç”¨æˆ·å®šä¹‰çš„ç›¸ä¼¼æ€§é˜ˆå€¼å¯¹è›‹ç™½è´¨åºåˆ—è¿›è¡Œèšç±», éœ€è¦è¾“å…¥ä¸€ç»„<code>FASTA</code>æ ¼å¼çš„åºåˆ—,å¹¶è¿”å›ä¸¤ä¸ªæ–‡ä»¶: ä¸€ä¸ªæ˜¯èšç±»åˆ—è¡¨, ä¸€ä¸ªæ˜¯æ‰€èšå„ç±»çš„ä»£è¡¨åºåˆ—.</p>
<p>@<a target="_blank" rel="noopener" href="http://weizhongli-lab.org/cd-hit/">CD-HIT Official Website (weizhongli-lab.org)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://github.com/weizhongli/cdhit">weizhongli/cdhit: Automatically exported from code.google.com/p/cdhit (github.com)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/190353140_785442">æ•™ç¨‹ | å¦‚ä½•ç”¨cd-hitå»é™¤å†—ä½™åºåˆ—ï¼Ÿ (sohu.com)</a></p>
</li>
</ul>
<h3><a href="#di-7-zhang-guan-li-biao-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-7-zhang-guan-li-biao-shu-ju">ç¬¬7ç«  ç®¡ç†è¡¨æ•°æ®</span></h3>
<h4><a href="#7-2-an-li-que-ding-dan-bai-nong-du" class="header-anchor"> âš“ï¸</a><span id="7-2-an-li-que-ding-dan-bai-nong-du">7.2 æ¡ˆä¾‹ï¼š ç¡®å®šè›‹ç™½æµ“åº¦</span></h4>
<ul>
<li>
<p><code>è‹¯é…šç¦æ—è¯•å‰‚</code>æµ‹å®š<code>è›‹ç™½è´¨å«é‡</code>çš„æµç¨‹:</p>
<ul>
<li>
<p>ç›¸å…³è®ºæ–‡: <a target="_blank" rel="noopener" href="https://www.jbc.org/content/193/1/265.long">PROTEIN MEASUREMENT WITH THE FOLIN PHENOL REAGENT (jbc.org)</a></p>
</li>
<li>
<p>å¯¹æ¯”äº<code>å‡¯æ°å®šæ°®æ³•</code>çš„ä¼˜åŠ¿,å¯¹å¤šä¸ªæ ·å“å¯ä»¥ä½¿ç”¨å…‰åº¦è®¡å¿«é€Ÿåœ°è¿›è¡Œæµ‹é‡.</p>
<p>@<a target="_blank" rel="noopener" href="http://people.umass.edu/~mcclemen/581Proteins.html">6. ANALYSIS OF PROTEINS (umass.edu)</a></p>
</li>
</ul>
</li>
<li>
<p><code>zip()</code>å†…ç½®å‡½æ•°: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#zip"><code>zip</code>([<em>iterable</em>, <em>â€¦</em>]) 2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>x = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>y = [<span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>zipped = <span class="built_in">zip</span>(x, y)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>zipped</span><br><span class="line">[(<span class="number">1</span>, <span class="number">4</span>), (<span class="number">2</span>, <span class="number">5</span>), (<span class="number">3</span>, <span class="number">6</span>)]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>x2, y2 = <span class="built_in">zip</span>(*zipped)  <span class="comment"># *å·çš„åŠŸèƒ½æ˜¯å»é™¤zippedå…ƒç»„çš„å¤–æ‹¬å·(ç›¸å½“äºæ‹†è§£)</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>x == <span class="built_in">list</span>(x2) <span class="keyword">and</span> y == <span class="built_in">list</span>(y2)</span><br><span class="line"><span class="literal">True</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># æ•ˆæœåŒè½¬ç½®çŸ©é˜µ:</span></span><br><span class="line">zipped = <span class="built_in">zip</span>(*zipped)</span><br><span class="line"></span><br><span class="line"><span class="comment"># æ ·ä¾‹</span></span><br><span class="line">table = [</span><br><span class="line">    [<span class="string">'protein'</span>, <span class="string">'ext1'</span>, <span class="string">'ext2'</span>, <span class="string">'ext3'</span>],</span><br><span class="line">    [<span class="number">0.16</span>, <span class="number">0.038</span>, <span class="number">0.044</span>, <span class="number">0.040</span>],</span><br><span class="line">    [<span class="number">0.33</span>, <span class="number">0.089</span>, <span class="number">0.095</span>, <span class="number">0.091</span>],</span><br><span class="line">    [<span class="number">0.66</span>, <span class="number">0.184</span>, <span class="number">0.191</span>, <span class="number">0.191</span>],</span><br><span class="line">    [<span class="number">1.00</span>, <span class="number">0.280</span>, <span class="number">0.292</span>, <span class="number">0.283</span>],</span><br><span class="line">    [<span class="number">1.32</span>, <span class="number">0.365</span>, <span class="number">0.367</span>, <span class="number">0.365</span>],</span><br><span class="line">    [<span class="number">1.66</span>, <span class="number">0.441</span>, <span class="number">0.443</span>, <span class="number">0.444</span>]</span><br><span class="line">    ]</span><br><span class="line"></span><br><span class="line"><span class="comment"># remove the first row</span></span><br><span class="line">table = table[<span class="number">1</span>:]</span><br><span class="line"></span><br><span class="line">protein, ext1, ext2, ext3 = <span class="built_in">zip</span>(*table)</span><br><span class="line"></span><br><span class="line"><span class="comment"># create a single column for `ext` or concatenate each `ext`</span></span><br><span class="line"><span class="comment"># extend (`* 3`) the `protein` column to match the `ext` column.</span></span><br><span class="line">extinction = ext1 + ext2 + ext3</span><br><span class="line">protein = protein * <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># create four tuples for each column</span></span><br><span class="line">table = <span class="built_in">zip</span>(protein, extinction)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> prot, ext <span class="keyword">in</span> table:</span><br><span class="line">    <span class="built_in">print</span> prot, ext</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>åˆ—è¡¨åˆ›å»ºæ³¨æ„: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/faq/programming.html#id43">å¦‚ä½•åˆ›å»ºå¤šç»´åˆ—è¡¨ï¼Ÿç¼–ç¨‹å¸¸è§é—®é¢˜ â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li><code>[0]*3</code>è¿™æ ·çš„è¯­æ³•æ˜¯åˆ›å»ºäº†<code>[0]</code>çš„ä¸‰ä¸ªå¼•ç”¨, æ”¹å˜ä¸€ä¸ªåˆ—è¡¨çš„å†…å®¹å°†æ”¹å˜æ‰€æœ‰å†…å®¹</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">table = [</span><br><span class="line">    [<span class="string">'protein'</span>, <span class="string">'ext'</span>],</span><br><span class="line">    [<span class="number">0.16</span>, <span class="number">0.038</span>],</span><br><span class="line">    [<span class="number">0.33</span>, <span class="number">0.089</span>],</span><br><span class="line">    [<span class="number">0.66</span>, <span class="number">0.184</span>],</span><br><span class="line">    [<span class="number">1.00</span>, <span class="number">0.280</span>],</span><br><span class="line">    [<span class="number">1.32</span>, <span class="number">0.365</span>],</span><br><span class="line">    [<span class="number">1.66</span>, <span class="number">0.441</span>]</span><br><span class="line">]</span><br><span class="line"></span><br><span class="line"><span class="comment"># convert nested list to nested dict</span></span><br><span class="line">nested_dict = {}</span><br><span class="line">n = <span class="number">0</span></span><br><span class="line">key = table[<span class="number">0</span>]</span><br><span class="line"><span class="keyword">for</span> row <span class="keyword">in</span> table[<span class="number">1</span>:]: <span class="comment"># for row in table[0:]  includes the header</span></span><br><span class="line">    n += <span class="number">1</span></span><br><span class="line">    entry = {key[<span class="number">0</span>]: row[<span class="number">0</span>], key[<span class="number">1</span>]: row[<span class="number">1</span>]}</span><br><span class="line">    nested_dict[<span class="string">'row'</span>+<span class="built_in">str</span>(n)] = entry</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> nested_dict</span><br></pre></td></tr></tbody></table></figure>
<h4><a href="#7-5-zi-ce-ti" class="header-anchor"> âš“ï¸</a><span id="7-5-zi-ce-ti">7.5è‡ªæµ‹é¢˜</span></h4>
<ul>
<li>
<p>RNAç¢±åŸºç›¸ä¼¼æ€§çŸ©é˜µ:</p>
<table>
<thead>
<tr>
<th style="text-align:center"></th>
<th style="text-align:center">A</th>
<th style="text-align:center">G</th>
<th style="text-align:center">C</th>
<th style="text-align:center">U</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">A</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
</tr>
<tr>
<td style="text-align:center">G</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
</tr>
<tr>
<td style="text-align:center">C</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.5</td>
</tr>
<tr>
<td style="text-align:center">U</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">1.0</td>
</tr>
</tbody>
</table>
</li>
</ul>
<h3><a href="#di-8-zhang-shu-ju-pai-xu" class="header-anchor"> âš“ï¸</a><span id="di-8-zhang-shu-ju-pai-xu">ç¬¬8ç«  æ•°æ®æ’åº</span></h3>
<h4><a href="#8-2-an-li-shu-ju-biao-pai-xu" class="header-anchor"> âš“ï¸</a><span id="8-2-an-li-shu-ju-biao-pai-xu">8.2æ¡ˆä¾‹ï¼š æ•°æ®è¡¨æ’åº</span></h4>
<ul>
<li>
<p><code>sorted</code>(<em>iterable</em>[, <em>cmp</em>[, <em>key</em>[, <em>reverse</em>]]]) å†…ç½®å‡½æ•°: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=sorted#sorted">sorted 2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li>
<p>æ ¹æ® <em>iterable</em> ä¸­çš„é¡¹è¿”å›ä¸€ä¸ªæ–°çš„å·²æ’åºåˆ—è¡¨ã€‚</p>
</li>
<li>
<p>å‚æ•°<code>cmp</code>(<em>x</em>, <em>y</em>): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=sorted#cmp"> 2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<blockquote>
<p>sortedæ–¹æ³•ä¼šéšå¼çš„è°ƒç”¨cmp()å‡½æ•°è¿›è¡Œä¸¤ä¸¤æ¯”å¯¹, å¯ä»¥é€šè¿‡è‡ªå®šä¹‰å‡½æ•°ä¼ ç»™cmpå‚æ•°æ§åˆ¶æ’åº</p>
<p>Compare the two objects <em>x</em> and <em>y</em> and return an integer according to the outcome. The return value is negative if <code>x &lt; y</code>, zero if <code>x == y</code> and strictly positive if <code>x &gt; y</code>.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">def</span> <span class="title function_">numeric_compare</span>(<span class="params">x, y</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">return</span> x - y</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">sorted</span>([<span class="number">5</span>, <span class="number">2</span>, <span class="number">4</span>, <span class="number">1</span>, <span class="number">3</span>], cmp=numeric_compare) </span><br><span class="line">[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>]</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">my_cmp</span>(<span class="params">a,b</span>):  <span class="comment"># è¿™ä¸ªcmpå¯ä»¥åœ¨æ¯”å¯¹aå’Œbçš„ç¬¬xé¡¹ç›¸ç­‰æ—¶,å¯»æ‰¾x+1é¡¹è¿›è¡Œæ¯”å¯¹</span></span><br><span class="line">    x = <span class="number">0</span></span><br><span class="line">    <span class="keyword">if</span> a[x] &lt; b[x]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> a[x] == b[x]:</span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] &lt; b[x+<span class="number">1</span>]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] == b[x+<span class="number">1</span>]:</span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] &lt; b[x+<span class="number">2</span>]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] == b[x+<span class="number">2</span>]: <span class="keyword">return</span> <span class="number">0</span></span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] &gt; b[x+<span class="number">2</span>]: <span class="keyword">return</span> -<span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] &gt; b[x+<span class="number">1</span>]: <span class="keyword">return</span> -<span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> a[x] &gt; b[x]: <span class="keyword">return</span> -<span class="number">1</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>æ’åºé»˜è®¤æŒ‰ç…§<code>ASCII</code>é¡ºåºè¿›è¡Œæ’åº</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>space</td>
<td>0</td>
<td>:</td>
<td>A</td>
<td>Q</td>
<td>[</td>
<td>a</td>
<td>q</td>
<td>}</td>
</tr>
<tr>
<td>!</td>
<td>1</td>
<td>;</td>
<td>B</td>
<td>R</td>
<td>\</td>
<td>b</td>
<td>r</td>
<td>tick</td>
</tr>
<tr>
<td>â€œ</td>
<td>2</td>
<td>&lt;</td>
<td>C</td>
<td>S</td>
<td>]</td>
<td>c</td>
<td>s</td>
<td>{</td>
</tr>
<tr>
<td>#</td>
<td>3</td>
<td>=</td>
<td>D</td>
<td>T</td>
<td>^</td>
<td>d</td>
<td>t</td>
<td></td>
</tr>
<tr>
<td>$</td>
<td>4</td>
<td>&gt;</td>
<td>E</td>
<td>U</td>
<td>_</td>
<td>e</td>
<td>u</td>
<td>DEL</td>
</tr>
<tr>
<td>%</td>
<td>5</td>
<td>?</td>
<td>F</td>
<td>V</td>
<td>`</td>
<td>f</td>
<td>v</td>
<td></td>
</tr>
<tr>
<td>&amp;</td>
<td>6</td>
<td>@</td>
<td>G</td>
<td>W</td>
<td></td>
<td>g</td>
<td>w</td>
<td></td>
</tr>
<tr>
<td>â€˜</td>
<td>7</td>
<td></td>
<td>H</td>
<td>X</td>
<td></td>
<td>h</td>
<td>x</td>
<td></td>
</tr>
<tr>
<td>(</td>
<td>8</td>
<td></td>
<td>I</td>
<td>Y</td>
<td></td>
<td>i</td>
<td>y</td>
<td></td>
</tr>
<tr>
<td>)</td>
<td>9</td>
<td></td>
<td>J</td>
<td>Z</td>
<td></td>
<td>j</td>
<td>z</td>
<td></td>
</tr>
<tr>
<td>*</td>
<td></td>
<td></td>
<td>K</td>
<td></td>
<td></td>
<td>k</td>
<td></td>
<td></td>
</tr>
<tr>
<td>+</td>
<td></td>
<td></td>
<td>L</td>
<td></td>
<td></td>
<td>l</td>
<td></td>
<td></td>
</tr>
<tr>
<td>,</td>
<td></td>
<td></td>
<td>M</td>
<td></td>
<td></td>
<td>m</td>
<td></td>
<td></td>
</tr>
<tr>
<td>-</td>
<td></td>
<td></td>
<td>N</td>
<td></td>
<td></td>
<td>n</td>
<td></td>
<td></td>
</tr>
<tr>
<td>.</td>
<td></td>
<td></td>
<td>O</td>
<td></td>
<td></td>
<td>o</td>
<td></td>
<td></td>
</tr>
<tr>
<td>/</td>
<td></td>
<td></td>
<td>P</td>
<td></td>
<td></td>
<td>p</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
<li>
<p><code>openator</code>æ¨¡å—çš„<code>itemgetter</code>:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/operator.html?highlight=itemgetter#operator.itemgetter">9.9. operator â€” æ ‡å‡†è¿ç®—ç¬¦æ›¿ä»£å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>inventory = [(<span class="string">'apple'</span>, <span class="number">3</span>), (<span class="string">'banana'</span>, <span class="number">2</span>), (<span class="string">'pear'</span>, <span class="number">5</span>), (<span class="string">'orange'</span>, <span class="number">1</span>)]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>getcount = itemgetter(<span class="number">1</span>) <span class="comment"># æ„å»ºä¸€ä¸ªè·å–`ç´¢å¼•1çš„å€¼`çš„itemgetterå‡½æ•°</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">map</span>(getcount, inventory) <span class="comment"># å°†inventoryæŒ‰é¡¹é€ä¸ªä¼ å…¥getcount, è·å–æ¯é¡¹ç´¢å¼•å·ä¸º1çš„å€¼</span></span><br><span class="line">[<span class="number">3</span>, <span class="number">2</span>, <span class="number">5</span>, <span class="number">1</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">sorted</span>(inventory, key=getcount) <span class="comment"># ä¹Ÿå¯ä»¥å°†getcountä½œä¸ºkey, ä½¿ç”¨sortedæ’åº</span></span><br><span class="line"><span class="comment"># è¿™é‡ŒæŒ‰ç…§inventoryçš„æ¯ä¸€é¡¹çš„ç´¢å¼•å·ä¸º1çš„å€¼è¿›è¡Œæ’åº</span></span><br><span class="line">[(<span class="string">'orange'</span>, <span class="number">1</span>), (<span class="string">'banana'</span>, <span class="number">2</span>), (<span class="string">'apple'</span>, <span class="number">3</span>), (<span class="string">'pear'</span>, <span class="number">5</span>)]</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>æ›´å¤šæ’åºå‚è§: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/howto/sorting.html?highlight=cmp">æ’åºæŒ‡å— â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p>é€šè¿‡<code>shell</code>ç»ˆç«¯å‘½ä»¤æ’åº: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/c/sort.html">Linux sort å‘½ä»¤è¯¦è§£ï¼šå¯¹æ–‡æœ¬æ–‡ä»¶ä¸­æ‰€æœ‰è¡Œè¿›è¡Œæ’åºã€‚ - Linux å‘½ä»¤æœç´¢å¼•æ“ (wangchujiang.com)</a></p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> myfile.txt <span class="comment"># æŒ‰ç…§å­—æ¯è¡¨é¡ºåºæ’åº</span></span></span><br><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> -n myfile.txt <span class="comment"># æŒ‰ç…§æ•°å€¼æ’åº</span></span></span><br><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> myfile.txt -k2n -kl <span class="comment"># å¤šåˆ—æ’åº</span></span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>å¯¹<code>.tsv</code>åˆ¶è¡¨ç¬¦åˆ†éš”ç¬¦æ–‡ä»¶çš„æ¯è¡Œ,æŒ‰ç…§ä¸ƒä¸ªä¼˜å…ˆçº§æ’åº</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> operator <span class="keyword">import</span> itemgetter</span><br><span class="line"></span><br><span class="line"><span class="comment"># read table</span></span><br><span class="line">in_file = <span class="built_in">open</span>(<span class="string">"random_distribution.tsv"</span>)</span><br><span class="line">table = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    columns = line.split()</span><br><span class="line">    columns = [<span class="built_in">float</span>(x) <span class="keyword">for</span> x <span class="keyword">in</span> columns]</span><br><span class="line">    table.append(columns)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># æå–tableæ¯é¡¹çš„0~6å½¢æˆçš„å…ƒç»„è¿›è¡Œé€é¡¹æ¯”å¯¹æ’åº</span></span><br><span class="line">table_sorted = <span class="built_in">sorted</span>(table, key=itemgetter(<span class="number">0</span>, <span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>)) </span><br><span class="line"><span class="built_in">print</span> table_sorted</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>å¯¹BLASTçš„<code>.csv</code>é€—å·åˆ†éš”ç¬¦æ–‡ä»¶æ¯è¡Œçš„æŸä¸€é¡¹è¿›è¡Œæ’åº</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> operator <span class="keyword">import</span> itemgetter</span><br><span class="line"></span><br><span class="line">input_file = <span class="built_in">open</span>(<span class="string">"BlastOut.csv"</span>)</span><br><span class="line">output_file = <span class="built_in">open</span>(<span class="string">"BlastOutSorted.csv"</span>,<span class="string">"w"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># read BLAST output table</span></span><br><span class="line">table = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> input_file:</span><br><span class="line">    col = line.split(<span class="string">','</span>)</span><br><span class="line">    col[<span class="number">2</span>] = <span class="built_in">float</span>(col[<span class="number">2</span>])</span><br><span class="line">    table.append(col)</span><br><span class="line"></span><br><span class="line">table_sorted = <span class="built_in">sorted</span>(table, key=itemgetter(<span class="number">2</span>), reverse=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># write sorted table to an output file</span></span><br><span class="line"><span class="keyword">for</span> row <span class="keyword">in</span> table_sorted:</span><br><span class="line">    row = [<span class="built_in">str</span>(x) <span class="keyword">for</span> x <span class="keyword">in</span> row]</span><br><span class="line">    output_file.write(<span class="string">"\t"</span>.join(row) + <span class="string">'\n'</span>)</span><br><span class="line"></span><br><span class="line">input_file.close()</span><br><span class="line">output_file.close()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>blastp(è›‹ç™½è´¨æ¯”å¯¹è›‹ç™½æ•°æ®åº“)</code>å·¥å…·:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/wh-ff-ly520/p/11758999.html">blast+æœ¬åœ°åŒ–ä¸­blastpæ“ä½œ(åŸºäºPDBåº“)â€”linux - wangéœéœ - åšå®¢å›­ (cnblogs.com)</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/de28be1a3bea">è¿™æˆ–è®¸æ˜¯æˆ‘å†™çš„æœ€å…¨çš„BLASTæ•™ç¨‹ - ç®€ä¹¦ (jianshu.com)</a></li>
</ul>
</li>
<li>
<p><code>Aé“¾è¡€çº¢è›‹ç™½PDB</code>è¡¨æŠ¥å‘Š:</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://www.rcsb.org/search?request=%7B%22query%22%3A%7B%22parameters%22%3A%7B%22value%22%3A%22hemoglobin%22%7D%2C%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22node_id%22%3A0%7D%2C%22return_type%22%3A%22entry%22%2C%22request_options%22%3A%7B%22pager%22%3A%7B%22start%22%3A0%2C%22rows%22%3A100%7D%2C%22scoring_strategy%22%3A%22combined%22%2C%22sort%22%3A%5B%7B%22sort_by%22%3A%22score%22%2C%22direction%22%3A%22desc%22%7D%5D%7D%2C%22request_info%22%3A%7B%22src%22%3A%22ui%22%2C%22query_id%22%3A%22cb5954b71eff55f454db70f597aec338%22%7D%7D">RCSB PDB: Search</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-9-zhang-mo-shi-pi-pei-he-wen-ben-wa-jue" class="header-anchor"> âš“ï¸</a><span id="di-9-zhang-mo-shi-pi-pei-he-wen-ben-wa-jue">ç¬¬9ç«  æ¨¡å¼åŒ¹é…å’Œæ–‡æœ¬æŒ–æ˜</span></h3>
<h4><a href="#9-2-an-li-zai-dan-bai-zhi-xu-lie-zhong-sou-suo-lin-suan-hua-mo-ti" class="header-anchor"> âš“ï¸</a><span id="9-2-an-li-zai-dan-bai-zhi-xu-lie-zhong-sou-suo-lin-suan-hua-mo-ti">9.2 æ¡ˆä¾‹ï¼š åœ¨è›‹ç™½è´¨åºåˆ—ä¸­æœç´¢ç£·é…¸åŒ–æ¨¡ä½“</span></h4>
<ul>
<li>
<p><code>æ­£åˆ™è¡¨è¾¾å¼(regular expression, regexp)</code>: <strong>ä»£è¡¨</strong>ä¸€ç»„å­—ç¬¦ä¸²çš„å­—ç¬¦ä¸²è¯­æ³•</p>
<ul>
<li>
<p>ç±»ä¼¼çš„: ç”Ÿç‰©å­¦å®¶ä½¿ç”¨<code>N</code>æ¥ä»£æ›¿ä»»æ„ç¢±åŸº(ATCG)</p>
</li>
<li>
<p><code>[</code>å’Œ<code>]</code>è¢«ç§°ä¸º<code>å…ƒå­—ç¬¦</code>,ç”±å®ƒå’Œ<code>å­—ç¬¦</code>æ¥ç¼–ç ä¸€ç»„å­—ç¬¦ä¸²çš„è¡¨è¾¾,ä¸º<code>æ­£åˆ™è¡¨è¾¾å¼</code></p>
<ul>
<li><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex">7.2. re â€” æ­£åˆ™è¡¨è¾¾å¼æ“ä½œ â€” Python 2.7.18 æ–‡æ¡£</a> <strong>è¯·ä»”ç»†é˜…è¯»</strong></li>
</ul>
</li>
<li>
<p><code>re.compile</code>(<em>pattern</em>, <em>flags=0</em>): Compile a regular expression pattern into a regular expression object, which can be used for matching using its <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex#re.RegexObject.match"><code>match()</code></a> and <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex#re.RegexObject.search"><code>search()</code></a> methods.<br>
å‚æ•°è§ä¸Šé¢çš„æ–‡æ¡£</p>
</li>
<li>
<p><code>re.compile()</code>è¿”å›<code>RegexpObject</code>å¯¹è±¡, ä½¿ç”¨è¯¥å¯¹è±¡çš„<code>.search()</code>æ–¹æ³•å°†è¿”å›<code>Match Object</code>å¯¹è±¡, è¯¥å¯¹è±¡è¿˜å¯ä»¥ä½¿ç”¨<code>.group()</code>/<code>.span()</code>/<code>.start()</code>/<code>.end()</code>æ–¹æ³•. å¦å¤–<code>RegexpObject</code>å¯¹è±¡è¿˜æœ‰<code>.findall()</code>/<code>.match()</code>/<code>.finditer()</code>ç­‰æ–¹æ³•å¯ä»¥ç”¨, è¯¦è§æ–‡æ¡£</p>
</li>
<li>
<p><code>ç”±(</code>å’Œ<code>)</code>é…åˆåŒ¹é…åˆ°çš„å­—ç¬¦ä¸²å¯ä»¥ç”¨<code>group()</code>è·å–è¯†åˆ«åˆ°çš„æ›´å°çš„å­ç»„</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">seq = <span class="string">'zzabcdzz'</span></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'(a(b)c)d'</span>)</span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">0</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">1</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">2</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.groups()</span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
</ul>
<p><img src="https://ugoproto.github.io/ugo_py_doc/img/managing_your_biological/regex.png" alt="Regex"></p>
<ul>
<li>
<p>æ£€ç´¢åŒ¹é…æŒ‡å®šåºåˆ—:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re  <span class="comment"># æ­£åˆ™è¡¨è¾¾å¼æ¨¡å—</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># define a string with occurrences of regex:</span></span><br><span class="line">seq = <span class="string">'VSVLTMFRYAGWLDRLYMLVGTQLAAIIHGVALPLMMLI'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># compile a pattern and assign it to a variable</span></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'[ST]Q'</span>) <span class="comment"># è½¬æ¢å­—ç¬¦ä¸²ä¸ºæ­£åˆ™è¡¨è¾¾å¼å¯¹è±¡</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># search for the pattern in the string</span></span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="keyword">if</span> <span class="keyword">match</span>:</span><br><span class="line">    <span class="comment"># print the first match along the sequence with the group() method</span></span><br><span class="line">    <span class="comment"># 4 characters before and after the pattern</span></span><br><span class="line">    <span class="built_in">print</span> <span class="string">'%10s'</span> %(seq[<span class="keyword">match</span>.start() - <span class="number">4</span>:<span class="keyword">match</span>.end() + <span class="number">4</span>])</span><br><span class="line">    <span class="built_in">print</span> <span class="string">'%6s'</span> % <span class="keyword">match</span>.group()</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    <span class="built_in">print</span> <span class="string">"no match"</span></span><br><span class="line"><span class="comment"># ç»“æœ:   </span></span><br><span class="line"><span class="comment"># MLVGTQLAAI</span></span><br><span class="line"><span class="comment">#     TQ</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>åŠŸèƒ½æ€§æ¨¡ä½“è¡¨è¾¾:</p>
<ul>
<li>
<p><code>ELM</code>: <a target="_blank" rel="noopener" href="http://elm.eu.org/">elm.eu.org</a></p>
<blockquote>
<p>This computational biology resource mainly focuses on <strong>annotation and detection of eukaryotic linear motifs (ELMs) by providing both a repository of annotated motif data and an exploratory tool for motif prediction</strong>. <code>ELMs, or short linear motifs (SLiMs)</code>, are compact protein interaction sites composed of short stretches of adjacent amino acids. They are enriched in intrinsically disordered regions of the proteome and provide a wide range of functionality to proteins (<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/21909575">Davey,2011</a>,<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/24926813">Van Roey,2014</a>) They play crucial roles in cell regulation and are also of clinical importance, as aberrant SLiM function has been associated with several diseases and SLiM mimics are often used by pathogens to manipulate their hostsâ€™ cellular machinery (<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/21146412">Davey,2011</a>, <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/25057855">Uyar,2014</a>)</p>
</blockquote>
</li>
<li>
<p><code>PROSITE</code>: <a target="_blank" rel="noopener" href="https://prosite.expasy.org/">ExPASy - PROSITE</a></p>
<blockquote>
<p>PROSITE consists of <strong>documentation entries describing protein domains, families and functional sites as well as associated patterns and profiles to identify them</strong> [<a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_details.html">Moreâ€¦</a> / <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_ref.html">References</a> / <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_license.html">Commercial users</a> ].<br>
PROSITE is complemented by <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prorule.html">ProRule</a> , a collection of rules based on profiles and patterns, which increases the discriminatory power of profiles and patterns by providing additional information about functionally and/or structurally critical amino acids [<a target="_blank" rel="noopener" href="https://prosite.expasy.org/prorule_details.html">Moreâ€¦</a>].</p>
</blockquote>
</li>
</ul>
</li>
<li>
<p><code>.findall()</code>å’Œ<code>.finditer()</code>æ–¹æ³•:</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'R.[ST][^P]'</span>)</span><br><span class="line">seq = <span class="string">'RQSAMGSNKSKPKDASQRRRSLEPAENVHGAGGGAFPASQRPSKP'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># findall returns a list of all matches</span></span><br><span class="line">matches = pattern.findall(seq)</span><br><span class="line"><span class="built_in">print</span> matches  <span class="comment"># ç»“æœ: ['RQSA', 'RRSL', 'RPSK']</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># finditer returns an iterator of match objects</span></span><br><span class="line">match_iter = pattern.finditer(seq)  <span class="comment"># è¿”å›ä¸€ä¸ªç¬¦åˆæ¡ä»¶çš„è¿­ä»£å™¨</span></span><br><span class="line"><span class="keyword">for</span> <span class="keyword">match</span> <span class="keyword">in</span> match_iter:</span><br><span class="line">    <span class="built_in">print</span> <span class="keyword">match</span>.group(), <span class="keyword">match</span>.span(), <span class="keyword">match</span>.start(), <span class="keyword">match</span>.end()</span><br><span class="line"></span><br><span class="line">    <span class="comment"># ç»“æœ:</span></span><br><span class="line">    <span class="comment"># RQSA (0, 4) 0 4</span></span><br><span class="line">    <span class="comment"># RRSL (18, 22) 18 22</span></span><br><span class="line">    <span class="comment"># RPSK (40, 44) 40 44</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li><code>å­ç»„group()</code>çš„ä½¿ç”¨</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">seq = <span class="string">'QSAMGSNKSKPKDASQRRRSLEPAENVHGAGGGAFPASQRPSKP'</span></span><br><span class="line"></span><br><span class="line">pattern1 = re.<span class="built_in">compile</span>(<span class="string">'R(.)[ST][^P]'</span>)</span><br><span class="line">match1 = pattern1.search(seq)</span><br><span class="line"><span class="built_in">print</span> match1.group()  <span class="comment"># ç»“æœ: RRSL</span></span><br><span class="line"><span class="built_in">print</span> match1.group(<span class="number">1</span>) <span class="comment"># ç»“æœ: R</span></span><br><span class="line"></span><br><span class="line">pattern2 = re.<span class="built_in">compile</span>(<span class="string">'R(.{0,3})[ST][^P]'</span>)</span><br><span class="line">match2 = pattern2.search(seq)</span><br><span class="line"><span class="built_in">print</span> match2.group()  <span class="comment"># ç»“æœ: RRRSL</span></span><br><span class="line"><span class="built_in">print</span> match2.group(<span class="number">1</span>) <span class="comment"># ç»“æœ: RR</span></span><br><span class="line"></span><br><span class="line">seq = <span class="string">'zzabcdzz'</span></span><br><span class="line"></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'(?P&lt;w1&gt;a(?P&lt;w2&gt;b)c)d'</span>)</span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">0</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="string">'w1'</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="string">'w2'</span>)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>Linuxå‘½ä»¤-<code>grep</code>: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/c/grep.html">Linux grep å‘½ä»¤è¯¦è§£ï¼šå¼ºå¤§çš„æ–‡æœ¬æœç´¢å·¥å…· (wangchujiang.com)</a></li>
</ul>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep ^'&gt;' 3G5U.fasta # ç­›é€‰å¸¦æœ‰ `&gt;` çš„è¡Œ</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>PROSITE</code>æ­£åˆ™è¡¨è¾¾å¼è¯­æ³•:  <a target="_blank" rel="noopener" href="https://prosite.expasy.org/scanprosite/scanprosite_doc.html#pattern_syntax">ScanProsite user manual (expasy.org)</a></p>
<ul>
<li>
<p>å¯é€šè¿‡å­—ç¬¦ä¸²çš„<code>.replace()</code>æ–¹æ³•å°†å…¶è½¬æ¢ä¸ºPythonçš„<code>re</code>æ¨¡å—çš„è¯­æ³•</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># æ­¤æ­£åˆ™è¡¨è¾¾å¼ç”¨äºå¯»æ‰¾ é’™ç»“åˆçš„ç±»EGFç»“æ„åŸŸç­¾å(PROSITE ID: EGF_CA; PROSITE AC: PS01187)</span></span><br><span class="line">pattern = <span class="string">'[DEQN]-x-[DEQN](2)-C-x(3,14)-C-x(3,7)\</span></span><br><span class="line"><span class="string">      -C-x-[DN]-x(4)-[FY]-x-C'</span></span><br><span class="line">pattern = pattern.replace(<span class="string">'{'</span>,<span class="string">'[^'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'}'</span>,<span class="string">']'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'('</span>,<span class="string">'{'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">')'</span>,<span class="string">'}'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'-'</span>,<span class="string">''</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'x'</span>,<span class="string">'.'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'&gt;'</span>,<span class="string">'$'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'&lt;'</span>,<span class="string">'^'</span>)</span><br><span class="line"><span class="built_in">print</span> pattern</span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
<li>
<p>åœ¨åŸºå› ç»„åºåˆ—ä¸­æ‰¾åˆ°è½¬å½•å› å­çš„ç»“åˆä½ç‚¹:</p>
<ul>
<li><code>è½¬å½•å› å­ç»“åˆä½ç‚¹åˆ—è¡¨(TFBS)</code>ä½œä¸ºæ£€ç´¢è¡¨è¾¾å¼, æ£€ç´¢åŸºå› ç»„åºåˆ—å³å¯</li>
<li>è½¬å½•å› å­: <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1613417">æ‰¹é‡é¢„æµ‹è½¬å½•å› å­(TF)å’Œè½¬å½•å› å­ç»“åˆä½ç‚¹(TFBS) (tencent.com)</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">genome_seq = <span class="built_in">open</span>(<span class="string">'genome.txt'</span>).read()</span><br><span class="line"></span><br><span class="line"><span class="comment"># read transcription factor binding site patterns</span></span><br><span class="line">sites = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">'TFBS.txt'</span>):</span><br><span class="line">    fields = line.split()</span><br><span class="line">    tf = fields[<span class="number">0</span>]</span><br><span class="line">    site = fields[<span class="number">1</span>]</span><br><span class="line">    sites.append((tf, site))</span><br><span class="line"></span><br><span class="line"><span class="comment"># match all TF's to the genome and print matches</span></span><br><span class="line"><span class="keyword">for</span> tf, site <span class="keyword">in</span> sites:</span><br><span class="line">    tfbs_regexp = re.<span class="built_in">compile</span>(site)</span><br><span class="line">    all_matches = tfbs_regexp.findall(genome_seq)</span><br><span class="line">    matches = tfbs_regexp.finditer(genome_seq)</span><br><span class="line">    <span class="keyword">if</span> all_matches:</span><br><span class="line">        <span class="built_in">print</span> tf, <span class="string">':'</span></span><br><span class="line">        <span class="keyword">for</span> tfbs <span class="keyword">in</span> matches:</span><br><span class="line">            <span class="built_in">print</span> <span class="string">'\t'</span>, tfbs.group(), tfbs.start(), tfbs.end()</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>ä»PubMedçš„HTMLé¡µé¢æå–æ ‡é¢˜å’Œæ‘˜è¦æ–‡æœ¬:</p>
<ul>
<li>é€šè¿‡è®¿é—®æƒ³è¦çš„ç½‘é¡µç„¶åä½¿ç”¨æµè§ˆå™¨çš„<code>å¼€å‘è€…æ¨¡å¼</code>, å¾—åˆ°æƒ³è¦è·å–çš„å†…å®¹çš„æ ‡ç­¾, è¿™æ ·å¯ä»¥æ„å»ºæ­£åˆ™è¡¨è¾¾å¼æ¥æ£€ç´¢æˆ‘ä»¬æƒ³è¦çš„å†…å®¹.</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> urllib2  <span class="comment"># ç½‘ç»œè®¿é—®åº“</span></span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">pmid = <span class="string">'18235848'</span></span><br><span class="line">url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed?term=%s'</span> % pmid</span><br><span class="line">handler = urllib2.urlopen(url)  <span class="comment"># ä½¿ç”¨ urllib2çš„urlopenæ–¹æ³•æ‰“å¼€ç›®æ ‡ç½‘é¡µ</span></span><br><span class="line">html = handler.read()</span><br><span class="line"></span><br><span class="line">title_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;h1&gt;.{5,400}&lt;/h1&gt;'</span>)  <span class="comment"># åŒ¹é…æ ¼å¼ä¸ºä¸€çº§æ ‡é¢˜çš„å†…å®¹å­—ç¬¦5~400çš„æ ‡ç­¾</span></span><br><span class="line">title_text = title_regexp.search(html)</span><br><span class="line"></span><br><span class="line">abstract_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;AbstractText&gt;.{20,3000}&lt;/AbstractText&gt;'</span>)</span><br><span class="line">abstract_text = abstract_regexp.search(html)</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> <span class="string">'TITLE:'</span>, title_text.group() </span><br><span class="line"><span class="built_in">print</span> <span class="string">'ABSTRACT:'</span>, abstract_text.group()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>æ£€æµ‹ç§‘å­¦æ‘˜è¦ä¸­ç‰¹å®šçš„è¯æˆ–è¯ç»„:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> urllib2</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line"><span class="comment"># word to be searched</span></span><br><span class="line">word_regexp = re.<span class="built_in">compile</span>(<span class="string">'schistosoma'</span>)  <span class="comment"># æ„å»ºæƒ³è¦æ£€ç´¢çš„è¯ç¼–è¯‘å‡ºçš„æ­£åˆ™å¯¹è±¡</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># list of PMIDs where we want to search the word</span></span><br><span class="line">pmids = [<span class="string">'18235848'</span>, <span class="string">'22607149'</span>, <span class="string">'22405002'</span>, <span class="string">'21630672'</span>]  <span class="comment"># å¤šç¯‡æ–‡ç« çš„id</span></span><br><span class="line"><span class="keyword">for</span> pmid <span class="keyword">in</span> pmids:</span><br><span class="line">    url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed?term='</span> + pmid</span><br><span class="line">    handler = urllib2.urlopen(url)</span><br><span class="line">    html = handler.read()</span><br><span class="line">    title_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;h1&gt;.{5,400}&lt;/h1&gt;'</span>)</span><br><span class="line">    title = title_regexp.search(html)</span><br><span class="line">    title = title.group() </span><br><span class="line">    abstract_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;AbstractText&gt;.{20,3000}&lt;/AbstractText&gt;'</span>)</span><br><span class="line">    abstract = abstract_regexp.search(html)</span><br><span class="line">    abstract = abstract.group()</span><br><span class="line">    word = word_regexp.search(abstract, re.IGNORECASE) <span class="comment"># åŒ¹é…æƒ³è¦çš„å­—è¯</span></span><br><span class="line">    <span class="keyword">if</span> word:</span><br><span class="line">        <span class="comment"># display title and where the keyword was found</span></span><br><span class="line">        <span class="built_in">print</span> title</span><br><span class="line">        <span class="built_in">print</span> word.group(), word.start(), word.end()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>æ°¨åŸºé…¸ç£·é…¸åŒ–ä½ç‚¹:
<ul>
<li>ç®€å•çš„åŒ¹é…å¯ç”¨æ­£åˆ™è¡¨è¾¾å¼: <code>'R.[ST][^P]'</code></li>
<li>å‚è€ƒ: <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/beab18393d0a">NetPhos 3.1â€“è›‹ç™½ç£·é…¸åŒ–ä½ç‚¹é¢„æµ‹ - ç®€ä¹¦ (jianshu.com)</a></li>
</ul>
</li>
</ul>
<h2><a href="#di-san-bu-fen-mo-kuai-hua-bian-cheng" class="header-anchor"> âš“ï¸</a><span id="di-san-bu-fen-mo-kuai-hua-bian-cheng">ç¬¬ä¸‰éƒ¨åˆ† æ¨¡å—åŒ–ç¼–ç¨‹</span></h2>
<p>è¿™ä¸€éƒ¨åˆ†ä½œè€…å¸Œæœ›è¯»è€…å¯ä»¥æŒæ¡ç¼–å†™ç”±åŠŸèƒ½è¾ƒä¸ºç‹¬ç«‹çš„å®Œå–„çš„æ¨¡å—ç»„æˆçš„ç¨‹åº</p>
<h3><a href="#di-10-zhang-jiang-cheng-xu-hua-fen-wei-han-shu" class="header-anchor"> âš“ï¸</a><span id="di-10-zhang-jiang-cheng-xu-hua-fen-wei-han-shu">ç¬¬10ç«  å°†ç¨‹åºåˆ’åˆ†ä¸ºå‡½æ•°</span></h3>
<h4><a href="#10-2-an-li-chu-li-san-wei-zuo-biao-wen-jian" class="header-anchor"> âš“ï¸</a><span id="10-2-an-li-chu-li-san-wei-zuo-biao-wen-jian">10.2 æ¡ˆä¾‹ï¼š å¤„ç†ä¸‰ç»´åæ ‡æ–‡ä»¶</span></h4>
<ul>
<li>
<p><code>.PDB</code>æ ¼å¼è§£è¯»:</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html">Atomic Coordinate Entry Format Version 3.3 (wwpdb.org) </a>  <strong>æ¨èé˜…è¯»</strong></li>
</ul>
</li>
<li>
<p><a target="_blank" rel="noopener" href="http://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/primary-sequences-and-the-pdb-format">PDB-101: Learn: Guide to Understanding PDB Data: Primary Sequences and the PDB Format (rcsb.org)</a></p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/aff39489bd80">PDBæ–‡ä»¶è¯´æ˜ - ç®€ä¹¦ (jianshu.com)</a>  <strong>æ¨èé˜…è¯»</strong></li>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/klausage/p/10803308.html">PDB(Protein Data Bank)æ•°æ®æ ¼å¼è¯¦è§£ - Raymone1125 - åšå®¢å›­ (cnblogs.com)</a></li>
<li>è¯¥æ ¼å¼æ–‡ä»¶è·å–æ¥æº: <a target="_blank" rel="noopener" href="http://www.rcsb.org/">RCSB PDB: Homepage</a></li>
</ul>
<blockquote>
<p>(æ­¤è®°å½•ä¸ºä¹¦ä¸­ä¾‹å­è¿™é‡Œæ‘˜å½•ä½œä¸ºä¸€ä¸ªç®€å•çš„ä¾‹å­)</p>
<p>è®°å½•çš„ç¬¬ä¸€éƒ¨åˆ†æ­¤ä¸º<strong>æ ‡å¤´</strong>, ç”±å‡ ä¸ªè§£é‡Šè¡Œç»„æˆ, åŒ…æ‹¬æ¥æºç”Ÿç‰©ä½“/å®éªŒæŠ€æœ¯(Xå°„çº¿/æ ¸ç£å…±æŒ¯)/äº¤å‰å¼•ç”¨/å®éªŒç»†èŠ‚/åŸå§‹åˆ†å­åºåˆ—å’Œå˜å¼‚æ®‹åŸº(å¦‚æœæœ‰)ç­‰. è®°å½•çš„ç¬¬äºŒéƒ¨åˆ†ä»‹ç»äº†æ ‡å‡†ç»„çš„åŸå­åæ ‡è¡Œ. æ¯ä¸€ä¸ªåŸå­åæ ‡è¡Œå‡†ç¡®åœ°æè¿°äº†ä¸€ä¸ªåŸå­. ä»–ä»¬ä»¥**â€˜ATOMâ€™**å­—ç¬¦ä¸²å¼€å§‹: 1<sub>6åˆ—ä¸ºâ€™ATOMâ€™, 7</sub>11åˆ—ä¸ºâ€™åŸå­åºå·â€™, 13<sub>16åˆ—ä¸ºâ€™åŸå­åâ€™, 17åˆ—â€™å¦å­˜ä½ç½®æŒ‡ç¤ºâ€™, 18</sub>20åˆ—ä¸ºâ€™æ®‹åŸºåâ€™, 22åˆ—ä¸ºâ€™é“¾æ ‡è¯†ç¬¦â€™, 23<sub>26åˆ—ä¸ºâ€™æ®‹åŸºåºå·â€™, 27åˆ—ä¸ºâ€™æ®‹åŸºæ’å…¥ç â€™, 31</sub>38åˆ—ä¸ºxç¬›å¡å°”åæ ‡(xè½´å•ä½ä¸ºåŸƒ), 39<sub>46åˆ—ä¸ºâ€™yâ€¦â€˜, 47</sub>54åˆ—ä¸ºâ€™zâ€¦â€™, 55<sub>60åˆ—ä¸ºâ€™å¸ƒå±€â€™, 61</sub>66åˆ—ä¸ºâ€™æ¸©åº¦ç³»æ•°â€™, 77<sub>78åˆ—ä¸ºâ€™å…ƒç´ ç¬¦å·â€™, 79</sub>81åˆ—ä¸ºâ€™åŸå­ç”µè·â€™</p>
<p><strong>(å€¼å¾—è¯´æ˜çš„æ˜¯<code>åŸå­å</code> <code>CA</code>å’Œ<code>CB</code>æŒ‡çš„éƒ½æ˜¯ç¢³åŸå­, <code>A</code>å’Œ<code>B</code>ç”¨äºè¿œç¨‹æ ‡è¯†ç¬¦, è¡¨ç¤ºç¦»æ°¨åŸºç¢³åŸå­çš„è¿œè¿‘, <code>A</code>: Î±,<code>B</code>: Î²,<code>G</code>: Î³ç­‰,åŸå­åç§°çš„æœ€åä¸€ä¸ªå­—ç¬¦å¯ä»¥ä»£è¡¨åˆ†æ”¯æ ‡è¯†ç¬¦)</strong></p>
</blockquote>
</li>
<li>
<p><code>struct</code>æ¨¡å—: [7.3. struct â€” Interpret strings as packed binary data â€” Python 2.7.18 æ–‡æ¡£](<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/struct.html?highlight=struct">https://docs.python.org/zh-cn/2.7/library/struct.html?highlight=struct</a> pack#module-struct)</p>
<ul>
<li>
<p><code>struct.pack</code>(<em>fmt</em>, <em>v1</em>, <em>v2</em>, <em>â€¦</em>)</p>
<p>Return a string containing the values <code>v1, v2, ...</code> packed according to <code>the given format</code>. The arguments must match the values required by the format exactly.</p>
</li>
<li>
<p><code>struct.unpack</code>(<em>fmt</em>, <em>string</em>)</p>
<p>Unpack the string (presumably packed by <code>pack(fmt, ...)</code>) according to the given format. The result is a tuple even if it contains exactly one item. The string must contain exactly the amount of data required by the format (<code>len(string)</code> must equal <code>calcsize(fmt)</code>).</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> struct</span><br><span class="line"></span><br><span class="line"><span class="comment"># pack() method; creates a string</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'2s1s1s1s1s'</span></span><br><span class="line">group = struct.pack(<span class="built_in">format</span>, <span class="string">'10'</span>, <span class="string">'2'</span>, <span class="string">'3'</span>, <span class="string">'4'</span>, <span class="string">'5'</span>) <span class="comment"># ç»“æœ: 102345</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># unpack() method; parses the string to a tuple</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'1s2s1s1s'</span></span><br><span class="line">line = <span class="string">'12345'</span></span><br><span class="line">col = struct.unpack(<span class="built_in">format</span>, line)  <span class="comment"># ç»“æœ: ('1', '23', '4', '5')</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># calcsize() returns the number of characters</span></span><br><span class="line"><span class="comment"># in a given format string</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'30s30s20s1s'</span></span><br><span class="line"><span class="built_in">print</span> struct.calcsize(<span class="built_in">format</span>)  <span class="comment"># ç»“æœ: 81</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p><code>å‡½æ•°</code>å®šä¹‰ä¸ä½¿ç”¨: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=%E5%87%BD%E6%95%B0#defining-functions">å®šä¹‰å‡½æ•° 4. å…¶ä»–æµç¨‹æ§åˆ¶å·¥å…· â€” Python 2.7.18 æ–‡æ¡£</a>  <strong>è¯¦ç»†é˜…è¯»</strong></p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">def</span> <span class="title function_">fib</span>(<span class="params">n</span>):    <span class="comment"># write Fibonacci series up to n</span></span><br><span class="line"><span class="meta">... </span>    <span class="string">"""Print a Fibonacci series up to n."""</span></span><br><span class="line"><span class="meta">... </span>    a, b = <span class="number">0</span>, <span class="number">1</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">while</span> a &lt; n:</span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> a,</span><br><span class="line"><span class="meta">... </span>        a, b = b, a+b</span><br><span class="line">...</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="comment"># Now call the function we just defined:</span></span><br><span class="line"><span class="meta">... </span>fib(<span class="number">2000</span>)</span><br><span class="line"><span class="number">0</span> <span class="number">1</span> <span class="number">1</span> <span class="number">2</span> <span class="number">3</span> <span class="number">5</span> <span class="number">8</span> <span class="number">13</span> <span class="number">21</span> <span class="number">34</span> <span class="number">55</span> <span class="number">89</span> <span class="number">144</span> <span class="number">233</span> <span class="number">377</span> <span class="number">610</span> <span class="number">987</span> <span class="number">1597</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>å¤„ç†PDBæ•°æ®é‡Œçš„åæ ‡æ•°æ®å˜æˆæ°¨åŸºé…¸åºåˆ—:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">'''</span></span><br><span class="line"><span class="string">Extract a FASTA sequence from a PDB file.</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">(c) 2013 Allegra Via and Kristian Rother</span></span><br><span class="line"><span class="string">    Licensed under the conditions of the Python License</span></span><br><span class="line"><span class="string">    This code appears in section 10.2.2 of the book</span></span><br><span class="line"><span class="string">    "Managing Biological Data with Python".</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> struct</span><br><span class="line"></span><br><span class="line">pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s10s2s3s'</span> <span class="comment"># é‡‡ç”¨`format`æ ¼å¼ %6s</span></span><br><span class="line"></span><br><span class="line">amino_acids = {</span><br><span class="line">    <span class="string">'ALA'</span>:<span class="string">'A'</span>, <span class="string">'CYS'</span>:<span class="string">'C'</span>, <span class="string">'ASP'</span>:<span class="string">'D'</span>, <span class="string">'GLU'</span>:<span class="string">'E'</span>,</span><br><span class="line">    <span class="string">'PHE'</span>:<span class="string">'F'</span>, <span class="string">'GLY'</span>:<span class="string">'G'</span>, <span class="string">'HIS'</span>:<span class="string">'H'</span>, <span class="string">'LYS'</span>:<span class="string">'K'</span>,</span><br><span class="line">    <span class="string">'ILE'</span>:<span class="string">'I'</span>, <span class="string">'LEU'</span>:<span class="string">'L'</span>, <span class="string">'MET'</span>:<span class="string">'M'</span>, <span class="string">'ASN'</span>:<span class="string">'N'</span>,</span><br><span class="line">    <span class="string">'PRO'</span>:<span class="string">'P'</span>, <span class="string">'GLN'</span>:<span class="string">'Q'</span>, <span class="string">'ARG'</span>:<span class="string">'R'</span>, <span class="string">'SER'</span>:<span class="string">'S'</span>,</span><br><span class="line">    <span class="string">'THR'</span>:<span class="string">'T'</span>, <span class="string">'VAL'</span>:<span class="string">'V'</span>, <span class="string">'TYR'</span>:<span class="string">'Y'</span>, <span class="string">'TRP'</span>:<span class="string">'W'</span></span><br><span class="line">    }</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">threeletter2oneletter</span>(<span class="params">residues</span>):   <span class="comment"># æ°¨åŸºé…¸ç¼©å†™ä»ä¸‰ä¸ªå­—æ¯ 2(to,åˆ°) å•ä¸ªå­—æ¯ </span></span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Converts the first element of each residue</span></span><br><span class="line"><span class="string">    to a one letter amino acid symbol</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    <span class="keyword">for</span> i, threeletter <span class="keyword">in</span> <span class="built_in">enumerate</span>(residues):</span><br><span class="line">        residues[i][<span class="number">0</span>] = amino_acids[threeletter[<span class="number">0</span>]]</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_residues</span>(<span class="params">pdb_file</span>):  </span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Reads the PDB input file, extracts the</span></span><br><span class="line"><span class="string">    residue type and chain from the CA lines</span></span><br><span class="line"><span class="string">    and appends both to the residues list</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    residues = []</span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> pdb_file:</span><br><span class="line">        <span class="keyword">if</span> line[<span class="number">0</span>:<span class="number">4</span>] == <span class="string">"ATOM"</span>:</span><br><span class="line">            tmp = struct.unpack(pdb_format, line)</span><br><span class="line">            ca = tmp[<span class="number">3</span>].strip()</span><br><span class="line">            <span class="keyword">if</span> ca == <span class="string">'CA'</span>:</span><br><span class="line">                res_type = tmp[<span class="number">5</span>].strip()</span><br><span class="line">                chain = tmp[<span class="number">7</span>]</span><br><span class="line">                residues.append([res_type, chain])</span><br><span class="line">    <span class="keyword">return</span> residues</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_fasta_records</span>(<span class="params">residues, pdb_id, fasta_file</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Write a FASTA record for each PDB chain</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    seq = <span class="string">''</span></span><br><span class="line">    chain = residues[<span class="number">0</span>][<span class="number">1</span>]</span><br><span class="line">    <span class="keyword">for</span> aa, new_chain <span class="keyword">in</span> residues:</span><br><span class="line">        <span class="keyword">if</span> new_chain == chain:</span><br><span class="line">            seq = seq + aa</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># write sequence in FASTA format</span></span><br><span class="line">            fasta_file.write(<span class="string">"&gt;%s_%s\n%s\n"</span> % (pdb_id, chain, seq))</span><br><span class="line">            seq = aa</span><br><span class="line">            chain = new_chain</span><br><span class="line">    <span class="comment"># write the last PDB chain</span></span><br><span class="line">    fasta_file.write(<span class="string">"&gt;%s_%s\n%s\n"</span> % (pdb_id, chain, seq))</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">extract_sequence</span>(<span class="params">pdb_id</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Main function: Opens files, writes files</span></span><br><span class="line"><span class="string">    and calls other functions.</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    pdb_file = <span class="built_in">open</span>(pdb_id + <span class="string">".pdb"</span>)</span><br><span class="line">    fasta_file = <span class="built_in">open</span>(pdb_id + <span class="string">".fasta"</span>, <span class="string">"w"</span>)</span><br><span class="line">    residues = get_residues(pdb_file)</span><br><span class="line">    threeletter2oneletter(residues)</span><br><span class="line">    write_fasta_records(residues, pdb_id, fasta_file)</span><br><span class="line">    pdb_file.close()</span><br><span class="line">    fasta_file.close()</span><br><span class="line">    </span><br><span class="line"><span class="comment"># call the main function</span></span><br><span class="line">extract_sequence(<span class="string">"3G5U"</span>)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>range()</code>è¿­ä»£åˆ©å™¨: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#range">range()  2. å†…ç½®å‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p><code>å¯å˜å‚æ•°</code>ä¸<code>ä¸å¯å˜å‚æ•°</code>:</p>
<ul>
<li>ä¸è¦ä½¿ç”¨ä¸å¯å˜å‚æ•°ä½œä¸ºå‡½æ•°çš„é»˜è®¤å‚æ•°</li>
</ul>
</li>
<li>
<p><code>lambdaåŒ¿åå‡½æ•°</code>: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=lambda#lambda-expressions">4. å…¶ä»–æµç¨‹æ§åˆ¶å·¥å…· â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
</ul>
<h3><a href="#di-11-zhang-yong-lei-hua-fan-wei-jian" class="header-anchor"> âš“ï¸</a><span id="di-11-zhang-yong-lei-hua-fan-wei-jian">ç¬¬11ç«  ç”¨ç±»åŒ–ç¹ä¸ºç®€</span></h3>
<h4><a href="#11-2-an-li-meng-de-er-yi-chuan" class="header-anchor"> âš“ï¸</a><span id="11-2-an-li-meng-de-er-yi-chuan">11.2 æ¡ˆä¾‹ï¼š å­Ÿå¾·å°”é—ä¼ </span></h4>
<ul>
<li><code>Class(ç±»)</code>çš„æ„å»º:
<ul>
<li><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/classes.html?highlight=class">9. ç±» â€” Python 2.7.18 æ–‡æ¡£</a>  <strong>ä»”ç»†é˜…è¯»</strong></li>
<li><code>.__repr__()</code> ç±»æ–¹æ³•: å½“<code>print</code>å‘½ä»¤æ‰“å°ç±»æ—¶, å°†æŒ‰ç…§æ­¤ç±»æ–¹æ³•è¿”å›çš„æ ¼å¼è¿›è¡Œè¾“å‡º</li>
<li>ç±»åœ¨å¤šäººåˆä½œç¼–ç¨‹çš„ä¼˜åŠ¿</li>
<li>å­ç±»ä¸ç»§æ‰¿</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Pea</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, genotype</span>):  <span class="comment"># åˆå§‹åŒ–çš„æ„é€ å‡½æ•°</span></span><br><span class="line">        self.genotype = genotype</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">PeaStrain</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, peas</span>):</span><br><span class="line">        self.peas = peas</span><br><span class="line"></span><br><span class="line">yellow = Pea(<span class="string">'GG'</span>)  <span class="comment"># ç±»çš„å®ä¾‹åŒ–</span></span><br><span class="line">green = Pea(<span class="string">'gg'</span>)</span><br><span class="line">strain = PeaStrain([yellow, green])</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> Pea                 <span class="comment"># __main__.Pea</span></span><br><span class="line"><span class="built_in">print</span> PeaStrain           <span class="comment"># __main__.PeaStrain</span></span><br><span class="line"><span class="built_in">print</span> Pea.__init__        <span class="comment"># &lt;unbound method Pea.__init__&gt;</span></span><br><span class="line"><span class="built_in">print</span> PeaStrain.__init__  <span class="comment"># &lt;unbound method PeaStrain.__init__&gt;</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> yellow              <span class="comment"># &lt;__main__.Pea instance at 0x0000000001DC6748&gt;</span></span><br><span class="line"><span class="built_in">print</span> green               <span class="comment"># &lt;__main__.Pea instance at 0x000000000216BE08&gt;</span></span><br><span class="line"><span class="built_in">print</span> strain              <span class="comment"># &lt;__main__.PeaStrain instance at 0x000000000216BE48&gt;</span></span><br><span class="line"></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"><span class="string">Classes can inherit from other classes</span></span><br><span class="line"><span class="string">and extend their functionality.</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">(c) 2013 Allegra Via and Kristian Rother</span></span><br><span class="line"><span class="string">    Licensed under the conditions of the Python License</span></span><br><span class="line"><span class="string">    This code appears in section 11.4.3 of the book</span></span><br><span class="line"><span class="string">    "Managing Biological Data with Python".</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> pea <span class="keyword">import</span> Pea</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">CommentedPea</span>(<span class="title class_ inherited__">Pea</span>):</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, genotype, comment</span>):</span><br><span class="line">        Pea.__init__(self, genotype)        <span class="comment"># è½½å…¥çˆ¶ç±»æ„é€ å‡½æ•°</span></span><br><span class="line">        self.comment = comment</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__repr__</span>(<span class="params">self</span>):</span><br><span class="line">        <span class="keyword">return</span>  <span class="string">'%s [%s] (%s)'</span> % (self.get_phenotype(), self.genotype, self.comment)</span><br><span class="line"></span><br><span class="line">yellow1 = CommentedPea(<span class="string">'GG'</span>, <span class="string">'homozygote'</span>)</span><br><span class="line">yellow2 = CommentedPea(<span class="string">'Gg'</span>, <span class="string">'heterozygote'</span>)</span><br><span class="line"><span class="built_in">print</span> yellow1</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç±»çš„è®¾è®¡æŒ‡å—:
<ul>
<li><code>å•ä¸€èŒè´£åŸåˆ™</code>å’Œ<code>ä¸é‡å¤è‡ªå·±åŸåˆ™</code></li>
<li><code>è®¾è®¡æ¨¡å¼</code> : <a target="_blank" rel="noopener" href="https://sourcemaking.com/">Design Patterns &amp; Refactoring (sourcemaking.com)</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-12-zhang-diao-shi" class="header-anchor"> âš“ï¸</a><span id="di-12-zhang-diao-shi">ç¬¬12ç«  è°ƒè¯•</span></h3>
<h4><a href="#12-2-an-li-cheng-xu-wu-fa-yun-xing-shi-ying-gai-zen-yang-chu-li" class="header-anchor"> âš“ï¸</a><span id="12-2-an-li-cheng-xu-wu-fa-yun-xing-shi-ying-gai-zen-yang-chu-li">12.2 æ¡ˆä¾‹ï¼š ç¨‹åºæ— æ³•è¿è¡Œæ—¶åº”è¯¥æ€æ ·å¤„ç†</span></h4>
<ul>
<li>
<p>ä¸‰ç§é”™è¯¯ç±»å‹</p>
<ul>
<li><code>è¯­æ³•é”™è¯¯(SyntaxError)</code>: Pythonè§„å®šè¯­æ³•ä¹‹å¤–çš„ä¸åˆæ³•çš„å†™æ³•</li>
<li><code>é€»è¾‘é”™è¯¯</code>: ç¨‹åºè¿è¡Œä¸ä¼šæŠ¥é”™, ä½†æ˜¯ç¨‹åºçš„è¿è¡Œç»“æœä¸æ˜¯ä½ æƒ³è¦çš„ç»“æœ</li>
<li><code>è¿è¡Œæ—¶é”™è¯¯</code>: è¯­æ³•æ£€æŸ¥é€šè¿‡, ä½†æ˜¯è¿è¡Œæ—¶å‡ºç°äº†æ„æƒ³ä¸åˆ°çš„æƒ…å†µçš„</li>
</ul>
</li>
<li>
<p>å¼‚å¸¸å¤„ç†: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/errors.html">8. é”™è¯¯å’Œå¼‚å¸¸ â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li><code>try: ... except:</code> è¯­å¥:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">while</span> <span class="literal">True</span>:</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">try</span>:</span><br><span class="line"><span class="meta">... </span>        x = <span class="built_in">int</span>(raw_input(<span class="string">"Please enter a number: "</span>))</span><br><span class="line"><span class="meta">... </span>        <span class="keyword">break</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">except</span> ValueError:</span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> <span class="string">"Oops!  That was no valid number.  Try again..."</span></span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
<blockquote>
<p>æˆ‘è¿˜æ²¡æœ‰æ•°æ®. åœ¨æŒæ¡æ•°æ®å‰å»ºç«‹ç†è®ºæ˜¯ä¸€ä¸ªå·¨å¤§çš„é”™è¯¯. ä¸çŸ¥ä¸è§‰åœ°, å°±å¼€å§‹æ‰­æ›²äº‹å®ä»¥é€‚åº”ç†è®º, è€Œä¸æ˜¯è®©ç†è®ºç¬¦åˆäº‹å®.</p>
<p>@ <em>The Adventures of Sherlock Holmes</em>, Arthur Conan Doyle</p>
</blockquote>
<ul>
<li><code>pdb</code>è°ƒè¯•å·¥å…·: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/pdb.html">26.2. pdb â€” Pythonçš„è°ƒè¯•å™¨ â€” Python 2.7.18 æ–‡æ¡£</a></li>
</ul>
<h3><a href="#di-13-zhang-shi-yong-wai-bu-mo-kuai-r-yu-yan-de-python-diao-yong-jie-kou" class="header-anchor"> âš“ï¸</a><span id="di-13-zhang-shi-yong-wai-bu-mo-kuai-r-yu-yan-de-python-diao-yong-jie-kou">ç¬¬13ç«  ä½¿ç”¨å¤–éƒ¨æ¨¡å—:Rè¯­è¨€çš„Pythonè°ƒç”¨æ¥å£</span></h3>
<h4><a href="#13-2-an-li-cong-wen-jian-zhong-du-qu-shu-ju-bing-tong-guo-python-shi-yong-r-ji-suan-qi-ping-jun-zhi" class="header-anchor"> âš“ï¸</a><span id="13-2-an-li-cong-wen-jian-zhong-du-qu-shu-ju-bing-tong-guo-python-shi-yong-r-ji-suan-qi-ping-jun-zhi">13.2 æ¡ˆä¾‹ï¼š ä»æ–‡ä»¶ä¸­è¯»å–æ•°æ®ï¼Œ å¹¶é€šè¿‡Pythonä½¿ç”¨Rè®¡ç®—å…¶å¹³å‡å€¼</span></h4>
<ul>
<li><code>Rè¯­è¨€</code></li>
</ul>
<h3><a href="#di-14-zhang-gou-jian-cheng-xu-liu-cheng" class="header-anchor"> âš“ï¸</a><span id="di-14-zhang-gou-jian-cheng-xu-liu-cheng">ç¬¬14ç«  æ„å»ºç¨‹åºæµç¨‹</span></h3>
<h4><a href="#14-2-an-li-gou-jian-ngs-liu-cheng" class="header-anchor"> âš“ï¸</a><span id="14-2-an-li-gou-jian-ngs-liu-cheng">14.2 æ¡ˆä¾‹ï¼š æ„å»ºNGSæµç¨‹</span></h4>
<ul>
<li>
<p><code>NGS</code>æŠ€æœ¯åº”ç”¨:</p>
<ul>
<li>å·®å¼‚è¡¨è¾¾åŸºå› </li>
<li>éç¼–ç RNA</li>
<li>çªå˜</li>
</ul>
</li>
<li>
<p><code>Top Hat</code></p>
</li>
<li>
<p>RNA-seq:</p>
<ul>
<li>è¾“å‡ºåŒ…å«<code>å°cDNAæµ‹åºç‰‡æ®µ</code> (ç§°ä¸º<code>è¯»é•¿</code>,<code>reads</code>)çš„æ–‡ä»¶, éœ€å›è´´è‡³å‚è€ƒåŸºå› ç»„å¹¶è¿›è¡Œæ‹¼è£…, ä»¥è·å¾—å…¨éƒ¨çš„æ ·æœ¬åŸºå› åºåˆ—</li>
<li>æŸç§æµç¨‹å®ä¾‹:
<ul>
<li>ç”±<code>Illumina(GAIIx)</code>(æˆ–å…¶ä»–NGS)å¹³å°ç”Ÿæˆå¹¶å­˜å‚¨<code>è¯»é•¿</code>è‡³<code>æ–‡æœ¬æ–‡ä»¶(å¦‚sample.fastq)</code>ä¸­.</li>
<li><code>TopHat</code>å¯ä»¥ä½œä¸ºRNA-seqæ•°æ®çš„å‰ªåˆ‡æ¥å¤´å›è´´å·¥å…·, å®ƒå¯ä»¥æŠŠRNA-seqè¯»é•¿å¯¹æ¯”è‡³å“ºä¹³åŠ¨ç‰©æ•°é‡çº§çš„åŸºå› ç»„ä¸­, å‚è€ƒåŸºå› ç»„å¿…é¡»ä¿å­˜è‡³æœ¬åœ°ä»¥ä¾¿äº<code>TopHat</code>è®¿é—®, ç„¶ååˆ†æå›è´´ç»“æœ, æ‰¾åˆ°å¤–æ˜¾å­ä¹‹é—´çš„å‰ªåˆ‡æ¥å¤´. <code>TopHat</code>è¯»å–<code>sample.fastq</code>å¾—åˆ°æ¯”å¯¹è¯»é•¿åˆ—è¡¨, å‚¨å­˜åœ¨<code>.bam</code>æ–‡ä»¶ä¸­.</li>
<li><code>Cufflinks</code>åŒ…å«æœ‰ä¸‰ä¸ªç¨‹åº: è½¬å½•æœ¬è£…é…(Cufflinks)/è½¬å½•æœ¬æ¯”è¾ƒ(Cuffcompare)/æ£€æµ‹è°ƒæ§å’Œè¡¨è¾¾å·®å¼‚(Cuffdiff). <code>Cufflinks</code>è¿›è¡Œ<code>è½¬å½•æœ¬çš„æ‹¼æ¥</code>, ç”Ÿæˆæ ¼å¼ä¸º<code>transcripts.gtf</code>çš„è½¬å½•æ–‡ä»¶, ä»¥ä¾¿ä¹‹åæ‰€ç”¨. å¯¹äºä¸åŒçš„æ ·æœ¬(é‡ç”Ÿå‹/ç™Œç—‡ç»†èƒ, é‡å¤å®éªŒ)è¿˜å¯ä»¥ç­›é€‰è½¬å½•ç»„æ–‡ä»¶, ç„¶åè¿›è¡Œç›¸äº’æ¯”è¾ƒ, æŒ‰ç…§å”¯ä¸€çš„å‚è€ƒè½¬å½•ç»„è¿›è¡Œæ‹¼æ¥</li>
<li>æ‹¼æ¥å¥½åçš„æ–‡ä»¶ä¸º<code>transcripts.gtf</code>,å¯ä»¥ä½¿ç”¨Cufflinkå·¥å…·åŒ…ä¸­çš„Cuffcompareç¨‹åºä¸å…¶ä»–è½¬å½•ç»„è¿›è¡Œæ¯”è¾ƒ</li>
</ul>
</li>
</ul>
</li>
<li>
<p><code>os</code>æ¨¡å—: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/os.html?highlight=os#module-os">15.1. os â€” æ“ä½œç³»ç»Ÿæ¥å£æ¨¡å— â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p><code>åŒ…è£…å™¨(wrapper)</code>: å¯ä»¥è¿è¡Œå…¶ä»–ç¨‹åºçš„ç¨‹åº</p>
</li>
<li>
<p>ç¨‹åºé—´ä¼ é€’æ•°æ®:</p>
<ul>
<li><code>raw_input()</code>ç›´æ¥æ‰‹åŠ¨è¾“å…¥æ•°æ®</li>
<li>ä½¿ç”¨<code>Python</code>å˜é‡</li>
<li>å°†æ•°æ®å­˜å‚¨åœ¨å¦ä¸€ä¸ªæ–‡ä»¶é‡Œ</li>
<li>ä½¿ç”¨å‘½ä»¤è¡Œå‚æ•°, æ¯”å¦‚<code>sys.argv</code>åˆ—è¡¨å˜é‡, å‚è§<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/sys.html?highlight=sys#module-sys">28.1. sys â€” ç³»ç»Ÿç›¸å…³çš„å‚æ•°å’Œå‡½æ•° â€” Python 2.7.18 æ–‡æ¡£</a></li>
</ul>
</li>
<li>
<p>å…³é—­æ–‡ä»¶çš„å»¶æ—¶é—®é¢˜: é€šè¿‡å¢åŠ ä¸€ä¸ªä¸­é—´æ­¥éª¤(æ¯”å¦‚æ‰“å¼€ä¸€ä¸ªæ–‡ä»¶ç„¶åå†™å…¥ä¸€äº›å­—ç¬¦ä¿å­˜, æ¥ç¡®ä¿è¿™ä¸ªä¸­é—´æ­¥éª¤çš„ä¸Šä¸‹ä¸¤ä¸ªæ­¥éª¤äº’ä¸å†²çª</p>
</li>
<li>
<p><code>os.path</code> æ“æ§æ–‡ä»¶ç›®å½•: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/os.path.html#module-os.path">10.1. os.path â€” å¸¸è§è·¯å¾„æ“ä½œ â€” Python 2.7.18 æ–‡æ¡£</a></p>
</li>
<li>
<p><code>MSA(multiple sequence alignment, å¤šåºåˆ—æ¯”å¯¹)</code>:</p>
<blockquote>
<p><strong>Pairwise Sequence Alignment</strong> is used to identify regions of similarity that may indicate functional, structural and/or evolutionary relationships between two biological sequences (protein or nucleic acid).</p>
<p>By contrast, <strong><a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/Tools/msa/">Multiple Sequence Alignment</a> (MSA)</strong> is the alignment of three or more biological sequences of similar length. From the output of MSA applications, homology can be inferred and the evolutionary relationship between the sequences studied.</p>
</blockquote>
<ul>
<li>
<p>MSAå·¥å…·: <a target="_blank" rel="noopener" href="https://www.genome.jp/tools-bin/clustalw">Multiple Sequence Alignment - CLUSTALW (genome.jp)</a></p>
</li>
<li>
<p><code>T-COFFEE</code>ç®—æ³•: <a target="_blank" rel="noopener" href="http://www.tcoffee.org/Projects/tcoffee/">T-Coffee Home Page (tcoffee.org)</a></p>
<blockquote>
<p>T-Coffee is a multiple sequence alignment package. You can use T-Coffee to align sequences or to combine the output of your favorite alignment methods (Clustal, Mafft, Probcons, Muscleâ€¦) into one unique alignment (<a target="_blank" rel="noopener" href="http://www.tcoffee.org/Projects/mcoffee/index.html">M-Coffee</a>).</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<h3><a href="#di-15-zhang-bian-xie-liang-hao-de-cheng-xu" class="header-anchor"> âš“ï¸</a><span id="di-15-zhang-bian-xie-liang-hao-de-cheng-xu">ç¬¬15ç«  ç¼–å†™è‰¯å¥½çš„ç¨‹åº</span></h3>
<h4><a href="#15-2-wen-ti-miao-shu-bu-que-ding-xing" class="header-anchor"> âš“ï¸</a><span id="15-2-wen-ti-miao-shu-bu-que-ding-xing">15.2 é—®é¢˜æè¿°ï¼š ä¸ç¡®å®šæ€§</span></h4>
<ul>
<li>
<p>ã€Šç¼–ç¨‹çš„å‡†åˆ™ã€‹<em>(Dogma of Programming)</em>:</p>
<ul>
<li>è‘—: Donald Knuth</li>
</ul>
</li>
<li>
<p>è½¯ä»¶å·¥ç¨‹å‘¨æœŸ:</p>
<ul>
<li>æœé›†éœ€æ±‚, å½¢æˆä»¥ç”¨æˆ·ä¸ºè§†è§’çš„<code>ç”¨æˆ·æ•…äº‹</code></li>
<li>æ„å»ºç®€å•æœ‰æ•ˆçš„æ¨¡å—</li>
<li>æ„æˆæ•´ä¸ªç¨‹åº, æŒ‰ç…§æ–°éœ€æ±‚ä¿®æ”¹</li>
</ul>
</li>
<li>
<p><code>PEP8</code>ä»£ç è§„èŒƒ:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://pep8.org/">PEP 8: The Style Guide for Python Code (pep8.org)</a></li>
<li><a target="_blank" rel="noopener" href="https://www.python.org/dev/peps/pep-0008/">PEP 8 â€“ Style Guide for Python Code | Python.org</a></li>
</ul>
</li>
<li>
<p><code>pylint</code>:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://pylint.org/">Pylint - code analysis for Python | www.pylint.org</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c0bd637f706d">ã€pythonã€‘pylintåœ¨é¡¹ç›®ä¸­çš„ä½¿ç”¨ - ç®€ä¹¦ (jianshu.com)</a></li>
</ul>
</li>
<li>
<p><code>ä»£ç ç®¡ç†å·¥å…·</code>:</p>
<ul>
<li><code>Git</code>:  <a href="/git-shou-ce/" title="Gitæ‰‹å†Œ">Gitæ‰‹å†Œ</a></li>
</ul>
</li>
<li>
<p>pythonä»£ç è‡ªåŠ¨æ‰“åŒ…å·¥å…·</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/goldsunshine/p/8872623.html">pythonæ‰“åŒ…å·¥å…·distutilsã€setuptoolsåˆ†æ - é‡‘è‰²æ—­å…‰ - åšå®¢å›­ (cnblogs.com)</a></li>
<li><code>distutils</code>: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/distutils.html?highlight=distutils#module-distutils">27.1. distutils â€” æ„å»ºå’Œå®‰è£… Python æ¨¡å— â€” Python 2.7.18 æ–‡æ¡£</a></li>
<li><code>setuptools</code>:<a target="_blank" rel="noopener" href="https://setuptools.readthedocs.io/en/latest/">Documentation â€” setuptools 50.3.1 documentation</a></li>
</ul>
</li>
<li>
<p>pythonä»£ç ç”Ÿæˆå¯æ‰§è¡Œæ–‡ä»¶</p>
<ul>
<li><code>py2exe</code>: <a target="_blank" rel="noopener" href="http://www.py2exe.org/">FrontPage - py2exe.org</a></li>
<li><code>pyinstaller</code>: <a target="_blank" rel="noopener" href="http://www.pyinstaller.org/">PyInstaller bundles Python applications</a></li>
</ul>
</li>
<li>
<p>æ•æ·å¼€å‘: <a target="_blank" rel="noopener" href="http://agilemanifesto.org/iso/zhchs/manifesto.html">æ•æ·è½¯ä»¶å¼€å‘å®£è¨€ (agilemanifesto.org)</a></p>
<blockquote>
<p><strong>æ•æ·è½¯ä»¶å¼€å‘å®£è¨€</strong></p>
<p>æˆ‘ä»¬ä¸€ç›´åœ¨å®è·µä¸­æ¢å¯»æ›´å¥½çš„è½¯ä»¶å¼€å‘æ–¹æ³•ï¼Œ<br>
èº«ä½“åŠ›è¡Œçš„åŒæ—¶ä¹Ÿå¸®åŠ©ä»–äººã€‚ç”±æ­¤æˆ‘ä»¬å»ºç«‹äº†å¦‚ä¸‹ä»·å€¼è§‚ï¼š</p>
<p>ä¸ªä½“å’Œäº’åŠ¨ é«˜äº æµç¨‹å’Œå·¥å…·<br>
å·¥ä½œçš„è½¯ä»¶ é«˜äº è¯¦å°½çš„æ–‡æ¡£<br>
å®¢æˆ·åˆä½œ é«˜äº åˆåŒè°ˆåˆ¤<br>
å“åº”å˜åŒ– é«˜äº éµå¾ªè®¡åˆ’</p>
<p>ä¹Ÿå°±æ˜¯è¯´ï¼Œå°½ç®¡å³é¡¹æœ‰å…¶ä»·å€¼ï¼Œ<br>
æˆ‘ä»¬æ›´é‡è§†å·¦é¡¹çš„ä»·å€¼ã€‚</p>
<table>
<thead>
<tr>
<th>Kent Beck Mike Beedle Arie van Bennekum Alistair Cockburn Ward Cunningham Martin Fowler</th>
<th>James Grenning Jim Highsmith Andrew Hunt Ron Jeffries Jon Kern Brian Marick</th>
<th>Robert C. Martin Steve Mellor Ken Schwaber Jeff Sutherland Dave Thomas</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>è‘—ä½œæƒä¸ºä¸Šè¿°ä½œè€…æ‰€æœ‰ï¼Œ2001å¹´</p>
</blockquote>
<ul>
<li>é€‚äºç§‘ç ”ç•Œçš„å·¥ä¸šåŒ–å®ä¾‹: <a target="_blank" rel="noopener" href="https://www.perforce.com/products/helix-alm">Helix ALM | ALM Tools | Perforce</a></li>
<li>å¯¹ç”Ÿç‰©ä¿¡æ¯å­¦ç ”ç©¶çš„å„ç§è½¯ä»¶å·¥ç¨‹æŠ€æœ¯(å¦‚è‡ªåŠ¨åŒ–æµ‹è¯•/ä»£ç è¯„å®¡/ç”¨æˆ·æ•…äº‹)éƒ½æœ‰æè¿°çš„å·¥å…·ç®±:  <a target="_blank" rel="noopener" href="https://www.researchgate.net/publication/51535066_A_toolbox_for_developing_bioinformatics_software">A toolbox for developing bioinformatics software | Request PDF (researchgate.net)</a></li>
</ul>
</li>
</ul>
<h2><a href="#di-si-bu-fen-shu-ju-ke-shi-hua" class="header-anchor"> âš“ï¸</a><span id="di-si-bu-fen-shu-ju-ke-shi-hua">ç¬¬å››éƒ¨åˆ† æ•°æ®å¯è§†åŒ–</span></h2>
<h3><a href="#di-16-zhang-chuang-jian-ke-xue-tu-biao" class="header-anchor"> âš“ï¸</a><span id="di-16-zhang-chuang-jian-ke-xue-tu-biao">ç¬¬16ç«  åˆ›å»ºç§‘å­¦å›¾è¡¨</span></h3>
<h4><a href="#16-2-an-li-he-tang-ti-de-he-gan-suan-pin-lu" class="header-anchor"> âš“ï¸</a><span id="16-2-an-li-he-tang-ti-de-he-gan-suan-pin-lu">16.2 æ¡ˆä¾‹ï¼š æ ¸ç³–ä½“çš„æ ¸è‹·é…¸é¢‘ç‡</span></h4>
<blockquote>
<p>2009å¹´çš„è¯ºè´å°”åŒ–å­¦å¥–æˆäºˆäº†Venkatraman RamaKrishnan, Thomas Steitz å’Œ Ada Yonath, ä»¥è¡¨å½°ä»–ä»¬å¯¹æ ¸ç³–ä½“ç»“æ„å’ŒåŠŸèƒ½çš„ç ”ç©¶. æ ¸ç³–ä½“æ˜¯å·²çŸ¥çš„æœ€å¤§çš„åå­æœºå™¨ä¹‹ä¸€, ç”±ä¸‰ç§RNAç»„åˆ†(åŸæ ¸ç”Ÿç‰©ä¸­æ˜¯23S, 16S å’Œ 5S rRNA) å’Œè®¸å¤šçš„è›‹ç™½è´¨ç»„æˆ. RNAçš„æ„æˆåŒ…æ‹¬å››ç§åŸºæœ¬çš„æ ¸ç³–æ ¸è‹·é…¸å’Œä¸€äº›èµ·å¾®è°ƒæ ¸ç³–ä½“åŠŸèƒ½ä½œç”¨çš„ä¿®æ˜¯æ ¸ç³–ä½“</p>
</blockquote>
<ul>
<li><code>matplotlib</code>:
<ul>
<li>å®˜æ–¹æ•™ç¨‹: <a target="_blank" rel="noopener" href="https://matplotlib.org/">Matplotlib: Python plotting â€” Matplotlib 3.3.2 documentation</a></li>
<li>å®˜æ–¹æ¨¡æ¿: <a target="_blank" rel="noopener" href="https://matplotlib.org/gallery/index.html#gallery">Gallery â€” Matplotlib 3.3.2 documentation</a></li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> pylab <span class="keyword">import</span> figure, plot, savefig</span><br><span class="line"></span><br><span class="line">xdata = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>]</span><br><span class="line">ydata = [<span class="number">1.25</span>, <span class="number">2.5</span>, <span class="number">5.0</span>, <span class="number">10.0</span>]</span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line">plot(xdata, ydata)</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'figure1.png'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># -----</span></span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line"></span><br><span class="line">xdata = [<span class="number">0.1</span> * i <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">100</span>)]</span><br><span class="line">ydata = [math.sin(j) <span class="keyword">for</span> j <span class="keyword">in</span> xdata]</span><br><span class="line"></span><br><span class="line">plot(xdata, ydata, <span class="string">'kd'</span>, linewidth = <span class="number">1</span>)</span><br><span class="line">text(<span class="number">4.8</span>, <span class="number">0</span>, <span class="string">"$y = sin(x)$"</span>, horizontalalignment = <span class="string">'center'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">axis([<span class="number">0</span>, <span class="number">3</span> * math.pi, -<span class="number">1.2</span>, <span class="number">1.2</span>])</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'sinfunc.png'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ---</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> pylab <span class="keyword">import</span> figure, title, xlabel, ylabel, hist, axis, grid, savefig</span><br><span class="line"></span><br><span class="line">data = [<span class="number">1</span>, <span class="number">1</span>, <span class="number">9</span>, <span class="number">1</span>, <span class="number">3</span>, <span class="number">5</span>, <span class="number">8</span>, <span class="number">2</span>, <span class="number">1</span>, <span class="number">5</span>, <span class="number">11</span>, <span class="number">8</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">2</span>, <span class="number">5</span>]</span><br><span class="line">n_bins = <span class="number">5</span></span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line">num, bins, patches = hist(data, n_bins, normed = <span class="number">1.0</span>, histtype = <span class="string">'bar'</span>, facecolor = <span class="string">'green'</span>, alpha = <span class="number">0.75</span>)</span><br><span class="line"></span><br><span class="line">title(<span class="string">'Histogram'</span>)</span><br><span class="line">xlabel(<span class="string">'value'</span>)</span><br><span class="line">ylabel(<span class="string">'frequency'</span>)</span><br><span class="line">axis()</span><br><span class="line">grid(<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'histogram.png'</span>)</span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#di-17-zhang-shi-yong-pymol-chuang-jian-fen-zi-tu-xiang" class="header-anchor"> âš“ï¸</a><span id="di-17-zhang-shi-yong-pymol-chuang-jian-fen-zi-tu-xiang">ç¬¬17ç«  ä½¿ç”¨PyMOLåˆ›å»ºåˆ†å­å›¾åƒ</span></h3>
<h4><a href="#17-2-shi-li-xin-zhi" class="header-anchor"> âš“ï¸</a><span id="17-2-shi-li-xin-zhi">17.2 ç¤ºä¾‹ï¼š é”ŒæŒ‡</span></h4>
<ul>
<li>
<p>é”ŒæŒ‡:</p>
<blockquote>
<p>é”ŒæŒ‡(è‹±è¯­: Zinc finger),åˆç§°é”Œæ‰‹æŒ‡, æ˜¯ä¸€ç§å°çš„è›‹ç™½è´¨ç»“æ„æ¨¡ä½“, å…¶ç‰¹å¾åœ¨äºé…åˆä¸€ä¸ªæˆ–å¤šä¸ªé”Œç¦»å­(Zn<sup>2+</sup>)ä»¥ç¨³å®šæŠ˜å . æœ€åˆç”¨æ¥æè¿°éæ´²çˆªèŸ¾åµæ¯ç»†èƒè½¬å½•å› å­IIIAå‡è¯´ç»“æ„çš„æ‰‹æŒ‡çŠ¶å¤–è§‚</p>
<p>ä¸€ä¸ªé”ŒæŒ‡åŒ…å«ç”±ä¸¤ä¸ªèºæ—‹, å®ƒä»¬é€šè¿‡ä¸€ä¸ªé”ŒåŸå­å½¢æˆç‰¹æ®Šçš„æ„è±¡, å¯¼è‡´ä¸€ä¸ªèºæ—‹ä¸DNAçš„å¤§æ²ŸåŒ¹é…</p>
<p>é”ŒæŒ‡åç§°ç°åœ¨å·²ç»åŒ…å«äº†å„ç§ä¸åŒçš„è›‹ç™½è´¨ç»“æ„</p>
<p>@<a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/blog-635619-765781.html">ç§‘å­¦ç½‘â€”ä»€ä¹ˆæ˜¯é”ŒæŒ‡è›‹ç™½ï¼Ÿ - å€ªå¸…çš„åšæ–‡ (sciencenet.cn)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-hans/%E9%94%8C%E6%8C%87">é”ŒæŒ‡ - ç»´åŸºç™¾ç§‘ï¼Œè‡ªç”±çš„ç™¾ç§‘å…¨ä¹¦ (wikipedia.org)</a></p>
</blockquote>
</li>
<li>
<p><code>PyMol</code>: <a target="_blank" rel="noopener" href="https://pymol.org/2/">PyMOL | pymol.org</a></p>
<blockquote>
<p>PyMOL is a user-sponsored molecular visualization system on an <a target="_blank" rel="noopener" href="https://pymol.org/2/#opensource">open-source foundation</a>, maintained and distributed by <a target="_blank" rel="noopener" href="https://www.schrodinger.com/">SchrÃ¶dinger</a>.</p>
</blockquote>
<ul>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/58803491">å…è´¹å¼€æºç‰ˆPyMOLå®‰è£… - çŸ¥ä¹ (zhihu.com)</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/schrodinger/pymol-open-source">schrodinger/pymol-open-source: Open-source foundation of the user-sponsored PyMOL molecular visualization system. (github.com)</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-18-zhang-chu-li-tu-xiang" class="header-anchor"> âš“ï¸</a><span id="di-18-zhang-chu-li-tu-xiang">ç¬¬18ç«  å¤„ç†å›¾åƒ</span></h3>
<h4><a href="#18-2-an-li-hua-yi-ge-zhi-li" class="header-anchor"> âš“ï¸</a><span id="18-2-an-li-hua-yi-ge-zhi-li">18.2 æ¡ˆä¾‹ï¼š ç”»ä¸€ä¸ªè´¨ç²’</span></h4>
<blockquote>
<p>1977å¹´ Francisco Bolivar å’Œ Raymond Rodriguez æ„å»ºäº†æœ€æ—©çš„äººå·¥è´¨ç²’. å®ƒåŒ…å«äº†4361ä¸ªDNAç¢±åŸºå¯¹, ä¸€ä¸ªå¤åˆ¶èµ·ç‚¹, ä¸€ä¸ªæ°¨è‹„é’éœ‰ç´ æŠ—æ€§åŸºå› å’Œä¸€ä¸ªå››ç¯ç´ æŠ—æ€§åŸºå› . è¯¥è´¨ç²’ç”±è®¸å¤šé™åˆ¶æ€§ä½ç‚¹, ä»è€Œæˆä¸ºæ„å»ºé—ä¼ è½½ä½“çš„åŸºç¡€.</p>
</blockquote>
<ul>
<li>
<p><code>PIL(Python Imaging Library)</code>å›¾å½¢åº“:</p>
<blockquote>
<p>The <strong>Python Imaging Library (PIL)</strong> adds image processing capabilities to your Python interpreter. This library supports many file formats, and provides powerful image processing and graphics capabilities.</p>
<p>PIL ä»…æ”¯æŒPython2</p>
</blockquote>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="http://pythonware.com/products/pil/">Python Imaging Library (PIL) (pythonware.com)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://www.liaoxuefeng.com/wiki/897692888725344/966759628285152">PIL - å»–é›ªå³°çš„å®˜æ–¹ç½‘ç«™ (liaoxuefeng.com)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://pillow.readthedocs.io/en/stable/">Pillow â€” Pillow (PIL Fork) 8.0.0 documentation</a></p>
<blockquote>
<p>Pillow is the friendly PIL fork by <a target="_blank" rel="noopener" href="https://github.com/python-pillow/Pillow/graphs/contributors">Alex Clark and Contributors</a>. PIL is the Python Imaging Library by Fredrik Lundh and Contributors.</p>
<p>Pillow æ”¯æŒPython3</p>
</blockquote>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://github.com/python-pillow/Pillow">python-pillow/Pillow: The friendly PIL fork (Python Imaging Library) (github.com)</a></p>
</li>
</ul>
</li>
<li>
<p>è›‹ç™½è´¨ç»“æ„åŸŸæ¶æ„:</p>
<ul>
<li>
<p><code>InterPro</code>æ•°æ®åº“: <a target="_blank" rel="noopener" href="http://www.ebi.ac.uk/interpro/">InterPro (ebi.ac.uk)</a></p>
<blockquote>
<p>InterPro provides functional analysis of proteins by classifying them into families and predicting domains and important sites. To classify proteins in this way, InterPro uses predictive models, known as signatures, provided by several different databases (referred to as member databases) that make up the InterPro consortium. We combine protein signatures from these member databases into a single searchable resource, capitalising on their individual strengths to produce a powerful integrated database and diagnostic tool.</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<h2><a href="#di-wu-bu-fen-biopython" class="header-anchor"> âš“ï¸</a><span id="di-wu-bu-fen-biopython">ç¬¬äº”éƒ¨åˆ† Biopython</span></h2>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://biopython.org/">Biopython</a> æ˜¯ä¸€ä¸ªè®¡ç®—åˆ†å­ç”Ÿç‰©å­¦æ¨¡å—çš„é›†åˆ, ç”¨å®ƒå¯ä»¥å®ç°è®¸å¤šç”Ÿç‰©ä¿¡æ¯å­¦é¡¹ç›®ä¸­çš„åŸºæœ¬ä»»åŠ¡</p>
<blockquote>
<p>Biopython is a set of freely available tools for biological computation written in <a target="_blank" rel="noopener" href="http://www.python.org/">Python</a> by an international team of developers.</p>
</blockquote>
<ul>
<li>
<p>è§£æç”Ÿç‰©æ•°æ®æ ¼å¼</p>
</li>
<li>
<p>ä»èµ„æºåº“ä¸‹è½½èµ„æº</p>
</li>
<li>
<p>è¿è¡Œå¸¸ç”¨çš„ç”Ÿç‰©ä¿¡æ¯å­¦ç®—æ³•</p>
</li>
<li>
<p>è¿è¡ŒBiopythonå®ç°çš„ç®—æ³•, è¿›è¡Œèšç±», æœºå™¨å­¦ä¹ , æ•°æ®åˆ†æå’Œå¯è§†åŒ–</p>
</li>
<li>
<p>å„ç§æ–‡æ¡£ç´¢å¼•: <a target="_blank" rel="noopener" href="https://biopython.org/wiki/Documentation">Documentation Â· Biopython</a></p>
</li>
<li>
<p>ä¸Šæ‰‹æŒ‡å—: <a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">Biopython Tutorial and Cookbook</a></p>
</li>
</ul>
</li>
</ul>
<h3><a href="#di-19-zhang-shi-yong-xu-lie-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-19-zhang-shi-yong-xu-lie-shu-ju">ç¬¬19ç«  ä½¿ç”¨åºåˆ—æ•°æ®</span></h3>
<h4><a href="#19-2-an-li-ru-he-jiang-yi-tiao-dna-bian-ma-xu-lie-fan-yi-cheng-dui-ying-de-dan-bai-zhi-xu-lie-bing-ba-ta-xie-ru-fasta-wen-jian" class="header-anchor"> âš“ï¸</a><span id="19-2-an-li-ru-he-jiang-yi-tiao-dna-bian-ma-xu-lie-fan-yi-cheng-dui-ying-de-dan-bai-zhi-xu-lie-bing-ba-ta-xie-ru-fasta-wen-jian">19.2 æ¡ˆä¾‹ï¼š å¦‚ä½•å°†ä¸€æ¡DNAç¼–ç åºåˆ—ç¿»è¯‘æˆå¯¹åº”çš„è›‹ç™½è´¨åºåˆ—å¹¶æŠŠå®ƒå†™å…¥FASTAæ–‡ä»¶</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec17">Chapter 3 Sequence objects (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio.Seq <span class="keyword">import</span> Seq</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq = Seq(<span class="string">"AGTACACTGGT"</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq</span><br><span class="line">Seq(<span class="string">'AGTACACTGGT'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">print</span>(my_seq)</span><br><span class="line">AGTACACTGGT</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq</span><br><span class="line">Seq(<span class="string">'AGTACACTGGT'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq.complement()</span><br><span class="line">Seq(<span class="string">'TCATGTGACCA'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq.reverse_complement()</span><br><span class="line">Seq(<span class="string">'ACCAGTGTACT'</span>)</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> seq_record <span class="keyword">in</span> SeqIO.parse(<span class="string">"ls_orchid.fasta"</span>, <span class="string">"fasta"</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(seq_record.<span class="built_in">id</span>)</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="built_in">repr</span>(seq_record.seq))</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="built_in">len</span>(seq_record))</span><br><span class="line">gi|<span class="number">2765658</span>|emb|Z78533<span class="number">.1</span>|CIZ78533</span><br><span class="line">Seq(<span class="string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span>)</span><br><span class="line"><span class="number">740</span></span><br><span class="line">...</span><br><span class="line">gi|<span class="number">2765564</span>|emb|Z78439<span class="number">.1</span>|PBZ78439</span><br><span class="line">Seq(<span class="string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span>)</span><br><span class="line"><span class="number">592</span></span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio.Seq <span class="keyword">import</span> Seq</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq = Seq(<span class="string">"GATCG"</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> index, letter <span class="keyword">in</span> <span class="built_in">enumerate</span>(my_seq):</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="string">"%i %s"</span> % (index, letter))</span><br><span class="line"><span class="number">0</span> G</span><br><span class="line"><span class="number">1</span> A</span><br><span class="line"><span class="number">2</span> T</span><br><span class="line"><span class="number">3</span> C</span><br><span class="line"><span class="number">4</span> G</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">print</span>(<span class="built_in">len</span>(my_seq))</span><br><span class="line"><span class="number">5</span></span><br><span class="line"></span><br><span class="line">...</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>IUPACå­—ç¬¦é›†</code>:</p>
<blockquote>
<p>é¢„ç¼–è¯‘å¥½çš„å­—ç¬¦é›†, è¦†ç›–äº†æ‰€æœ‰ç”Ÿç‰©åºåˆ—. åŒ…æ‹¬ IUPACUnamiguousDNA(åŸºæœ¬çš„ATCGå­—æ¯), IUPACAmbiguousDNA(åŒ…å«äºŒä¹‰å­—æ¯), ExtendedIUPACDNA(åŒ…å«ä¿®é¥°çš„ç¢±åŸº), IUPACUnamiguousRNA,  IUPACAmbiguousRNA, ExtendedIUPACRNA, IUPACProtein(IPUACæ ‡å‡†æ°¨åŸºé…¸)å’ŒExtendedIUPACProtein(åŒ…æ‹¬ç¡’ä»£åŠèƒ±æ°¨é…¸, Xç­‰)</p>
</blockquote>
</li>
<li>
<p>è¿­ä»£å™¨:</p>
<blockquote>
<p>ä¸€ä¸ªè¿­ä»£å™¨æ˜¯äº§ç”Ÿä¸€ç³»åˆ—çš„é¡¹(å¦‚SeqRecordå¯¹è±¡)çš„ä¸€ç§æ•°æ®ç»“æ„. å®ƒå¯ä»¥åƒåˆ—è¡¨ä¸€æ ·ä½¿ç”¨åœ¨å¾ªç¯ä¸­, ä½†æŠ€æœ¯ä¸Šå®ƒä¸æ˜¯ä¸€ä¸ªåˆ—è¡¨. ä¸€ä¸ªè¿­ä»£å™¨æ²¡æœ‰é•¿åº¦, ä¹Ÿä¸èƒ½è¢«ç´¢å¼•æˆ–è€…æ˜¯åˆ‡ç‰‡. ç”¨æˆ·åªèƒ½å‘å®ƒç”³è¯·ä¸‹ä¸€ä¸ªå…ƒç´ , å½“ä½ è¿™æ ·åšæ—¶, è¿­ä»£å™¨å¯»æ‰¾æ˜¯å¦å­˜åœ¨æ›´å¤šçš„å¯ç”¨çš„è®°å½•. è¿™ç§æ–¹å¼ä¸‹, è¿­ä»£å™¨æ— éœ€æŠŠæ‰€æœ‰çš„è®°å½•ä¸€è‡´æ”¾åœ¨å†…å­˜ä¸­.</p>
</blockquote>
</li>
<li>
<p>NCBIæ ¸é…¸ç¿»è¯‘ç¼–ç æ ‡å‡†: <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">The Genetic Codes</a></p>
</li>
</ul>
<h3><a href="#di-20-zhang-cong-wang-luo-zi-yuan-zhong-jian-suo-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-20-zhang-cong-wang-luo-zi-yuan-zhong-jian-suo-shu-ju">ç¬¬20ç«  ä»ç½‘ç»œèµ„æºä¸­æ£€ç´¢æ•°æ®</span></h3>
<h4><a href="#20-2-an-li-zai-pubmed-zhong-yong-guan-jian-ci-sou-suo-wen-xian-xia-zai-bing-jie-xi-dui-ying-de-ji-lu" class="header-anchor"> âš“ï¸</a><span id="20-2-an-li-zai-pubmed-zhong-yong-guan-jian-ci-sou-suo-wen-xian-xia-zai-bing-jie-xi-dui-ying-de-ji-lu">20.2 æ¡ˆä¾‹ï¼š åœ¨PubMedä¸­ç”¨å…³é”®è¯æœç´¢æ–‡çŒ®ï¼Œ ä¸‹è½½å¹¶è§£æå¯¹åº”çš„è®°å½•</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec139">Chapter 9 Accessing NCBIâ€™s Entrez databases (biopython.org)</a></p>
<h3><a href="#di-21-zhang-shi-yong-san-wei-jie-gou-shu-ju" class="header-anchor"> âš“ï¸</a><span id="di-21-zhang-shi-yong-san-wei-jie-gou-shu-ju">ç¬¬21ç«  ä½¿ç”¨ä¸‰ç»´ç»“æ„æ•°æ®</span></h3>
<h4><a href="#21-2-an-li-cong-pdb-wen-jian-zhong-ti-qu-yuan-zi-ming-ji-qi-san-wei-zuo-biao" class="header-anchor"> âš“ï¸</a><span id="21-2-an-li-cong-pdb-wen-jian-zhong-ti-qu-yuan-zi-ming-ji-qi-san-wei-zuo-biao">21.2 æ¡ˆä¾‹ï¼š ä»PDBæ–‡ä»¶ä¸­æå–åŸå­ååŠå…¶ä¸‰ç»´åæ ‡</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec178">Chapter 11 Going 3D: The PDB module (biopython.org)</a></p>
<h2><a href="#di-liu-bu-fen-bian-cheng-mi-ji" class="header-anchor"> âš“ï¸</a><span id="di-liu-bu-fen-bian-cheng-mi-ji">ç¬¬å…­éƒ¨åˆ† ç¼–ç¨‹ç§˜ç¬ˆ</span></h2>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ01: PyCogentåº“</li>
</ul>
<p><code>PyCogent</code>: <a target="_blank" rel="noopener" href="https://github.com/pycogent/pycogent/">pycogent/pycogent: PyCogent: Official repository for software and unit tests (github.com)</a></p>
<blockquote>
<p>PyCogent includes connectors to remote databases, built-in generalized probabilistic techniques for working with biological sequences, and controllers for 3rd party applications.</p>
<p>é¡¹ç›®ç–‘ä¼¼ä¸å†ç»´æŠ¤</p>
</blockquote>
<p>å‚è€ƒæ–‡çŒ®: <a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2007-8-8-r171">PyCogent: a toolkit for making sense from sequence | Genome Biology | Full Text (biomedcentral.com)</a></p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ02: åå‘äº’è¡¥å’ŒéšæœºåŒ–åºåˆ—</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/random.html?highlight=random#module-random">9.6. random â€” ç”Ÿæˆä¼ªéšæœºæ•° â€” Python 2.7.18 æ–‡æ¡£</a></p>
<p>ä¸ºä»€ä¹ˆè¦éšæœºåŒ–åºåˆ—:</p>
<blockquote>
<p>æ¯ä¸ªåºåˆ—æ¨¡å¼å¯èƒ½åœ¨åºåˆ—ä¸­éšæœºå‡ºç°, åœ¨è›‹ç™½è´¨å’Œæ ¸é…¸ä¸­å¯»æ‰¾åŒ¹é…çš„åŠŸèƒ½æ¨¡ä½“ä¸èƒ½ä¿è¯è¯¥åŒ¹é…ä¸€å®šæœ‰ç”Ÿç‰©æ„ä¹‰. ä¸ºäº†è¯„ä¼°ä¸€ä¸ªæ¨¡ä½“å­˜åœ¨çš„ç”Ÿç‰©å­¦æ˜¾è‘—æ€§, éœ€è¦æ£€æŸ¥çš„åºåˆ—ä¸€å•Šä¸èƒ½è¦å’Œéšæœºé›†åˆè¿›è¡Œæ¯”è¾ƒ. åœ¨ç»™å®šçš„åºåˆ—é›†åˆä¸­ç›¸å¯¹äºéšæœºåºåˆ—é›†åˆæœ‰æ›´æ˜¾è‘—è¡¨è¾¾çš„åºåˆ—æ¨¡ä½“, æ›´å€¾å‘äºç¼–ç ä¸€ä¸ªæœ‰åŠŸèƒ½çš„æ€§è´¨(æ›´æœ‰ç”Ÿç‰©å­¦æ„ä¹‰). ä¸€ä¸ªå¥½çš„éšæœºåºåˆ—æ˜¯ç¼ºä¹ç”Ÿç‰©å­¦æ„ä¹‰çš„, ä½†æ˜¯å´ä¸ç”Ÿç‰©å¯¹è±¡æœ‰ç›¸åŒçš„æ°¨åŸºé…¸/æ ¸è‹·é…¸ç»„åˆ†, å¯ä»¥é€šè¿‡å¯¹åŸåºåˆ—åå‘,æ´—ç‰Œ,éšæœºåŒ–ç­‰æ–¹å¼åˆ›å»º.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">seq = <span class="string">'ABCDEFGHIJKLMNOPQRSTUVWXYZ'</span></span><br><span class="line"></span><br><span class="line">seq_list = <span class="built_in">list</span>(seq)    </span><br><span class="line">seq_list.reverse()</span><br><span class="line">rev_seq = <span class="string">''</span>.join(seq_list)</span><br><span class="line"><span class="built_in">print</span>(rev_seq)   <span class="comment"># åˆ—è¡¨åŒ–ååè½¬å†æ‹¼æ¥å›å­—ç¬¦ä¸²</span></span><br><span class="line"></span><br><span class="line">rev_seq = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> s <span class="keyword">in</span> <span class="built_in">reversed</span>(seq):</span><br><span class="line">    rev_seq = rev_seq + s</span><br><span class="line"><span class="built_in">print</span>(rev_seq)  <span class="comment"># ä½¿ç”¨reversed()è¿­ä»£å™¨å¾ªç¯åå‘æ‹¼æ¥å­—ç¬¦ä¸²</span></span><br><span class="line"></span><br><span class="line">rev_seq = seq[::-<span class="number">1</span>] <span class="comment"># å­—ç¬¦ä¸²åˆ‡ç‰‡[start:end:step], è®¾æ­¥é•¿ä¸º-1, å¾—åå‘</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> random</span><br><span class="line"><span class="comment"># è¿”å›ç”±seqå†…å…ƒç´ ç»„æˆçš„é•¿ä¸ºlen(seq)çš„åˆ—è¡¨</span></span><br><span class="line">ran_seq = random.sample(seq, <span class="built_in">len</span>(seq)) </span><br><span class="line"><span class="comment"># è½¬å­—ç¬¦ä¸²</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join(random.sample(seq, <span class="built_in">len</span>(seq)))</span><br><span class="line"><span class="comment"># æ§åˆ¶é•¿åº¦</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join(random.sample(seq, <span class="built_in">len</span>(seq)-<span class="number">10</span>))</span><br><span class="line"><span class="comment"># ä½¿ç”¨random.choiceæ–¹æ³•åœ¨seqä¸­æ¯æ¬¡éšæœºè¿”å›ä¸€ä¸ªå…ƒç´ , forå¾ªç¯æ§åˆ¶æ¬¡æ•°</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join([random.choice(seq) <span class="keyword">for</span> x <span class="keyword">in</span> <span class="built_in">range</span>(<span class="built_in">len</span>(seq))])</span><br><span class="line"></span><br><span class="line">data = <span class="built_in">list</span>(seq)</span><br><span class="line">random.shuffle(data) <span class="comment"># åºåˆ—æ´—ç‰Œ</span></span><br><span class="line">shuffled_seq = data</span><br><span class="line">shuffled_seq = <span class="string">''</span>.join(data)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ03: ç”¨æ¦‚ç‡åˆ›å»ºéšæœºåºåˆ—</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> random</span><br><span class="line"></span><br><span class="line">nucleotides = <span class="built_in">list</span>(<span class="string">'ACGT'</span>)</span><br><span class="line">dna = <span class="string">''</span></span><br><span class="line"><span class="keyword">while</span> <span class="built_in">len</span>(dna) &lt; <span class="number">100</span>:</span><br><span class="line">    dna += random.choice(nucleotides) <span class="comment"># éšæœºç”Ÿæˆ100ä¸ªç¢±åŸºç»„æˆçš„DNAåºåˆ—</span></span><br><span class="line"></span><br><span class="line">nucleotides = <span class="built_in">list</span>(<span class="string">'ACGT'</span>)</span><br><span class="line">probs = {<span class="string">'A'</span>: <span class="number">0.3</span>, <span class="string">'C'</span>: <span class="number">0.2</span>, <span class="string">'G'</span>: <span class="number">0.2</span>, <span class="string">'T'</span>: <span class="number">0.3</span>} <span class="comment"># æ€»æ¦‚ç‡ä¸º1</span></span><br><span class="line"><span class="keyword">assert</span> <span class="built_in">sum</span>(probs.values()) == <span class="number">1.0</span> </span><br><span class="line"><span class="comment"># ä¸Šè¿°assertè¯­å¥ç­‰ä»·äºä¸‹é¢çš„ä¸¤è¡Œ</span></span><br><span class="line"><span class="keyword">if</span> <span class="built_in">sum</span>(probs.values()) != <span class="number">1.0</span>:</span><br><span class="line">  <span class="keyword">raise</span> Exception(<span class="string">'Sum of probabilites is not 1.0!'</span>)</span><br><span class="line"></span><br><span class="line">dna = <span class="string">''</span></span><br><span class="line"><span class="keyword">while</span> <span class="built_in">len</span>(dna) &lt; <span class="number">100</span>:</span><br><span class="line">    nuc = random.choice(nucleotides)</span><br><span class="line">    dice = random.random()</span><br><span class="line">    <span class="keyword">if</span> dice &lt; probs[nuc]:</span><br><span class="line">        dna += nuc</span><br><span class="line"><span class="built_in">print</span> dna</span><br></pre></td></tr></tbody></table></figure>
<p><code>assert</code> è¯­å¥: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/simple_stmts.html?highlight=assert#assert">assert (python.org)</a></p>
<p>assert è¯­å¥æ˜¯åœ¨ç¨‹åºä¸­æ’å…¥è°ƒè¯•æ€§æ–­è¨€çš„ç®€ä¾¿æ–¹å¼:</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">assert_stmt ::=  "assert" expression ["," expression]</span><br></pre></td></tr></tbody></table></figure>
<p>ç®€å•å½¢å¼ <code>assert expression</code> ç­‰ä»·äº</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">if __debug__:</span><br><span class="line">    if not expression: raise AssertionError</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ04: ç”¨Biopythonè§£æå¤šåºåˆ—è”é…</li>
</ul>
<p>å¤šåºåˆ—è”é…(MSA):</p>
<p><code>Pfam</code>æ•°æ®åº“: <a target="_blank" rel="noopener" href="https://pfam.xfam.org/">Pfam: Home page</a></p>
<p>å¸¸ç”¨æ ¼å¼ä¸º<code>.sth (Stockholm æ ¼å¼)</code></p>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/10bfcdf06693">biostar handbook(å…­)| åºåˆ—è”é… - ç®€ä¹¦ (jianshu.com)</a></p>
<p>ä¾‹å­: <a target="_blank" rel="noopener" href="https://pfam.xfam.org/family/PF00042#tabview=tab0">Pfam: Family: Globin (PF00042) (xfam.org)</a></p>
<p>é€šè¿‡Biopythonçš„Bio.AlignIOæ¨¡å—è§£æå¤šè”é…:</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec79">Chapter 6 Multiple Sequence Alignment objects (biopython.org)</a></p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ05: ä»å¤šåºåˆ—è”é…ä¸­è®¡ç®—å…±æœ‰åºåˆ—</li>
</ul>
<p>å…±æœ‰åºåˆ—(Consensus Sequence): å¤šè”é…ä¸­æ¯åˆ—ä¸­é¢‘ç‡æœ€é«˜çš„æ®‹åŸº(æœ€ä¿å®ˆçš„)</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">seqs = [</span><br><span class="line">     <span class="string">'ATCCAGCT'</span>,</span><br><span class="line">     <span class="string">'GGGCAACT'</span>,</span><br><span class="line">     <span class="string">'ATGGATCT'</span>,</span><br><span class="line">     <span class="string">'AAGCAACC'</span>,</span><br><span class="line">     <span class="string">'ATGCCATT'</span>,</span><br><span class="line">     <span class="string">'ATGGCACT'</span>]</span><br><span class="line"></span><br><span class="line">n = <span class="built_in">len</span>(seqs[<span class="number">0</span>])</span><br><span class="line">profile = {<span class="string">'A'</span>:[<span class="number">0</span>]*n, <span class="string">'C'</span>:[<span class="number">0</span>]*n, <span class="string">'G'</span>:[<span class="number">0</span>]*n, <span class="string">'T'</span>:[<span class="number">0</span>]*n}</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> seq <span class="keyword">in</span> seqs:</span><br><span class="line">    <span class="keyword">for</span> i, char <span class="keyword">in</span> <span class="built_in">enumerate</span>(seq):</span><br><span class="line">        profile[char][i] += <span class="number">1</span></span><br><span class="line">consensus = <span class="string">''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(n):</span><br><span class="line">    col = [(profile[nt][i], nt) <span class="keyword">for</span> nt <span class="keyword">in</span> <span class="string">'ATCG'</span>]</span><br><span class="line">    consensus += <span class="built_in">max</span>(col)[<span class="number">1</span>]</span><br><span class="line">    </span><br><span class="line"><span class="built_in">print</span> consensus</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ06: è®¡ç®—ç³»ç»Ÿå‘ç”Ÿæ ‘çš„èŠ‚ç‚¹é—´çš„è·ç¦»</li>
</ul>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec249">Chapter 13 Phylogenetics with Bio.Phylo</a></p>
<p><code>Newick</code>æ ¼å¼æ˜¯è¡¨ç¤ºç³»ç»Ÿå‘ç”Ÿæ ‘çš„ä¸€ç§ç»Ÿä¸€æ ¼å¼. å®ƒå¯ä»¥è¢«å¤§å¤šæ•°çš„ç¨‹åºè¯»å–, åŒ…æ‹¬MrBayes, GARLI, PHYLIP, TREE-PUZZLEç­‰. Newickæ ‘åŒ…æ‹¬ç‰©ç§å’Œç¥–å…ˆèŠ‚ç‚¹çš„åå­—, å®ƒä»¬çš„ç›¸äº’å…³ç³», ä»¥åŠæ¯ä¸ªèŠ‚ç‚¹èµ‹äºˆçš„å€¼(é€šå¸¸ç”¨äºçˆ¶èŠ‚ç‚¹çš„è·ç¦»)</p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ07: æ ¸è‹·é…¸åºåˆ—çš„å¯†ç å­é¢‘ç‡</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># This program calculates the codon frequency in a RNA sequence</span></span><br><span class="line"><span class="comment"># (it could also be an entire genome)</span></span><br><span class="line">aa_codon = {</span><br><span class="line"><span class="string">'A'</span>:[<span class="string">'GCU'</span>,<span class="string">'GCC'</span>,<span class="string">'GCA'</span>,<span class="string">'GCG'</span>], <span class="string">'C'</span>:[<span class="string">'UGU'</span>,<span class="string">'UGC'</span>],</span><br><span class="line"><span class="string">'D'</span>:[<span class="string">'GAU'</span>,<span class="string">'GAC'</span>],<span class="string">'E'</span>:[<span class="string">'GAA'</span>,<span class="string">'GAG'</span>],<span class="string">'F'</span>:[<span class="string">'UUU'</span>,<span class="string">'UUC'</span>],</span><br><span class="line"><span class="string">'G'</span>:[<span class="string">'GGU'</span>,<span class="string">'GGC'</span>,<span class="string">'GGA'</span>,<span class="string">'GGG'</span>], <span class="string">'H'</span>:[<span class="string">'CAU'</span>,<span class="string">'CAC'</span>],</span><br><span class="line"><span class="string">'K'</span>:[<span class="string">'AAA'</span>,<span class="string">'AAG'</span>],<span class="string">'I'</span>:[<span class="string">'AUU'</span>,<span class="string">'AUC'</span>,<span class="string">'AUA'</span>,<span class="string">'AUU'</span>,<span class="string">'AUC'</span>,<span class="string">'AUA'</span>],</span><br><span class="line"><span class="string">'L'</span>:[<span class="string">'UUA'</span>,<span class="string">'UUG'</span>,<span class="string">'CUU'</span>,<span class="string">'CUC'</span>,<span class="string">'CUA'</span>,<span class="string">'CUG'</span>],<span class="string">'M'</span>:[<span class="string">'AUG'</span>],</span><br><span class="line"><span class="string">'N'</span>:[<span class="string">'AAU'</span>,<span class="string">'AAC'</span>],<span class="string">'P'</span>:[<span class="string">'CCU'</span>,<span class="string">'CCC'</span>,<span class="string">'CCA'</span>,<span class="string">'CCG'</span>],</span><br><span class="line"><span class="string">'Q'</span>:[<span class="string">'CAA'</span>,<span class="string">'CAG'</span>],<span class="string">'R'</span>:[<span class="string">'CGU'</span>,<span class="string">'CGC'</span>,<span class="string">'CGA'</span>,<span class="string">'CGG'</span>,<span class="string">'AGA'</span>,<span class="string">'AGG'</span>],</span><br><span class="line"><span class="string">'S'</span>:[<span class="string">'UCU'</span>,<span class="string">'UCC'</span>,<span class="string">'UCA'</span>,<span class="string">'UCG'</span>,<span class="string">'AGU'</span>,<span class="string">'AGC'</span>,],</span><br><span class="line"><span class="string">'Y'</span>:[<span class="string">'UAU'</span>,<span class="string">'UAC'</span>],<span class="string">'T'</span>:[<span class="string">'ACU'</span>,<span class="string">'ACC'</span>,<span class="string">'ACA'</span>,<span class="string">'ACG'</span>],</span><br><span class="line"><span class="string">'V'</span>:[<span class="string">'GUU'</span>,<span class="string">'GUC'</span>,<span class="string">'GUA'</span>,<span class="string">'GUG'</span>],<span class="string">'W'</span>:[<span class="string">'UGG'</span>],</span><br><span class="line"><span class="string">'STOP'</span>:[<span class="string">'UAG'</span>,<span class="string">'UGA'</span>,<span class="string">'UAA'</span>]}</span><br><span class="line"></span><br><span class="line">codon_count = {</span><br><span class="line"><span class="string">'GCU'</span>:<span class="number">0</span>,<span class="string">'GCC'</span>:<span class="number">0</span>,<span class="string">'GCA'</span>:<span class="number">0</span>,<span class="string">'GCG'</span>:<span class="number">0</span>,<span class="string">'CGU'</span>:<span class="number">0</span>,<span class="string">'CGC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CGA'</span>:<span class="number">0</span>,<span class="string">'CGG'</span>:<span class="number">0</span>,<span class="string">'AGA'</span>:<span class="number">0</span>,<span class="string">'AGG'</span>:<span class="number">0</span>,<span class="string">'UCU'</span>:<span class="number">0</span>,<span class="string">'UCC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UCA'</span>:<span class="number">0</span>,<span class="string">'UCG'</span>:<span class="number">0</span>,<span class="string">'AGU'</span>:<span class="number">0</span>,<span class="string">'AGC'</span>:<span class="number">0</span>,<span class="string">'AUU'</span>:<span class="number">0</span>,<span class="string">'AUC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'AUA'</span>:<span class="number">0</span>,<span class="string">'AUU'</span>:<span class="number">0</span>,<span class="string">'AUC'</span>:<span class="number">0</span>,<span class="string">'AUA'</span>:<span class="number">0</span>,<span class="string">'UUA'</span>:<span class="number">0</span>,<span class="string">'UUG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CUU'</span>:<span class="number">0</span>,<span class="string">'CUC'</span>:<span class="number">0</span>,<span class="string">'CUA'</span>:<span class="number">0</span>,<span class="string">'CUG'</span>:<span class="number">0</span>,<span class="string">'GGU'</span>:<span class="number">0</span>,<span class="string">'GGC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'GGA'</span>:<span class="number">0</span>,<span class="string">'GGG'</span>:<span class="number">0</span>,<span class="string">'GUU'</span>:<span class="number">0</span>,<span class="string">'GUC'</span>:<span class="number">0</span>,<span class="string">'GUA'</span>:<span class="number">0</span>,<span class="string">'GUG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'ACU'</span>:<span class="number">0</span>,<span class="string">'ACC'</span>:<span class="number">0</span>,<span class="string">'ACA'</span>:<span class="number">0</span>,<span class="string">'ACG'</span>:<span class="number">0</span>,<span class="string">'CCU'</span>:<span class="number">0</span>,<span class="string">'CCC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CCA'</span>:<span class="number">0</span>,<span class="string">'CCG'</span>:<span class="number">0</span>,<span class="string">'AAU'</span>:<span class="number">0</span>,<span class="string">'AAC'</span>:<span class="number">0</span>,<span class="string">'GAU'</span>:<span class="number">0</span>,<span class="string">'GAC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UGU'</span>:<span class="number">0</span>,<span class="string">'UGC'</span>:<span class="number">0</span>,<span class="string">'CAA'</span>:<span class="number">0</span>,<span class="string">'CAG'</span>:<span class="number">0</span>,<span class="string">'GAA'</span>:<span class="number">0</span>,<span class="string">'GAG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CAU'</span>:<span class="number">0</span>,<span class="string">'CAC'</span>:<span class="number">0</span>,<span class="string">'AAA'</span>:<span class="number">0</span>,<span class="string">'AAG'</span>:<span class="number">0</span>,<span class="string">'UUU'</span>:<span class="number">0</span>,<span class="string">'UUC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UAU'</span>:<span class="number">0</span>,<span class="string">'UAC'</span>:<span class="number">0</span>,<span class="string">'AUG'</span>:<span class="number">0</span>,<span class="string">'UGG'</span>:<span class="number">0</span>,<span class="string">'UAG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UGA'</span>:<span class="number">0</span>,<span class="string">'UAA'</span>:<span class="number">0</span>}</span><br><span class="line"></span><br><span class="line"><span class="comment"># Writes the frequency of each codon to a file</span></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">calc_freq</span>(<span class="params">codon_count, out_file</span>):</span><br><span class="line">    count_tot = {}</span><br><span class="line">    <span class="keyword">for</span> aa <span class="keyword">in</span> aa_codon.keys():</span><br><span class="line">        n = <span class="number">0</span></span><br><span class="line">        <span class="keyword">for</span> codon <span class="keyword">in</span> aa_codon[aa]:</span><br><span class="line">            n = n + codon_count[codon]</span><br><span class="line">        count_tot[aa] = <span class="built_in">float</span>(n)</span><br><span class="line">    <span class="keyword">for</span> aa <span class="keyword">in</span> aa_codon.keys():</span><br><span class="line">        <span class="keyword">for</span> codon <span class="keyword">in</span> aa_codon[aa]:</span><br><span class="line">            <span class="keyword">if</span> count_tot[aa] != <span class="number">0.0</span>:</span><br><span class="line">                freq = codon_count[codon] / count_tot[aa]</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                freq = <span class="number">0.0</span></span><br><span class="line">            out_file.write(<span class="string">'%4s\t%5s\t%4d\t%9.3f\n'</span>% \</span><br><span class="line">                (aa,codon,codon_count[codon], freq))</span><br><span class="line">in_file = <span class="built_in">open</span>(<span class="string">'A06662.1.fasta'</span>)</span><br><span class="line">out_file = <span class="built_in">open</span>(<span class="string">'CodonFrequency.txt'</span>, <span class="string">'w'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Reads the RNA sequence into a single string</span></span><br><span class="line">rna = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span>:</span><br><span class="line">        rna = rna + line.strip()</span><br><span class="line">        </span><br><span class="line"><span class="comment"># Scans the sequence frame by frame,</span></span><br><span class="line"><span class="comment"># counts the number of occurrences</span></span><br><span class="line"><span class="comment"># of each codon, and stores it in codon_count dictionary.</span></span><br><span class="line"><span class="comment"># Then calls calc_freq()</span></span><br><span class="line"><span class="keyword">for</span> j <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">3</span>):</span><br><span class="line">    out_file.write(<span class="string">'!!Codon frequency in frame %d\n'</span> %(j+<span class="number">1</span>))</span><br><span class="line">    out_file.write(<span class="string">' AA\tcodon\thits\tfrequency\n'</span>)</span><br><span class="line">    prot = <span class="string">''</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(j, <span class="built_in">len</span>(rna), <span class="number">3</span>):</span><br><span class="line">        codon = rna[i:i + <span class="number">3</span>]</span><br><span class="line">        <span class="keyword">if</span> codon <span class="keyword">in</span> codon_count:</span><br><span class="line">            codon_count[codon] = codon_count[codon] + <span class="number">1</span></span><br><span class="line">    calc_freq(codon_count, out_file)</span><br><span class="line">    </span><br><span class="line">out_file.close()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ08: è§£æViennaæ ¼å¼çš„RNAäºŒçº§ç»“æ„</li>
</ul>
<p>RNAåºåˆ—æœ€é‡è¦çš„æ€§è´¨ä¹‹ä¸€æ˜¯å…¶æŠ˜å å’Œé…å¯¹çš„èƒ½åŠ›. ä¸DNAä¸åŒ, RNAçš„ç¢±åŸºé…å¯¹ä¸é™äºåŒèºæ—‹, è€Œæ˜¯èƒ½å¤Ÿå½¢æˆå¤æ‚çš„ç»“æ„. é¢„æµ‹è¿™äº›RNAçš„ç¢±åŸºé…å¯¹å’ŒäºŒçº§ç»“æ„æ˜¯RNAç”Ÿç‰©ä¿¡æ¯å­¦çš„å¸¸è§ä»»åŠ¡.</p>
<p><code>Vienna</code>åŒ…: <a target="_blank" rel="noopener" href="http://rna.tbi.univie.ac.at/">ViennaRNA Web Services (univie.ac.at)</a></p>
<p>ä¸€ä¸ªå‘½ä»¤è¡Œå’Œç½‘ç»œçš„å·¥å…·é›†, å¯ä»¥è¿›è¡Œå¦‚ä»RNAåºåˆ—é¢„æµ‹ç¢±åŸºé…å¯¹ç­‰åŸºç¡€ä»»åŠ¡</p>
<p>è¿™äº›ç»“æœé€šå¸¸è¢«ä¿å­˜ä¸ºViennaæ ¼å¼</p>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt; two hairpin loops</span><br><span class="line">AAACCCCGUUUCGGGGAACCACCA</span><br><span class="line">((((...)))).((((..)).)).</span><br></pre></td></tr></tbody></table></figure>
<p>ç¬¬ä¸€è¡ŒåŒ…å«äº†RNAåºåˆ—åç§°, ç¬¬äºŒè¡Œæ˜¯åºåˆ—æœ¬èº«</p>
<p>ç¬¬ä¸‰è¡Œçš„ç‚¹-æ‹¬å·ç»“æ„è¡¨è¾¾RNAäºŒçº§ç»“æ„</p>
<p>å¯¹åº”çš„æ‹¬å·æŒ‡ç¤ºä¸€ä¸ªç¢±åŸºå¯¹, è€Œä¸é…å¯¹çš„ç¢±åŸºå¯¹ç”¨ç‚¹è¡¨ç¤º</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">RNAStructure</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, vienna</span>):</span><br><span class="line">        lines = vienna.split(<span class="string">'\n'</span>)</span><br><span class="line">        self.name = lines[<span class="number">0</span>].strip()</span><br><span class="line">        self.sequence = lines[<span class="number">1</span>].strip()</span><br><span class="line">        self.basepairs = \</span><br><span class="line">        <span class="built_in">sorted</span>(self.parse_basepairs(lines[<span class="number">2</span>].strip()))</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">parse_basepairs</span>(<span class="params">self, dotbracket</span>):</span><br><span class="line">        stack = []</span><br><span class="line">        <span class="keyword">for</span> i, char <span class="keyword">in</span> <span class="built_in">enumerate</span>(dotbracket):</span><br><span class="line">            <span class="keyword">if</span> char == <span class="string">'('</span>:</span><br><span class="line">                stack.append(i)</span><br><span class="line">            <span class="keyword">elif</span> char == <span class="string">')'</span>:</span><br><span class="line">                j = stack.pop()</span><br><span class="line">        <span class="keyword">yield</span> j, i</span><br><span class="line">        </span><br><span class="line">vienna = <span class="string">'''&gt; two hairpin loops</span></span><br><span class="line"><span class="string">AAACCCCGUUUCGGGGAACCACCA</span></span><br><span class="line"><span class="string">((((...)))).((((..)).)).</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line">rna = RNAStructure(vienna)</span><br><span class="line"><span class="built_in">print</span> rna.name</span><br><span class="line"><span class="built_in">print</span> rna.sequence</span><br><span class="line"><span class="keyword">for</span> base1, base2 <span class="keyword">in</span> rna.basepairs:</span><br><span class="line">    <span class="built_in">print</span> <span class="string">'(%i, %i)'</span>%(base1, base2)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ09: è§£æBLASTçš„XMLè¾“å‡º</li>
</ul>
<p>Biopython BLAST XMLè§£æå™¨è¯»å–XML BLASTè¾“å‡ºæ–‡ä»¶</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec125">7.3 Parsing BLAST output (biopython.org)</a></p>
<p>NCBIXML.parse()å‡½æ•°è¿”å›ä¸€ä¸ªblast_outå®ä¾‹, å®ƒåŒ…å«ä¸€åˆ°å¤šä¸ªè®°å½•å¯¹è±¡(BLASTç»“æœåœ¨ä¸€ä¸ªè®°å½•é‡Œ, è€Œå¦‚PSI-BLASTåˆ™æ˜¯å¯¹æ¯ä¸ªæŸ¥è¯¢æœ‰ä¸€ä¸ªçºªå½•). æ¯ä¸ªæŸ¥è¯¢éƒ½åŒ…å«ä¸€åˆ°å¤šä¸ªåŒ¹é…(hit)æˆ–è”é…(alignment), ä¾æ¬¡åŒ…å«ä¸€åˆ°å¤šä¸ªå±€éƒ¨è”é…(HSP).</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>blastp -query P05480.fasta -db nr<span class="number">.00</span> -out blast_output.xml</span><br><span class="line">-outfmt <span class="number">5</span></span><br><span class="line"></span><br><span class="line">xml_file = <span class="built_in">open</span>(<span class="string">"blast_output.xml"</span>)</span><br><span class="line">blast_out = NCBIXML.parse(xml_file)</span><br><span class="line"><span class="keyword">for</span> record <span class="keyword">in</span> blast_out:</span><br><span class="line">    <span class="keyword">for</span> alignment <span class="keyword">in</span> record.alignments:</span><br><span class="line">        <span class="built_in">print</span> alignment.title</span><br><span class="line">        <span class="keyword">for</span> hsp <span class="keyword">in</span> alignment.hsps:</span><br><span class="line">            <span class="comment"># filter by e-value</span></span><br><span class="line">            <span class="keyword">if</span> hsp.expect &lt; <span class="number">0.0001</span>:</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"score:"</span>, hsp.score</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"query:"</span>, hsp.query</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"match:"</span>, hsp.<span class="keyword">match</span></span><br><span class="line">                <span class="built_in">print</span> <span class="string">"sbjct:"</span>, hsp.sbjct</span><br><span class="line">                <span class="built_in">print</span> <span class="string">'#'</span> * <span class="number">70</span></span><br><span class="line">                <span class="built_in">print</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ10: è§£æSBMLæ–‡ä»¶</li>
</ul>
<p><strong>Systems Biology Markup Language</strong> (<strong>SBML</strong>), a free and open interchange format for computer models of biological processes. SBML is useful for models of metabolism, cell signaling, and more. It <strong>continues to be evolved and expanded</strong> by an international community.</p>
<p><a target="_blank" rel="noopener" href="http://sbml.org/Main_Page">Main Page - SBML.caltech.edu</a></p>
<p>SBMLæ ¼å¼ç”¨æ¥å‚¨å­˜é€šè·¯,ååº”å’Œè°ƒæ§ç½‘ç»œä¿¡æ¯çš„æ ‡å‡†æ ¼å¼</p>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/xml/xml-tutorial.html">XML æ•™ç¨‹ | èœé¸Ÿæ•™ç¨‹ (runoob.com)</a></p>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/xml.html?highlight=xml#module-xml">19.5. XMLå¤„ç†æ¨¡å— â€” Python 2.7.18 æ–‡æ¡£</a></p>
<figure class="highlight xml"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&lt;?xml version=<span class="string">"1.0"</span> encoding=<span class="string">"UTF-8"</span>?&gt;</span></span><br><span class="line"><span class="tag">&lt;<span class="name">sbml</span> <span class="attr">xmlns</span>=<span class="string">"http://www.sbml.org/sbml/level2"</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;<span class="name">model</span> <span class="attr">name</span>=<span class="string">"SBML file with three metabolites"</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;<span class="name">listOfSpecies</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m78"</span> <span class="attr">name</span>=<span class="string">"Inosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O5<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m79"</span> <span class="attr">name</span>=<span class="string">"Xanthosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O6<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m80"</span> <span class="attr">name</span>=<span class="string">"Xanthosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O6<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;/<span class="name">listOfSpecies</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;/<span class="name">model</span>&gt;</span></span><br><span class="line"><span class="tag">&lt;/<span class="name">sbml</span>&gt;</span></span><br></pre></td></tr></tbody></table></figure>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> xml.dom.minidom <span class="keyword">import</span> parse</span><br><span class="line"></span><br><span class="line">document = parse(<span class="string">'sbml_example.xml'</span>)</span><br><span class="line">species_list = document.getElementsByTagName(<span class="string">"species"</span>)</span><br><span class="line"><span class="keyword">for</span> species <span class="keyword">in</span> species_list:</span><br><span class="line">    species_id = species.getAttribute(<span class="string">'id'</span>)</span><br><span class="line">    name = species.getAttribute(<span class="string">'name'</span>)</span><br><span class="line">    p_list = species.getElementsByTagName(<span class="string">"p"</span>)</span><br><span class="line">    p = p_list[<span class="number">0</span>]</span><br><span class="line">    text = p.childNodes[<span class="number">0</span>]</span><br><span class="line">    formula = text.nodeValue</span><br><span class="line">    <span class="built_in">print</span> <span class="string">"%-20s\t%5s\t%s"</span>%(name, species_id, formula)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># Inosine      M_m78 FORMULA: C10H12N4O5</span></span><br><span class="line"><span class="comment"># Xanthosine   M_m79 FORMULA: C10H12N4O6</span></span><br><span class="line"><span class="comment"># Xanthosine   M_m80 FORMULA: C10H12N4O6</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ11: è¿è¡ŒBLAST</li>
</ul>
<blockquote>
<p>The NCBI provides <strong>a suite of command-line tools to run BLAST called BLAST+</strong>. This allows users to perform BLAST searches on their own server without size, volume and database restrictions. BLAST+ can be used with a command line so it can be integrated directly into your workflow.</p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">Download BLAST Software and Databases Documentation (nih.gov)</a></p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec121">7.2 Running BLAST locally (Biopython)</a></p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ12: è®¿é—®ã€ ä¸‹è½½å’Œè¯»å–ç½‘é¡µ</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/urllib.html?highlight=urllib">20.5. urllib â€” Open arbitrary resources by URL â€” Python 2.7.18 æ–‡æ¡£</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> urllib <span class="keyword">import</span> urlopen</span><br><span class="line">url = urlopen(<span class="string">'http://www.uniprot.org/uniprot/P01308.fasta'</span>)</span><br><span class="line">doc = url.read()</span><br><span class="line"><span class="built_in">print</span> doc</span><br></pre></td></tr></tbody></table></figure>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/urllib2.html">20.6. urllib2 â€” extensible library for opening URLs â€” Python 2.7.18 æ–‡æ¡£</a></p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ13: è§£æHTMLæ–‡ä»¶</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/htmlparser.html?highlight=htmlparser#module-HTMLParser">19.1. HTMLParser â€” Simple HTML and XHTML parser â€” Python 2.7.18 æ–‡æ¡£</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> HTMLParser <span class="keyword">import</span> HTMLParser</span><br><span class="line"><span class="keyword">import</span> urllib</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">MyHTMLParser</span>(<span class="title class_ inherited__">HTMLParser</span>):</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_starttag</span>(<span class="params">self, tag, attrs</span>):</span><br><span class="line">        self.start_tag = tag</span><br><span class="line">        <span class="built_in">print</span> <span class="string">"Start tag:"</span>, self.start_tag</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_endtag</span>(<span class="params">self, tag</span>):</span><br><span class="line">        self.end_tag = tag</span><br><span class="line">        <span class="built_in">print</span> <span class="string">"End tag :"</span>, self.end_tag</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_data</span>(<span class="params">self, data</span>):</span><br><span class="line">        self.data = data.strip()</span><br><span class="line">        <span class="keyword">if</span> self.data:</span><br><span class="line">            <span class="built_in">print</span> <span class="string">"Data :"</span>, self.data</span><br><span class="line">            </span><br><span class="line">parser = MyHTMLParser()</span><br><span class="line">url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed/21998472'</span></span><br><span class="line">page = urllib.urlopen(url)</span><br><span class="line">data = page.read()</span><br><span class="line">parser.feed(data)</span><br><span class="line"></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ14: å°†PDBæ–‡ä»¶åˆ†å‰²æˆPDBé“¾æ–‡ä»¶</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> struct <span class="keyword">import</span> unpack</span><br><span class="line"><span class="keyword">import</span> os.path</span><br><span class="line"></span><br><span class="line">filename = <span class="string">'2H8L.pdb'</span></span><br><span class="line">in_file = <span class="built_in">open</span>(filename)</span><br><span class="line">pdb_id = filename.split(<span class="string">'.'</span>)[<span class="number">0</span>]</span><br><span class="line">pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s6s4s2s3s'</span></span><br><span class="line">chain_old = <span class="string">'@'</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    <span class="keyword">if</span> line[<span class="number">0</span>:<span class="number">4</span>] == <span class="string">"ATOM"</span>:</span><br><span class="line">        col = unpack(pdb_format, line)</span><br><span class="line">        chain = col[<span class="number">7</span>].strip()</span><br><span class="line">        <span class="keyword">if</span> chain != chain_old:</span><br><span class="line">            <span class="keyword">if</span> os.path.exists(pdb_id+chain_old+<span class="string">'.pdb'</span>):</span><br><span class="line">                chain_file.close()</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"closed:"</span>, pdb_id+chain_old+<span class="string">'.pdb'</span></span><br><span class="line">            chain_file = <span class="built_in">open</span>(pdb_id+chain+<span class="string">'.pdb'</span>,<span class="string">'w'</span>)</span><br><span class="line">            chain_file.write(line)</span><br><span class="line">            chain_old = chain</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            chain_file.write(line)</span><br><span class="line">chain_file.close()</span><br><span class="line"><span class="built_in">print</span> <span class="string">"closed:"</span>, pdb_id+chain_old+<span class="string">'.pdb'</span></span><br></pre></td></tr></tbody></table></figure>
<p>ä¹Ÿå¯ä»¥ä½¿ç”¨ç»ˆç«¯çš„<code>&gt;</code>å°†ç»“æœå¯¼å‡ºåˆ°æ–‡ä»¶</p>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ15: åœ¨PDBç»“æ„ä¸Šæ‰¾åˆ°ä¸¤ä¸ªæœ€é è¿‘çš„CÎ±åŸå­</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> math <span class="keyword">import</span> sqrt</span><br><span class="line"><span class="keyword">from</span> struct <span class="keyword">import</span> unpack</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">calc_dist</span>(<span class="params">p1, p2</span>):</span><br><span class="line">    <span class="string">'''returns the distance between two 3D points'''</span></span><br><span class="line">    tmp = <span class="built_in">pow</span>(p1[<span class="number">0</span>] - p2[<span class="number">0</span>], <span class="number">2</span>) + \</span><br><span class="line">        <span class="built_in">pow</span>(p1[<span class="number">1</span>] - p2[<span class="number">1</span>], <span class="number">2</span>) + \</span><br><span class="line">        <span class="built_in">pow</span>(p1[<span class="number">2</span>] - p2[<span class="number">2</span>], <span class="number">2</span>)</span><br><span class="line">    tmp = sqrt(tmp)</span><br><span class="line">    <span class="keyword">return</span> tmp</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">min_dist</span>(<span class="params">arglist</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    returns the closest residue pair and their</span></span><br><span class="line"><span class="string">    CA_CA distance</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    <span class="comment"># initialize variables</span></span><br><span class="line">    maxval = <span class="number">10000</span></span><br><span class="line">    residue_pair = ()</span><br><span class="line">    <span class="comment"># read arglist starting from the 1st position</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="built_in">len</span>(arglist)):</span><br><span class="line">        <span class="comment"># save x,y,z coordinates from the arglist</span></span><br><span class="line">        <span class="comment"># i-element into the atom1 variable</span></span><br><span class="line">        atom1 = arglist[i][<span class="number">1</span>:]</span><br><span class="line">        <span class="comment"># run over all other elements</span></span><br><span class="line">        <span class="keyword">for</span> j <span class="keyword">in</span> <span class="built_in">range</span>(i + <span class="number">1</span>, <span class="built_in">len</span>(arglist)):</span><br><span class="line">            atom2 = arglist[j][<span class="number">1</span>:]</span><br><span class="line">            <span class="comment"># calculate the distance</span></span><br><span class="line">            tmp = calc_dist(atom1, atom2)</span><br><span class="line">            <span class="comment"># check if the distance is lower than</span></span><br><span class="line">            <span class="comment"># the previously recorded lowest value</span></span><br><span class="line">            <span class="keyword">if</span> tmp &lt; maxval :</span><br><span class="line">                <span class="comment"># save the new data</span></span><br><span class="line">                residue_pair = (arglist[i][<span class="number">0</span>], \</span><br><span class="line">                                arglist[j][<span class="number">0</span>])</span><br><span class="line">                maxval = tmp</span><br><span class="line">    <span class="keyword">return</span> residue_pair, maxval</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_list_ca_atoms</span>(<span class="params">pdb_file, chain</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    returns a list of CA atoms, the residues</span></span><br><span class="line"><span class="string">    they belong to, and their x,y,z coordinates</span></span><br><span class="line"><span class="string">    from the input PDB file</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    in_file = <span class="built_in">open</span>(pdb_file)</span><br><span class="line">    CA_list = []</span><br><span class="line">    pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s6s4s2s3s'</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">        tmp = unpack(pdb_format, line)</span><br><span class="line">        tmp = [i.strip() <span class="keyword">for</span> i <span class="keyword">in</span> tmp]</span><br><span class="line">        <span class="comment"># only save CA coords belonging to input chain</span></span><br><span class="line">        <span class="keyword">if</span> tmp[<span class="number">0</span>] ==<span class="string">"ATOM"</span> <span class="keyword">and</span> tmp[<span class="number">7</span>] == chain <span class="keyword">and</span> \</span><br><span class="line">            tmp[<span class="number">3</span>] == <span class="string">"CA"</span>:</span><br><span class="line">            <span class="comment"># create a tuple (aa_number, x, y, x)</span></span><br><span class="line">            tmp = (tmp[<span class="number">5</span>]+tmp[<span class="number">8</span>], <span class="built_in">float</span>(tmp[<span class="number">11</span>]), \</span><br><span class="line">            <span class="built_in">float</span>(tmp[<span class="number">12</span>]), <span class="built_in">float</span>(tmp[<span class="number">13</span>]))</span><br><span class="line">            <span class="comment"># add the tuple to the list</span></span><br><span class="line">            CA_list.append(tmp)</span><br><span class="line">    in_file.close()</span><br><span class="line">    <span class="keyword">return</span> CA_list</span><br><span class="line"></span><br><span class="line"><span class="comment"># obtain the list of CA atoms of Chain A</span></span><br><span class="line">CA_list = get_list_ca_atoms(<span class="string">"2H8L.pdb"</span>, <span class="string">"A"</span>)</span><br><span class="line"><span class="comment"># identify the closest atoms</span></span><br><span class="line">res_pair, dist = min_dist(CA_list)</span><br><span class="line"><span class="built_in">print</span> <span class="string">'The distance between'</span>, res_pair, <span class="string">'is:'</span>, dist</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ16: æå–ä¸¤ä¸ªPDBé“¾é—´çš„ç•Œé¢</li>
</ul>
<p>åœ¨ä¸€ä¸ªä¸‰ç»´ç»“æ„ä¸­, ä¸¤æ¡å¤šè‚½é“¾ä¹‹é—´çš„ç•Œé¢è¢«å®šä¹‰ä¸ºæ»¡è¶³è¿™ä¸¤ä¸ªæ¡ä»¶çš„æ‰€æœ‰æ®‹åŸºå¯¹:1.ä¸¤ä¸ªæ®‹åŸºåˆ†åˆ«å±äºä¸åŒçš„é“¾;2.å®ƒä»¬ä¹‹é—´çš„è·ç¦»å°äºæŸä¸ªé˜ˆå€¼</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec189">11.2 Structure representation (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> PDB</span><br><span class="line">parser = PDB.PDBParser()</span><br><span class="line">s = parser.get_structure(<span class="string">"2H8L"</span>,<span class="string">"2H8L.pdb"</span>)</span><br><span class="line">first_model = s[<span class="number">0</span>]</span><br><span class="line">chain_A = first_model[<span class="string">"A"</span>]</span><br><span class="line">chain_B = first_model[<span class="string">"B"</span>]</span><br><span class="line"><span class="keyword">for</span> res1 <span class="keyword">in</span> chain_A:</span><br><span class="line">    <span class="keyword">for</span> res2 <span class="keyword">in</span> chain_B:</span><br><span class="line">        d = res1[<span class="string">"CA"</span>]-res2[<span class="string">"CA"</span>]</span><br><span class="line">        <span class="keyword">if</span> d &lt;= <span class="number">6.0</span>:</span><br><span class="line">            <span class="built_in">print</span> res1.resname,res1.get_id()[<span class="number">1</span>], res2.resname,\</span><br><span class="line">                res2.get_id()[<span class="number">1</span>], d</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ17: ç”¨Modellerå»ºç«‹åŒæºæ¨¡å‹</li>
</ul>
<p><code>Modeller</code> æ˜¯ä¸€ä¸ªä¸ºå»ºç«‹è›‹ç™½è´¨ä¸‰ç»´ç»“æ„çš„åŒæºæ¨¡å‹è®¾è®¡çš„è½¯ä»¶åŒ…</p>
<p><a target="_blank" rel="noopener" href="https://salilab.org/modeller/">About MODELLER (salilab.org)</a></p>
<blockquote>
<p>MODELLER is used for homology or comparative modeling of protein three-dimensional structures (1,2). The user provides an alignment of a sequence to be modeled with known related structures and MODELLER automatically calculates a model containing all non-hydrogen atoms. MODELLER implements comparative protein structure modeling by satisfaction of spatial restraints (3,4), and can perform many additional tasks, including de novo modeling of loops in protein structures, optimization of various models of protein structure with respect to a flexibly defined objective function, multiple alignment of protein sequences and/or structures, clustering, searching of sequence databases, comparison of protein structures, etc.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> modeller <span class="keyword">import</span> *</span><br><span class="line"><span class="keyword">from</span> modeller.automodel <span class="keyword">import</span> *</span><br><span class="line">log.verbose()</span><br><span class="line">env = environ()</span><br><span class="line">env.io.atom_files_directory = [<span class="string">'.'</span>, <span class="string">'../atom_files'</span>]</span><br><span class="line">a = automodel(env,</span><br><span class="line">            alnfile = <span class="string">'alignment.ali'</span>,</span><br><span class="line">            knowns = <span class="string">'1eq9A'</span>,</span><br><span class="line">            sequence = <span class="string">'MyTarget_Seq'</span>)</span><br><span class="line">a.starting_model = <span class="number">1</span></span><br><span class="line">a.ending_model = <span class="number">1</span></span><br><span class="line">a.make()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ18: ç”¨ModeRNAåˆ†æRNAä¸‰ç»´åŒæºæ¨¡å‹</li>
</ul>
<p><code>ModeRNA</code> åº“: <a target="_blank" rel="noopener" href="https://pypi.org/project/moderna/">moderna Â· PyPI</a> æœåŠ¡ç»´æŠ¤åœæ­¢</p>
<p>åˆ†æ/æ“çºµ/å»ºæ¨¡RNAä¸‰ç»´ç»“æ„</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> moderna <span class="keyword">import</span> *</span><br><span class="line"></span><br><span class="line">ehz = load_model(<span class="string">'1ehz.ent'</span>, <span class="string">'A'</span>)</span><br><span class="line">clean_structure(ehz)</span><br><span class="line"><span class="built_in">print</span> get_sequence(ehz)</span><br><span class="line"><span class="built_in">print</span> get_secstruc(ehz)</span><br><span class="line"></span><br><span class="line">m = create_model()</span><br><span class="line">copy_some_residues(ehz[<span class="string">'1'</span>:<span class="string">'15'</span>], m)</span><br><span class="line">write_model(m, <span class="string">'1ehz_15r.ent'</span>)</span><br><span class="line">temp = load_template(<span class="string">'1ehz_15r.ent'</span>)</span><br><span class="line">ali = load_alignment(<span class="string">'''&gt; model sequence</span></span><br><span class="line"><span class="string">ACUGUGAYUA[UACCU#PG</span></span><br><span class="line"><span class="string">&gt; template: first 15 bases from 1ehz</span></span><br><span class="line"><span class="string">GCGGA--UUUALCUCAG'''</span>)</span><br><span class="line">model = create_model(temp, ali)</span><br><span class="line"><span class="built_in">print</span> get_sequence(model)</span><br><span class="line"></span><br><span class="line">GCGGAUUUALCUCAGDDGGGAGAGCRCCAGABU<span class="comment">#AAYAP?UGGAG7UC?UGUGTPCG"UCC</span></span><br><span class="line">  ACAGAAUUCGCACCA</span><br><span class="line">(((((((..((((........)))).((((.........)))).....</span><br><span class="line">   (((((.......))))))))))))....</span><br><span class="line">ACUGUGAYUA[UACCU<span class="comment">#PG</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ19: ä»ä¸‰çº§ç»“æ„è®¡ç®—RNAç¢±åŸºé…å¯¹</li>
</ul>
<p><code>RNAView</code>: è®¡ç®—ä¸€ä¸ªRNAåˆ†å­PDBç»“æ„ä¸­çš„ç¢±åŸºé…å¯¹</p>
<p><a target="_blank" rel="noopener" href="http://ndbserver.rutgers.edu/ndbmodule/services/download/rnaview.html">NDB | Tools | Downloads (rutgers.edu)</a></p>
<blockquote>
<p><strong><a target="_blank" rel="noopener" href="http://ndbserver.rutgers.edu/ndbmodule/services/rna_viewer/rnaViewerIndex.html">RNA Viewer</a></strong>, a tool for viewing RNA 2-dimensional structure, base pairs and RNA patterns. 2D structure is fully annotated according to Leontis and Westhof (2001) <em>RNA</em> <strong>7</strong>:499-512.</p>
<p>Yang, H., Jossinet, F., Leontis, N., Chen, L., Westbrook, J., Berman, H.M. and Westhof, E. (2003) Tools for the automatic identification and classification of RNA base pairs. Nucleic Acids Res, 31, 3450-3460.</p>
</blockquote>
<p>PyCogentåº“</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>rnaview lehz.pdb &gt; basepair.out</span><br><span class="line"><span class="number">1_72</span>, A:  <span class="number">1</span> G-C <span class="number">72</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">2_71</span>, A:  <span class="number">2</span> C-G <span class="number">71</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">3_70</span>, A:  <span class="number">3</span> G-C <span class="number">70</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">4_69</span>, A:  <span class="number">4</span> G-C <span class="number">69</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">5_68</span>, A:  <span class="number">5</span> G-C <span class="number">68</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">6_67</span>, A:  <span class="number">6</span> A-U <span class="number">67</span>  A: W/W cis  n/a</span><br><span class="line"><span class="number">7_66</span>, A:  <span class="number">7</span> A-U <span class="number">66</span>  A: -/- cis  XX</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> cogent.app.rnaview <span class="keyword">import</span> RnaView</span><br><span class="line"><span class="keyword">from</span> cogent.parse.rnaview <span class="keyword">import</span> RnaviewParser</span><br><span class="line">rna_prog = RnaView()</span><br><span class="line">result = rna_prog(<span class="string">'1ehz.pdb'</span>)</span><br><span class="line">bpairs = result[<span class="string">'base_pairs'</span>]</span><br><span class="line">errors = result[<span class="string">'StdErr'</span>].read()</span><br><span class="line">stdout = result[<span class="string">'StdOut'</span>].read()</span><br><span class="line"></span><br><span class="line">bp_dict = RnaviewParser(bpairs)</span><br><span class="line"><span class="built_in">print</span> <span class="string">'INFORMATION:'</span></span><br><span class="line">sys.stderr.write(errors)</span><br><span class="line"><span class="built_in">print</span> stdout</span><br><span class="line"><span class="built_in">print</span> <span class="string">'BASE PAIRS:'</span></span><br><span class="line"><span class="keyword">for</span> key <span class="keyword">in</span> bp_dict:</span><br><span class="line">    <span class="built_in">print</span> key, bp_dict[key]</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>ç¼–ç¨‹ç§˜ç¬ˆ20: ç»“æ„é‡å çš„çœŸå®å®ä¾‹: ä¸æ°¨é…¸è›‹ç™½é…¶å‚¬åŒ–ä¸‰åˆ†å­</li>
</ul>
<p>ç»“æ„é‡å : åœ¨é‡å çš„è¿‡ç¨‹ä¸­, ä¸¤ä¸ªç»“æ„ä¹‹ä¸€çš„(é¶æ ‡)è¢«æ”¾ç½®åœ¨ä¸€ä¸ªå›ºå®šçš„ä½ç½®ä¸Š,å¦ä¸€ä¸ª(æ¢é’ˆ)è¢«å¹³ç§»å’Œæ—‹è½¬, ç›´åˆ°è¿ä¸ªç»“æ„çš„å‡æ–¹å·®è¾¾åˆ°æœ€å°. ä¸€ä¸ªé‡å å¿…é¡»è¦æ±‚ä¸¤ä¸ªé›†åˆçš„åŸå­æ•°ç›¸åŒæ‰èƒ½å·¥ä½œ, æ‰€ä»¥åœ¨é‡å ä¸¤ä¸ªç»“æ„ä¹‹å‰, éœ€è¦å†³å®šæ¯ä¸ªç»“æ„ä¸Šå“ªäº›åŸå­éœ€è¦é‡å , è¿™ç”šè‡³å¯ä»¥è¯†æ•´ä¸ªç»“æ„çš„æ‰€æœ‰åŸå­.</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec238">11.8 Accessing the Protein Data Bank (biopython.org)</a></p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec223">11.6.6 Superimposing two structures (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Superimpose the catalytic triads of two different serine</span></span><br><span class="line"><span class="comment"># proteases(on CA and N atoms of res H57, D102, and S195 of chain A)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> PDB</span><br><span class="line"></span><br><span class="line"><span class="comment"># Retrieve PDB files   ç½‘ç»œè·å–ä¸‰ä¸ªç»“æ„æ–‡ä»¶</span></span><br><span class="line">pdbl = PDB.PDBList()</span><br><span class="line">pdbl.retrieve_pdb_file(<span class="string">"1EQ9"</span>)</span><br><span class="line">pdbl.retrieve_pdb_file(<span class="string">"1FXY"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Parse the two structures  åˆ°å¯¹åº”ç›®å½•è¯»å–æ–‡ä»¶ç»“æ„ä¿¡æ¯</span></span><br><span class="line"><span class="keyword">from</span> Bio.PDB <span class="keyword">import</span> PDBParser, Superimposer, PDBIO</span><br><span class="line">parser = PDB.PDBParser()</span><br><span class="line">struct_1 = parser.get_structure(<span class="string">"1EQ9"</span>, <span class="string">"eq/pdb1eq9.ent"</span>)</span><br><span class="line">struct_2 = parser.get_structure(<span class="string">"1FXY"</span>, <span class="string">"fx/pdb1fxy.ent"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># get the catalytic triads  æå–ä¸‰ä¸ªå‚¬åŒ–ä½ç‚¹æ°¨åŸºé…¸ä¿¡æ¯</span></span><br><span class="line">res_57_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">57</span>]</span><br><span class="line">res_102_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">102</span>]</span><br><span class="line">res_195_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">195</span>]</span><br><span class="line">res_57_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">57</span>]</span><br><span class="line">res_102_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">102</span>]</span><br><span class="line">res_195_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">195</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># Build 2 lists of atoms for calculating a rot.-trans. matrix</span></span><br><span class="line"><span class="comment"># (target and probe). æ¯ä¸ªä½ç‚¹æ°¨åŸºé…¸çš„Î±ç¢³ä»¥åŠæ°®åŸå­æ”¾å…¥ç›®æ ‡å’Œæ¢é’ˆåˆ—è¡¨</span></span><br><span class="line">target = []</span><br><span class="line">backbone_names = [<span class="string">'CA'</span>, <span class="string">'N'</span>]</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> backbone_names:</span><br><span class="line">    target.append(res_57_struct_1[name])</span><br><span class="line">    target.append(res_102_struct_1[name])</span><br><span class="line">    target.append(res_195_struct_1[name])</span><br><span class="line">    </span><br><span class="line">probe = []</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> backbone_names:</span><br><span class="line">    probe.append(res_57_struct_2[name])</span><br><span class="line">    probe.append(res_102_struct_2[name])</span><br><span class="line">    probe.append(res_195_struct_2[name])</span><br><span class="line">    </span><br><span class="line"><span class="comment"># Check whether target and probe lists are equal in size.</span></span><br><span class="line"><span class="comment"># This is needed for calculating a rot.-trans. matrix</span></span><br><span class="line"><span class="keyword">assert</span> <span class="built_in">len</span>(target) == <span class="built_in">len</span>(probe)  <span class="comment"># ç¡®å®šåŸå­æ•°é‡ç›¸ç­‰</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Calculate the rotation-translation matrix.</span></span><br><span class="line">sup = Superimposer()   <span class="comment"># æ—‹è½¬å¹³ç§»çŸ©é˜µè®¡ç®—</span></span><br><span class="line">sup.set_atoms(target, probe)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Apply the matrix. Remember that it can be applied only on</span></span><br><span class="line"><span class="comment"># lists of atoms.  åº”ç”¨çŸ©é˜µ, æ³¨æ„åªèƒ½æŒ‰ç…§åˆ—è¡¨æäº¤</span></span><br><span class="line">struct_2_atoms = [at <span class="keyword">for</span> at <span class="keyword">in</span> struct_2.get_atoms()]</span><br><span class="line">sup.apply(struct_2_atoms)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Write the rotation-translated structure ä¿å­˜æ—‹è½¬å¹³ç§»ç»“æ„</span></span><br><span class="line">out = PDBIO()</span><br><span class="line">out.set_structure(struct_2)</span><br><span class="line">out.save(<span class="string">'1FXY-superimposed.pdb'</span>)</span><br></pre></td></tr></tbody></table></figure>
<h2><a href="#fu-lu" class="header-anchor"> âš“ï¸</a><span id="fu-lu">é™„å½•</span></h2>
<h3><a href="#fu-lu-a-ming-ling-gai-lan" class="header-anchor"> âš“ï¸</a><span id="fu-lu-a-ming-ling-gai-lan">é™„å½•A å‘½ä»¤æ¦‚è§ˆ</span></h3>
<p>Linuxå‘½ä»¤è§: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/">Linuxå‘½ä»¤æœç´¢å¼•æ“ å‘½ä»¤ï¼ŒLinux Linuxå‘½ä»¤æœç´¢å¼•æ“ å‘½ä»¤è¯¦è§£ï¼šæœ€ä¸“ä¸šçš„Linuxå‘½ä»¤å¤§å…¨ï¼Œå†…å®¹åŒ…å«Linuxå‘½ä»¤æ‰‹å†Œã€è¯¦è§£ã€å­¦ä¹ ï¼Œå€¼å¾—æ”¶è—çš„Linuxå‘½ä»¤é€ŸæŸ¥æ‰‹å†Œã€‚ - Linux å‘½ä»¤æœç´¢å¼•æ“</a></p>
<p>Pythonå‘½ä»¤è§: <a target="_blank" rel="noopener" href="https://www.runoob.com/python3/python3-tutorial.html">Python3 æ•™ç¨‹ | èœé¸Ÿæ•™ç¨‹ (runoob.com)</a></p>
<h3><a href="#fu-lu-b-python-zi-yuan" class="header-anchor"> âš“ï¸</a><span id="fu-lu-b-python-zi-yuan">é™„å½•B Pythonèµ„æº</span></h3>
<p>Pythonå®˜ç½‘: <a target="_blank" rel="noopener" href="https://www.python.org/">Welcome to Python.org</a></p>
<p>Pythonå®˜æ–¹æ–‡æ¡£é›†åˆ: <a target="_blank" rel="noopener" href="https://docs.python.org/3/">3.9.0 Documentation (python.org)</a></p>
<p>Pythonæ ‡å‡†åº“: <a target="_blank" rel="noopener" href="https://docs.python.org/3/library/index.html#library-index">The Python Standard Library</a></p>
<p>Biopythonæ•™ç¨‹: <a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec249">Biopython Tutorial and Cookbook</a></p>
<h3><a href="#fu-lu-c-ji-lu-yang-ban" class="header-anchor"> âš“ï¸</a><span id="fu-lu-c-ji-lu-yang-ban">é™„å½•C è®°å½•æ ·æ¿</span></h3>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br></pre></td><td class="code"><pre><span class="line"># C.<span class="number">1</span> A SINGLE PROTEIN SEQUENCE FILE <span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;sp |P03372|ESR1_HUMAN Estrogen receptor OS = Homo sapiens GN =</span><br><span class="line">ESR1 PE = <span class="number">1</span> SV = <span class="number">2</span></span><br><span class="line">MTM TLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY</span><br><span class="line">EFN AAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF</span><br><span class="line">LQP HGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK</span><br><span class="line">ETR YCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC</span><br><span class="line">RLR KCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR</span><br><span class="line">SKK NSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW</span><br><span class="line">AKR VPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG</span><br><span class="line">MVE IFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD</span><br><span class="line">KIT DTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL</span><br><span class="line">LEM LDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</span><br><span class="line"></span><br><span class="line"># C.<span class="number">2</span> A SINGLE NUCLEOTIDE SEQUENCE FILE <span class="keyword">IN</span> FASTA</span><br><span class="line"><span class="built_in">FORMAT</span></span><br><span class="line">&gt;ENSG00000188536|hemoglobin alpha <span class="number">2</span></span><br><span class="line">ATGGTGCTGTCTCCTGCCGACAAGACCAACGTCAAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCT</span><br><span class="line">GGCGAGTATGGTGCGGAGGCCCTGGAGAGGATGTTCCTGTCCTTCCCCACCACCAAGACCTACTTC</span><br><span class="line">CCGCACTTCGACCTGAGCCACGGCTCTGCCCAGGTTAAGGGCCACGGCAAGAAGGTGGCCGACGCG</span><br><span class="line">CTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCAC</span><br><span class="line">GCGCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGCTCCTAAGCCACTGCCTGCTGGTGACCCTG</span><br><span class="line">GCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCCTCCCTGGACAAGTTCCTGGCTTCT</span><br><span class="line">GTGAGCACCGTGCTGACCTCCAAATACCGTTAA</span><br><span class="line"></span><br><span class="line"># C.<span class="number">3</span> AN EXAMPLE OF AN RNA SEQUENCE</span><br><span class="line"><span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;ENSG00000188536|hemoglobin alpha <span class="number">2</span></span><br><span class="line">ATGGTGCTGTCTCCTGCCGACAAGACCAACGTCAAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCT</span><br><span class="line">GGCGAGTATGGTGCGGAGGCCCTGGAGAGGATGTTCCTGTCCTTCCCCACCACCAAGACCTACTTC</span><br><span class="line">CCGCACTTCGACCTGAGCCACGGCTCTGCCCAGGTTAAGGGCCACGGCAAGAAGGTGGCCGACGCG</span><br><span class="line">CTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCAC</span><br><span class="line">GCGCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGCTCCTAAGCCACTGCCTGCTGGTGACCCTG</span><br><span class="line">GCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCCTCCCTGGACAAGTTCCTGGCTTCT</span><br><span class="line">GTGAGCACCGTGCTGACCTCCAAATACCGTTAA</span><br><span class="line"></span><br><span class="line"># C.<span class="number">4</span> A MULTIPLE SEQUENCE FILE <span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;sp |P03372|ESR1_HUMAN Estrogen receptor OS = Homo sapiens GN =</span><br><span class="line">ESR1 PE = <span class="number">1</span> SV = <span class="number">2</span></span><br><span class="line">MTM TLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY</span><br><span class="line">EFN AAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF</span><br><span class="line">LQP HGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK</span><br><span class="line">ETR YCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC</span><br><span class="line">RLR KCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR</span><br><span class="line">SKK NSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW</span><br><span class="line">AKR VPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG</span><br><span class="line">MVE IFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD</span><br><span class="line">KIT DTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL</span><br><span class="line">LEM LDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</span><br><span class="line">&gt;sp |P62333|PRS10_HUMAN <span class="number">26</span>S protease regulatory subunit</span><br><span class="line"><span class="number">10</span>B OS = Homo sapiens GN = PSMC6 PE = <span class="number">1</span> SV = <span class="number">1</span></span><br><span class="line">MAD PRDKALQDYRKKLLEHKEIDGRLKELREQLKELTKQYEKSENDLKALQSVGQIVGEV</span><br><span class="line">LKQ LTEEKFIVKATNGPRYVVGCRRQLDKSKLKPGTRVALDMTTLTIMRYLPREVDPLVY</span><br><span class="line">NMS HEDPGNVSYSEIGGLSEQIRELREVIELPLTNPELFQRVGIIPPKGCLLYGPPGTGK</span><br><span class="line">TLL ARAVASQLDCNFLKVVSSSIVDKYIGESARLIREMFNYARDHQPCIIFMDEIDAIGG</span><br><span class="line">RRF SEGTSADREIQRTLMELLNQMDGFDTLHRVKMIMATNRPDTLDPALLRPGRLDRKIH</span><br><span class="line">IDL PNEQARLDILKIHAGPITKHGEIDYEAIVKLSDGFNGADLRNVCTEAGMFAIRADHD</span><br><span class="line">FVV QEDFMKAVRKVADSKKLESKLDYKPV</span><br><span class="line">&gt;sp |P62509|ERR3_MOUSE Estrogen-related receptor gamma OS = Mus</span><br><span class="line">musculus GN = Esrrg PE = <span class="number">1</span> SV = <span class="number">1</span></span><br><span class="line">MDS VELCLPESFSLHYEEELLCRMSNKDRHIDSSCSSFIKTEPSSPASLTDSVNHHSPGG</span><br><span class="line">SSD ASGSYSSTMNGHQNGLDSPPLYPSAPILGGSGPVRKLYDDCSSTIVEDPQTKCEYML</span><br><span class="line">NSM PKRLCLVCGDIASGYHYGVASCEACKAFFKRTIQGNIEYSCPATNECEITKRRRKSC</span><br><span class="line">QACR FMKCLKVGMLKEGVRLDRVRGGRQKYKRRIDAENSPYLNPQLVQPAKKPYNKIVSH</span><br><span class="line">LLV AEPEKIYAMPDPTVPDSDIKALTTLCDLADRELVVIIGWAKHIPGFSTLSLADQMSL</span><br><span class="line">LQS AWMEILILGVVYRSLSFEDELVYADDYIMDEDQSKLAGLLDLNNAILQLVKKYKSMK</span><br><span class="line">LEK EEFVTLKAIALANSDSMHIEDVEAVQKLQDVLHEALQDYEAGQHMEDPRRAGKMLMT</span><br><span class="line">LPLLRQTSTKAVQHFYNIKLEGKVPMHKLFLEMLEAKV</span><br><span class="line"></span><br><span class="line"># C.<span class="number">8</span> AN EXAMPLE OF THE SEQRES LINES OF</span><br><span class="line">A PDB FILE (FROM FILE <span class="number">1</span>TDL)</span><br><span class="line">SEQ RES <span class="number">1</span> A <span class="number">223</span> ILE VAL GLY GLY TYR THR CYS GLY ALA ASN THR VAL PRO</span><br><span class="line">SEQ RES <span class="number">2</span> A <span class="number">223</span> TYR GLN VAL SER LEU ASN SER GLY TYR HIS PHE CYS GLY</span><br><span class="line">SEQ RES <span class="number">3</span> A <span class="number">223</span> GLY SER LEU ILE ASN SER GLN TRP VAL VAL SER ALA ALA</span><br><span class="line">SEQ RES <span class="number">4</span> A <span class="number">223</span> HIS CYS TYR LYS SER GLY ILE GLN VAL ARG LEU GLY GLU</span><br><span class="line">SEQ RES <span class="number">5</span> A <span class="number">223</span> ASP ASN ILE ASN VAL VAL GLU GLY ASN GLU GLN PHE ILE</span><br><span class="line">SEQ RES <span class="number">6</span> A <span class="number">223</span> SER ALA SER LYS SER ILE VAL HIS PRO SER TYR ASN SER</span><br><span class="line">SEQ RES <span class="number">7</span> A <span class="number">223</span> ASN THR LEU ASN ASN ASP ILE MET LEU ILE LYS LEU LYS</span><br><span class="line">SEQ RES <span class="number">8</span> A <span class="number">223</span> SER ALA ALA SER LEU ASN SER ARG VAL ALA SER ILE SER</span><br><span class="line">SEQ RES <span class="number">9</span> A <span class="number">223</span> LEU PRO THR SER CYS ALA SER ALA GLY THR GLN CYS LEU</span><br><span class="line">SEQ RES <span class="number">10</span> A <span class="number">223</span> ILE SER GLY TRP GLY ASN THR LYS SER SER GLY THR SER</span><br><span class="line">SEQ RES <span class="number">11</span> A <span class="number">223</span> TYR PRO ASP VAL LEU LYS CYS LEU LYS ALA PRO ILE LEU</span><br><span class="line">SEQ RES <span class="number">12</span> A <span class="number">223</span> SER ASP SER SER CYS LYS SER ALA TYR PRO GLY GLN ILE</span><br><span class="line">SEQ RES <span class="number">13</span> A <span class="number">223</span> THR SER ASN MET PHE CYS ALA GLY TYR LEU GLU GLY GLY</span><br><span class="line">SEQ RES <span class="number">14</span> A <span class="number">223</span> LYS ASP SER CYS GLN GLY ASP SER GLY GLY PRO VAL VAL</span><br><span class="line">SEQ RES <span class="number">15</span> A <span class="number">223</span> CYS SER GLY LYS LEU GLN GLY ILE VAL SER TRP GLY SER</span><br><span class="line">SEQ RES <span class="number">16</span> A <span class="number">223</span> GLY CYS ALA GLN LYS ASN LYS PRO GLY VAL TYR THR LYS</span><br><span class="line">SEQ RES <span class="number">17</span> A <span class="number">223</span> VAL CYS ASN TYR VAL SER TRP ILE LYS GLN THR ILE ALA</span><br><span class="line">SEQRES <span class="number">18</span> A <span class="number">223</span> SER ASN</span><br><span class="line"></span><br><span class="line"># C.<span class="number">9</span> AN EXAMPLE OF THE CUFFCOMPARE OUTPUT <span class="keyword">FOR</span> THREE SAMPLES (Q1, Q2, AND Q3)</span><br><span class="line"># Because of the length of each line, the first field of each new line is <span class="keyword">in</span> bold,and different lines are highlighted using indentation.</span><br><span class="line">Med ullo-Diff_00000001 XLOC_000001 Lypla1|uc007afh.<span class="number">1</span></span><br><span class="line"><span class="function">    q1:<span class="title">NSC.P419</span>.228|<span class="title">uc007afh</span>.1 |100|35.109496|34.188903</span></span><br><span class="line"><span class="function">    |36.030089|397.404732|2433 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.18|<span class="title">uc007afh</span>.1|100 |15.885823|15.240240</span></span><br><span class="line"><span class="function">    |16.531407|171.011325 |2433 <span class="title">q3:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P437</span>.15|<span class="title">uc007afh</span>.1|100 |18.338541|17.704857|18.97222</span></span><br><span class="line"><span class="function">    4|181.643949|2433</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000002</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.228|<span class="title">uc007afi</span>.2|</span></span><br><span class="line"><span class="function">    18|1.653393|1.409591|1.897195 |18.587029|2671</span></span><br><span class="line"><span class="function">    - <span class="title">q3:NSC</span>.<span class="title">P437</span>.108|<span class="title">uc007afi</span>.2|100 |4.624079|4.258801|4</span></span><br><span class="line"><span class="function">    .989356|45.379750|2671</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000003</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc011wht</span>.1</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.228|<span class="title">uc011wht</span>.1|100 |9.011253|8.560848</span></span><br><span class="line"><span class="function">    |9.461657|101.302266|2668 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.116|<span class="title">uc011wht</span>.1|100 |6.889020|6.503460|7.27458</span></span><br><span class="line"><span class="function">    0|73.238938 |2668<span class="title">q3:NSC</span>.<span class="title">P437</span>.108 |<span class="title">uc011wht</span>.1|90</span></span><br><span class="line"><span class="function">    |4.170527 |3.817430|4.523625|40.928694|2668</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000004</span> <span class="title">XLOC_000003</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.231|<span class="title">NSC.P419</span>.231.1</span></span><br><span class="line"><span class="function">    |100|31.891396|30.892103 |32.890690|379.023601</span></span><br><span class="line"><span class="function">    |1568<span class="title">q2:NSC</span>.<span class="title">P429</span>.121|<span class="title">NSC.P429</span>.121.1 |100|27.991543</span></span><br><span class="line"><span class="function">    |27.007869|28.975218 |313.481210|1532 -</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000005</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.236 |<span class="title">NSC.P419</span>.236.1</span></span><br><span class="line"><span class="function">    |100|1.164739|0.868895 |1.460583|13.879881|- - -</span></span><br><span class="line"><span class="function">    <span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000006</span> <span class="title">XLOC_000004</span> <span class="title">Atp6v1h</span>|<span class="title">uc007afn</span>.1</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.55|<span class="title">uc007afn</span>.1 |100|39.526818 |38.58510</span></span><br><span class="line"><span class="function">    2|40.468533|455.599775|1976 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.43|<span class="title">uc007afn</span>.1 |100|25.034945 |24.182398|25.887</span></span><br><span class="line"><span class="function">    493|271.738343|1976 <span class="title">q3:NSC</span>.<span class="title">P437</span>.37|<span class="title">uc007afn</span>.1</span></span><br><span class="line"><span class="function">    |100|20.848047 |20.043989|21.652104|205.866771|1976</span></span><br><span class="line"><span class="function">    </span></span><br><span class="line"><span class="function"></span></span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#fu-lu-d-chu-li-mu-lu-he-yong-unix-bian-cheng" class="header-anchor"> âš“ï¸</a><span id="fu-lu-d-chu-li-mu-lu-he-yong-unix-bian-cheng">é™„å½•D å¤„ç†ç›®å½•å’Œç”¨UNIXç¼–ç¨‹</span></h3>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/linux/linux-shell.html">Shell æ•™ç¨‹ | èœé¸Ÿæ•™ç¨‹ (runoob.com)</a></p>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/linux/linux-tutorial.html">Linux æ•™ç¨‹ | èœé¸Ÿæ•™ç¨‹ (runoob.com)</a></p>
<br>
<br>
        </div>
        <footer class="article-footer">
            



    <a data-url="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/" data-id="clkva0d1f0066owuu4gnm2qrw" class="article-share-link"><i class="fa fa-share"></i>åˆ†äº«åˆ°</a>
<script>
    (function ($) {
        $('body').on('click', function() {
            $('.article-share-box.on').removeClass('on');
        }).on('click', '.article-share-link', function(e) {
            e.stopPropagation();

            var $this = $(this),
                url = $this.attr('data-url'),
                encodedUrl = encodeURIComponent(url),
                id = 'article-share-box-' + $this.attr('data-id'),
                offset = $this.offset(),
                box;

            if ($('#' + id).length) {
                box = $('#' + id);

                if (box.hasClass('on')){
                    box.removeClass('on');
                    return;
                }
            } else {
                var html = [
                    '<div id="' + id + '" class="article-share-box">',
                        '<input class="article-share-input" value="' + url + '">',
                        '<div class="article-share-links">',
                            '<a href="https://twitter.com/intent/tweet?url=' + encodedUrl + '" class="article-share-twitter" target="_blank" title="Twitter"></a>',
                            '<a href="https://www.facebook.com/sharer.php?u=' + encodedUrl + '" class="article-share-facebook" target="_blank" title="Facebook"></a>',
                            '<a href="http://pinterest.com/pin/create/button/?url=' + encodedUrl + '" class="article-share-pinterest" target="_blank" title="Pinterest"></a>',
                            '<a href="https://plus.google.com/share?url=' + encodedUrl + '" class="article-share-google" target="_blank" title="Google+"></a>',
                        '</div>',
                    '</div>'
                ].join('');

              box = $(html);

              $('body').append(box);
            }

            $('.article-share-box.on').hide();

            box.css({
                top: offset.top + 25,
                left: offset.left
            }).addClass('on');

        }).on('click', '.article-share-box', function (e) {
            e.stopPropagation();
        }).on('click', '.article-share-box-input', function () {
            $(this).select();
        }).on('click', '.article-share-box-link', function (e) {
            e.preventDefault();
            e.stopPropagation();

            window.open(this.href, 'article-share-box-window-' + Date.now(), 'width=500,height=450');
        });
    })(jQuery);
</script>

        </footer>
    </div>
    <script type="application/ld+json">
    {
        "@context": "https://schema.org",
        "@type": "BlogPosting",
        "author": {
            "@type": "Person",
            "name": "è‹‘é•¿"
        },
        "headline": "è¯»ä¹¦ç¬”è®°-Pythonç”Ÿç‰©ä¿¡æ¯å­¦æ•°æ®ç®¡ç†",
        "image": "https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png",
        "keywords": "Python æ•°æ®å¤„ç†",
        "genre": "ç§‘æŠ€ ç”Ÿç‰©",
        "datePublished": "2020-10-04",
        "dateCreated": "2020-10-04",
        "dateModified": "2021-06-21",
        "url": "https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/",
        "description": "
å½“ç”Ÿç‰©ä¸Pythonäº¤ç»‡ä¹‹æ—¶
æ•…äº‹å°†å°±æ­¤å±•å¼€
",
        "wordCount": 17848
    }
</script>

</article>

    <section id="comments">
    
        
<div id="comment-container"></div>


    
    </section>



                        </div>
                    </section>
                    <aside id="sidebar">
    <a class="sidebar-toggle" title="Expand Sidebar"><i class="toggle icon"></i></a>
    <div class="sidebar-top">
        <p>å…³æ³¨æˆ‘ :</p>
        <ul class="social-links">
            
                
                <li>
                    <a class="social-tooltip" title="zhihu" href="https://www.zhihu.com/people/WoodYuan" target="_blank" rel="noopener">
                        <i class="fab fa-zhihu"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="github" href="https://github.com/YuanZ1949" target="_blank" rel="noopener">
                        <i class="fab fa-github"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="bilibili" href="https://space.bilibili.com/12549559" target="_blank" rel="noopener">
                        <i class="fab fa-bimobject"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="rss" href="/atom.xml" target="_blank" rel="noopener">
                        <i class="fas fa-rss"></i>
                    </a>
                </li>
                
            
        </ul>
    </div>
    
        
<nav id="article-nav">
    
        <a href="/ri-chang-za-ji/" id="article-nav-newer" class="article-nav-link-wrap">
        <strong class="article-nav-caption">ä¸‹ä¸€ç¯‡</strong>
        <p class="article-nav-title">
        
            æ—¥å¸¸æ‚è®°
        
        </p>
        <i class="icon fa fa-chevron-right" id="icon-chevron-right"></i>
    </a>
    
    
        <a href="/justscoresummary-ge-ren-cheng-ji-gai-lan/" id="article-nav-older" class="article-nav-link-wrap">
        <strong class="article-nav-caption">ä¸Šä¸€ç¯‡</strong>
        <p class="article-nav-title">JUSTScoreSummary-ä¸ªäººæˆç»©æ¦‚è§ˆ</p>
        <i class="icon fa fa-chevron-left" id="icon-chevron-left"></i>
        </a>
    
</nav>

    
    <div class="widgets-container">
        
            
                

            
                
    <div class="widget-wrap">
        <h3 class="widget-title">æœ€æ–°æ–‡ç« </h3>
        <div class="widget">
            <ul id="recent-post" class="">
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/di-gui-wei-xing-xing-zuo-xiang-guan-xin-xi-hui-zong/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">ç©ºå¤©</a></p>
                            <p class="item-title"><a href="/di-gui-wei-xing-xing-zuo-xiang-guan-xin-xi-hui-zong/" class="title">ä½è½¨å«æ˜Ÿæ˜Ÿåº§ç›¸å…³ä¿¡æ¯æ±‡æ€»</a></p>
                            <p class="item-date"><time datetime="2023-07-25T14:38:28.000Z" itemprop="datePublished">2023-07-25</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/ubuntu-wei-hu-bei-wang-lu/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">ä¿¡æ¯æŠ€æœ¯</a></p>
                            <p class="item-title"><a href="/ubuntu-wei-hu-bei-wang-lu/" class="title">Ubuntuç»´æŠ¤å¤‡å¿˜å½•</a></p>
                            <p class="item-date"><time datetime="2023-07-16T10:13:23.000Z" itemprop="datePublished">2023-07-16</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/qbittorrent-xiang-guan/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"></p>
                            <p class="item-title"><a href="/qbittorrent-xiang-guan/" class="title">Qbittorrentç›¸å…³</a></p>
                            <p class="item-date"><time datetime="2023-05-04T12:45:37.000Z" itemprop="datePublished">2023-05-04</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/wang-luo-fu-shu-cun-chu-nas-pei-zhi-zhi-nan/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">ä¿¡æ¯æŠ€æœ¯</a></p>
                            <p class="item-title"><a href="/wang-luo-fu-shu-cun-chu-nas-pei-zhi-zhi-nan/" class="title">ç½‘ç»œé™„å±å­˜å‚¨(NAS)é…ç½®æŒ‡å—</a></p>
                            <p class="item-date"><time datetime="2023-02-23T13:05:48.000Z" itemprop="datePublished">2023-02-23</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/shu-qian-suo-yin/" class="thumbnail">
    
    
        <span style="background-image:url(/shu-qian-suo-yin/logocat.JPG)" alt="ä¹¦ç­¾ç´¢å¼•" class="thumbnail-image"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%B4%A2%E5%BC%95/">ç´¢å¼•</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">èµ„æºç´¢å¼•</a></p>
                            <p class="item-title"><a href="/shu-qian-suo-yin/" class="title">ä¹¦ç­¾ç´¢å¼•</a></p>
                            <p class="item-date"><time datetime="2023-02-19T07:47:44.399Z" itemprop="datePublished">2023-02-19</time></p>
                        </div>
                    </li>
                
            </ul>
        </div>
    </div>

            
                
    <div class="widget-wrap widget-list">
        <h3 class="widget-title">åˆ†ç±»</h3>
        <div class="widget">
            <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E6%B4%BB/">ç”Ÿæ´»</a><span class="category-list-count">7</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E6%B4%BB/%E6%90%9E%E4%BA%8B%E6%83%85%E5%90%96/">æäº‹æƒ…å–</a><span class="category-list-count">6</span></li></ul></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/">ç§‘æŠ€</a><span class="category-list-count">22</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">ä¿¡æ¯æŠ€æœ¯</a><span class="category-list-count">12</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E5%A4%9A%E5%AA%92%E4%BD%93/">å¤šåª’ä½“</a><span class="category-list-count">4</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">ç”Ÿç‰©</a><span class="category-list-count">4</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">ç©ºå¤©</a><span class="category-list-count">2</span></li></ul></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/">ç´¢å¼•</a><span class="category-list-count">5</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%AE%9E%E6%97%B6%E4%BF%A1%E6%81%AF/">å®æ—¶ä¿¡æ¯</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%B4%A2%E5%BC%95/">å·¥ä½œæµç´¢å¼•</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E6%94%BF%E7%AD%96%E7%B4%A2%E5%BC%95/">æ”¿ç­–ç´¢å¼•</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">èµ„æºç´¢å¼•</a><span class="category-list-count">2</span></li></ul></li></ul>
        </div>
    </div>


            
                
    <div class="widget-wrap widget-float">
        <h3 class="widget-title">æ ‡ç­¾äº‘</h3>
        <div class="widget tagcloud">
            <a href="/tags/3D/" style="font-size: 10px;">3D</a> <a href="/tags/Blender/" style="font-size: 10px;">Blender</a> <a href="/tags/Blog/" style="font-size: 10px;">Blog</a> <a href="/tags/Chagas/" style="font-size: 10px;">Chagas</a> <a href="/tags/DIY/" style="font-size: 10px;">DIY</a> <a href="/tags/Excel/" style="font-size: 10px;">Excel</a> <a href="/tags/Github/" style="font-size: 10px;">Github</a> <a href="/tags/Hexo/" style="font-size: 10px;">Hexo</a> <a href="/tags/IT/" style="font-size: 10px;">IT</a> <a href="/tags/NAS/" style="font-size: 10px;">NAS</a> <a href="/tags/Python/" style="font-size: 20px;">Python</a> <a href="/tags/Ubuntu/" style="font-size: 10px;">Ubuntu</a> <a href="/tags/VBA/" style="font-size: 10px;">VBA</a> <a href="/tags/VPN/" style="font-size: 10px;">VPN</a> <a href="/tags/python/" style="font-size: 10px;">python</a> <a href="/tags/%E4%B9%A6%E7%AD%BE/" style="font-size: 10px;">ä¹¦ç­¾</a> <a href="/tags/%E4%BB%AA%E5%99%A8/" style="font-size: 10px;">ä»ªå™¨</a> <a href="/tags/%E4%BD%8E%E8%BD%A8%E5%8D%AB%E6%98%9F%E6%98%9F%E5%BA%A7/" style="font-size: 10px;">ä½è½¨å«æ˜Ÿæ˜Ÿåº§</a> <a href="/tags/%E5%85%B3%E4%BA%8E/" style="font-size: 10px;">å…³äº</a> <a href="/tags/%E5%88%86%E4%BA%AB/" style="font-size: 20px;">åˆ†äº«</a> <a href="/tags/%E5%8C%96%E5%AD%A6/" style="font-size: 10px;">åŒ–å­¦</a> <a href="/tags/%E5%8C%96%E5%AD%A6%E5%8F%91%E5%85%89/" style="font-size: 10px;">åŒ–å­¦å‘å…‰</a> <a href="/tags/%E5%8D%95%E5%8F%8D/" style="font-size: 10px;">å•å</a> <a href="/tags/%E5%91%BD%E4%BB%A4%E8%A1%8C/" style="font-size: 10px;">å‘½ä»¤è¡Œ</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 10px;">å›¾åƒå¤„ç†</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 10px;">å¤©æ–‡</a> <a href="/tags/%E5%AE%9E%E9%AA%8C/" style="font-size: 10px;">å®éªŒ</a> <a href="/tags/%E5%B7%A5%E4%BD%9C/" style="font-size: 10px;">å·¥ä½œ</a> <a href="/tags/%E5%B7%A5%E4%BD%9C%E6%B5%81/" style="font-size: 10px;">å·¥ä½œæµ</a> <a href="/tags/%E5%B7%A5%E5%85%B7/" style="font-size: 13.33px;">å·¥å…·</a> <a href="/tags/%E5%BC%80%E5%8F%91/" style="font-size: 10px;">å¼€å‘</a> <a href="/tags/%E6%91%84%E5%BD%B1/" style="font-size: 10px;">æ‘„å½±</a> <a href="/tags/%E6%94%BF%E7%AD%96/" style="font-size: 10px;">æ”¿ç­–</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" style="font-size: 10px;">æ•°æ®å¤„ç†</a> <a href="/tags/%E6%98%9F%E7%BD%91/" style="font-size: 10px;">æ˜Ÿç½‘</a> <a href="/tags/%E6%98%9F%E9%93%BE/" style="font-size: 10px;">æ˜Ÿé“¾</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">æœåŠ¡å™¨</a> <a href="/tags/%E6%B1%9F%E7%A7%91%E5%A4%A7/" style="font-size: 10px;">æ±Ÿç§‘å¤§</a> <a href="/tags/%E6%B5%8B%E8%AF%95/" style="font-size: 10px;">æµ‹è¯•</a> <a href="/tags/%E7%83%AD%E7%82%B9/" style="font-size: 10px;">çƒ­ç‚¹</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 10px;">ç‰ˆæœ¬æ§åˆ¶</a> <a href="/tags/%E7%94%9F%E6%B4%BB/" style="font-size: 10px;">ç”Ÿæ´»</a> <a href="/tags/%E7%96%BE%E7%97%85/" style="font-size: 10px;">ç–¾ç—…</a> <a href="/tags/%E7%9E%8E%E9%97%B2%E9%80%9B/" style="font-size: 10px;">çé—²é€›</a> <a href="/tags/%E7%AC%94%E8%AE%B0/" style="font-size: 16.67px;">ç¬”è®°</a> <a href="/tags/%E7%BB%88%E7%AB%AF/" style="font-size: 10px;">ç»ˆç«¯</a> <a href="/tags/%E7%BB%A9%E7%82%B9/" style="font-size: 10px;">ç»©ç‚¹</a> <a href="/tags/%E7%BD%91%E7%AB%99/" style="font-size: 10px;">ç½‘ç«™</a> <a href="/tags/%E7%BD%91%E7%BB%9C%E9%99%84%E5%B1%9E%E5%AD%98%E5%82%A8/" style="font-size: 10px;">ç½‘ç»œé™„å±å­˜å‚¨</a> <a href="/tags/%E8%87%AA%E5%8A%A8%E5%8C%96/" style="font-size: 13.33px;">è‡ªåŠ¨åŒ–</a> <a href="/tags/%E8%A7%86%E9%A2%91/" style="font-size: 10px;">è§†é¢‘</a> <a href="/tags/%E8%AE%A1%E5%88%92/" style="font-size: 10px;">è®¡åˆ’</a> <a href="/tags/%E8%AE%BE%E5%A4%87/" style="font-size: 13.33px;">è®¾å¤‡</a> <a href="/tags/%E9%80%89%E8%B4%AD/" style="font-size: 10px;">é€‰è´­</a>
        </div>
    </div>


            
                
    <div class="widget-wrap widget-list">
        <h3 class="widget-title">é“¾æ¥</h3>
        <div class="widget">
            <ul>
                
                    <li>
                        <a href="/sheng-huo/gao-shi-qing-a/zhe-shi-yi-pian-testblog/">ç½‘ç«™æµ‹è¯•é¡µ</a>
                    </li>
                
                    <li>
                        <a target="_blank" rel="noopener" href="https://lib.just.edu.cn/">æ±Ÿè‹ç§‘æŠ€å¤§å­¦å›¾ä¹¦é¦†</a>
                    </li>
                
                    <li>
                        <a target="_blank" rel="noopener" href="http://hexo.io">Hexoå®˜ç½‘</a>
                    </li>
                
            </ul>
        </div>
    </div>


            
        
    </div>
</aside>

                </div>
            </div>
        </div>
        <footer id="footer">
    <div class="container">
        <div class="container-inner">
            <a id="back-to-top" href="javascript:;"><i class="icon fa fa-angle-up"></i></a>
            <div class="credit">
                <h1 class="logo-wrap">
                    <a href="/" class="logo"></a>
                </h1>
                <p>&copy; 2023 è‹‘é•¿</p>
                
                <p>Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>. Theme by <a href="https://github.com/ppoffice" target="_blank">PPOffice</a></p>
                
            </div>
            <div class="footer-plugins">
              
    


            </div>
        </div>
    </div>
</footer>

        
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css">
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>
<script>
    var gitalk = new Gitalk({
        clientID: 'ffffe2a96b4c373c1d4d',
        clientSecret: 'fc0ec8dd07cfb61e09dd06962678251d943bef37',
        id: '05acbd50fa835f8f9bf44ef4b390abd3',
        repo: 'yuanz1949.github.io',
        owner: 'YuanZ1949',
        admin: "YuanZ1949"
    })
    gitalk.render('comment-container')
</script>





    
        
<script src="/libs/lightgallery/js/lightgallery.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-thumbnail.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-pager.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-autoplay.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-fullscreen.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-zoom.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-hash.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-share.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-video.min.js"></script>

    
    
        
<script src="/libs/justified-gallery/jquery.justifiedGallery.min.js"></script>

    
    
         <script type="text/x-mathjax-config">
            MathJax.Hub.Config({
                showProcessingMessages: false,
                messageStyle: "none",
                extensions: ["tex2jax.js"],
                jax: ["input/TeX", "output/HTML-CSS"],
                tex2jax: {
                    inlineMath:  [ ["$", "$"] ],
                    displayMath: [ ["$$","$$"] ],
                    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre','code','a'],
                    ignoreClass:"comment-content"
                },
                "HTML-CSS": {
                    availableFonts: ["STIX","TeX"],
                    showMathMenu: true
                }
            });
            MathJax.Hub.Queue(["Typeset",MathJax.Hub]);
         </script>
         <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
         <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    

    



<!-- Custom Scripts -->

<script src="/js/main.js"></script>


    </div>
</body>
</html>
