<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />

    

    
    <title>读书笔记-Python生物信息学数据管理 | 苑长的Index</title>
    
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    
    <meta name="keywords" content="Python,数据处理" />
    
    <meta name="description" content="当生物与Python交织之时 故事将就此展开">
<meta property="og:type" content="article">
<meta property="og:title" content="读书笔记-Python生物信息学数据管理">
<meta property="og:url" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/">
<meta property="og:site_name" content="苑长的Index">
<meta property="og:description" content="当生物与Python交织之时 故事将就此展开">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png">
<meta property="article:published_time" content="2020-10-04T10:00:05.000Z">
<meta property="article:modified_time" content="2021-06-21T12:47:59.553Z">
<meta property="article:author" content="苑长">
<meta property="article:tag" content="Python">
<meta property="article:tag" content="数据处理">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png">
    

    
        <link rel="alternate" href="/atom.xml" title="苑长的Index" type="application/atom+xml" />
    

    
        <link rel="icon" href="/css/images/favicon.ico" />
    

    
<link rel="stylesheet" href="/libs/font-awesome/css/all.css">

    
<link rel="stylesheet" href="/libs/titillium-web/styles.css">

    
<link rel="stylesheet" href="/libs/source-code-pro/styles.css">


    
<link rel="stylesheet" href="/css/style.css">


    
<script src="/libs/jquery/3.4.1/jquery.min.js"></script>

    
    
        
<link rel="stylesheet" href="/libs/lightgallery/css/lightgallery.min.css">

    
    
        
<link rel="stylesheet" href="/libs/justified-gallery/justifiedGallery.min.css">

    
    
    
        <script>
var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = "//hm.baidu.com/hm.js?67fc4a038906ef55e9307dac98a80fe2";
  var s = document.getElementsByTagName("script")[0];
  s.parentNode.insertBefore(hm, s);
})();
</script>

    


<meta name="generator" content="Hexo 6.3.0">
<style>.github-emoji { position: relative; display: inline-block; width: 1.2em; min-height: 1.2em; overflow: hidden; vertical-align: top; color: transparent; }  .github-emoji > span { position: relative; z-index: 10; }  .github-emoji img, .github-emoji .fancybox { margin: 0 !important; padding: 0 !important; border: none !important; outline: none !important; text-decoration: none !important; user-select: none !important; cursor: auto !important; }  .github-emoji img { height: 1.2em !important; width: 1.2em !important; position: absolute !important; left: 50% !important; top: 50% !important; transform: translate(-50%, -50%) !important; user-select: none !important; cursor: auto !important; } .github-emoji-fallback { color: inherit; } .github-emoji-fallback img { opacity: 0 !important; }</style>
</head>

<body>
    <div id="wrap">
        <header id="header">
    <div id="header-outer" class="outer">
        <div class="container">
            <div class="container-inner">
                <div id="header-title">
                    <h1 class="logo-wrap">
                        <a href="/" class="logo"></a>
                    </h1>
                    
                        <h2 class="subtitle-wrap">
                            <p class="subtitle">Life &amp; Science &amp; Technology</p>
                        </h2>
                    
                </div>
                <div id="header-inner" class="nav-container">
                    <a id="main-nav-toggle" class="nav-icon fa fa-bars"></a>
                    <div class="nav-container-inner">
                        <ul id="main-nav">
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/">主页</a>
                                </li>
                            
                                        <ul class="main-nav-list"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%94%9F%E6%B4%BB/">生活</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%94%9F%E6%B4%BB/%E6%90%9E%E4%BA%8B%E6%83%85%E5%90%96/">搞事情吖</a></li></ul></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">信息技术</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E5%A4%9A%E5%AA%92%E4%BD%93/">多媒体</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">生物</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">空天</a></li></ul></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/">索引</a><ul class="main-nav-list-child"><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%AE%9E%E6%97%B6%E4%BF%A1%E6%81%AF/">实时信息</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%B4%A2%E5%BC%95/">工作流索引</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E6%94%BF%E7%AD%96%E7%B4%A2%E5%BC%95/">政策索引</a></li><li class="main-nav-list-item"><a class="main-nav-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">资源索引</a></li></ul></li></ul>
                                    
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/suo-yin/zi-yuan-suo-yin/chang-yong-zi-yuan-suo-yin/">常用索引</a>
                                </li>
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/sheng-huo/gao-shi-qing-a/blog-gu-gu-gu/">Blog咕咕咕</a>
                                </li>
                            
                                <li class="main-nav-list-item" >
                                    <a class="main-nav-list-link" href="/sheng-huo/about/">关于</a>
                                </li>
                            
                        </ul>
                        <nav id="sub-nav">
                            <div id="search-form-wrap">

    <form class="search-form">
        <input type="text" class="ins-search-input search-form-input" placeholder="搜索" />
        <button type="submit" class="search-form-submit"></button>
    </form>
    <div class="ins-search">
    <div class="ins-search-mask"></div>
    <div class="ins-search-container">
        <div class="ins-input-wrapper">
            <input type="text" class="ins-search-input" placeholder="想要查找什么..." />
            <span class="ins-close ins-selectable"><i class="fa fa-times-circle"></i></span>
        </div>
        <div class="ins-section-wrapper">
            <div class="ins-section-container"></div>
        </div>
    </div>
</div>
<script>
(function (window) {
    var INSIGHT_CONFIG = {
        TRANSLATION: {
            POSTS: '文章',
            PAGES: '页面',
            CATEGORIES: '分类',
            TAGS: '标签',
            UNTITLED: '(未命名)',
        },
        ROOT_URL: '/',
        CONTENT_URL: '/content.json',
    };
    window.INSIGHT_CONFIG = INSIGHT_CONFIG;
})(window);
</script>

<script src="/js/insight.js"></script>


</div>
                        </nav>
                    </div>
                </div>
            </div>
        </div>
    </div>
</header>
        <div class="container">
            <div class="main-body container-inner">
                <div class="main-body-inner">
                    <section id="main">
                        <div class="main-body-header">
    <h1 class="header">
    
    <a class="page-title-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><i class="icon fa fa-angle-right"></i><a class="page-title-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">生物</a>
    </h1>
</div>

                        <div class="main-body-content">
                            <article id="post-读书笔记-Python生物信息学数据管理" class="article article-single article-type-post" itemscope itemprop="blogPost">
    <div class="article-inner">
        
            <header class="article-header">
                
    
        <h1 class="article-title" itemprop="name">
        读书笔记-Python生物信息学数据管理
        </h1>
    

            </header>
        
        
            <div class="article-meta">
                
    <div class="article-date">
      <i class="fa fa-calendar"></i>
      <a href="/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/" class="article-date">
         <time datetime="2020-10-04T10:00:05.000Z" itemprop="datePublished">2020-10-04</time>
      </a>
    </div>


                

                
    <div class="article-tag">
        <i class="fa fa-tag"></i>
        <a class="tag-link-link" href="/tags/Python/" rel="tag">Python</a>, <a class="tag-link-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" rel="tag">数据处理</a>
    </div>

                

                
   <div class="article-counter">
      <i class="fa fa-file-word"></i>
      <span class="post-count-item-text">本文字数: </span>
      <span class="post-count">18.3k</span>
      <i class="fa fa-clock"></i>
      <span class="post-time-item-text">阅读时长: </span>
      <span class="post-count">87 分钟 </span>
   </div>


            </div>
        
        
        <br>
        <div class="article-entry" itemprop="articleBody">
            <img src="/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png" class>
<blockquote><p><span style="font-size:24px"><strong>当生物与Python交织之时</strong></span></p>
<p><span style="font-size:24px"><strong>故事将就此展开</strong></span></p>
</blockquote>
<span id="more"></span>
<div class="toc">
<!-- toc -->
<ul>
<li><a href="#shu-ji-jian-jie-books">书籍简介 <span class="github-emoji"><span>📚</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4da.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a></li>
<li><a href="#xie-zai-qian-mian-writing-hand">写在前面 <span class="github-emoji"><span>✍</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/270d.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a></li>
<li><a href="#gu-shi-jiu-ci-zhan-kai-book">故事就此展开 <span class="github-emoji"><span>📖</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4d6.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></a>
<ul>
<li><a href="#di-yi-bu-fen-ru-men">第一部分 入门</a></li>
<li><a href="#di-er-bu-fen-shu-ju-guan-li">第二部分 数据管理</a></li>
<li><a href="#di-san-bu-fen-mo-kuai-hua-bian-cheng">第三部分 模块化编程</a></li>
<li><a href="#di-si-bu-fen-shu-ju-ke-shi-hua">第四部分 数据可视化</a></li>
<li><a href="#di-wu-bu-fen-biopython">第五部分 Biopython</a></li>
<li><a href="#di-liu-bu-fen-bian-cheng-mi-ji">第六部分 编程秘笈</a></li>
<li><a href="#fu-lu">附录</a></li>
</ul>
</li>
</ul>
<!-- tocstop -->
</div>
<p>接触 <code>计算机技术(IT)</code> 也已经有些年头了</p>
<p>计算机技术表现出来的 <code>高效</code> 的生产力着实让人赞叹, 让人有无穷的动力参与其中</p>
<p>如今入了生物的坑, 自然一直有探索各种生物和信息技术交融的知识领域</p>
<p>但是浅尝辄止的兴趣式的游玩, 终究是不太够用的</p>
<p>倒不如静下心来, 实际地系统地学习一些内容</p>
<h1><a href="#shu-ji-jian-jie" class="header-anchor"> ⚓️</a><span id="shu-ji-jian-jie">书籍简介 <span class="github-emoji"><span>📚</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4da.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<p>挑选的书是从 <code>江科大长山校区图书馆</code> 里翻找到的</p>
<p>位于 <code>自然科学图书第2阅览室(5楼)</code>, 5号架B面5层, 索书号: <code>Q811.4/51</code></p>
<blockquote>
<p>@信息来自<a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/search_adv.php#/index">江科大图书馆v5.6书目检索系统</a></p>
<ul>
<li>
<p>题名/责任者: <a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/item.php?marc_no=4b33443333726c7851522b2b4e57577445694d3150513d3d&amp;list=1">Python生物信息学数据管理</a>/(意) Allegra Via, (德) Kristian Rother, (意) Anna Tramontano著 卢宏超, 陈一情, 李绍娟译</p>
</li>
<li>
<p>出版发行项: 北京:电子工业出版社,2017.01</p>
</li>
<li>
<p>ISBN及定价: 978-7-121-30382-1/CNY69.00</p>
</li>
<li>
<p>载体形态项: 17, 318页:图;26cm</p>
</li>
<li>
<p>统一题名: <a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/openlink.php?title=Managing+your+biological+data+with+Python">Managing your biological data with Python</a></p>
</li>
<li>
<p>中图法分类号:<a target="_blank" rel="noopener" href="http://lib.just.edu.cn:8080/opac/openlink.php?coden=Q811.4-39">Q811.4-39</a></p>
</li>
<li>
<p>版本附注:由Taylor &amp; Francis出版集团旗下，CRC出版公司授权翻译出版</p>
</li>
<li>
<p>提要文摘附注:</p>
</li>
</ul>
<p>本书由六部分组成：Python语言基本介绍，语言所有成分介绍，高级编程，数据可视化，生物信息通用包Biopython，最后给出20个"编程秘笈”，范围涵盖了从二级结构预测、多序列比对到蛋白质三维结构的广泛话题。此外，附录还包括了大量的生物信息常用资源的信息。</p>
</blockquote>
<p>不过找也费了番功夫, 而且发现学校关于<code>生物信息学</code>方面的书还真的不是非常多</p>
<p>可见<code>生物信息学</code>的确还是比较<strong>新</strong>的一门学科</p>
<p><a target="_blank" rel="noopener" href="http://libgen.rs/book/index.php?md5=B980D099C8E284CBEE2241C4A17319BC">Library Genesis: Via, Allegra; Rother, Kristian; Tramontano, Anna - Managing Your Biological Data with Python (libgen.rs)</a></p>
<h1><a href="#xie-zai-qian-mian" class="header-anchor"> ⚓️</a><span id="xie-zai-qian-mian">写在前面 <span class="github-emoji"><span>✍</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/270d.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<p>此读书笔记会忽略亦或者增加一些内容, 比如</p>
<ul>
<li>忽略基础的<code>Python语法</code>: 本书还是使用了大量的例子来讲解Python的基础语法的, 但是对于笔者来说已无必要. 需要补充基础语法的小伙伴可以看这里 =&gt; <a target="_blank" rel="noopener" href="https://www.runoob.com/python3/python3-tutorial.html">Python3 教程 | 菜鸟教程 (runoob.com)</a></li>
<li>增加一些涉及到的生物学理论细节: 这主要是补充书中的内容, 使得易于理解相关的例子</li>
</ul>
<p>当然, 其实主要是什么值得补充就会补充的 :happy:</p>
<p>本笔记按照原书的目录顺序展开, 但是仅保留少部分的标题, 主要以<code>技术流程</code>来整合笔记</p>
<p>本书的代码使用的Python的版本为 <code>Python 2.6.6</code></p>
<p>笔者使用由 <code>Anaconda</code> 构建的Python版本为 <code>python-2.7.13</code> 的虚拟环境来复现书中的代码</p>
<p>期间可能会尝试使用 <code>Python3.8</code> 的版本来重构里面的代码</p>
<p>没有特别说明, 笔记里的Python版本则为 <code>python-2.7.13</code></p>
<p>事实上, 掌握充足的生物学知识是非常重要的</p>
<p>计算机技术只是帮助我们自动化地完成了基于生物学认识的工作</p>
<p>本书也只是讲解了如何将Python引入分析以及数据管理的工作流</p>
<p>当你掌握真正的算法和理论实验背景, 你将可以更好地利用Python这个强大的工具</p>
<p><strong>多多做实验, 看看理论书吧</strong></p>
<h1><a href="#gu-shi-jiu-ci-zhan-kai" class="header-anchor"> ⚓️</a><span id="gu-shi-jiu-ci-zhan-kai">故事就此展开 <span class="github-emoji"><span>📖</span><img src="https://github.githubassets.com/images/icons/emoji/unicode/1f4d6.png?v8" aria-hidden="true" onerror="this.parent.classList.add('github-emoji-fallback')"></span></span></h1>
<h2><a href="#di-yi-bu-fen-ru-men" class="header-anchor"> ⚓️</a><span id="di-yi-bu-fen-ru-men">第一部分 入门</span></h2>
<p>本书的作者描述了他希望这本书带给读者什么</p>
<blockquote>
<p>我们希望本书能成为帮助读者每天攀登数据之山的地图, 希望读者能让编程为工作提供便利, 而不必先成为一个专业的软件开发者</p>
</blockquote>
<p>作者建议像学习德语一样, 学习Python语法, 把<code>Python标准库</code> 当作字典来查</p>
<p><code>Python标准库</code> :</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/3.7/library/index.html">Python 标准库 — Python 3.7.8 文档</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/index.html">Python 标准库 — Python 2.7.18 文档</a></p>
</li>
</ul>
<p><code>书中的代码</code>和<code>一些总结</code>参考:</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://ugoproto.github.io/ugo_py_doc/managing_your_biological_data_with_python/">Managing Your Biological Data with Python - ugo_py_doc (ugoproto.github.io)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://github.com/wqhf/Managing-Your-Biological-Data-with-Python">Managing-Your-Biological-Data-with-Python (github.com)</a></p>
</li>
</ul>
<h3><a href="#di-1-zhang-python-shell" class="header-anchor"> ⚓️</a><span id="di-1-zhang-python-shell">第1章 Python shell</span></h3>
<p>关于<code>Python</code>的安装可以参考这篇文章 <a href="/anaconda-zhi-lu/" title="Anaconda之路">Anaconda之路</a></p>
<p>主要介绍了如何使用<code>Anaconda</code>来构建<code>Python</code>虚拟环境</p>
<blockquote>
<p><code>Anaconda</code>是一个可以管理<code>Python包</code>和<code>Python环境</code>的软件</p>
<p>所谓<code>Python包</code>指的是<strong>别人写</strong>(第三方)的有一定功能的<code>Python代码</code></p>
<p>所谓<code>Python环境</code>其实就是<code>Python代码</code>运行时调用的具体版本的<code>Python解释器</code></p>
<p>所谓<code>管理</code>指的就是我们可以通过<code>Anaconda</code>的<code>conda</code>指令, 随意地创建包含有我们指定的<code>Python包</code>的指定版本的<code>Python环境</code>, 这样做的目的是解决兼容性问题(相互依赖问题)</p>
<p>比如这本书的代码是<code>Python2</code>时代编写的, 我自然是不可以用<code>Python3</code>的解释器来运行的.</p>
<p>那就可以创建一个<code>Python2</code>的环境来复现书中的代码, 然后再创建一个<code>Python3</code>的环境来重新编写他们(重构)</p>
</blockquote>
<p>这样我们就可以复现书中的代码了, 而且还可以轻松的重构老旧的代码</p>
<h4><a href="#1-2-an-li-ji-suan-atp-shui-jie-de-dg" class="header-anchor"> ⚓️</a><span id="1-2-an-li-ji-suan-atp-shui-jie-de-dg">1.2 案例： 计算ATP水解的ΔG</span></h4>
<ul>
<li><code>吉布斯自由能</code>:</li>
</ul>
<blockquote>
<p>在<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%83%AD%E5%8A%9B%E5%AD%A6">热力学</a>里，<strong>吉布斯能</strong>（英语：<strong>Gibbs Free Energy</strong>），又称<strong>吉布斯自由能</strong>、<strong>吉布斯函数</strong>、<strong>自由焓</strong>，常用英文字母<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/f5f3c8921a3b352de45446a6789b104458c9f90b" alt>标记。吉布斯能是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%9C%8B%E9%9A%9B%E7%B4%94%E5%8C%96%E5%AD%B8%E5%92%8C%E6%87%89%E7%94%A8%E5%8C%96%E5%AD%B8%E8%81%AF%E5%90%88%E6%9C%83">国际化学联会</a>建议采用的名称。吉布斯能是描述系统的热力性质的一种<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%86%B1%E5%8A%9B%E5%8B%A2">热力势</a>，定义为[<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-cn/%E5%90%89%E5%B8%83%E6%96%AF%E8%83%BD#cite_note-Greiner1995-1">1]</a>:101[<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-cn/%E5%90%89%E5%B8%83%E6%96%AF%E8%83%BD#cite_note-Perrot-2">2]</a>:128-129</p>
<p><img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/6212e95f1f023c632b04fa0600e5a37259383a6c" alt></p>
<p>其中，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/458a728f53b9a0274f059cd695e067c430956025" alt>是系统的<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%86%85%E8%83%BD">内能</a>，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/ec7200acd984a1d3a3d7dc455e262fbe54f7f6e0" alt>是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%B5%95%E5%B0%8D%E6%BA%AB%E5%BA%A6">绝对温度</a>，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/4611d85173cd3b508e67077d4a1252c9c05abca2" alt>是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%86%B5">熵</a>，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/81eac1e205430d1f40810df36a0edffdc367af36" alt>是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%8E%8B%E5%BC%BA">压强</a>，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/af0f6064540e84211d0ffe4dac72098adfa52845" alt>是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E4%BD%93%E7%A7%AF">体积</a>，<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/75a9edddcca2f782014371f75dca39d7e13a9c1b" alt>是<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E7%84%93">焓</a>。</p>
</blockquote>
<p>$$<br>
\Delta G=\Delta G^{0}+R T * \ln \left([A D P] *\left[P_{i}\right] /[A T P]\right)<br>
$$</p>
<p>$$<br>
[?]: 指 ? 的摩尔浓度   \\  ΔG : 吉布斯自由能  \\  ΔG^0 : 标准反应吉布斯能   \\  R : 摩尔气体常数  \\  T : 温度  \\<br>
$$</p>
<ul>
<li>python的 <code>math</code> 模块:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">import</span> math  <span class="comment"># 导入数学计算模块</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">dir</span>(math)    <span class="comment"># dir()函数可以获取传入的模块的内部的函数</span></span><br><span class="line">[<span class="string">'__doc__'</span>, <span class="string">'__name__'</span>, <span class="string">'__package__'</span>, <span class="string">'acos'</span>, <span class="string">'acosh'</span>, <span class="string">'asin'</span>, <span class="string">'asinh'</span>, <span class="string">'atan'</span>, <span class="string">'atan2'</span>, <span class="string">'atanh'</span>, <span class="string">'ceil'</span>, <span class="string">'copysign'</span>, <span class="string">'cos'</span>, <span class="string">'cosh'</span>, <span class="string">'degrees'</span>, <span class="string">'e'</span>, <span class="string">'erf'</span>, <span class="string">'erfc'</span>, <span class="string">'exp'</span>, <span class="string">'expm1'</span>, <span class="string">'fabs'</span>, <span class="string">'factorial'</span>, <span class="string">'floor'</span>, <span class="string">'fmod'</span>, <span class="string">'frexp'</span>, <span class="string">'fsum'</span>, <span class="string">'gamma'</span>, <span class="string">'hypot'</span>, <span class="string">'isinf'</span>, <span class="string">'isnan'</span>, <span class="string">'ldexp'</span>, <span class="string">'lgamma'</span>, <span class="string">'log'</span>, <span class="string">'log10'</span>, <span class="string">'log1p'</span>, <span class="string">'modf'</span>, <span class="string">'pi'</span>, <span class="string">'pow'</span>, <span class="string">'radians'</span>, <span class="string">'sin'</span>, <span class="string">'sinh'</span>, <span class="string">'sqrt'</span>, <span class="string">'tan'</span>, <span class="string">'tanh'</span>, <span class="string">'trunc'</span>]</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>math.log(...)  <span class="comment"># 传入参数, 返回它的自然对数</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">help</span>(math.log) <span class="comment"># 使用`帮助`这个内建函数查看某个函数的使用说明</span></span><br><span class="line">Help on built-<span class="keyword">in</span> function log <span class="keyword">in</span> module math:</span><br><span class="line">log(...)</span><br><span class="line">  log(x[, base])</span><br><span class="line">    Return the logarithm of x to the given base.</span><br><span class="line">    If the base <span class="keyword">not</span> specified, returns the natural logarithm (base e) of x.</span><br></pre></td></tr></tbody></table></figure>
<h4><a href="#1-3-4-ji-suan" class="header-anchor"> ⚓️</a><span id="1-3-4-ji-suan">1.3.4 计算</span></h4>
<p>这里引入一篇博文来补充各种<code>符号</code>的意思</p>
<p><a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_38981172/article/details/90176525">python中的各种符号（欢迎补充）_杉木人-CSDN博客</a></p>
<p>特别之处在Python中, <code>^</code> 符号</p>
<table>
<thead>
<tr>
<th style="text-align:center">符号</th>
<th style="text-align:center">功能</th>
<th style="text-align:center">例子</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center"><code>&lt;sup&gt;</code></td>
<td style="text-align:center">按位异或运算符：当两对应的二进位相异时，结果为1</td>
<td style="text-align:center">(a  b) 输出结果 49 ，二进制解释： 0011 0001</td>
</tr>
<tr>
<td style="text-align:center"><code>**</code></td>
<td style="text-align:center">幂 - 返回x的y次幂</td>
<td style="text-align:center">a**b 为10的20次方， 输出结果 100000000000000000000</td>
</tr>
</tbody>
</table>
<p>参见 <code>Python2.7.18文档</code> @<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/operator.html?highlight=%E8%BF%90%E7%AE%97%E7%AC%A6#mapping-operators-to-functions">9.9.1. 将运算符映射到函数</a></p>
<ul>
<li>计算流程:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>ATP = <span class="number">3.5</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>ADP = <span class="number">1.8</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>Pi = <span class="number">5.0</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>R = <span class="number">0.00831</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>T = <span class="number">298</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>deltaG0 = -<span class="number">30.5</span></span><br><span class="line">&gt;&gt;&gt;</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">import</span> math</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>deltaG0 + R * T * math.log(ADP * Pi / ATP)</span><br><span class="line">-<span class="number">28.161154161098693</span></span><br></pre></td></tr></tbody></table></figure>
<p>通过导入各种参数, 可以直接计算出<code>吉布斯自由能</code></p>
<h3><a href="#di-2-zhang-di-yi-ge-python-cheng-xu" class="header-anchor"> ⚓️</a><span id="di-2-zhang-di-yi-ge-python-cheng-xu">第2章 第一个Python程序</span></h3>
<h4><a href="#2-2-an-li-ru-he-ji-suan-yi-dao-su-xu-lie-zhong-de-an-ji-suan-pin-lu" class="header-anchor"> ⚓️</a><span id="2-2-an-li-ru-he-ji-suan-yi-dao-su-xu-lie-zhong-de-an-ji-suan-pin-lu">2.2案例： 如何计算胰岛素序列中的氨基酸频率 *</span></h4>
<ul>
<li>计算流程:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># insulin [Homo sapiens] GI:386828</span></span><br><span class="line"><span class="comment"># extracted 51 amino acids of A+B chain</span></span><br><span class="line">insulin = <span class="string">"GIVEQCCTSICSLYQLENYCNFVNQHLCGSHLVEALYLVCGERGFFYTPKT"</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> amino_acid <span class="keyword">in</span> <span class="string">"ACDEFGHIKLMNPQRSTVWY"</span>:</span><br><span class="line">    number = insulin.count(amino_acid)</span><br><span class="line">    <span class="built_in">print</span> amino_acid,number</span><br><span class="line"></span><br><span class="line">A <span class="number">1</span></span><br><span class="line">C <span class="number">6</span></span><br><span class="line">D <span class="number">0</span></span><br><span class="line">E <span class="number">4</span></span><br><span class="line">F <span class="number">3</span></span><br><span class="line">G <span class="number">4</span></span><br><span class="line">H <span class="number">2</span></span><br><span class="line">I <span class="number">2</span></span><br><span class="line">K <span class="number">1</span></span><br><span class="line">L <span class="number">6</span></span><br><span class="line">M <span class="number">0</span></span><br><span class="line">N <span class="number">3</span></span><br><span class="line">P <span class="number">1</span></span><br><span class="line">Q <span class="number">3</span></span><br><span class="line">R <span class="number">1</span></span><br><span class="line">S <span class="number">3</span></span><br><span class="line">T <span class="number">3</span></span><br><span class="line">V <span class="number">4</span></span><br><span class="line">W <span class="number">0</span></span><br><span class="line">Y <span class="number">4</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>“字符串1”.count(“字符串2”) : 返回<code>字符串1</code>中含有多少<code>字符串2</code>的数量</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/string.html?highlight=count#string.count">7.1. string — 常见的字符串操作 — Python 2.7.18 文档</a></p>
</li>
<li>
<p><code>string.``count</code>(<em>s</em>, <em>sub</em>[, <em>start</em>[, <em>end</em>]])</p>
<p>Return the number of (non-overlapping) occurrences of substring <em>sub</em> in string <code>s[start:end]</code>. Defaults for <em>start</em> and <em>end</em> and interpretation of negative values are the same as for slices.</p>
</li>
</ul>
</li>
<li>
<p>注意缩进错误：IndentationError: unindent does not match any outer indentation level</p>
</li>
</ul>
<p>切不可混用 <code>空格缩进</code>和<code>制表符缩进</code>, 否则你会疯的</p>
<ul>
<li>Python 的反斜杠换行:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="number">3</span> + <span class="number">5</span> + \</span><br><span class="line"><span class="meta">... </span><span class="number">7</span></span><br><span class="line"><span class="number">15</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>内建函数(build-in): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#zip">2. 内置函数 — Python 2.7.18 文档</a></p>
<p>比如 <code>len()</code> 可以直接调用, 并返回结果</p>
</li>
<li>
<p>对象方法: 某个对象(类的实例)的内部的方法(函数), 通过<code>.</code>符号调用, 如 <code>" ".count(' ')</code></p>
</li>
<li>
<p><code>in</code>迭代:</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> ... <span class="keyword">in</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="comment"># `in` 可以迭代字符串等</span></span><br></pre></td></tr></tbody></table></figure>
<p>补充 in</p>
<ul>
<li><code>,</code> 对 <code>print</code> 输出的影响:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span> <span class="number">1</span>,<span class="number">2</span>,   <span class="comment"># 注意末尾的逗号可以抑制换行符</span></span><br><span class="line"><span class="built_in">print</span> <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="number">1</span> <span class="number">2</span> <span class="number">3</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p>转义字符<code>\</code>:</p>
</li>
<li>
<p>引号之间的嵌套:</p>
</li>
<li>
<p><code>random模块</code>:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/random.html#module-random">9.6. random — 生成伪随机数 — Python 2.7.18 文档</a></p>
</li>
</ul>
<h4><a href="#2-5-zi-ce-ti" class="header-anchor"> ⚓️</a><span id="2-5-zi-ce-ti">2.5自测题</span></h4>
<blockquote>
<p>2009年, Elizabeth H. Blackburn, Carol W. Greide 和 Jack W. Szostak 因发现端粒酶功能而获得诺贝尔奖, 端粒酶负责延伸染色体的末端. 通过NCBI的蛋白质数据库检索人类亚型I的端粒的11132个残基序列, 试问其中哪种氨基酸出现的最为频繁?</p>
</blockquote>
<h2><a href="#di-er-bu-fen-shu-ju-guan-li" class="header-anchor"> ⚓️</a><span id="di-er-bu-fen-shu-ju-guan-li">第二部分 数据管理</span></h2>
<h3><a href="#di-3-zhang-fen-xi-shu-ju-lie" class="header-anchor"> ⚓️</a><span id="di-3-zhang-fen-xi-shu-ju-lie">第3章 分析数据列</span></h3>
<p>译者推荐标准库模块 <code>collections</code> :</p>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/collections.html#module-collections">8.3. collections — High-performance container datatypes — Python 2.7.18 文档</a></p>
<h4><a href="#3-2-an-li-shu-tu-chang-du" class="header-anchor"> ⚓️</a><span id="3-2-an-li-shu-tu-chang-du">3.2案例： 树突长度</span></h4>
<p>作者使用了 <code>ImageJ</code> 来计算树突的长度并把值传入文本文件</p>
<p>软件指路: <a target="_blank" rel="noopener" href="https://imagej.net/Welcome">ImageJ</a></p>
<ul>
<li><code>in</code> 迭代:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># for循环中按行迭代文件的方法</span></span><br><span class="line"><span class="keyword">for</span> ... <span class="keyword">in</span> <span class="built_in">open</span>(...):</span><br><span class="line">    ...</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>内建函数(内置函数): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=max#max">2. 内置函数 — Python 2.7.18 文档</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 内建函数 </span></span><br><span class="line"><span class="built_in">sum</span>(); <span class="built_in">min</span>(); <span class="built_in">max</span>(); <span class="built_in">len</span>()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>%</code> 格式化输出::</p>
</li>
<li>
<p>文件的<code>readlines</code> 方法: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=readlines#file.readlines">file.readlines([sizehint]) — Python 2.7.18 文档</a></p>
</li>
</ul>
<p>此方法可以将文件的内容<code>按行</code>组合成<code>列表</code>返回</p>
<ul>
<li>
<p><code>硬编码(hard-coding)</code>:</p>
</li>
<li>
<p><strong>不关闭文件</strong>的隐患:</p>
</li>
</ul>
<p>当在其他时候读取文件的时候, 之前写入的部分内容因为没有超过缓冲区大小, 还处于内存内.</p>
<p>这将导致有时候你将无法获取想要的内容, 切记及时关闭文件.</p>
<ul>
<li>
<p><code>缓冲区</code>:</p>
</li>
<li>
<p>文件的 <code>writelines</code> 方法:</p>
</li>
</ul>
<p>将<code>列表</code>按<code>项</code>无空格连续写入<code>文件</code>, 列表内只支持 <code>字符串</code> .</p>
<blockquote>
<p><code>file.writelines</code>(<em>sequence</em>)<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.writelines">¶</a></p>
<p>Write a sequence of strings to the file. The sequence can be any iterable object producing strings, typically a list of strings. There is no return value. (The name is intended to match <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.readlines"><code>readlines()</code></a>; <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=writelines#file.writelines"><code>writelines()</code></a> does not add line separators.)</p>
</blockquote>
<ul>
<li><code>" ".join()</code>方法:</li>
</ul>
<p>返回一个由<code>可迭代对象</code>中的字符串<code>串联" "</code>组成的字符串。</p>
<blockquote>
<p>str.join(<em>iterable</em>)<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=join#str.join">¶</a></p>
<p>Return a string which is the concatenation of the strings in <em>iterable</em>. If there is any Unicode object in <em>iterable</em>, return a Unicode instead. A <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/exceptions.html#exceptions.TypeError"><code>TypeError</code></a> will be raised if there are any non-string or non Unicode object values in <em>iterable</em>. The separator between elements is the string providing this method.</p>
</blockquote>
<h3><a href="#di-4-zhang-jie-xi-shu-ju-ji-lu" class="header-anchor"> ⚓️</a><span id="di-4-zhang-jie-xi-shu-ju-ji-lu">第4章 解析数据记录</span></h3>
<h4><a href="#4-2-an-li-zheng-he-zhi-pu-shu-ju-zhuan-hua-dao-dai-xie-tong-lu-zhong" class="header-anchor"> ⚓️</a><span id="4-2-an-li-zheng-he-zhi-pu-shu-ju-zhuan-hua-dao-dai-xie-tong-lu-zhong">4.2 案例： 整合质谱数据， 转化到代谢通路中</span></h4>
<ul>
<li>
<p><code>UniProt</code> 形式:  <a target="_blank" rel="noopener" href="https://www.uniprot.org/">UniProt</a></p>
<blockquote>
<p>The mission of <a target="_blank" rel="noopener" href="https://www.uniprot.org/help/about">UniProt</a> is to provide the scientific community with a comprehensive, high-quality and freely accessible resource of protein sequence and functional information.</p>
</blockquote>
<ul>
<li>~ AC:</li>
</ul>
</li>
<li>
<p><code>CSV</code> 逗号分隔文件:</p>
</li>
<li>
<p><code>TSV</code> 制表符分隔文件:</p>
</li>
<li>
<p><code>MS</code> 质谱:</p>
</li>
<li>
<p>作者提到的软件与网站:</p>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="http://www.matrixscience.com/">Mascot search engine</a>: Protein identification software for <strong>mass spec</strong>(质谱) data</p>
<blockquote>
<p>Welcome to the home of Mascot software, the benchmark for identification, characterisation and quantitation of proteins using mass spectrometry data. Here, you can learn more about the tools developed by Matrix Science to get the best out of your data, whatever your chosen instrument.</p>
</blockquote>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://reactome.org/">Reactome Pathway Database</a>:  <code>代谢通路</code>资源获取</p>
<blockquote>
<p>Reactome is a free, open-source, curated and peer-reviewed pathway database. Our goal is to provide intuitive bioinformatics tools for the visualization, interpretation and analysis of pathway knowledge to support basic research, genome analysis, modeling, systems biology and education.</p>
<p>If you use Reactome in Asia, we suggest using our Chinese mirror site at <a target="_blank" rel="noopener" href="http://reactome.ncpsb.org/">reactome.ncpsb.org</a>.</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># proteins participating in cell cycle 参与细胞通路的蛋白</span></span><br><span class="line">list_a =[]</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">"cell_cycle_proteins.txt"</span>) :</span><br><span class="line">    list_a.append(line.strip())</span><br><span class="line"><span class="built_in">print</span> list_a</span><br><span class="line"></span><br><span class="line"><span class="comment"># proteins expressed in a given cancer cell 在给定的细胞中表达的蛋白</span></span><br><span class="line">list_b =[]</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">"cancer_cell_proteins.txt"</span>) :</span><br><span class="line">    list_b.append(line.strip())</span><br><span class="line"><span class="built_in">print</span> list_b</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> protein <span class="keyword">in</span> list_a:  <span class="comment">#迭代列表a</span></span><br><span class="line">    <span class="keyword">if</span> protein <span class="keyword">in</span> list_b:  <span class="comment">#判断是不是在列表b中</span></span><br><span class="line">        <span class="built_in">print</span> protein, <span class="string">'detected in the cancer cell'</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">print</span> protein, <span class="string">'not observed'</span></span><br></pre></td></tr></tbody></table></figure>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#if">7.1. The <code>if</code> statement</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># if 语句</span></span><br><span class="line"><span class="keyword">if</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">elif</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">elif</span> ...:</span><br><span class="line">    ...</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    ....</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>列表推导式: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/datastructures.html?highlight=%E5%88%97%E8%A1%A8#list-comprehensions">5.1.4. 列表推导式</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">bases = [<span class="string">'A'</span>,<span class="string">'C'</span>,<span class="string">'T'</span>,<span class="string">'G'</span>]</span><br><span class="line">seq = <span class="string">'GGACXCAGXXGATT'</span></span><br><span class="line">seqlist = [base <span class="keyword">for</span> base <span class="keyword">in</span> seq <span class="keyword">if</span> base <span class="keyword">in</span> bases]</span><br><span class="line"><span class="built_in">print</span> seqlist</span><br><span class="line"></span><br><span class="line">[<span class="string">'G'</span>, <span class="string">'G'</span>, <span class="string">'A'</span>, <span class="string">'C'</span>, <span class="string">'C'</span>, <span class="string">'A'</span>, <span class="string">'G'</span>, <span class="string">'G'</span>, <span class="string">'A'</span>, <span class="string">'T'</span>, <span class="string">'T'</span>]</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>ESR1</code>蛋白:</p>
<blockquote>
<p>Protein       : Estrogen receptor  雌激素受体</p>
<p>Gene          : ESR1</p>
<p>Organism  : <em>Homo sapiens (Human)</em></p>
</blockquote>
<ul>
<li>
<p>蛋白详情<a target="_blank" rel="noopener" href="https://www.uniprot.org/uniprot/P03372">ESR1 - Estrogen receptor - Homo sapiens (Human) - ESR1 gene &amp; protein (uniprot.org)</a></p>
</li>
<li>
<p><code>ESR1</code>蛋白序列: <a target="_blank" rel="noopener" href="https://www.uniprot.org/uniprot/P03372.fasta">https://www.uniprot.org/uniprot/P03372.fasta</a></p>
<blockquote>
<p>&gt;sp|P03372|ESR1_HUMAN Estrogen receptor OS=Homo sapiens OX=9606 GN=ESR1 PE=1 SV=2</p>
<p>MTMTLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY<br>
EFNAAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF<br>
LQPHGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK<br>
ETRYCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC<br>
RLRKCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR<br>
SKKNSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW<br>
AKRVPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG<br>
MVEIFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD<br>
KITDTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL<br>
LEMLDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 读取整个FASTA记录, 找到里面所属为'Homo sapiens'的序列</span></span><br><span class="line"></span><br><span class="line">fasta_file = <span class="built_in">open</span>(<span class="string">'SwissProt.fasta'</span>,<span class="string">'r'</span>)  <span class="comment"># 读取多序列Fasta文件</span></span><br><span class="line">out_file = <span class="built_in">open</span>(<span class="string">'SwissProtHuman.fasta'</span>,<span class="string">'w'</span>) <span class="comment"># 输出文件准备</span></span><br><span class="line"></span><br><span class="line">seq = <span class="string">''</span>  <span class="comment"># 序列存贮变量</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> fasta_file:  <span class="comment">#按行迭代</span></span><br><span class="line">    <span class="keyword">if</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span> <span class="keyword">and</span> seq == <span class="string">''</span>:  <span class="comment"># 判断是不是当前在读记录的第一行</span></span><br><span class="line">        header = line</span><br><span class="line">    <span class="keyword">elif</span> line[<span class="number">0</span>] != <span class="string">'&gt;'</span>:  <span class="comment"># 判断不是首行(没有'&gt;'), 直接将此行放入序列</span></span><br><span class="line">        seq = seq + line</span><br><span class="line">    <span class="keyword">elif</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span> <span class="keyword">and</span> seq != <span class="string">''</span>: <span class="comment"># 如果是首行且序列是空的话, 说明上一条的记录读完了</span></span><br><span class="line">        <span class="keyword">if</span> <span class="string">"Homo sapiens"</span> <span class="keyword">in</span> header:   <span class="comment"># 这里先判断上一条记录的记录是不是我们要的,是就存起来</span></span><br><span class="line">            out_file.write(header + seq)</span><br><span class="line">        seq = <span class="string">''</span>                       <span class="comment"># 清空</span></span><br><span class="line">        header = line                  <span class="comment"># 上一条纪录搞完啦, 把现在在读的这一行保存至haeder</span></span><br><span class="line">        </span><br><span class="line"><span class="keyword">if</span> <span class="string">"Homo sapiens"</span> <span class="keyword">in</span> header:           <span class="comment"># 全部读完了, 但是最后一条记录还没处理, 这里处理掉</span></span><br><span class="line">    out_file.write(header + seq)</span><br><span class="line">out_file.close()</span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#di-5-zhang-sou-suo-shu-ju" class="header-anchor"> ⚓️</a><span id="di-5-zhang-sou-suo-shu-ju">第5章 搜索数据</span></h3>
<h4><a href="#5-2-an-li-jiang-rna-xu-lie-fan-yi-wei-xiang-ying-de-dan-bai-zhi-xu-lie" class="header-anchor"> ⚓️</a><span id="5-2-an-li-jiang-rna-xu-lie-fan-yi-wei-xiang-ying-de-dan-bai-zhi-xu-lie">5.2案例： 将RNA序列翻译为相应的蛋白质序列</span></h4>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/nuccore/A06662.1?from=470&amp;to=577&amp;report=fasta">Synthetic nucleotide sequence of the human GSH transferase pi gene - Nucleotide - NCBI (nih.gov)</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">codon_table={</span><br><span class="line"><span class="string">'GCU'</span>:<span class="string">'A'</span>,<span class="string">'GCC'</span>:<span class="string">'A'</span>,<span class="string">'GCA'</span>:<span class="string">'A'</span>,<span class="string">'GCG'</span>:<span class="string">'A'</span>,<span class="string">'CGU'</span>:<span class="string">'R'</span>,</span><br><span class="line"><span class="string">'CGC'</span>:<span class="string">'R'</span>,<span class="string">'CGA'</span>:<span class="string">'R'</span>,<span class="string">'CCS'</span>:<span class="string">'R'</span>,<span class="string">'AGA'</span>:<span class="string">'R'</span>,<span class="string">'AGG'</span>:<span class="string">'R'</span>,</span><br><span class="line"><span class="string">'UCU'</span>:<span class="string">'S'</span>,<span class="string">'UCC'</span>:<span class="string">'S'</span>,<span class="string">'UCA'</span>:<span class="string">'S'</span>,<span class="string">'UCG'</span>:<span class="string">'S'</span>,<span class="string">'AGU'</span>:<span class="string">'S'</span>,</span><br><span class="line"><span class="string">'AGC'</span>:<span class="string">'S'</span>,<span class="string">'AUU'</span>:<span class="string">'T'</span>,<span class="string">'AUC'</span>:<span class="string">'I'</span>,<span class="string">'AUA'</span>:<span class="string">'T'</span>,<span class="string">'AUU'</span>:<span class="string">'I'</span>,</span><br><span class="line"><span class="string">'AUC'</span>:<span class="string">'I'</span>,<span class="string">'AUA'</span>:<span class="string">'I'</span>,<span class="string">'UUA'</span>:<span class="string">'L'</span>,<span class="string">'UUG'</span>:<span class="string">'L'</span>,<span class="string">'CUU'</span>:<span class="string">'L'</span>,</span><br><span class="line"><span class="string">'CUC'</span>:<span class="string">'L'</span>,<span class="string">'CUA'</span>:<span class="string">'L'</span>,<span class="string">'CUG'</span>:<span class="string">'L'</span>,<span class="string">'GGU'</span>:<span class="string">'G'</span>,<span class="string">'GGC'</span>:<span class="string">'G'</span>,</span><br><span class="line"><span class="string">'GGA'</span>:<span class="string">'G'</span>,<span class="string">'GSG'</span>:<span class="string">'G'</span>,<span class="string">'GUU'</span>:<span class="string">'V'</span>,<span class="string">'GUC'</span>:<span class="string">'V'</span>,<span class="string">'GUA'</span>:<span class="string">'V'</span>,</span><br><span class="line"><span class="string">'GUG'</span>:<span class="string">'V'</span>,<span class="string">'ACU'</span>:<span class="string">'T'</span>,<span class="string">'ACC'</span>:<span class="string">'T'</span>,<span class="string">'ACA'</span>:<span class="string">'T'</span>,<span class="string">'ACG'</span>:<span class="string">'T'</span>,</span><br><span class="line"><span class="string">'CCU'</span>:<span class="string">'P'</span>,<span class="string">'CCC'</span>:<span class="string">'P'</span>,<span class="string">'CCA'</span>:<span class="string">'P'</span>,<span class="string">'CCG'</span>:<span class="string">'P'</span>,<span class="string">'AAU'</span>:<span class="string">'N'</span>,</span><br><span class="line"><span class="string">'AAC'</span>:<span class="string">'N'</span>,<span class="string">'GAU'</span>:<span class="string">'D'</span>,<span class="string">'GAC'</span>:<span class="string">'D'</span>,<span class="string">'UGU'</span>:<span class="string">'C'</span>,<span class="string">'UGC'</span>:<span class="string">'C'</span>,</span><br><span class="line"><span class="string">'CAA'</span>:<span class="string">'Q'</span>,<span class="string">'CAG'</span>:<span class="string">'Q'</span>,<span class="string">'GAA'</span>:<span class="string">'E'</span>,<span class="string">'GAG'</span>:<span class="string">'E'</span>,<span class="string">'CAU'</span>:<span class="string">'H'</span>,</span><br><span class="line"><span class="string">'CAC'</span>:<span class="string">'H'</span>,<span class="string">'AAA'</span>:<span class="string">'K'</span>,<span class="string">'AAG'</span>:<span class="string">'K'</span>,<span class="string">'UUU'</span>:<span class="string">'F'</span>,<span class="string">'UUC'</span>:<span class="string">'F'</span>,</span><br><span class="line"><span class="string">'UAU'</span>:<span class="string">'Y'</span>,<span class="string">'UAC'</span>:<span class="string">'Y'</span>,<span class="string">'AUG'</span>:<span class="string">'M'</span>,<span class="string">'UGG'</span>:<span class="string">'W'</span>,</span><br><span class="line"><span class="string">'UAG'</span>:<span class="string">'STOP'</span>,<span class="string">'UGA'</span>:<span class="string">'STOP'</span>,<span class="string">'UAA'</span>:<span class="string">'STOP'</span>}</span><br><span class="line"></span><br><span class="line">rna = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">'A06662-RNA.fasta'</span>):</span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> line.startswith(<span class="string">'&gt;'</span>):</span><br><span class="line">        rna = rna +line.strip()</span><br><span class="line"><span class="comment"># 一次翻译进行一次阅读框读取</span></span><br><span class="line"><span class="keyword">for</span> frame <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">3</span>):</span><br><span class="line">    prot = <span class="string">''</span></span><br><span class="line">    <span class="built_in">print</span> <span class="string">'Reading frame '</span> + <span class="built_in">str</span>(frame + <span class="number">1</span>)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(frame, <span class="built_in">len</span>(rna), <span class="number">3</span>):</span><br><span class="line">        codon = rna[i:i+<span class="number">3</span>]</span><br><span class="line">        <span class="keyword">if</span> codon <span class="keyword">in</span> codon_table:</span><br><span class="line">            <span class="keyword">if</span> codon_table[codon] == <span class="string">'STOP'</span>:</span><br><span class="line">                prot= prot + <span class="string">'*'</span></span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                prot = prot + codon_table[codon]</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># 如果要识别的短序列不在table里, 则用'-'替代</span></span><br><span class="line">            prot = prot + <span class="string">'-'</span></span><br><span class="line">        </span><br><span class="line">    <span class="comment"># 格式化输出成48列的块</span></span><br><span class="line">    i = <span class="number">0</span></span><br><span class="line">    <span class="keyword">while</span> i &lt; <span class="built_in">len</span>(prot):</span><br><span class="line">        <span class="built_in">print</span> prot[i:i + <span class="number">48</span>]</span><br><span class="line">        i = i + <span class="number">48</span></span><br><span class="line">        </span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>str.startswith</code>(<em>prefix</em>[, <em>start</em>[, <em>end</em>]]) <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=startswi#str.startswith">5. 内置类型 — Python 2.7.18 文档</a></p>
<p>如果字符串以指定的 <em>prefix</em> 开始则返回 <code>True</code>，否则返回 <code>False</code>。 <em>prefix</em> 也可以为由多个供查找的前缀构成的元组。 如果有可选项 <em>start</em>，将从所指定位置开始检查。 如果有可选项 <em>end</em>，将在所指定位置停止比较。</p>
</li>
<li>
<p>核酸序列<code>(开放)阅读框(Open Reading Frame)</code>:</p>
<ul>
<li>
<p>由于核糖体翻译mRNA时, <strong>三个碱基翻译为一个氨基酸</strong>, 那么同一个序列理论上存在三个不同的阅读框</p>
</li>
<li>
<p>对应于<code>移码突变</code></p>
<blockquote>
<p>对于 序列 <code>ATCGATCG</code> 有 <code>ATC GAT CG</code>, <code>A TCG ATC G</code>, <code>AT CGA TCG</code> 三种方法解读</p>
</blockquote>
</li>
<li>
<p><code>ORF</code>指的是DNA上的序列，从5’端翻译起始密码子ATG到终止密码子(TAA,TAG,TGA)的蛋白质编码序列。对于任意给定的一段DNA，有两个问题需要考虑，</p>
<ul>
<li>
<p>一是DNA双链中的哪条是编码链</p>
</li>
<li>
<p>二是编码区究竟从第一个碱基开始进行编码<br>
所以每条链都有潜在的3种ORF，而对于双链DNA来说就有6种可能的ORF。也就是说先从给定的DNA单链为模版，分别从5’-3’方向第123个碱基开始翻译，再以互补链为模版，分别从3’-5’方向的第123个碱基开始翻译，得到另外3种翻译结果。正链上的正向读码框forward，负链的为反向读码框reverse。<strong>6个潜在ORF中，一般选择中间没有被终止密码子隔开的最大的读码框为正确的结果</strong>。</p>
<p>@<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/5789365faa7e">4️⃣ 核酸序列特征分析(1):开放阅读框识别 - 简书 (jianshu.com)</a></p>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>文件读取的<code>.next()</code>方法: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next">file.next  2.7文档</a></p>
<blockquote>
<p><code>file.next</code>()</p>
<p>A file object is its own iterator, for example <code>iter(f)</code> returns <em>f</em> (unless <em>f</em> is closed). When a file is used as an iterator, typically in a <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#for"><code>for</code></a> loop (for example, <code>for line in f: print line.strip()</code>), the <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> method is called repeatedly. This method returns the next input line, or raises <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/exceptions.html#exceptions.StopIteration"><code>StopIteration</code></a> when EOF is hit when the file is open for reading (behavior is undefined when the file is open for writing). In order to make a <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/compound_stmts.html#for"><code>for</code></a> loop the most efficient way of looping over the lines of a file (a very common operation), the <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> method uses a hidden read-ahead buffer. As a consequence of using a read-ahead buffer, combining <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.next"><code>next()</code></a> with other file methods (like <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.readline"><code>readline()</code></a>) does not work right. However, using <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=next#file.seek"><code>seek()</code></a> to reposition the file to an absolute position will flush the read-ahead buffer.</p>
</blockquote>
</li>
<li>
<p><code>for</code>循环中的 <code>else</code> :   <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=else#break-and-continue-statements-and-else-clauses-on-loops">break-and-continue-statements-and-else-clauses-on-loops</a></p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> n <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">2</span>, <span class="number">10</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">for</span> x <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">2</span>, n):</span><br><span class="line"><span class="meta">... </span>        <span class="keyword">if</span> n % x == <span class="number">0</span>:</span><br><span class="line"><span class="meta">... </span>            <span class="built_in">print</span> n, <span class="string">'equals'</span>, x, <span class="string">'*'</span>, n/x</span><br><span class="line"><span class="meta">... </span>            <span class="keyword">break</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">else</span>:   <span class="comment"># 此else与第二个for对应, 当发生了 break 后, 执行这里</span></span><br><span class="line"><span class="meta">... </span>        <span class="comment"># loop fell through without finding a factor</span></span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> n, <span class="string">'is a prime number'</span></span><br><span class="line">...</span><br><span class="line"><span class="number">2</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">3</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">4</span> equals <span class="number">2</span> * <span class="number">2</span></span><br><span class="line"><span class="number">5</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">6</span> equals <span class="number">2</span> * <span class="number">3</span></span><br><span class="line"><span class="number">7</span> <span class="keyword">is</span> a prime number</span><br><span class="line"><span class="number">8</span> equals <span class="number">2</span> * <span class="number">4</span></span><br><span class="line"><span class="number">9</span> equals <span class="number">3</span> * <span class="number">3</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>预测蛋白质序列中的无序(环)区域: 既不是<code>α螺旋</code>也不是<code>β折叠</code>的区域, 就是无序的.
<ul>
<li><code>核心思想</code>: 每个氨基酸都有一个特定的二级结构元件倾向, 通过大量已知的<code>蛋白质结构(PDB)</code>的二级结构元件中各类氨基酸类型出现的<code>频率f</code>,来估计倾向.</li>
<li>使用<code>1-f</code>计算: <code>f</code>是以有二级结构元件的部分为基数的(就是有序的), 那么在这样的基数里某种氨基酸出现的越多(频数接近1)则倾向于有序的可能性越大.</li>
<li>对20个氨基酸倾向的值进行标准化和引入负值, 设定一个是否为无序的<code>阈值</code>, 大于阈值为<code>无序(环)</code>, 小于为<code>有序(二级结构)</code>.</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">propensities ={</span><br><span class="line"><span class="string">'N'</span>: <span class="number">0.2299</span>, <span class="string">'P'</span>:  <span class="number">0.5523</span>, <span class="string">'Q'</span>:-<span class="number">0.18770</span>, <span class="string">'A'</span>:-<span class="number">0.2615</span>,</span><br><span class="line"><span class="string">'R'</span>:-<span class="number">0.1766</span>, <span class="string">'S'</span>:  <span class="number">0.1429</span>, <span class="string">'C'</span>:-<span class="number">0.01515</span>, <span class="string">'T'</span>: <span class="number">0.0089</span>,</span><br><span class="line"><span class="string">'D'</span>: <span class="number">0.2276</span>, <span class="string">'E'</span>: -<span class="number">0.2047</span>, <span class="string">'V'</span>:-<span class="number">0.38620</span>, <span class="string">'F'</span>:-<span class="number">0.2256</span>,</span><br><span class="line"><span class="string">'W'</span>:-<span class="number">0.2434</span>, <span class="string">'G'</span>:  <span class="number">0.4332</span>, <span class="string">'H'</span>:-<span class="number">0.00120</span>, <span class="string">'Y'</span>:-<span class="number">0.2075</span>,</span><br><span class="line"><span class="string">'I'</span>:-<span class="number">0.4222</span>, <span class="string">'K'</span>: -<span class="number">0.1001</span>, <span class="string">'L'</span>: <span class="number">0.33793</span>, <span class="string">'M'</span>:-<span class="number">0.2259</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://www.rcsb.org/">RCSB PDB: </a></p>
<ul>
<li></li>
<li>
<p>数据示例概览: <a target="_blank" rel="noopener" href="https://www.rcsb.org/structure/6XM5">RCSB PDB - 6XM5: Structure of SARS-CoV-2 spike at pH 5.5, all RBDs down</a></p>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br></pre></td><td class="code"><pre><span class="line">HEADER    VIRAL PROTEIN                           <span class="number">29</span>-JUN-<span class="number">20</span>   <span class="number">6</span>XM5              </span><br><span class="line"><span class="built_in">TITLE</span>     STRUCTURE OF SARS-COV-<span class="number">2</span> SPIKE <span class="built_in">AT</span> PH <span class="number">5</span>.<span class="number">5</span>, ALL RBDS DOWN                </span><br><span class="line">COMPND    MOL_ID: <span class="number">1</span>;                                                            </span><br><span class="line">COMPND   <span class="number">2</span> MOLECULE: SPIKE GLYCOPROTEIN;                                        </span><br><span class="line">...                    </span><br><span class="line">COMPND   <span class="number">6</span> MUTATION: YES                                                        </span><br><span class="line">SOURCE    MOL_ID: <span class="number">1</span>;                                                            </span><br><span class="line">SOURCE   <span class="number">2</span> ORGANISM_SCIENTIFIC: SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS   </span><br><span class="line">...                          </span><br><span class="line">SOURCE   <span class="number">9</span> EXPRESSION_SYSTEM_CELL_LINE: FREESTYLE <span class="number">293</span>F                          </span><br><span class="line">KEYWDS    SARS-COV-<span class="number">2</span> SPIKE, COVID19, VIRAL PROTEIN                              </span><br><span class="line">EXPDTA    ELECTRON MICROSCOPY                                                   </span><br><span class="line">AUTHOR    T.ZHOU,Y.TSYBOVSKY,A.OLIA,P.D.KWONG                                   </span><br><span class="line">REVDAT   <span class="number">2</span>   <span class="number">12</span>-AUG-<span class="number">20</span> <span class="number">6</span>XM5    <span class="number">1</span>       JRNL   HETSYN                            </span><br><span class="line">REVDAT   <span class="number">1</span>   <span class="number">29</span>-JUL-<span class="number">20</span> <span class="number">6</span>XM5    <span class="number">0</span>                                                </span><br><span class="line">JRNL        AUTH   T.ZHOU,Y.TSYBOVSKY,A.S.OLIA,J.GORMAN,M.A.RAPP,G.CERUTTI,     </span><br><span class="line">...  </span><br><span class="line">JRNL        TITL   CRYO-EM STRUCTURES DELINEATE A PH-DEPENDENT SWITCH THAT         ...</span><br><span class="line">JRNL        PMID   <span class="number">32637958</span>                                                     </span><br><span class="line">JRNL        DOI    <span class="number">10</span>.<span class="number">1101</span>/<span class="number">2020</span>.<span class="number">07</span>.<span class="number">04</span>.<span class="number">187989</span>                                    </span><br><span class="line">REMARK   <span class="number">2</span>                                                                      </span><br><span class="line">...       </span><br><span class="line">REMARK   <span class="number">3</span>   SOFTWARE PACKAGES      : LATITUDE, CTFFIND, UCSF CHIMERA, COOT,    </span><br><span class="line">REMARK   <span class="number">3</span>                            PHENIX, RELION, RELION, RELION, RELION    </span><br><span class="line">...         </span><br><span class="line">REMARK   <span class="number">3</span>   OVERALL ANISOTROPIC B VALUE  : NULL                                </span><br><span class="line">...</span><br><span class="line">REMARK <span class="number">900</span> STRUCTURE OF SARS-COV-<span class="number">2</span> SPIKE <span class="built_in">AT</span> PH <span class="number">5</span>.<span class="number">5</span>, ALL RBDS DOWN               </span><br><span class="line">DBREF  <span class="number">6</span>XM5 A   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">DBREF  <span class="number">6</span>XM5 B   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">DBREF  <span class="number">6</span>XM5 C   <span class="number">14</span>  <span class="number">1208</span>  UNP    P0DTC2   SPIKE_SARS2     <span class="number">14</span>   <span class="number">1208</span>             </span><br><span class="line">SEQADV <span class="number">6</span>XM5 GLY A  <span class="number">682</span>  UNP  P0DTC2    ARG   <span class="number">682</span> ENGINEERED MUTATION            </span><br><span class="line">SEQADV <span class="number">6</span>XM5 SER A  <span class="number">683</span>  UNP  P0DTC2    ARG   <span class="number">683</span> ENGINEERED MUTATION    </span><br><span class="line">...</span><br><span class="line">SEQADV <span class="number">6</span>XM5 LYS C <span class="number">1288</span>  UNP  P0DTC2              EXPRESSION TAG                 </span><br><span class="line">SEQRES   <span class="number">1</span> A <span class="number">1275</span>  GLN CYS VAL ASN LEU THR THR ARG THR GLN LEU PRO PRO          </span><br><span class="line">SEQRES   <span class="number">2</span> A <span class="number">1275</span>  ALA TYR THR ASN SER PHE THR ARG GLY VAL TYR TYR PRO       </span><br><span class="line">...</span><br><span class="line">SEQRES  <span class="number">99</span> C <span class="number">1275</span>  LYS                                                          </span><br><span class="line">HET    NAG  D   <span class="number">1</span>      <span class="number">14</span>                                                       </span><br><span class="line">HET    NAG  D   <span class="number">2</span>      <span class="number">14</span>                                                    </span><br><span class="line">...</span><br><span class="line">HETNAM     NAG <span class="number">2</span>-ACETAMIDO-<span class="number">2</span>-DEOXY-BETA-D-GLUCOPYRANOSE                         </span><br><span class="line">...</span><br><span class="line">FORMUL   <span class="number">4</span>  NAG    <span class="number">58</span>(C8 H15 N O6)                                              </span><br><span class="line">HELIX    <span class="number">1</span> AA1 ASP A  <span class="number">294</span>  LYS A  <span class="number">304</span>  <span class="number">1</span>                                  <span class="number">11</span>    </span><br><span class="line">...      </span><br><span class="line">SHEET    <span class="number">3</span> AA1 <span class="number">8</span> TYR A <span class="number">265</span>  TYR A <span class="number">269</span> -<span class="number">1</span>  O  TYR A <span class="number">265</span>   N  PHE A  <span class="number">65</span>           </span><br><span class="line">...</span><br><span class="line">LINK         O4  NAG S   <span class="number">1</span>                 C1  NAG S   <span class="number">2</span>     <span class="number">1555</span>   <span class="number">1555</span>  <span class="number">1</span>.<span class="number">44</span>  </span><br><span class="line">CRYST1    <span class="number">1</span>.<span class="number">000</span>    <span class="number">1</span>.<span class="number">000</span>    <span class="number">1</span>.<span class="number">000</span>  <span class="number">90</span>.<span class="number">00</span>  <span class="number">90</span>.<span class="number">00</span>  <span class="number">90</span>.<span class="number">00</span> P <span class="number">1</span>                      </span><br><span class="line">ORIGX1      <span class="number">1</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>        <span class="number">0</span>.<span class="number">00000</span>                         </span><br><span class="line">...</span><br><span class="line">SCALE3      <span class="number">0</span>.<span class="number">000000</span>  <span class="number">0</span>.<span class="number">000000</span>  <span class="number">1</span>.<span class="number">000000</span>        <span class="number">0</span>.<span class="number">00000</span>                         </span><br><span class="line">ATOM      <span class="number">1</span>  N   ALA A  <span class="number">27</span>     <span class="number">188</span>.<span class="number">584</span> <span class="number">130</span>.<span class="number">357</span> <span class="number">168</span>.<span class="number">091</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           N  </span><br><span class="line">ATOM      <span class="number">2</span>  CA  ALA A  <span class="number">27</span>     <span class="number">189</span>.<span class="number">571</span> <span class="number">130</span>.<span class="number">261</span> <span class="number">169</span>.<span class="number">159</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           C  </span><br><span class="line">ATOM      <span class="number">3</span>  C   ALA A  <span class="number">27</span>     <span class="number">190</span>.<span class="number">053</span> <span class="number">131</span>.<span class="number">641</span> <span class="number">169</span>.<span class="number">567</span>  <span class="number">1</span>.<span class="number">00</span> <span class="number">91</span>.<span class="number">79</span>           C  </span><br><span class="line">ATOM   <span class="number">8249</span>  OG  SER A1147     <span class="number">191</span>.<span class="number">132</span> <span class="number">184</span>.<span class="number">224</span> <span class="number">284</span>.<span class="number">141</span>  <span class="number">1</span>.<span class="number">00152</span>.<span class="number">29</span>           O  </span><br><span class="line">TER    <span class="number">8250</span>      SER A1147                                                      </span><br><span class="line">ATOM   <span class="number">8251</span>  N   ALA B  <span class="number">27</span>     <span class="number">132</span>.<span class="number">579</span> <span class="number">208</span>.<span class="number">265</span> <span class="number">174</span>.<span class="number">154</span>  <span class="number">1</span>.<span class="number">00140</span>.<span class="number">96</span>           N  </span><br><span class="line">...</span><br><span class="line">TER   <span class="number">24584</span>      SER C1147                                                      </span><br><span class="line">HETATM24585  C1  NAG D   <span class="number">1</span>     <span class="number">184</span>.<span class="number">739</span> <span class="number">152</span>.<span class="number">280</span> <span class="number">150</span>.<span class="number">742</span>  <span class="number">1</span>.<span class="number">00114</span>.<span class="number">03</span>           C  </span><br><span class="line">HETATM24586  C2  NAG D   <span class="number">1</span>     <span class="number">183</span>.<span class="number">632</span> <span class="number">153</span>.<span class="number">136</span> <span class="number">150</span>.<span class="number">126</span>  <span class="number">1</span>.<span class="number">00114</span>.<span class="number">03</span>           C  </span><br><span class="line">...</span><br><span class="line">CONECT  <span class="number">29225033</span>                                                                </span><br><span class="line">CONECT  <span class="number">70225047</span>                                                                </span><br><span class="line">...</span><br><span class="line">CONECT2539625389                                                                </span><br><span class="line">MASTER      <span class="number">907</span>    <span class="number">0</span>   <span class="number">58</span>   <span class="number">69</span>  <span class="number">173</span>    <span class="number">0</span>    <span class="number">0</span>    <span class="number">625393</span>    <span class="number">3</span>  <span class="number">936</span>  <span class="number">297</span>          </span><br><span class="line">END                                                                             </span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
</ul>
<h4><a href="#5-5-zi-ce-ti" class="header-anchor"> ⚓️</a><span id="5-5-zi-ce-ti">5.5 自测题</span></h4>
<ul>
<li>
<p><code>NCBI</code>下载<code>FASTA格式</code>RNA序列:</p>
</li>
<li>
<p><code>PubMed</code>文献检索:</p>
</li>
<li>
<p>二级结构元素预测:</p>
<ul>
<li>氨基酸二级结构偏好表: <a target="_blank" rel="noopener" href="http://www.bmrb.wisc.edu/referenc/choufas.html">Amino Acid Secondary Structure Preferences</a></li>
</ul>
</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align:center"><strong>A.A.</strong></th>
<th style="text-align:center"><strong>Helix</strong></th>
<th style="text-align:center"><strong>Helix</strong></th>
<th style="text-align:center"><strong>Sheet</strong></th>
<th style="text-align:center"><strong>Sheet</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center"><strong>A.A.</strong></td>
<td style="text-align:center"><strong>Designation</strong></td>
<td style="text-align:center"><strong>P</strong></td>
<td style="text-align:center"><strong>Designation</strong></td>
<td style="text-align:center"><strong>P</strong></td>
</tr>
<tr>
<td style="text-align:center"><strong>A</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.45</td>
<td style="text-align:center">I</td>
<td style="text-align:center">0.97</td>
</tr>
<tr>
<td style="text-align:center"><strong>C</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.77</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.30</td>
</tr>
<tr>
<td style="text-align:center"><strong>D</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.98</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.80</td>
</tr>
<tr>
<td style="text-align:center"><strong>E</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.53</td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.26</td>
</tr>
<tr>
<td style="text-align:center"><strong>F</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.12</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.28</td>
</tr>
<tr>
<td style="text-align:center"><strong>G</strong></td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.53</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.81</td>
</tr>
<tr>
<td style="text-align:center"><strong>H</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.24</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.71</td>
</tr>
<tr>
<td style="text-align:center"><strong>I</strong></td>
<td style="text-align:center">I</td>
<td style="text-align:center">1.00</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.60</td>
</tr>
<tr>
<td style="text-align:center"><strong>K</strong></td>
<td style="text-align:center">I</td>
<td style="text-align:center">1.07</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.74</td>
</tr>
<tr>
<td style="text-align:center"><strong>L</strong></td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.34</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.22</td>
</tr>
<tr>
<td style="text-align:center"><strong>M</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.20</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.67</td>
</tr>
<tr>
<td style="text-align:center"><strong>N</strong></td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.73</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.65</td>
</tr>
<tr>
<td style="text-align:center"><strong>P</strong></td>
<td style="text-align:center">B</td>
<td style="text-align:center">0.59</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.62</td>
</tr>
<tr>
<td style="text-align:center"><strong>Q</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.17</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.23</td>
</tr>
<tr>
<td style="text-align:center"><strong>R</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.79</td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.90</td>
</tr>
<tr>
<td style="text-align:center"><strong>S</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.79</td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.72</td>
</tr>
<tr>
<td style="text-align:center"><strong>T</strong></td>
<td style="text-align:center">i</td>
<td style="text-align:center">0.82</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.20</td>
</tr>
<tr>
<td style="text-align:center"><strong>V</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.14</td>
<td style="text-align:center">H</td>
<td style="text-align:center">1.65</td>
</tr>
<tr>
<td style="text-align:center"><strong>W</strong></td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.14</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.19</td>
</tr>
<tr>
<td style="text-align:center"><strong>Y</strong></td>
<td style="text-align:center">b</td>
<td style="text-align:center">0.61</td>
<td style="text-align:center">h</td>
<td style="text-align:center">1.29</td>
</tr>
</tbody>
</table>
<ul>
<li>
<p><code>Designations</code>:</p>
<ul>
<li>H = Strong Former, h = Former, I = Weak Former, i = Indifferent, B = Strong Breaker, b = Breaker;</li>
</ul>
</li>
<li>
<p><code>P</code> : Conformational Parameter</p>
</li>
</ul>
<h3><a href="#di-6-zhang-guo-lu-shu-ju" class="header-anchor"> ⚓️</a><span id="di-6-zhang-guo-lu-shu-ju">第6章 过滤数据</span></h3>
<h4><a href="#6-2-an-li-shi-yong-rnaseq-shu-chu-shu-ju" class="header-anchor"> ⚓️</a><span id="6-2-an-li-shi-yong-rnaseq-shu-chu-shu-ju">6.2 案例： 使用RNAseq输出数据</span></h4>
<ul>
<li>
<p><strong>新一代测序(Next-generation sequencing，NGS)</strong>:</p>
<blockquote>
<p>最早广泛应用测序技术为 70 年代的 <strong>Sanger</strong> 法，这也是完成人类基因组计划的基础，因其测序通量低、费时费力，科学家们一直在寻求通量更高、速度更快、价格更便宜、自动化程度更高的测序技术。自 2005 年以来，以Roche 公司的 <strong>454 技术</strong>、Illumina 公司的 <strong>Solexa</strong> 技术以及 ABI 公司的 <strong>SOLiD</strong> 技术为标志的高通量测序技术相继诞生 [6]。相较于传统方法，该技术主要特点是测序通量高、测序时间和成本显著下降，可以一次对几十万到几百万条 DNA 分子序列测定，这使某物种全基因组和转录组的全貌细致分析成为可能，又称为<strong>深度测序</strong>，很多文献中称其为新一代测序技术，足见其划时代意义 [7]。</p>
<p>@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54729244">转录组测序技术的应用及发展综述 - 知乎 (zhihu.com)</a> <strong>推荐阅读</strong></p>
</blockquote>
</li>
</ul>
<p><img src="https://pic2.zhimg.com/v2-d4a7d909a5aea6dd5b31f4072b5eac51_r.jpg" alt="四种转录平台的比较"></p>
<ul>
<li>
<p><strong>转录组</strong>(<strong>Transcriptome</strong>)测序技术|<strong>RNA-Seq(RNA sequencing)技术</strong> :</p>
<ul>
<li>
<p><code>转录本</code>就是基因通过转录形成的通过剪切生成的一种或多种可供编码蛋白质的成熟的mRNA。剪切的过程中可能会剪切掉外显子，也有可能保留部分内含子，这样就形成了多种mRNA即多个转录本。<br>
@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/136389445_170798">基因研究不知道转录本，做什么引物设计？！ (sohu.com)</a></p>
</li>
<li>
<p>RNA-Seq利用高通量测序技术对组织或细胞中<strong>所有RNA 反转录而成cDNA文库</strong>进行测序，通过统计相关读段(reads)数计算出不同RNA的表达量，发现新的转录本；如果有基因组参考序列，可以<strong>把转录本映射回基因组</strong>，确定转录本位置、剪切情况等更为全面的遗传信息，已广泛应用于生物学研究、医学研究、临床研究和药物研发等。<br>
@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54729244">转录组测序技术的应用及发展综述 - 知乎 (zhihu.com)</a> <strong>推荐阅读</strong></p>
</li>
<li>
<p>大致流程:</p>
<p>比对read到基因组，转录本组装，差异表达分析和差异基因注释</p>
<p>转录本组装主要就是将比对到参考基因组的read组装成转录本接着进行下游的新转录本鉴定和转录本定量等分析，常见软件有StringTie和cufflinks。而对于无参转录组分析来说，转录本组装主要是将测序reads从头组装成转录本，常见软件有Trinity,Oases和SOAPdenovo-Trans。</p>
<p>@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/413528571_652735">转录组测序（转录本组装篇）_分析 (sohu.com)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/687766131808">转录组分析流程 - 简书 (jianshu.com)</a></p>
</li>
<li>
<p><strong>无参转录组和有参转录组的区别</strong>: 如果所研究的物种有组装注释质量较好基因组序列，且和该基因组序列比对效率较高，那么可以采用有参转录组的分析策略，直接进行分析。反之，则需要按照无参转录组的分析策略进行转录本组装，构建unigene(基因簇)库，然后进行后续分析。<br>
@<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/190736664">转录组测序：常见基础知识准备一 - 知乎 (zhihu.com)</a></p>
</li>
</ul>
</li>
<li>
<p>转录本文件 <code>.gtf</code> : 记录了各个样本装配好的转录本</p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="http://cole-trapnell-lab.github.io/cufflinks/">Cufflinks (cole-trapnell-lab.github.io)</a></p>
<blockquote>
<p>Cufflinks assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</p>
</blockquote>
<ul>
<li><a target="_blank" rel="noopener" href="http://cole-trapnell-lab.github.io/cufflinks/cuffcompare/">Cufflinks <strong>cuffcompare</strong></a></li>
</ul>
<blockquote>
<p>Cufflinks includes a program that you can use to help analyze the transfrags you assemble. The program <strong>cuffcompare</strong> helps you:</p>
<ul>
<li>Compare your assembled transcripts to a reference annotation</li>
<li>Track Cufflinks transcripts across multiple experiments (e.g. across a time course)</li>
</ul>
</blockquote>
</li>
<li>
<p>转录本比对文件 <code>.tracking</code> : 导入转录本文件(<code>.gtf</code>)到 <code>cuffcompare</code> 生成</p>
</li>
<li>
<p>元组方法:</p>
<ul>
<li>
<p><code>intersection</code>(*<em>others</em>) : 返回一个新集合，其中包含原集合以及 others 指定的所有集合中共有的元素。</p>
<p>@<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=intersection#frozenset.intersection">intersection 5. 内置类型 — Python 2.7.18 文档</a></p>
</li>
<li>
<p><code>difference</code>(*<em>others</em>)  : 返回一个新集合，其中包含原集合中在 others 指定的其他集合中不存在的元素。</p>
<p>@<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/stdtypes.html?highlight=intersection#frozenset.difference">difference 5. 内置类型 — Python 2.7.18 文档</a></p>
</li>
<li>
<p>更多方法参见文档</p>
</li>
</ul>
</li>
<li>
<p>列表的<code>.count()</code>方法:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/array.html?highlight=count#array.array.count">8.6. array — 高效的数值数组 — Python 2.7.18 文档</a></p>
<ul>
<li><code>array.count</code>(<em>x</em>) : 返回 <em>x</em> 在数组中的出现次数。</li>
</ul>
</li>
<li>
<p>内置函数相关: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#built-in-functions">2. 内置函数 — Python 2.7.18 文档</a></p>
<ul>
<li><code>pop()</code>:</li>
<li><code>reduce()</code>:</li>
<li><code>del()</code>:</li>
<li><code>remove()</code>:</li>
<li><code>enumerate()</code>:</li>
<li></li>
</ul>
</li>
<li>
<p><code>.readlines()</code> 返回的列表可以索引:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">lines = <span class="built_in">open</span>(<span class="string">'text.txt'</span>).readlines()</span><br><span class="line"><span class="built_in">open</span>(<span class="string">'new.txt'</span>, <span class="string">'w'</span>).writelines(lines[<span class="number">2</span>:<span class="number">4</span>]+lines[<span class="number">6</span>:])</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>使用<code>列表推导式</code>筛选:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">data_a = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>]</span><br><span class="line">data = [x <span class="keyword">for</span> x <span class="keyword">in</span> data_a <span class="keyword">if</span> x != <span class="number">3</span>]</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>去重妙法-<code>元组去重</code>:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">input_file = <span class="built_in">open</span>(<span class="string">'UniprotID.txt'</span>)</span><br><span class="line">unique = <span class="built_in">set</span>(input_file)</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p><code>CD-HIT</code>(可容错的高同源性聚类数据库): 可以快速地的基于用户定义的相似性阈值对蛋白质序列进行聚类, 需要输入一组<code>FASTA</code>格式的序列,并返回两个文件: 一个是聚类列表, 一个是所聚各类的代表序列.</p>
<p>@<a target="_blank" rel="noopener" href="http://weizhongli-lab.org/cd-hit/">CD-HIT Official Website (weizhongli-lab.org)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://github.com/weizhongli/cdhit">weizhongli/cdhit: Automatically exported from code.google.com/p/cdhit (github.com)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://www.sohu.com/a/190353140_785442">教程 | 如何用cd-hit去除冗余序列？ (sohu.com)</a></p>
</li>
</ul>
<h3><a href="#di-7-zhang-guan-li-biao-shu-ju" class="header-anchor"> ⚓️</a><span id="di-7-zhang-guan-li-biao-shu-ju">第7章 管理表数据</span></h3>
<h4><a href="#7-2-an-li-que-ding-dan-bai-nong-du" class="header-anchor"> ⚓️</a><span id="7-2-an-li-que-ding-dan-bai-nong-du">7.2 案例： 确定蛋白浓度</span></h4>
<ul>
<li>
<p><code>苯酚福林试剂</code>测定<code>蛋白质含量</code>的流程:</p>
<ul>
<li>
<p>相关论文: <a target="_blank" rel="noopener" href="https://www.jbc.org/content/193/1/265.long">PROTEIN MEASUREMENT WITH THE FOLIN PHENOL REAGENT (jbc.org)</a></p>
</li>
<li>
<p>对比于<code>凯氏定氮法</code>的优势,对多个样品可以使用光度计快速地进行测量.</p>
<p>@<a target="_blank" rel="noopener" href="http://people.umass.edu/~mcclemen/581Proteins.html">6. ANALYSIS OF PROTEINS (umass.edu)</a></p>
</li>
</ul>
</li>
<li>
<p><code>zip()</code>内置函数: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#zip"><code>zip</code>([<em>iterable</em>, <em>…</em>]) 2. 内置函数 — Python 2.7.18 文档</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>x = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>y = [<span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>zipped = <span class="built_in">zip</span>(x, y)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>zipped</span><br><span class="line">[(<span class="number">1</span>, <span class="number">4</span>), (<span class="number">2</span>, <span class="number">5</span>), (<span class="number">3</span>, <span class="number">6</span>)]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>x2, y2 = <span class="built_in">zip</span>(*zipped)  <span class="comment"># *号的功能是去除zipped元组的外括号(相当于拆解)</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>x == <span class="built_in">list</span>(x2) <span class="keyword">and</span> y == <span class="built_in">list</span>(y2)</span><br><span class="line"><span class="literal">True</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 效果同转置矩阵:</span></span><br><span class="line">zipped = <span class="built_in">zip</span>(*zipped)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 样例</span></span><br><span class="line">table = [</span><br><span class="line">    [<span class="string">'protein'</span>, <span class="string">'ext1'</span>, <span class="string">'ext2'</span>, <span class="string">'ext3'</span>],</span><br><span class="line">    [<span class="number">0.16</span>, <span class="number">0.038</span>, <span class="number">0.044</span>, <span class="number">0.040</span>],</span><br><span class="line">    [<span class="number">0.33</span>, <span class="number">0.089</span>, <span class="number">0.095</span>, <span class="number">0.091</span>],</span><br><span class="line">    [<span class="number">0.66</span>, <span class="number">0.184</span>, <span class="number">0.191</span>, <span class="number">0.191</span>],</span><br><span class="line">    [<span class="number">1.00</span>, <span class="number">0.280</span>, <span class="number">0.292</span>, <span class="number">0.283</span>],</span><br><span class="line">    [<span class="number">1.32</span>, <span class="number">0.365</span>, <span class="number">0.367</span>, <span class="number">0.365</span>],</span><br><span class="line">    [<span class="number">1.66</span>, <span class="number">0.441</span>, <span class="number">0.443</span>, <span class="number">0.444</span>]</span><br><span class="line">    ]</span><br><span class="line"></span><br><span class="line"><span class="comment"># remove the first row</span></span><br><span class="line">table = table[<span class="number">1</span>:]</span><br><span class="line"></span><br><span class="line">protein, ext1, ext2, ext3 = <span class="built_in">zip</span>(*table)</span><br><span class="line"></span><br><span class="line"><span class="comment"># create a single column for `ext` or concatenate each `ext`</span></span><br><span class="line"><span class="comment"># extend (`* 3`) the `protein` column to match the `ext` column.</span></span><br><span class="line">extinction = ext1 + ext2 + ext3</span><br><span class="line">protein = protein * <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># create four tuples for each column</span></span><br><span class="line">table = <span class="built_in">zip</span>(protein, extinction)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> prot, ext <span class="keyword">in</span> table:</span><br><span class="line">    <span class="built_in">print</span> prot, ext</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>列表创建注意: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/faq/programming.html#id43">如何创建多维列表？编程常见问题 — Python 2.7.18 文档</a></p>
<ul>
<li><code>[0]*3</code>这样的语法是创建了<code>[0]</code>的三个引用, 改变一个列表的内容将改变所有内容</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">table = [</span><br><span class="line">    [<span class="string">'protein'</span>, <span class="string">'ext'</span>],</span><br><span class="line">    [<span class="number">0.16</span>, <span class="number">0.038</span>],</span><br><span class="line">    [<span class="number">0.33</span>, <span class="number">0.089</span>],</span><br><span class="line">    [<span class="number">0.66</span>, <span class="number">0.184</span>],</span><br><span class="line">    [<span class="number">1.00</span>, <span class="number">0.280</span>],</span><br><span class="line">    [<span class="number">1.32</span>, <span class="number">0.365</span>],</span><br><span class="line">    [<span class="number">1.66</span>, <span class="number">0.441</span>]</span><br><span class="line">]</span><br><span class="line"></span><br><span class="line"><span class="comment"># convert nested list to nested dict</span></span><br><span class="line">nested_dict = {}</span><br><span class="line">n = <span class="number">0</span></span><br><span class="line">key = table[<span class="number">0</span>]</span><br><span class="line"><span class="keyword">for</span> row <span class="keyword">in</span> table[<span class="number">1</span>:]: <span class="comment"># for row in table[0:]  includes the header</span></span><br><span class="line">    n += <span class="number">1</span></span><br><span class="line">    entry = {key[<span class="number">0</span>]: row[<span class="number">0</span>], key[<span class="number">1</span>]: row[<span class="number">1</span>]}</span><br><span class="line">    nested_dict[<span class="string">'row'</span>+<span class="built_in">str</span>(n)] = entry</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> nested_dict</span><br></pre></td></tr></tbody></table></figure>
<h4><a href="#7-5-zi-ce-ti" class="header-anchor"> ⚓️</a><span id="7-5-zi-ce-ti">7.5自测题</span></h4>
<ul>
<li>
<p>RNA碱基相似性矩阵:</p>
<table>
<thead>
<tr>
<th style="text-align:center"></th>
<th style="text-align:center">A</th>
<th style="text-align:center">G</th>
<th style="text-align:center">C</th>
<th style="text-align:center">U</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center">A</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
</tr>
<tr>
<td style="text-align:center">G</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
</tr>
<tr>
<td style="text-align:center">C</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">1.0</td>
<td style="text-align:center">0.5</td>
</tr>
<tr>
<td style="text-align:center">U</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.0</td>
<td style="text-align:center">0.5</td>
<td style="text-align:center">1.0</td>
</tr>
</tbody>
</table>
</li>
</ul>
<h3><a href="#di-8-zhang-shu-ju-pai-xu" class="header-anchor"> ⚓️</a><span id="di-8-zhang-shu-ju-pai-xu">第8章 数据排序</span></h3>
<h4><a href="#8-2-an-li-shu-ju-biao-pai-xu" class="header-anchor"> ⚓️</a><span id="8-2-an-li-shu-ju-biao-pai-xu">8.2案例： 数据表排序</span></h4>
<ul>
<li>
<p><code>sorted</code>(<em>iterable</em>[, <em>cmp</em>[, <em>key</em>[, <em>reverse</em>]]]) 内置函数: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=sorted#sorted">sorted 2. 内置函数 — Python 2.7.18 文档</a></p>
<ul>
<li>
<p>根据 <em>iterable</em> 中的项返回一个新的已排序列表。</p>
</li>
<li>
<p>参数<code>cmp</code>(<em>x</em>, <em>y</em>): <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html?highlight=sorted#cmp"> 2. 内置函数 — Python 2.7.18 文档</a></p>
<blockquote>
<p>sorted方法会隐式的调用cmp()函数进行两两比对, 可以通过自定义函数传给cmp参数控制排序</p>
<p>Compare the two objects <em>x</em> and <em>y</em> and return an integer according to the outcome. The return value is negative if <code>x &lt; y</code>, zero if <code>x == y</code> and strictly positive if <code>x &gt; y</code>.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">def</span> <span class="title function_">numeric_compare</span>(<span class="params">x, y</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">return</span> x - y</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">sorted</span>([<span class="number">5</span>, <span class="number">2</span>, <span class="number">4</span>, <span class="number">1</span>, <span class="number">3</span>], cmp=numeric_compare) </span><br><span class="line">[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>]</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">my_cmp</span>(<span class="params">a,b</span>):  <span class="comment"># 这个cmp可以在比对a和b的第x项相等时,寻找x+1项进行比对</span></span><br><span class="line">    x = <span class="number">0</span></span><br><span class="line">    <span class="keyword">if</span> a[x] &lt; b[x]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> a[x] == b[x]:</span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] &lt; b[x+<span class="number">1</span>]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] == b[x+<span class="number">1</span>]:</span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] &lt; b[x+<span class="number">2</span>]: <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] == b[x+<span class="number">2</span>]: <span class="keyword">return</span> <span class="number">0</span></span><br><span class="line">            <span class="keyword">if</span> a[x+<span class="number">2</span>] &gt; b[x+<span class="number">2</span>]: <span class="keyword">return</span> -<span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> a[x+<span class="number">1</span>] &gt; b[x+<span class="number">1</span>]: <span class="keyword">return</span> -<span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> a[x] &gt; b[x]: <span class="keyword">return</span> -<span class="number">1</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>排序默认按照<code>ASCII</code>顺序进行排序</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>space</td>
<td>0</td>
<td>:</td>
<td>A</td>
<td>Q</td>
<td>[</td>
<td>a</td>
<td>q</td>
<td>}</td>
</tr>
<tr>
<td>!</td>
<td>1</td>
<td>;</td>
<td>B</td>
<td>R</td>
<td>\</td>
<td>b</td>
<td>r</td>
<td>tick</td>
</tr>
<tr>
<td>“</td>
<td>2</td>
<td>&lt;</td>
<td>C</td>
<td>S</td>
<td>]</td>
<td>c</td>
<td>s</td>
<td>{</td>
</tr>
<tr>
<td>#</td>
<td>3</td>
<td>=</td>
<td>D</td>
<td>T</td>
<td>^</td>
<td>d</td>
<td>t</td>
<td></td>
</tr>
<tr>
<td>$</td>
<td>4</td>
<td>&gt;</td>
<td>E</td>
<td>U</td>
<td>_</td>
<td>e</td>
<td>u</td>
<td>DEL</td>
</tr>
<tr>
<td>%</td>
<td>5</td>
<td>?</td>
<td>F</td>
<td>V</td>
<td>`</td>
<td>f</td>
<td>v</td>
<td></td>
</tr>
<tr>
<td>&amp;</td>
<td>6</td>
<td>@</td>
<td>G</td>
<td>W</td>
<td></td>
<td>g</td>
<td>w</td>
<td></td>
</tr>
<tr>
<td>‘</td>
<td>7</td>
<td></td>
<td>H</td>
<td>X</td>
<td></td>
<td>h</td>
<td>x</td>
<td></td>
</tr>
<tr>
<td>(</td>
<td>8</td>
<td></td>
<td>I</td>
<td>Y</td>
<td></td>
<td>i</td>
<td>y</td>
<td></td>
</tr>
<tr>
<td>)</td>
<td>9</td>
<td></td>
<td>J</td>
<td>Z</td>
<td></td>
<td>j</td>
<td>z</td>
<td></td>
</tr>
<tr>
<td>*</td>
<td></td>
<td></td>
<td>K</td>
<td></td>
<td></td>
<td>k</td>
<td></td>
<td></td>
</tr>
<tr>
<td>+</td>
<td></td>
<td></td>
<td>L</td>
<td></td>
<td></td>
<td>l</td>
<td></td>
<td></td>
</tr>
<tr>
<td>,</td>
<td></td>
<td></td>
<td>M</td>
<td></td>
<td></td>
<td>m</td>
<td></td>
<td></td>
</tr>
<tr>
<td>-</td>
<td></td>
<td></td>
<td>N</td>
<td></td>
<td></td>
<td>n</td>
<td></td>
<td></td>
</tr>
<tr>
<td>.</td>
<td></td>
<td></td>
<td>O</td>
<td></td>
<td></td>
<td>o</td>
<td></td>
<td></td>
</tr>
<tr>
<td>/</td>
<td></td>
<td></td>
<td>P</td>
<td></td>
<td></td>
<td>p</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
<li>
<p><code>openator</code>模块的<code>itemgetter</code>:  <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/operator.html?highlight=itemgetter#operator.itemgetter">9.9. operator — 标准运算符替代函数 — Python 2.7.18 文档</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>inventory = [(<span class="string">'apple'</span>, <span class="number">3</span>), (<span class="string">'banana'</span>, <span class="number">2</span>), (<span class="string">'pear'</span>, <span class="number">5</span>), (<span class="string">'orange'</span>, <span class="number">1</span>)]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>getcount = itemgetter(<span class="number">1</span>) <span class="comment"># 构建一个获取`索引1的值`的itemgetter函数</span></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">map</span>(getcount, inventory) <span class="comment"># 将inventory按项逐个传入getcount, 获取每项索引号为1的值</span></span><br><span class="line">[<span class="number">3</span>, <span class="number">2</span>, <span class="number">5</span>, <span class="number">1</span>]</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">sorted</span>(inventory, key=getcount) <span class="comment"># 也可以将getcount作为key, 使用sorted排序</span></span><br><span class="line"><span class="comment"># 这里按照inventory的每一项的索引号为1的值进行排序</span></span><br><span class="line">[(<span class="string">'orange'</span>, <span class="number">1</span>), (<span class="string">'banana'</span>, <span class="number">2</span>), (<span class="string">'apple'</span>, <span class="number">3</span>), (<span class="string">'pear'</span>, <span class="number">5</span>)]</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>更多排序参见: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/howto/sorting.html?highlight=cmp">排序指南 — Python 2.7.18 文档</a></p>
</li>
<li>
<p>通过<code>shell</code>终端命令排序: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/c/sort.html">Linux sort 命令详解：对文本文件中所有行进行排序。 - Linux 命令搜索引擎 (wangchujiang.com)</a></p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> myfile.txt <span class="comment"># 按照字母表顺序排序</span></span></span><br><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> -n myfile.txt <span class="comment"># 按照数值排序</span></span></span><br><span class="line"><span class="meta prompt_">%</span><span class="language-bash"><span class="built_in">sort</span> myfile.txt -k2n -kl <span class="comment"># 多列排序</span></span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>对<code>.tsv</code>制表符分隔符文件的每行,按照七个优先级排序</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> operator <span class="keyword">import</span> itemgetter</span><br><span class="line"></span><br><span class="line"><span class="comment"># read table</span></span><br><span class="line">in_file = <span class="built_in">open</span>(<span class="string">"random_distribution.tsv"</span>)</span><br><span class="line">table = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    columns = line.split()</span><br><span class="line">    columns = [<span class="built_in">float</span>(x) <span class="keyword">for</span> x <span class="keyword">in</span> columns]</span><br><span class="line">    table.append(columns)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># 提取table每项的0~6形成的元组进行逐项比对排序</span></span><br><span class="line">table_sorted = <span class="built_in">sorted</span>(table, key=itemgetter(<span class="number">0</span>, <span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">5</span>, <span class="number">6</span>)) </span><br><span class="line"><span class="built_in">print</span> table_sorted</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>对BLAST的<code>.csv</code>逗号分隔符文件每行的某一项进行排序</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> operator <span class="keyword">import</span> itemgetter</span><br><span class="line"></span><br><span class="line">input_file = <span class="built_in">open</span>(<span class="string">"BlastOut.csv"</span>)</span><br><span class="line">output_file = <span class="built_in">open</span>(<span class="string">"BlastOutSorted.csv"</span>,<span class="string">"w"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># read BLAST output table</span></span><br><span class="line">table = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> input_file:</span><br><span class="line">    col = line.split(<span class="string">','</span>)</span><br><span class="line">    col[<span class="number">2</span>] = <span class="built_in">float</span>(col[<span class="number">2</span>])</span><br><span class="line">    table.append(col)</span><br><span class="line"></span><br><span class="line">table_sorted = <span class="built_in">sorted</span>(table, key=itemgetter(<span class="number">2</span>), reverse=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># write sorted table to an output file</span></span><br><span class="line"><span class="keyword">for</span> row <span class="keyword">in</span> table_sorted:</span><br><span class="line">    row = [<span class="built_in">str</span>(x) <span class="keyword">for</span> x <span class="keyword">in</span> row]</span><br><span class="line">    output_file.write(<span class="string">"\t"</span>.join(row) + <span class="string">'\n'</span>)</span><br><span class="line"></span><br><span class="line">input_file.close()</span><br><span class="line">output_file.close()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>blastp(蛋白质比对蛋白数据库)</code>工具:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/wh-ff-ly520/p/11758999.html">blast+本地化中blastp操作(基于PDB库)—linux - wang霏霏 - 博客园 (cnblogs.com)</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/de28be1a3bea">这或许是我写的最全的BLAST教程 - 简书 (jianshu.com)</a></li>
</ul>
</li>
<li>
<p><code>A链血红蛋白PDB</code>表报告:</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://www.rcsb.org/search?request=%7B%22query%22%3A%7B%22parameters%22%3A%7B%22value%22%3A%22hemoglobin%22%7D%2C%22type%22%3A%22terminal%22%2C%22service%22%3A%22text%22%2C%22node_id%22%3A0%7D%2C%22return_type%22%3A%22entry%22%2C%22request_options%22%3A%7B%22pager%22%3A%7B%22start%22%3A0%2C%22rows%22%3A100%7D%2C%22scoring_strategy%22%3A%22combined%22%2C%22sort%22%3A%5B%7B%22sort_by%22%3A%22score%22%2C%22direction%22%3A%22desc%22%7D%5D%7D%2C%22request_info%22%3A%7B%22src%22%3A%22ui%22%2C%22query_id%22%3A%22cb5954b71eff55f454db70f597aec338%22%7D%7D">RCSB PDB: Search</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-9-zhang-mo-shi-pi-pei-he-wen-ben-wa-jue" class="header-anchor"> ⚓️</a><span id="di-9-zhang-mo-shi-pi-pei-he-wen-ben-wa-jue">第9章 模式匹配和文本挖掘</span></h3>
<h4><a href="#9-2-an-li-zai-dan-bai-zhi-xu-lie-zhong-sou-suo-lin-suan-hua-mo-ti" class="header-anchor"> ⚓️</a><span id="9-2-an-li-zai-dan-bai-zhi-xu-lie-zhong-sou-suo-lin-suan-hua-mo-ti">9.2 案例： 在蛋白质序列中搜索磷酸化模体</span></h4>
<ul>
<li>
<p><code>正则表达式(regular expression, regexp)</code>: <strong>代表</strong>一组字符串的字符串语法</p>
<ul>
<li>
<p>类似的: 生物学家使用<code>N</code>来代替任意碱基(ATCG)</p>
</li>
<li>
<p><code>[</code>和<code>]</code>被称为<code>元字符</code>,由它和<code>字符</code>来编码一组字符串的表达,为<code>正则表达式</code></p>
<ul>
<li><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex">7.2. re — 正则表达式操作 — Python 2.7.18 文档</a> <strong>请仔细阅读</strong></li>
</ul>
</li>
<li>
<p><code>re.compile</code>(<em>pattern</em>, <em>flags=0</em>): Compile a regular expression pattern into a regular expression object, which can be used for matching using its <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex#re.RegexObject.match"><code>match()</code></a> and <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/re.html?highlight=regex#re.RegexObject.search"><code>search()</code></a> methods.<br>
参数见上面的文档</p>
</li>
<li>
<p><code>re.compile()</code>返回<code>RegexpObject</code>对象, 使用该对象的<code>.search()</code>方法将返回<code>Match Object</code>对象, 该对象还可以使用<code>.group()</code>/<code>.span()</code>/<code>.start()</code>/<code>.end()</code>方法. 另外<code>RegexpObject</code>对象还有<code>.findall()</code>/<code>.match()</code>/<code>.finditer()</code>等方法可以用, 详见文档</p>
</li>
<li>
<p><code>由(</code>和<code>)</code>配合匹配到的字符串可以用<code>group()</code>获取识别到的更小的子组</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">seq = <span class="string">'zzabcdzz'</span></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'(a(b)c)d'</span>)</span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">0</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">1</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">2</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.groups()</span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
</ul>
<p><img src="https://ugoproto.github.io/ugo_py_doc/img/managing_your_biological/regex.png" alt="Regex"></p>
<ul>
<li>
<p>检索匹配指定序列:</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re  <span class="comment"># 正则表达式模块</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># define a string with occurrences of regex:</span></span><br><span class="line">seq = <span class="string">'VSVLTMFRYAGWLDRLYMLVGTQLAAIIHGVALPLMMLI'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># compile a pattern and assign it to a variable</span></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'[ST]Q'</span>) <span class="comment"># 转换字符串为正则表达式对象</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># search for the pattern in the string</span></span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="keyword">if</span> <span class="keyword">match</span>:</span><br><span class="line">    <span class="comment"># print the first match along the sequence with the group() method</span></span><br><span class="line">    <span class="comment"># 4 characters before and after the pattern</span></span><br><span class="line">    <span class="built_in">print</span> <span class="string">'%10s'</span> %(seq[<span class="keyword">match</span>.start() - <span class="number">4</span>:<span class="keyword">match</span>.end() + <span class="number">4</span>])</span><br><span class="line">    <span class="built_in">print</span> <span class="string">'%6s'</span> % <span class="keyword">match</span>.group()</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    <span class="built_in">print</span> <span class="string">"no match"</span></span><br><span class="line"><span class="comment"># 结果:   </span></span><br><span class="line"><span class="comment"># MLVGTQLAAI</span></span><br><span class="line"><span class="comment">#     TQ</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>功能性模体表达:</p>
<ul>
<li>
<p><code>ELM</code>: <a target="_blank" rel="noopener" href="http://elm.eu.org/">elm.eu.org</a></p>
<blockquote>
<p>This computational biology resource mainly focuses on <strong>annotation and detection of eukaryotic linear motifs (ELMs) by providing both a repository of annotated motif data and an exploratory tool for motif prediction</strong>. <code>ELMs, or short linear motifs (SLiMs)</code>, are compact protein interaction sites composed of short stretches of adjacent amino acids. They are enriched in intrinsically disordered regions of the proteome and provide a wide range of functionality to proteins (<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/21909575">Davey,2011</a>,<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/24926813">Van Roey,2014</a>) They play crucial roles in cell regulation and are also of clinical importance, as aberrant SLiM function has been associated with several diseases and SLiM mimics are often used by pathogens to manipulate their hosts’ cellular machinery (<a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/21146412">Davey,2011</a>, <a target="_blank" rel="noopener" href="http://www.ncbi.nlm.nih.gov/pubmed/25057855">Uyar,2014</a>)</p>
</blockquote>
</li>
<li>
<p><code>PROSITE</code>: <a target="_blank" rel="noopener" href="https://prosite.expasy.org/">ExPASy - PROSITE</a></p>
<blockquote>
<p>PROSITE consists of <strong>documentation entries describing protein domains, families and functional sites as well as associated patterns and profiles to identify them</strong> [<a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_details.html">More…</a> / <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_ref.html">References</a> / <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prosite_license.html">Commercial users</a> ].<br>
PROSITE is complemented by <a target="_blank" rel="noopener" href="https://prosite.expasy.org/prorule.html">ProRule</a> , a collection of rules based on profiles and patterns, which increases the discriminatory power of profiles and patterns by providing additional information about functionally and/or structurally critical amino acids [<a target="_blank" rel="noopener" href="https://prosite.expasy.org/prorule_details.html">More…</a>].</p>
</blockquote>
</li>
</ul>
</li>
<li>
<p><code>.findall()</code>和<code>.finditer()</code>方法:</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'R.[ST][^P]'</span>)</span><br><span class="line">seq = <span class="string">'RQSAMGSNKSKPKDASQRRRSLEPAENVHGAGGGAFPASQRPSKP'</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># findall returns a list of all matches</span></span><br><span class="line">matches = pattern.findall(seq)</span><br><span class="line"><span class="built_in">print</span> matches  <span class="comment"># 结果: ['RQSA', 'RRSL', 'RPSK']</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># finditer returns an iterator of match objects</span></span><br><span class="line">match_iter = pattern.finditer(seq)  <span class="comment"># 返回一个符合条件的迭代器</span></span><br><span class="line"><span class="keyword">for</span> <span class="keyword">match</span> <span class="keyword">in</span> match_iter:</span><br><span class="line">    <span class="built_in">print</span> <span class="keyword">match</span>.group(), <span class="keyword">match</span>.span(), <span class="keyword">match</span>.start(), <span class="keyword">match</span>.end()</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 结果:</span></span><br><span class="line">    <span class="comment"># RQSA (0, 4) 0 4</span></span><br><span class="line">    <span class="comment"># RRSL (18, 22) 18 22</span></span><br><span class="line">    <span class="comment"># RPSK (40, 44) 40 44</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li><code>子组group()</code>的使用</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">seq = <span class="string">'QSAMGSNKSKPKDASQRRRSLEPAENVHGAGGGAFPASQRPSKP'</span></span><br><span class="line"></span><br><span class="line">pattern1 = re.<span class="built_in">compile</span>(<span class="string">'R(.)[ST][^P]'</span>)</span><br><span class="line">match1 = pattern1.search(seq)</span><br><span class="line"><span class="built_in">print</span> match1.group()  <span class="comment"># 结果: RRSL</span></span><br><span class="line"><span class="built_in">print</span> match1.group(<span class="number">1</span>) <span class="comment"># 结果: R</span></span><br><span class="line"></span><br><span class="line">pattern2 = re.<span class="built_in">compile</span>(<span class="string">'R(.{0,3})[ST][^P]'</span>)</span><br><span class="line">match2 = pattern2.search(seq)</span><br><span class="line"><span class="built_in">print</span> match2.group()  <span class="comment"># 结果: RRRSL</span></span><br><span class="line"><span class="built_in">print</span> match2.group(<span class="number">1</span>) <span class="comment"># 结果: RR</span></span><br><span class="line"></span><br><span class="line">seq = <span class="string">'zzabcdzz'</span></span><br><span class="line"></span><br><span class="line">pattern = re.<span class="built_in">compile</span>(<span class="string">'(?P&lt;w1&gt;a(?P&lt;w2&gt;b)c)d'</span>)</span><br><span class="line"><span class="keyword">match</span> = pattern.search(seq)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="number">0</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="string">'w1'</span>)</span><br><span class="line"><span class="built_in">print</span> <span class="keyword">match</span>.group(<span class="string">'w2'</span>)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>Linux命令-<code>grep</code>: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/c/grep.html">Linux grep 命令详解：强大的文本搜索工具 (wangchujiang.com)</a></li>
</ul>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep ^'&gt;' 3G5U.fasta # 筛选带有 `&gt;` 的行</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>PROSITE</code>正则表达式语法:  <a target="_blank" rel="noopener" href="https://prosite.expasy.org/scanprosite/scanprosite_doc.html#pattern_syntax">ScanProsite user manual (expasy.org)</a></p>
<ul>
<li>
<p>可通过字符串的<code>.replace()</code>方法将其转换为Python的<code>re</code>模块的语法</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 此正则表达式用于寻找 钙结合的类EGF结构域签名(PROSITE ID: EGF_CA; PROSITE AC: PS01187)</span></span><br><span class="line">pattern = <span class="string">'[DEQN]-x-[DEQN](2)-C-x(3,14)-C-x(3,7)\</span></span><br><span class="line"><span class="string">      -C-x-[DN]-x(4)-[FY]-x-C'</span></span><br><span class="line">pattern = pattern.replace(<span class="string">'{'</span>,<span class="string">'[^'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'}'</span>,<span class="string">']'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'('</span>,<span class="string">'{'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">')'</span>,<span class="string">'}'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'-'</span>,<span class="string">''</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'x'</span>,<span class="string">'.'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'&gt;'</span>,<span class="string">'$'</span>)</span><br><span class="line">pattern = pattern.replace(<span class="string">'&lt;'</span>,<span class="string">'^'</span>)</span><br><span class="line"><span class="built_in">print</span> pattern</span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
</li>
<li>
<p>在基因组序列中找到转录因子的结合位点:</p>
<ul>
<li><code>转录因子结合位点列表(TFBS)</code>作为检索表达式, 检索基因组序列即可</li>
<li>转录因子: <a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1613417">批量预测转录因子(TF)和转录因子结合位点(TFBS) (tencent.com)</a></li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">genome_seq = <span class="built_in">open</span>(<span class="string">'genome.txt'</span>).read()</span><br><span class="line"></span><br><span class="line"><span class="comment"># read transcription factor binding site patterns</span></span><br><span class="line">sites = []</span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> <span class="built_in">open</span>(<span class="string">'TFBS.txt'</span>):</span><br><span class="line">    fields = line.split()</span><br><span class="line">    tf = fields[<span class="number">0</span>]</span><br><span class="line">    site = fields[<span class="number">1</span>]</span><br><span class="line">    sites.append((tf, site))</span><br><span class="line"></span><br><span class="line"><span class="comment"># match all TF's to the genome and print matches</span></span><br><span class="line"><span class="keyword">for</span> tf, site <span class="keyword">in</span> sites:</span><br><span class="line">    tfbs_regexp = re.<span class="built_in">compile</span>(site)</span><br><span class="line">    all_matches = tfbs_regexp.findall(genome_seq)</span><br><span class="line">    matches = tfbs_regexp.finditer(genome_seq)</span><br><span class="line">    <span class="keyword">if</span> all_matches:</span><br><span class="line">        <span class="built_in">print</span> tf, <span class="string">':'</span></span><br><span class="line">        <span class="keyword">for</span> tfbs <span class="keyword">in</span> matches:</span><br><span class="line">            <span class="built_in">print</span> <span class="string">'\t'</span>, tfbs.group(), tfbs.start(), tfbs.end()</span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p>从PubMed的HTML页面提取标题和摘要文本:</p>
<ul>
<li>通过访问想要的网页然后使用浏览器的<code>开发者模式</code>, 得到想要获取的内容的标签, 这样可以构建正则表达式来检索我们想要的内容.</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> urllib2  <span class="comment"># 网络访问库</span></span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line">pmid = <span class="string">'18235848'</span></span><br><span class="line">url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed?term=%s'</span> % pmid</span><br><span class="line">handler = urllib2.urlopen(url)  <span class="comment"># 使用 urllib2的urlopen方法打开目标网页</span></span><br><span class="line">html = handler.read()</span><br><span class="line"></span><br><span class="line">title_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;h1&gt;.{5,400}&lt;/h1&gt;'</span>)  <span class="comment"># 匹配格式为一级标题的内容字符5~400的标签</span></span><br><span class="line">title_text = title_regexp.search(html)</span><br><span class="line"></span><br><span class="line">abstract_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;AbstractText&gt;.{20,3000}&lt;/AbstractText&gt;'</span>)</span><br><span class="line">abstract_text = abstract_regexp.search(html)</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> <span class="string">'TITLE:'</span>, title_text.group() </span><br><span class="line"><span class="built_in">print</span> <span class="string">'ABSTRACT:'</span>, abstract_text.group()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>检测科学摘要中特定的词或词组:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> urllib2</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line"><span class="comment"># word to be searched</span></span><br><span class="line">word_regexp = re.<span class="built_in">compile</span>(<span class="string">'schistosoma'</span>)  <span class="comment"># 构建想要检索的词编译出的正则对象</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># list of PMIDs where we want to search the word</span></span><br><span class="line">pmids = [<span class="string">'18235848'</span>, <span class="string">'22607149'</span>, <span class="string">'22405002'</span>, <span class="string">'21630672'</span>]  <span class="comment"># 多篇文章的id</span></span><br><span class="line"><span class="keyword">for</span> pmid <span class="keyword">in</span> pmids:</span><br><span class="line">    url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed?term='</span> + pmid</span><br><span class="line">    handler = urllib2.urlopen(url)</span><br><span class="line">    html = handler.read()</span><br><span class="line">    title_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;h1&gt;.{5,400}&lt;/h1&gt;'</span>)</span><br><span class="line">    title = title_regexp.search(html)</span><br><span class="line">    title = title.group() </span><br><span class="line">    abstract_regexp = re.<span class="built_in">compile</span>(<span class="string">'&lt;AbstractText&gt;.{20,3000}&lt;/AbstractText&gt;'</span>)</span><br><span class="line">    abstract = abstract_regexp.search(html)</span><br><span class="line">    abstract = abstract.group()</span><br><span class="line">    word = word_regexp.search(abstract, re.IGNORECASE) <span class="comment"># 匹配想要的字词</span></span><br><span class="line">    <span class="keyword">if</span> word:</span><br><span class="line">        <span class="comment"># display title and where the keyword was found</span></span><br><span class="line">        <span class="built_in">print</span> title</span><br><span class="line">        <span class="built_in">print</span> word.group(), word.start(), word.end()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>氨基酸磷酸化位点:
<ul>
<li>简单的匹配可用正则表达式: <code>'R.[ST][^P]'</code></li>
<li>参考: <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/beab18393d0a">NetPhos 3.1–蛋白磷酸化位点预测 - 简书 (jianshu.com)</a></li>
</ul>
</li>
</ul>
<h2><a href="#di-san-bu-fen-mo-kuai-hua-bian-cheng" class="header-anchor"> ⚓️</a><span id="di-san-bu-fen-mo-kuai-hua-bian-cheng">第三部分 模块化编程</span></h2>
<p>这一部分作者希望读者可以掌握编写由功能较为独立的完善的模块组成的程序</p>
<h3><a href="#di-10-zhang-jiang-cheng-xu-hua-fen-wei-han-shu" class="header-anchor"> ⚓️</a><span id="di-10-zhang-jiang-cheng-xu-hua-fen-wei-han-shu">第10章 将程序划分为函数</span></h3>
<h4><a href="#10-2-an-li-chu-li-san-wei-zuo-biao-wen-jian" class="header-anchor"> ⚓️</a><span id="10-2-an-li-chu-li-san-wei-zuo-biao-wen-jian">10.2 案例： 处理三维坐标文件</span></h4>
<ul>
<li>
<p><code>.PDB</code>格式解读:</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html">Atomic Coordinate Entry Format Version 3.3 (wwpdb.org) </a>  <strong>推荐阅读</strong></li>
</ul>
</li>
<li>
<p><a target="_blank" rel="noopener" href="http://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/primary-sequences-and-the-pdb-format">PDB-101: Learn: Guide to Understanding PDB Data: Primary Sequences and the PDB Format (rcsb.org)</a></p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/aff39489bd80">PDB文件说明 - 简书 (jianshu.com)</a>  <strong>推荐阅读</strong></li>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/klausage/p/10803308.html">PDB(Protein Data Bank)数据格式详解 - Raymone1125 - 博客园 (cnblogs.com)</a></li>
<li>该格式文件获取来源: <a target="_blank" rel="noopener" href="http://www.rcsb.org/">RCSB PDB: Homepage</a></li>
</ul>
<blockquote>
<p>(此记录为书中例子这里摘录作为一个简单的例子)</p>
<p>记录的第一部分此为<strong>标头</strong>, 由几个解释行组成, 包括来源生物体/实验技术(X射线/核磁共振)/交叉引用/实验细节/原始分子序列和变异残基(如果有)等. 记录的第二部分介绍了标准组的原子坐标行. 每一个原子坐标行准确地描述了一个原子. 他们以**‘ATOM’**字符串开始: 1<sub>6列为’ATOM’, 7</sub>11列为’原子序号’, 13<sub>16列为’原子名’, 17列’另存位置指示’, 18</sub>20列为’残基名’, 22列为’链标识符’, 23<sub>26列为’残基序号’, 27列为’残基插入码’, 31</sub>38列为x笛卡尔坐标(x轴单位为埃), 39<sub>46列为’y…‘, 47</sub>54列为’z…’, 55<sub>60列为’布局’, 61</sub>66列为’温度系数’, 77<sub>78列为’元素符号’, 79</sub>81列为’原子电荷’</p>
<p><strong>(值得说明的是<code>原子名</code> <code>CA</code>和<code>CB</code>指的都是碳原子, <code>A</code>和<code>B</code>用于远程标识符, 表示离氨基碳原子的远近, <code>A</code>: α,<code>B</code>: β,<code>G</code>: γ等,原子名称的最后一个字符可以代表分支标识符)</strong></p>
</blockquote>
</li>
<li>
<p><code>struct</code>模块: [7.3. struct — Interpret strings as packed binary data — Python 2.7.18 文档](<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/struct.html?highlight=struct">https://docs.python.org/zh-cn/2.7/library/struct.html?highlight=struct</a> pack#module-struct)</p>
<ul>
<li>
<p><code>struct.pack</code>(<em>fmt</em>, <em>v1</em>, <em>v2</em>, <em>…</em>)</p>
<p>Return a string containing the values <code>v1, v2, ...</code> packed according to <code>the given format</code>. The arguments must match the values required by the format exactly.</p>
</li>
<li>
<p><code>struct.unpack</code>(<em>fmt</em>, <em>string</em>)</p>
<p>Unpack the string (presumably packed by <code>pack(fmt, ...)</code>) according to the given format. The result is a tuple even if it contains exactly one item. The string must contain exactly the amount of data required by the format (<code>len(string)</code> must equal <code>calcsize(fmt)</code>).</p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> struct</span><br><span class="line"></span><br><span class="line"><span class="comment"># pack() method; creates a string</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'2s1s1s1s1s'</span></span><br><span class="line">group = struct.pack(<span class="built_in">format</span>, <span class="string">'10'</span>, <span class="string">'2'</span>, <span class="string">'3'</span>, <span class="string">'4'</span>, <span class="string">'5'</span>) <span class="comment"># 结果: 102345</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># unpack() method; parses the string to a tuple</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'1s2s1s1s'</span></span><br><span class="line">line = <span class="string">'12345'</span></span><br><span class="line">col = struct.unpack(<span class="built_in">format</span>, line)  <span class="comment"># 结果: ('1', '23', '4', '5')</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># calcsize() returns the number of characters</span></span><br><span class="line"><span class="comment"># in a given format string</span></span><br><span class="line"><span class="built_in">format</span> = <span class="string">'30s30s20s1s'</span></span><br><span class="line"><span class="built_in">print</span> struct.calcsize(<span class="built_in">format</span>)  <span class="comment"># 结果: 81</span></span><br></pre></td></tr></tbody></table></figure>
</li>
<li>
<p><code>函数</code>定义与使用: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=%E5%87%BD%E6%95%B0#defining-functions">定义函数 4. 其他流程控制工具 — Python 2.7.18 文档</a>  <strong>详细阅读</strong></p>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">def</span> <span class="title function_">fib</span>(<span class="params">n</span>):    <span class="comment"># write Fibonacci series up to n</span></span><br><span class="line"><span class="meta">... </span>    <span class="string">"""Print a Fibonacci series up to n."""</span></span><br><span class="line"><span class="meta">... </span>    a, b = <span class="number">0</span>, <span class="number">1</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">while</span> a &lt; n:</span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> a,</span><br><span class="line"><span class="meta">... </span>        a, b = b, a+b</span><br><span class="line">...</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="comment"># Now call the function we just defined:</span></span><br><span class="line"><span class="meta">... </span>fib(<span class="number">2000</span>)</span><br><span class="line"><span class="number">0</span> <span class="number">1</span> <span class="number">1</span> <span class="number">2</span> <span class="number">3</span> <span class="number">5</span> <span class="number">8</span> <span class="number">13</span> <span class="number">21</span> <span class="number">34</span> <span class="number">55</span> <span class="number">89</span> <span class="number">144</span> <span class="number">233</span> <span class="number">377</span> <span class="number">610</span> <span class="number">987</span> <span class="number">1597</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>处理PDB数据里的坐标数据变成氨基酸序列:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">'''</span></span><br><span class="line"><span class="string">Extract a FASTA sequence from a PDB file.</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">(c) 2013 Allegra Via and Kristian Rother</span></span><br><span class="line"><span class="string">    Licensed under the conditions of the Python License</span></span><br><span class="line"><span class="string">    This code appears in section 10.2.2 of the book</span></span><br><span class="line"><span class="string">    "Managing Biological Data with Python".</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> struct</span><br><span class="line"></span><br><span class="line">pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s10s2s3s'</span> <span class="comment"># 采用`format`格式 %6s</span></span><br><span class="line"></span><br><span class="line">amino_acids = {</span><br><span class="line">    <span class="string">'ALA'</span>:<span class="string">'A'</span>, <span class="string">'CYS'</span>:<span class="string">'C'</span>, <span class="string">'ASP'</span>:<span class="string">'D'</span>, <span class="string">'GLU'</span>:<span class="string">'E'</span>,</span><br><span class="line">    <span class="string">'PHE'</span>:<span class="string">'F'</span>, <span class="string">'GLY'</span>:<span class="string">'G'</span>, <span class="string">'HIS'</span>:<span class="string">'H'</span>, <span class="string">'LYS'</span>:<span class="string">'K'</span>,</span><br><span class="line">    <span class="string">'ILE'</span>:<span class="string">'I'</span>, <span class="string">'LEU'</span>:<span class="string">'L'</span>, <span class="string">'MET'</span>:<span class="string">'M'</span>, <span class="string">'ASN'</span>:<span class="string">'N'</span>,</span><br><span class="line">    <span class="string">'PRO'</span>:<span class="string">'P'</span>, <span class="string">'GLN'</span>:<span class="string">'Q'</span>, <span class="string">'ARG'</span>:<span class="string">'R'</span>, <span class="string">'SER'</span>:<span class="string">'S'</span>,</span><br><span class="line">    <span class="string">'THR'</span>:<span class="string">'T'</span>, <span class="string">'VAL'</span>:<span class="string">'V'</span>, <span class="string">'TYR'</span>:<span class="string">'Y'</span>, <span class="string">'TRP'</span>:<span class="string">'W'</span></span><br><span class="line">    }</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">threeletter2oneletter</span>(<span class="params">residues</span>):   <span class="comment"># 氨基酸缩写从三个字母 2(to,到) 单个字母 </span></span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Converts the first element of each residue</span></span><br><span class="line"><span class="string">    to a one letter amino acid symbol</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    <span class="keyword">for</span> i, threeletter <span class="keyword">in</span> <span class="built_in">enumerate</span>(residues):</span><br><span class="line">        residues[i][<span class="number">0</span>] = amino_acids[threeletter[<span class="number">0</span>]]</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_residues</span>(<span class="params">pdb_file</span>):  </span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Reads the PDB input file, extracts the</span></span><br><span class="line"><span class="string">    residue type and chain from the CA lines</span></span><br><span class="line"><span class="string">    and appends both to the residues list</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    residues = []</span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> pdb_file:</span><br><span class="line">        <span class="keyword">if</span> line[<span class="number">0</span>:<span class="number">4</span>] == <span class="string">"ATOM"</span>:</span><br><span class="line">            tmp = struct.unpack(pdb_format, line)</span><br><span class="line">            ca = tmp[<span class="number">3</span>].strip()</span><br><span class="line">            <span class="keyword">if</span> ca == <span class="string">'CA'</span>:</span><br><span class="line">                res_type = tmp[<span class="number">5</span>].strip()</span><br><span class="line">                chain = tmp[<span class="number">7</span>]</span><br><span class="line">                residues.append([res_type, chain])</span><br><span class="line">    <span class="keyword">return</span> residues</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_fasta_records</span>(<span class="params">residues, pdb_id, fasta_file</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Write a FASTA record for each PDB chain</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    seq = <span class="string">''</span></span><br><span class="line">    chain = residues[<span class="number">0</span>][<span class="number">1</span>]</span><br><span class="line">    <span class="keyword">for</span> aa, new_chain <span class="keyword">in</span> residues:</span><br><span class="line">        <span class="keyword">if</span> new_chain == chain:</span><br><span class="line">            seq = seq + aa</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># write sequence in FASTA format</span></span><br><span class="line">            fasta_file.write(<span class="string">"&gt;%s_%s\n%s\n"</span> % (pdb_id, chain, seq))</span><br><span class="line">            seq = aa</span><br><span class="line">            chain = new_chain</span><br><span class="line">    <span class="comment"># write the last PDB chain</span></span><br><span class="line">    fasta_file.write(<span class="string">"&gt;%s_%s\n%s\n"</span> % (pdb_id, chain, seq))</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">extract_sequence</span>(<span class="params">pdb_id</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    Main function: Opens files, writes files</span></span><br><span class="line"><span class="string">    and calls other functions.</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    pdb_file = <span class="built_in">open</span>(pdb_id + <span class="string">".pdb"</span>)</span><br><span class="line">    fasta_file = <span class="built_in">open</span>(pdb_id + <span class="string">".fasta"</span>, <span class="string">"w"</span>)</span><br><span class="line">    residues = get_residues(pdb_file)</span><br><span class="line">    threeletter2oneletter(residues)</span><br><span class="line">    write_fasta_records(residues, pdb_id, fasta_file)</span><br><span class="line">    pdb_file.close()</span><br><span class="line">    fasta_file.close()</span><br><span class="line">    </span><br><span class="line"><span class="comment"># call the main function</span></span><br><span class="line">extract_sequence(<span class="string">"3G5U"</span>)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>range()</code>迭代利器: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/functions.html#range">range()  2. 内置函数 — Python 2.7.18 文档</a></p>
</li>
<li>
<p><code>可变参数</code>与<code>不可变参数</code>:</p>
<ul>
<li>不要使用不可变参数作为函数的默认参数</li>
</ul>
</li>
<li>
<p><code>lambda匿名函数</code>: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/controlflow.html?highlight=lambda#lambda-expressions">4. 其他流程控制工具 — Python 2.7.18 文档</a></p>
</li>
</ul>
<h3><a href="#di-11-zhang-yong-lei-hua-fan-wei-jian" class="header-anchor"> ⚓️</a><span id="di-11-zhang-yong-lei-hua-fan-wei-jian">第11章 用类化繁为简</span></h3>
<h4><a href="#11-2-an-li-meng-de-er-yi-chuan" class="header-anchor"> ⚓️</a><span id="11-2-an-li-meng-de-er-yi-chuan">11.2 案例： 孟德尔遗传</span></h4>
<ul>
<li><code>Class(类)</code>的构建:
<ul>
<li><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/classes.html?highlight=class">9. 类 — Python 2.7.18 文档</a>  <strong>仔细阅读</strong></li>
<li><code>.__repr__()</code> 类方法: 当<code>print</code>命令打印类时, 将按照此类方法返回的格式进行输出</li>
<li>类在多人合作编程的优势</li>
<li>子类与继承</li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Pea</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, genotype</span>):  <span class="comment"># 初始化的构造函数</span></span><br><span class="line">        self.genotype = genotype</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">PeaStrain</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, peas</span>):</span><br><span class="line">        self.peas = peas</span><br><span class="line"></span><br><span class="line">yellow = Pea(<span class="string">'GG'</span>)  <span class="comment"># 类的实例化</span></span><br><span class="line">green = Pea(<span class="string">'gg'</span>)</span><br><span class="line">strain = PeaStrain([yellow, green])</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> Pea                 <span class="comment"># __main__.Pea</span></span><br><span class="line"><span class="built_in">print</span> PeaStrain           <span class="comment"># __main__.PeaStrain</span></span><br><span class="line"><span class="built_in">print</span> Pea.__init__        <span class="comment"># &lt;unbound method Pea.__init__&gt;</span></span><br><span class="line"><span class="built_in">print</span> PeaStrain.__init__  <span class="comment"># &lt;unbound method PeaStrain.__init__&gt;</span></span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span> yellow              <span class="comment"># &lt;__main__.Pea instance at 0x0000000001DC6748&gt;</span></span><br><span class="line"><span class="built_in">print</span> green               <span class="comment"># &lt;__main__.Pea instance at 0x000000000216BE08&gt;</span></span><br><span class="line"><span class="built_in">print</span> strain              <span class="comment"># &lt;__main__.PeaStrain instance at 0x000000000216BE48&gt;</span></span><br><span class="line"></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"><span class="string">Classes can inherit from other classes</span></span><br><span class="line"><span class="string">and extend their functionality.</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">(c) 2013 Allegra Via and Kristian Rother</span></span><br><span class="line"><span class="string">    Licensed under the conditions of the Python License</span></span><br><span class="line"><span class="string">    This code appears in section 11.4.3 of the book</span></span><br><span class="line"><span class="string">    "Managing Biological Data with Python".</span></span><br><span class="line"><span class="string">-----------------------------------------------------------</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> pea <span class="keyword">import</span> Pea</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">CommentedPea</span>(<span class="title class_ inherited__">Pea</span>):</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, genotype, comment</span>):</span><br><span class="line">        Pea.__init__(self, genotype)        <span class="comment"># 载入父类构造函数</span></span><br><span class="line">        self.comment = comment</span><br><span class="line"></span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__repr__</span>(<span class="params">self</span>):</span><br><span class="line">        <span class="keyword">return</span>  <span class="string">'%s [%s] (%s)'</span> % (self.get_phenotype(), self.genotype, self.comment)</span><br><span class="line"></span><br><span class="line">yellow1 = CommentedPea(<span class="string">'GG'</span>, <span class="string">'homozygote'</span>)</span><br><span class="line">yellow2 = CommentedPea(<span class="string">'Gg'</span>, <span class="string">'heterozygote'</span>)</span><br><span class="line"><span class="built_in">print</span> yellow1</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>类的设计指南:
<ul>
<li><code>单一职责原则</code>和<code>不重复自己原则</code></li>
<li><code>设计模式</code> : <a target="_blank" rel="noopener" href="https://sourcemaking.com/">Design Patterns &amp; Refactoring (sourcemaking.com)</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-12-zhang-diao-shi" class="header-anchor"> ⚓️</a><span id="di-12-zhang-diao-shi">第12章 调试</span></h3>
<h4><a href="#12-2-an-li-cheng-xu-wu-fa-yun-xing-shi-ying-gai-zen-yang-chu-li" class="header-anchor"> ⚓️</a><span id="12-2-an-li-cheng-xu-wu-fa-yun-xing-shi-ying-gai-zen-yang-chu-li">12.2 案例： 程序无法运行时应该怎样处理</span></h4>
<ul>
<li>
<p>三种错误类型</p>
<ul>
<li><code>语法错误(SyntaxError)</code>: Python规定语法之外的不合法的写法</li>
<li><code>逻辑错误</code>: 程序运行不会报错, 但是程序的运行结果不是你想要的结果</li>
<li><code>运行时错误</code>: 语法检查通过, 但是运行时出现了意想不到的情况的</li>
</ul>
</li>
<li>
<p>异常处理: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/tutorial/errors.html">8. 错误和异常 — Python 2.7.18 文档</a></p>
<ul>
<li><code>try: ... except:</code> 语句:</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">while</span> <span class="literal">True</span>:</span><br><span class="line"><span class="meta">... </span>    <span class="keyword">try</span>:</span><br><span class="line"><span class="meta">... </span>        x = <span class="built_in">int</span>(raw_input(<span class="string">"Please enter a number: "</span>))</span><br><span class="line"><span class="meta">... </span>        <span class="keyword">break</span></span><br><span class="line"><span class="meta">... </span>    <span class="keyword">except</span> ValueError:</span><br><span class="line"><span class="meta">... </span>        <span class="built_in">print</span> <span class="string">"Oops!  That was no valid number.  Try again..."</span></span><br></pre></td></tr></tbody></table></figure>
</li>
</ul>
<blockquote>
<p>我还没有数据. 在掌握数据前建立理论是一个巨大的错误. 不知不觉地, 就开始扭曲事实以适应理论, 而不是让理论符合事实.</p>
<p>@ <em>The Adventures of Sherlock Holmes</em>, Arthur Conan Doyle</p>
</blockquote>
<ul>
<li><code>pdb</code>调试工具: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/pdb.html">26.2. pdb — Python的调试器 — Python 2.7.18 文档</a></li>
</ul>
<h3><a href="#di-13-zhang-shi-yong-wai-bu-mo-kuai-r-yu-yan-de-python-diao-yong-jie-kou" class="header-anchor"> ⚓️</a><span id="di-13-zhang-shi-yong-wai-bu-mo-kuai-r-yu-yan-de-python-diao-yong-jie-kou">第13章 使用外部模块:R语言的Python调用接口</span></h3>
<h4><a href="#13-2-an-li-cong-wen-jian-zhong-du-qu-shu-ju-bing-tong-guo-python-shi-yong-r-ji-suan-qi-ping-jun-zhi" class="header-anchor"> ⚓️</a><span id="13-2-an-li-cong-wen-jian-zhong-du-qu-shu-ju-bing-tong-guo-python-shi-yong-r-ji-suan-qi-ping-jun-zhi">13.2 案例： 从文件中读取数据， 并通过Python使用R计算其平均值</span></h4>
<ul>
<li><code>R语言</code></li>
</ul>
<h3><a href="#di-14-zhang-gou-jian-cheng-xu-liu-cheng" class="header-anchor"> ⚓️</a><span id="di-14-zhang-gou-jian-cheng-xu-liu-cheng">第14章 构建程序流程</span></h3>
<h4><a href="#14-2-an-li-gou-jian-ngs-liu-cheng" class="header-anchor"> ⚓️</a><span id="14-2-an-li-gou-jian-ngs-liu-cheng">14.2 案例： 构建NGS流程</span></h4>
<ul>
<li>
<p><code>NGS</code>技术应用:</p>
<ul>
<li>差异表达基因</li>
<li>非编码RNA</li>
<li>突变</li>
</ul>
</li>
<li>
<p><code>Top Hat</code></p>
</li>
<li>
<p>RNA-seq:</p>
<ul>
<li>输出包含<code>小cDNA测序片段</code> (称为<code>读长</code>,<code>reads</code>)的文件, 需回贴至参考基因组并进行拼装, 以获得全部的样本基因序列</li>
<li>某种流程实例:
<ul>
<li>由<code>Illumina(GAIIx)</code>(或其他NGS)平台生成并存储<code>读长</code>至<code>文本文件(如sample.fastq)</code>中.</li>
<li><code>TopHat</code>可以作为RNA-seq数据的剪切接头回贴工具, 它可以把RNA-seq读长对比至哺乳动物数量级的基因组中, 参考基因组必须保存至本地以便于<code>TopHat</code>访问, 然后分析回贴结果, 找到外显子之间的剪切接头. <code>TopHat</code>读取<code>sample.fastq</code>得到比对读长列表, 储存在<code>.bam</code>文件中.</li>
<li><code>Cufflinks</code>包含有三个程序: 转录本装配(Cufflinks)/转录本比较(Cuffcompare)/检测调控和表达差异(Cuffdiff). <code>Cufflinks</code>进行<code>转录本的拼接</code>, 生成格式为<code>transcripts.gtf</code>的转录文件, 以便之后所用. 对于不同的样本(野生型/癌症细胞, 重复实验)还可以筛选转录组文件, 然后进行相互比较, 按照唯一的参考转录组进行拼接</li>
<li>拼接好后的文件为<code>transcripts.gtf</code>,可以使用Cufflink工具包中的Cuffcompare程序与其他转录组进行比较</li>
</ul>
</li>
</ul>
</li>
<li>
<p><code>os</code>模块: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/os.html?highlight=os#module-os">15.1. os — 操作系统接口模块 — Python 2.7.18 文档</a></p>
</li>
<li>
<p><code>包装器(wrapper)</code>: 可以运行其他程序的程序</p>
</li>
<li>
<p>程序间传递数据:</p>
<ul>
<li><code>raw_input()</code>直接手动输入数据</li>
<li>使用<code>Python</code>变量</li>
<li>将数据存储在另一个文件里</li>
<li>使用命令行参数, 比如<code>sys.argv</code>列表变量, 参见<a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/sys.html?highlight=sys#module-sys">28.1. sys — 系统相关的参数和函数 — Python 2.7.18 文档</a></li>
</ul>
</li>
<li>
<p>关闭文件的延时问题: 通过增加一个中间步骤(比如打开一个文件然后写入一些字符保存, 来确保这个中间步骤的上下两个步骤互不冲突</p>
</li>
<li>
<p><code>os.path</code> 操控文件目录: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/os.path.html#module-os.path">10.1. os.path — 常见路径操作 — Python 2.7.18 文档</a></p>
</li>
<li>
<p><code>MSA(multiple sequence alignment, 多序列比对)</code>:</p>
<blockquote>
<p><strong>Pairwise Sequence Alignment</strong> is used to identify regions of similarity that may indicate functional, structural and/or evolutionary relationships between two biological sequences (protein or nucleic acid).</p>
<p>By contrast, <strong><a target="_blank" rel="noopener" href="https://www.ebi.ac.uk/Tools/msa/">Multiple Sequence Alignment</a> (MSA)</strong> is the alignment of three or more biological sequences of similar length. From the output of MSA applications, homology can be inferred and the evolutionary relationship between the sequences studied.</p>
</blockquote>
<ul>
<li>
<p>MSA工具: <a target="_blank" rel="noopener" href="https://www.genome.jp/tools-bin/clustalw">Multiple Sequence Alignment - CLUSTALW (genome.jp)</a></p>
</li>
<li>
<p><code>T-COFFEE</code>算法: <a target="_blank" rel="noopener" href="http://www.tcoffee.org/Projects/tcoffee/">T-Coffee Home Page (tcoffee.org)</a></p>
<blockquote>
<p>T-Coffee is a multiple sequence alignment package. You can use T-Coffee to align sequences or to combine the output of your favorite alignment methods (Clustal, Mafft, Probcons, Muscle…) into one unique alignment (<a target="_blank" rel="noopener" href="http://www.tcoffee.org/Projects/mcoffee/index.html">M-Coffee</a>).</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<h3><a href="#di-15-zhang-bian-xie-liang-hao-de-cheng-xu" class="header-anchor"> ⚓️</a><span id="di-15-zhang-bian-xie-liang-hao-de-cheng-xu">第15章 编写良好的程序</span></h3>
<h4><a href="#15-2-wen-ti-miao-shu-bu-que-ding-xing" class="header-anchor"> ⚓️</a><span id="15-2-wen-ti-miao-shu-bu-que-ding-xing">15.2 问题描述： 不确定性</span></h4>
<ul>
<li>
<p>《编程的准则》<em>(Dogma of Programming)</em>:</p>
<ul>
<li>著: Donald Knuth</li>
</ul>
</li>
<li>
<p>软件工程周期:</p>
<ul>
<li>搜集需求, 形成以用户为视角的<code>用户故事</code></li>
<li>构建简单有效的模块</li>
<li>构成整个程序, 按照新需求修改</li>
</ul>
</li>
<li>
<p><code>PEP8</code>代码规范:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://pep8.org/">PEP 8: The Style Guide for Python Code (pep8.org)</a></li>
<li><a target="_blank" rel="noopener" href="https://www.python.org/dev/peps/pep-0008/">PEP 8 – Style Guide for Python Code | Python.org</a></li>
</ul>
</li>
<li>
<p><code>pylint</code>:</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://pylint.org/">Pylint - code analysis for Python | www.pylint.org</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/c0bd637f706d">【python】pylint在项目中的使用 - 简书 (jianshu.com)</a></li>
</ul>
</li>
<li>
<p><code>代码管理工具</code>:</p>
<ul>
<li><code>Git</code>:  <a href="/git-shou-ce/" title="Git手册">Git手册</a></li>
</ul>
</li>
<li>
<p>python代码自动打包工具</p>
<ul>
<li><a target="_blank" rel="noopener" href="https://www.cnblogs.com/goldsunshine/p/8872623.html">python打包工具distutils、setuptools分析 - 金色旭光 - 博客园 (cnblogs.com)</a></li>
<li><code>distutils</code>: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/distutils.html?highlight=distutils#module-distutils">27.1. distutils — 构建和安装 Python 模块 — Python 2.7.18 文档</a></li>
<li><code>setuptools</code>:<a target="_blank" rel="noopener" href="https://setuptools.readthedocs.io/en/latest/">Documentation — setuptools 50.3.1 documentation</a></li>
</ul>
</li>
<li>
<p>python代码生成可执行文件</p>
<ul>
<li><code>py2exe</code>: <a target="_blank" rel="noopener" href="http://www.py2exe.org/">FrontPage - py2exe.org</a></li>
<li><code>pyinstaller</code>: <a target="_blank" rel="noopener" href="http://www.pyinstaller.org/">PyInstaller bundles Python applications</a></li>
</ul>
</li>
<li>
<p>敏捷开发: <a target="_blank" rel="noopener" href="http://agilemanifesto.org/iso/zhchs/manifesto.html">敏捷软件开发宣言 (agilemanifesto.org)</a></p>
<blockquote>
<p><strong>敏捷软件开发宣言</strong></p>
<p>我们一直在实践中探寻更好的软件开发方法，<br>
身体力行的同时也帮助他人。由此我们建立了如下价值观：</p>
<p>个体和互动 高于 流程和工具<br>
工作的软件 高于 详尽的文档<br>
客户合作 高于 合同谈判<br>
响应变化 高于 遵循计划</p>
<p>也就是说，尽管右项有其价值，<br>
我们更重视左项的价值。</p>
<table>
<thead>
<tr>
<th>Kent Beck Mike Beedle Arie van Bennekum Alistair Cockburn Ward Cunningham Martin Fowler</th>
<th>James Grenning Jim Highsmith Andrew Hunt Ron Jeffries Jon Kern Brian Marick</th>
<th>Robert C. Martin Steve Mellor Ken Schwaber Jeff Sutherland Dave Thomas</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>著作权为上述作者所有，2001年</p>
</blockquote>
<ul>
<li>适于科研界的工业化实例: <a target="_blank" rel="noopener" href="https://www.perforce.com/products/helix-alm">Helix ALM | ALM Tools | Perforce</a></li>
<li>对生物信息学研究的各种软件工程技术(如自动化测试/代码评审/用户故事)都有描述的工具箱:  <a target="_blank" rel="noopener" href="https://www.researchgate.net/publication/51535066_A_toolbox_for_developing_bioinformatics_software">A toolbox for developing bioinformatics software | Request PDF (researchgate.net)</a></li>
</ul>
</li>
</ul>
<h2><a href="#di-si-bu-fen-shu-ju-ke-shi-hua" class="header-anchor"> ⚓️</a><span id="di-si-bu-fen-shu-ju-ke-shi-hua">第四部分 数据可视化</span></h2>
<h3><a href="#di-16-zhang-chuang-jian-ke-xue-tu-biao" class="header-anchor"> ⚓️</a><span id="di-16-zhang-chuang-jian-ke-xue-tu-biao">第16章 创建科学图表</span></h3>
<h4><a href="#16-2-an-li-he-tang-ti-de-he-gan-suan-pin-lu" class="header-anchor"> ⚓️</a><span id="16-2-an-li-he-tang-ti-de-he-gan-suan-pin-lu">16.2 案例： 核糖体的核苷酸频率</span></h4>
<blockquote>
<p>2009年的诺贝尔化学奖授予了Venkatraman RamaKrishnan, Thomas Steitz 和 Ada Yonath, 以表彰他们对核糖体结构和功能的研究. 核糖体是已知的最大的反子机器之一, 由三种RNA组分(原核生物中是23S, 16S 和 5S rRNA) 和许多的蛋白质组成. RNA的构成包括四种基本的核糖核苷酸和一些起微调核糖体功能作用的修是核糖体</p>
</blockquote>
<ul>
<li><code>matplotlib</code>:
<ul>
<li>官方教程: <a target="_blank" rel="noopener" href="https://matplotlib.org/">Matplotlib: Python plotting — Matplotlib 3.3.2 documentation</a></li>
<li>官方模板: <a target="_blank" rel="noopener" href="https://matplotlib.org/gallery/index.html#gallery">Gallery — Matplotlib 3.3.2 documentation</a></li>
</ul>
</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> pylab <span class="keyword">import</span> figure, plot, savefig</span><br><span class="line"></span><br><span class="line">xdata = [<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>]</span><br><span class="line">ydata = [<span class="number">1.25</span>, <span class="number">2.5</span>, <span class="number">5.0</span>, <span class="number">10.0</span>]</span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line">plot(xdata, ydata)</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'figure1.png'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># -----</span></span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line"></span><br><span class="line">xdata = [<span class="number">0.1</span> * i <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">100</span>)]</span><br><span class="line">ydata = [math.sin(j) <span class="keyword">for</span> j <span class="keyword">in</span> xdata]</span><br><span class="line"></span><br><span class="line">plot(xdata, ydata, <span class="string">'kd'</span>, linewidth = <span class="number">1</span>)</span><br><span class="line">text(<span class="number">4.8</span>, <span class="number">0</span>, <span class="string">"$y = sin(x)$"</span>, horizontalalignment = <span class="string">'center'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">axis([<span class="number">0</span>, <span class="number">3</span> * math.pi, -<span class="number">1.2</span>, <span class="number">1.2</span>])</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'sinfunc.png'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ---</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> pylab <span class="keyword">import</span> figure, title, xlabel, ylabel, hist, axis, grid, savefig</span><br><span class="line"></span><br><span class="line">data = [<span class="number">1</span>, <span class="number">1</span>, <span class="number">9</span>, <span class="number">1</span>, <span class="number">3</span>, <span class="number">5</span>, <span class="number">8</span>, <span class="number">2</span>, <span class="number">1</span>, <span class="number">5</span>, <span class="number">11</span>, <span class="number">8</span>, <span class="number">3</span>, <span class="number">4</span>, <span class="number">2</span>, <span class="number">5</span>]</span><br><span class="line">n_bins = <span class="number">5</span></span><br><span class="line"></span><br><span class="line">figure()</span><br><span class="line">num, bins, patches = hist(data, n_bins, normed = <span class="number">1.0</span>, histtype = <span class="string">'bar'</span>, facecolor = <span class="string">'green'</span>, alpha = <span class="number">0.75</span>)</span><br><span class="line"></span><br><span class="line">title(<span class="string">'Histogram'</span>)</span><br><span class="line">xlabel(<span class="string">'value'</span>)</span><br><span class="line">ylabel(<span class="string">'frequency'</span>)</span><br><span class="line">axis()</span><br><span class="line">grid(<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line">savefig(<span class="string">'histogram.png'</span>)</span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#di-17-zhang-shi-yong-pymol-chuang-jian-fen-zi-tu-xiang" class="header-anchor"> ⚓️</a><span id="di-17-zhang-shi-yong-pymol-chuang-jian-fen-zi-tu-xiang">第17章 使用PyMOL创建分子图像</span></h3>
<h4><a href="#17-2-shi-li-xin-zhi" class="header-anchor"> ⚓️</a><span id="17-2-shi-li-xin-zhi">17.2 示例： 锌指</span></h4>
<ul>
<li>
<p>锌指:</p>
<blockquote>
<p>锌指(英语: Zinc finger),又称锌手指, 是一种小的蛋白质结构模体, 其特征在于配合一个或多个锌离子(Zn<sup>2+</sup>)以稳定折叠. 最初用来描述非洲爪蟾卵母细胞转录因子IIIA假说结构的手指状外观</p>
<p>一个锌指包含由两个螺旋, 它们通过一个锌原子形成特殊的构象, 导致一个螺旋与DNA的大沟匹配</p>
<p>锌指名称现在已经包含了各种不同的蛋白质结构</p>
<p>@<a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/blog-635619-765781.html">科学网—什么是锌指蛋白？ - 倪帅的博文 (sciencenet.cn)</a></p>
<p>@<a target="_blank" rel="noopener" href="https://zh.wikipedia.org/zh-hans/%E9%94%8C%E6%8C%87">锌指 - 维基百科，自由的百科全书 (wikipedia.org)</a></p>
</blockquote>
</li>
<li>
<p><code>PyMol</code>: <a target="_blank" rel="noopener" href="https://pymol.org/2/">PyMOL | pymol.org</a></p>
<blockquote>
<p>PyMOL is a user-sponsored molecular visualization system on an <a target="_blank" rel="noopener" href="https://pymol.org/2/#opensource">open-source foundation</a>, maintained and distributed by <a target="_blank" rel="noopener" href="https://www.schrodinger.com/">Schrödinger</a>.</p>
</blockquote>
<ul>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/58803491">免费开源版PyMOL安装 - 知乎 (zhihu.com)</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/schrodinger/pymol-open-source">schrodinger/pymol-open-source: Open-source foundation of the user-sponsored PyMOL molecular visualization system. (github.com)</a></li>
</ul>
</li>
</ul>
<h3><a href="#di-18-zhang-chu-li-tu-xiang" class="header-anchor"> ⚓️</a><span id="di-18-zhang-chu-li-tu-xiang">第18章 处理图像</span></h3>
<h4><a href="#18-2-an-li-hua-yi-ge-zhi-li" class="header-anchor"> ⚓️</a><span id="18-2-an-li-hua-yi-ge-zhi-li">18.2 案例： 画一个质粒</span></h4>
<blockquote>
<p>1977年 Francisco Bolivar 和 Raymond Rodriguez 构建了最早的人工质粒. 它包含了4361个DNA碱基对, 一个复制起点, 一个氨苄青霉素抗性基因和一个四环素抗性基因. 该质粒由许多限制性位点, 从而成为构建遗传载体的基础.</p>
</blockquote>
<ul>
<li>
<p><code>PIL(Python Imaging Library)</code>图形库:</p>
<blockquote>
<p>The <strong>Python Imaging Library (PIL)</strong> adds image processing capabilities to your Python interpreter. This library supports many file formats, and provides powerful image processing and graphics capabilities.</p>
<p>PIL 仅支持Python2</p>
</blockquote>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="http://pythonware.com/products/pil/">Python Imaging Library (PIL) (pythonware.com)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://www.liaoxuefeng.com/wiki/897692888725344/966759628285152">PIL - 廖雪峰的官方网站 (liaoxuefeng.com)</a></p>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://pillow.readthedocs.io/en/stable/">Pillow — Pillow (PIL Fork) 8.0.0 documentation</a></p>
<blockquote>
<p>Pillow is the friendly PIL fork by <a target="_blank" rel="noopener" href="https://github.com/python-pillow/Pillow/graphs/contributors">Alex Clark and Contributors</a>. PIL is the Python Imaging Library by Fredrik Lundh and Contributors.</p>
<p>Pillow 支持Python3</p>
</blockquote>
</li>
<li>
<p><a target="_blank" rel="noopener" href="https://github.com/python-pillow/Pillow">python-pillow/Pillow: The friendly PIL fork (Python Imaging Library) (github.com)</a></p>
</li>
</ul>
</li>
<li>
<p>蛋白质结构域架构:</p>
<ul>
<li>
<p><code>InterPro</code>数据库: <a target="_blank" rel="noopener" href="http://www.ebi.ac.uk/interpro/">InterPro (ebi.ac.uk)</a></p>
<blockquote>
<p>InterPro provides functional analysis of proteins by classifying them into families and predicting domains and important sites. To classify proteins in this way, InterPro uses predictive models, known as signatures, provided by several different databases (referred to as member databases) that make up the InterPro consortium. We combine protein signatures from these member databases into a single searchable resource, capitalising on their individual strengths to produce a powerful integrated database and diagnostic tool.</p>
</blockquote>
</li>
</ul>
</li>
</ul>
<h2><a href="#di-wu-bu-fen-biopython" class="header-anchor"> ⚓️</a><span id="di-wu-bu-fen-biopython">第五部分 Biopython</span></h2>
<ul>
<li>
<p><a target="_blank" rel="noopener" href="https://biopython.org/">Biopython</a> 是一个计算分子生物学模块的集合, 用它可以实现许多生物信息学项目中的基本任务</p>
<blockquote>
<p>Biopython is a set of freely available tools for biological computation written in <a target="_blank" rel="noopener" href="http://www.python.org/">Python</a> by an international team of developers.</p>
</blockquote>
<ul>
<li>
<p>解析生物数据格式</p>
</li>
<li>
<p>从资源库下载资源</p>
</li>
<li>
<p>运行常用的生物信息学算法</p>
</li>
<li>
<p>运行Biopython实现的算法, 进行聚类, 机器学习, 数据分析和可视化</p>
</li>
<li>
<p>各种文档索引: <a target="_blank" rel="noopener" href="https://biopython.org/wiki/Documentation">Documentation · Biopython</a></p>
</li>
<li>
<p>上手指南: <a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">Biopython Tutorial and Cookbook</a></p>
</li>
</ul>
</li>
</ul>
<h3><a href="#di-19-zhang-shi-yong-xu-lie-shu-ju" class="header-anchor"> ⚓️</a><span id="di-19-zhang-shi-yong-xu-lie-shu-ju">第19章 使用序列数据</span></h3>
<h4><a href="#19-2-an-li-ru-he-jiang-yi-tiao-dna-bian-ma-xu-lie-fan-yi-cheng-dui-ying-de-dan-bai-zhi-xu-lie-bing-ba-ta-xie-ru-fasta-wen-jian" class="header-anchor"> ⚓️</a><span id="19-2-an-li-ru-he-jiang-yi-tiao-dna-bian-ma-xu-lie-fan-yi-cheng-dui-ying-de-dan-bai-zhi-xu-lie-bing-ba-ta-xie-ru-fasta-wen-jian">19.2 案例： 如何将一条DNA编码序列翻译成对应的蛋白质序列并把它写入FASTA文件</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec17">Chapter 3 Sequence objects (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio.Seq <span class="keyword">import</span> Seq</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq = Seq(<span class="string">"AGTACACTGGT"</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq</span><br><span class="line">Seq(<span class="string">'AGTACACTGGT'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">print</span>(my_seq)</span><br><span class="line">AGTACACTGGT</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq</span><br><span class="line">Seq(<span class="string">'AGTACACTGGT'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq.complement()</span><br><span class="line">Seq(<span class="string">'TCATGTGACCA'</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq.reverse_complement()</span><br><span class="line">Seq(<span class="string">'ACCAGTGTACT'</span>)</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> seq_record <span class="keyword">in</span> SeqIO.parse(<span class="string">"ls_orchid.fasta"</span>, <span class="string">"fasta"</span>):</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(seq_record.<span class="built_in">id</span>)</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="built_in">repr</span>(seq_record.seq))</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="built_in">len</span>(seq_record))</span><br><span class="line">gi|<span class="number">2765658</span>|emb|Z78533<span class="number">.1</span>|CIZ78533</span><br><span class="line">Seq(<span class="string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span>)</span><br><span class="line"><span class="number">740</span></span><br><span class="line">...</span><br><span class="line">gi|<span class="number">2765564</span>|emb|Z78439<span class="number">.1</span>|PBZ78439</span><br><span class="line">Seq(<span class="string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span>)</span><br><span class="line"><span class="number">592</span></span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">from</span> Bio.Seq <span class="keyword">import</span> Seq</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span>my_seq = Seq(<span class="string">"GATCG"</span>)</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="keyword">for</span> index, letter <span class="keyword">in</span> <span class="built_in">enumerate</span>(my_seq):</span><br><span class="line"><span class="meta">... </span>    <span class="built_in">print</span>(<span class="string">"%i %s"</span> % (index, letter))</span><br><span class="line"><span class="number">0</span> G</span><br><span class="line"><span class="number">1</span> A</span><br><span class="line"><span class="number">2</span> T</span><br><span class="line"><span class="number">3</span> C</span><br><span class="line"><span class="number">4</span> G</span><br><span class="line"><span class="meta">&gt;&gt;&gt; </span><span class="built_in">print</span>(<span class="built_in">len</span>(my_seq))</span><br><span class="line"><span class="number">5</span></span><br><span class="line"></span><br><span class="line">...</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>
<p><code>IUPAC字符集</code>:</p>
<blockquote>
<p>预编译好的字符集, 覆盖了所有生物序列. 包括 IUPACUnamiguousDNA(基本的ATCG字母), IUPACAmbiguousDNA(包含二义字母), ExtendedIUPACDNA(包含修饰的碱基), IUPACUnamiguousRNA,  IUPACAmbiguousRNA, ExtendedIUPACRNA, IUPACProtein(IPUAC标准氨基酸)和ExtendedIUPACProtein(包括硒代半胱氨酸, X等)</p>
</blockquote>
</li>
<li>
<p>迭代器:</p>
<blockquote>
<p>一个迭代器是产生一系列的项(如SeqRecord对象)的一种数据结构. 它可以像列表一样使用在循环中, 但技术上它不是一个列表. 一个迭代器没有长度, 也不能被索引或者是切片. 用户只能向它申请下一个元素, 当你这样做时, 迭代器寻找是否存在更多的可用的记录. 这种方式下, 迭代器无需把所有的记录一致放在内存中.</p>
</blockquote>
</li>
<li>
<p>NCBI核酸翻译编码标准: <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">The Genetic Codes</a></p>
</li>
</ul>
<h3><a href="#di-20-zhang-cong-wang-luo-zi-yuan-zhong-jian-suo-shu-ju" class="header-anchor"> ⚓️</a><span id="di-20-zhang-cong-wang-luo-zi-yuan-zhong-jian-suo-shu-ju">第20章 从网络资源中检索数据</span></h3>
<h4><a href="#20-2-an-li-zai-pubmed-zhong-yong-guan-jian-ci-sou-suo-wen-xian-xia-zai-bing-jie-xi-dui-ying-de-ji-lu" class="header-anchor"> ⚓️</a><span id="20-2-an-li-zai-pubmed-zhong-yong-guan-jian-ci-sou-suo-wen-xian-xia-zai-bing-jie-xi-dui-ying-de-ji-lu">20.2 案例： 在PubMed中用关键词搜索文献， 下载并解析对应的记录</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec139">Chapter 9 Accessing NCBI’s Entrez databases (biopython.org)</a></p>
<h3><a href="#di-21-zhang-shi-yong-san-wei-jie-gou-shu-ju" class="header-anchor"> ⚓️</a><span id="di-21-zhang-shi-yong-san-wei-jie-gou-shu-ju">第21章 使用三维结构数据</span></h3>
<h4><a href="#21-2-an-li-cong-pdb-wen-jian-zhong-ti-qu-yuan-zi-ming-ji-qi-san-wei-zuo-biao" class="header-anchor"> ⚓️</a><span id="21-2-an-li-cong-pdb-wen-jian-zhong-ti-qu-yuan-zi-ming-ji-qi-san-wei-zuo-biao">21.2 案例： 从PDB文件中提取原子名及其三维坐标</span></h4>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec178">Chapter 11 Going 3D: The PDB module (biopython.org)</a></p>
<h2><a href="#di-liu-bu-fen-bian-cheng-mi-ji" class="header-anchor"> ⚓️</a><span id="di-liu-bu-fen-bian-cheng-mi-ji">第六部分 编程秘笈</span></h2>
<ul>
<li>编程秘笈01: PyCogent库</li>
</ul>
<p><code>PyCogent</code>: <a target="_blank" rel="noopener" href="https://github.com/pycogent/pycogent/">pycogent/pycogent: PyCogent: Official repository for software and unit tests (github.com)</a></p>
<blockquote>
<p>PyCogent includes connectors to remote databases, built-in generalized probabilistic techniques for working with biological sequences, and controllers for 3rd party applications.</p>
<p>项目疑似不再维护</p>
</blockquote>
<p>参考文献: <a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2007-8-8-r171">PyCogent: a toolkit for making sense from sequence | Genome Biology | Full Text (biomedcentral.com)</a></p>
<ul>
<li>编程秘笈02: 反向互补和随机化序列</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/random.html?highlight=random#module-random">9.6. random — 生成伪随机数 — Python 2.7.18 文档</a></p>
<p>为什么要随机化序列:</p>
<blockquote>
<p>每个序列模式可能在序列中随机出现, 在蛋白质和核酸中寻找匹配的功能模体不能保证该匹配一定有生物意义. 为了评估一个模体存在的生物学显著性, 需要检查的序列一啊不能要和随机集合进行比较. 在给定的序列集合中相对于随机序列集合有更显著表达的序列模体, 更倾向于编码一个有功能的性质(更有生物学意义). 一个好的随机序列是缺乏生物学意义的, 但是却与生物对象有相同的氨基酸/核苷酸组分, 可以通过对原序列反向,洗牌,随机化等方式创建.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">seq = <span class="string">'ABCDEFGHIJKLMNOPQRSTUVWXYZ'</span></span><br><span class="line"></span><br><span class="line">seq_list = <span class="built_in">list</span>(seq)    </span><br><span class="line">seq_list.reverse()</span><br><span class="line">rev_seq = <span class="string">''</span>.join(seq_list)</span><br><span class="line"><span class="built_in">print</span>(rev_seq)   <span class="comment"># 列表化后反转再拼接回字符串</span></span><br><span class="line"></span><br><span class="line">rev_seq = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> s <span class="keyword">in</span> <span class="built_in">reversed</span>(seq):</span><br><span class="line">    rev_seq = rev_seq + s</span><br><span class="line"><span class="built_in">print</span>(rev_seq)  <span class="comment"># 使用reversed()迭代器循环反向拼接字符串</span></span><br><span class="line"></span><br><span class="line">rev_seq = seq[::-<span class="number">1</span>] <span class="comment"># 字符串切片[start:end:step], 设步长为-1, 得反向</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> random</span><br><span class="line"><span class="comment"># 返回由seq内元素组成的长为len(seq)的列表</span></span><br><span class="line">ran_seq = random.sample(seq, <span class="built_in">len</span>(seq)) </span><br><span class="line"><span class="comment"># 转字符串</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join(random.sample(seq, <span class="built_in">len</span>(seq)))</span><br><span class="line"><span class="comment"># 控制长度</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join(random.sample(seq, <span class="built_in">len</span>(seq)-<span class="number">10</span>))</span><br><span class="line"><span class="comment"># 使用random.choice方法在seq中每次随机返回一个元素, for循环控制次数</span></span><br><span class="line">ran_seq = <span class="string">''</span>.join([random.choice(seq) <span class="keyword">for</span> x <span class="keyword">in</span> <span class="built_in">range</span>(<span class="built_in">len</span>(seq))])</span><br><span class="line"></span><br><span class="line">data = <span class="built_in">list</span>(seq)</span><br><span class="line">random.shuffle(data) <span class="comment"># 序列洗牌</span></span><br><span class="line">shuffled_seq = data</span><br><span class="line">shuffled_seq = <span class="string">''</span>.join(data)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈03: 用概率创建随机序列</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> random</span><br><span class="line"></span><br><span class="line">nucleotides = <span class="built_in">list</span>(<span class="string">'ACGT'</span>)</span><br><span class="line">dna = <span class="string">''</span></span><br><span class="line"><span class="keyword">while</span> <span class="built_in">len</span>(dna) &lt; <span class="number">100</span>:</span><br><span class="line">    dna += random.choice(nucleotides) <span class="comment"># 随机生成100个碱基组成的DNA序列</span></span><br><span class="line"></span><br><span class="line">nucleotides = <span class="built_in">list</span>(<span class="string">'ACGT'</span>)</span><br><span class="line">probs = {<span class="string">'A'</span>: <span class="number">0.3</span>, <span class="string">'C'</span>: <span class="number">0.2</span>, <span class="string">'G'</span>: <span class="number">0.2</span>, <span class="string">'T'</span>: <span class="number">0.3</span>} <span class="comment"># 总概率为1</span></span><br><span class="line"><span class="keyword">assert</span> <span class="built_in">sum</span>(probs.values()) == <span class="number">1.0</span> </span><br><span class="line"><span class="comment"># 上述assert语句等价于下面的两行</span></span><br><span class="line"><span class="keyword">if</span> <span class="built_in">sum</span>(probs.values()) != <span class="number">1.0</span>:</span><br><span class="line">  <span class="keyword">raise</span> Exception(<span class="string">'Sum of probabilites is not 1.0!'</span>)</span><br><span class="line"></span><br><span class="line">dna = <span class="string">''</span></span><br><span class="line"><span class="keyword">while</span> <span class="built_in">len</span>(dna) &lt; <span class="number">100</span>:</span><br><span class="line">    nuc = random.choice(nucleotides)</span><br><span class="line">    dice = random.random()</span><br><span class="line">    <span class="keyword">if</span> dice &lt; probs[nuc]:</span><br><span class="line">        dna += nuc</span><br><span class="line"><span class="built_in">print</span> dna</span><br></pre></td></tr></tbody></table></figure>
<p><code>assert</code> 语句: <a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/reference/simple_stmts.html?highlight=assert#assert">assert (python.org)</a></p>
<p>assert 语句是在程序中插入调试性断言的简便方式:</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">assert_stmt ::=  "assert" expression ["," expression]</span><br></pre></td></tr></tbody></table></figure>
<p>简单形式 <code>assert expression</code> 等价于</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">if __debug__:</span><br><span class="line">    if not expression: raise AssertionError</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈04: 用Biopython解析多序列联配</li>
</ul>
<p>多序列联配(MSA):</p>
<p><code>Pfam</code>数据库: <a target="_blank" rel="noopener" href="https://pfam.xfam.org/">Pfam: Home page</a></p>
<p>常用格式为<code>.sth (Stockholm 格式)</code></p>
<p><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/10bfcdf06693">biostar handbook(六)| 序列联配 - 简书 (jianshu.com)</a></p>
<p>例子: <a target="_blank" rel="noopener" href="https://pfam.xfam.org/family/PF00042#tabview=tab0">Pfam: Family: Globin (PF00042) (xfam.org)</a></p>
<p>通过Biopython的Bio.AlignIO模块解析多联配:</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec79">Chapter 6 Multiple Sequence Alignment objects (biopython.org)</a></p>
<ul>
<li>编程秘笈05: 从多序列联配中计算共有序列</li>
</ul>
<p>共有序列(Consensus Sequence): 多联配中每列中频率最高的残基(最保守的)</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">seqs = [</span><br><span class="line">     <span class="string">'ATCCAGCT'</span>,</span><br><span class="line">     <span class="string">'GGGCAACT'</span>,</span><br><span class="line">     <span class="string">'ATGGATCT'</span>,</span><br><span class="line">     <span class="string">'AAGCAACC'</span>,</span><br><span class="line">     <span class="string">'ATGCCATT'</span>,</span><br><span class="line">     <span class="string">'ATGGCACT'</span>]</span><br><span class="line"></span><br><span class="line">n = <span class="built_in">len</span>(seqs[<span class="number">0</span>])</span><br><span class="line">profile = {<span class="string">'A'</span>:[<span class="number">0</span>]*n, <span class="string">'C'</span>:[<span class="number">0</span>]*n, <span class="string">'G'</span>:[<span class="number">0</span>]*n, <span class="string">'T'</span>:[<span class="number">0</span>]*n}</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> seq <span class="keyword">in</span> seqs:</span><br><span class="line">    <span class="keyword">for</span> i, char <span class="keyword">in</span> <span class="built_in">enumerate</span>(seq):</span><br><span class="line">        profile[char][i] += <span class="number">1</span></span><br><span class="line">consensus = <span class="string">''</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(n):</span><br><span class="line">    col = [(profile[nt][i], nt) <span class="keyword">for</span> nt <span class="keyword">in</span> <span class="string">'ATCG'</span>]</span><br><span class="line">    consensus += <span class="built_in">max</span>(col)[<span class="number">1</span>]</span><br><span class="line">    </span><br><span class="line"><span class="built_in">print</span> consensus</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈06: 计算系统发生树的节点间的距离</li>
</ul>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec249">Chapter 13 Phylogenetics with Bio.Phylo</a></p>
<p><code>Newick</code>格式是表示系统发生树的一种统一格式. 它可以被大多数的程序读取, 包括MrBayes, GARLI, PHYLIP, TREE-PUZZLE等. Newick树包括物种和祖先节点的名字, 它们的相互关系, 以及每个节点赋予的值(通常用于父节点的距离)</p>
<ul>
<li>编程秘笈07: 核苷酸序列的密码子频率</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># This program calculates the codon frequency in a RNA sequence</span></span><br><span class="line"><span class="comment"># (it could also be an entire genome)</span></span><br><span class="line">aa_codon = {</span><br><span class="line"><span class="string">'A'</span>:[<span class="string">'GCU'</span>,<span class="string">'GCC'</span>,<span class="string">'GCA'</span>,<span class="string">'GCG'</span>], <span class="string">'C'</span>:[<span class="string">'UGU'</span>,<span class="string">'UGC'</span>],</span><br><span class="line"><span class="string">'D'</span>:[<span class="string">'GAU'</span>,<span class="string">'GAC'</span>],<span class="string">'E'</span>:[<span class="string">'GAA'</span>,<span class="string">'GAG'</span>],<span class="string">'F'</span>:[<span class="string">'UUU'</span>,<span class="string">'UUC'</span>],</span><br><span class="line"><span class="string">'G'</span>:[<span class="string">'GGU'</span>,<span class="string">'GGC'</span>,<span class="string">'GGA'</span>,<span class="string">'GGG'</span>], <span class="string">'H'</span>:[<span class="string">'CAU'</span>,<span class="string">'CAC'</span>],</span><br><span class="line"><span class="string">'K'</span>:[<span class="string">'AAA'</span>,<span class="string">'AAG'</span>],<span class="string">'I'</span>:[<span class="string">'AUU'</span>,<span class="string">'AUC'</span>,<span class="string">'AUA'</span>,<span class="string">'AUU'</span>,<span class="string">'AUC'</span>,<span class="string">'AUA'</span>],</span><br><span class="line"><span class="string">'L'</span>:[<span class="string">'UUA'</span>,<span class="string">'UUG'</span>,<span class="string">'CUU'</span>,<span class="string">'CUC'</span>,<span class="string">'CUA'</span>,<span class="string">'CUG'</span>],<span class="string">'M'</span>:[<span class="string">'AUG'</span>],</span><br><span class="line"><span class="string">'N'</span>:[<span class="string">'AAU'</span>,<span class="string">'AAC'</span>],<span class="string">'P'</span>:[<span class="string">'CCU'</span>,<span class="string">'CCC'</span>,<span class="string">'CCA'</span>,<span class="string">'CCG'</span>],</span><br><span class="line"><span class="string">'Q'</span>:[<span class="string">'CAA'</span>,<span class="string">'CAG'</span>],<span class="string">'R'</span>:[<span class="string">'CGU'</span>,<span class="string">'CGC'</span>,<span class="string">'CGA'</span>,<span class="string">'CGG'</span>,<span class="string">'AGA'</span>,<span class="string">'AGG'</span>],</span><br><span class="line"><span class="string">'S'</span>:[<span class="string">'UCU'</span>,<span class="string">'UCC'</span>,<span class="string">'UCA'</span>,<span class="string">'UCG'</span>,<span class="string">'AGU'</span>,<span class="string">'AGC'</span>,],</span><br><span class="line"><span class="string">'Y'</span>:[<span class="string">'UAU'</span>,<span class="string">'UAC'</span>],<span class="string">'T'</span>:[<span class="string">'ACU'</span>,<span class="string">'ACC'</span>,<span class="string">'ACA'</span>,<span class="string">'ACG'</span>],</span><br><span class="line"><span class="string">'V'</span>:[<span class="string">'GUU'</span>,<span class="string">'GUC'</span>,<span class="string">'GUA'</span>,<span class="string">'GUG'</span>],<span class="string">'W'</span>:[<span class="string">'UGG'</span>],</span><br><span class="line"><span class="string">'STOP'</span>:[<span class="string">'UAG'</span>,<span class="string">'UGA'</span>,<span class="string">'UAA'</span>]}</span><br><span class="line"></span><br><span class="line">codon_count = {</span><br><span class="line"><span class="string">'GCU'</span>:<span class="number">0</span>,<span class="string">'GCC'</span>:<span class="number">0</span>,<span class="string">'GCA'</span>:<span class="number">0</span>,<span class="string">'GCG'</span>:<span class="number">0</span>,<span class="string">'CGU'</span>:<span class="number">0</span>,<span class="string">'CGC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CGA'</span>:<span class="number">0</span>,<span class="string">'CGG'</span>:<span class="number">0</span>,<span class="string">'AGA'</span>:<span class="number">0</span>,<span class="string">'AGG'</span>:<span class="number">0</span>,<span class="string">'UCU'</span>:<span class="number">0</span>,<span class="string">'UCC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UCA'</span>:<span class="number">0</span>,<span class="string">'UCG'</span>:<span class="number">0</span>,<span class="string">'AGU'</span>:<span class="number">0</span>,<span class="string">'AGC'</span>:<span class="number">0</span>,<span class="string">'AUU'</span>:<span class="number">0</span>,<span class="string">'AUC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'AUA'</span>:<span class="number">0</span>,<span class="string">'AUU'</span>:<span class="number">0</span>,<span class="string">'AUC'</span>:<span class="number">0</span>,<span class="string">'AUA'</span>:<span class="number">0</span>,<span class="string">'UUA'</span>:<span class="number">0</span>,<span class="string">'UUG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CUU'</span>:<span class="number">0</span>,<span class="string">'CUC'</span>:<span class="number">0</span>,<span class="string">'CUA'</span>:<span class="number">0</span>,<span class="string">'CUG'</span>:<span class="number">0</span>,<span class="string">'GGU'</span>:<span class="number">0</span>,<span class="string">'GGC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'GGA'</span>:<span class="number">0</span>,<span class="string">'GGG'</span>:<span class="number">0</span>,<span class="string">'GUU'</span>:<span class="number">0</span>,<span class="string">'GUC'</span>:<span class="number">0</span>,<span class="string">'GUA'</span>:<span class="number">0</span>,<span class="string">'GUG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'ACU'</span>:<span class="number">0</span>,<span class="string">'ACC'</span>:<span class="number">0</span>,<span class="string">'ACA'</span>:<span class="number">0</span>,<span class="string">'ACG'</span>:<span class="number">0</span>,<span class="string">'CCU'</span>:<span class="number">0</span>,<span class="string">'CCC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CCA'</span>:<span class="number">0</span>,<span class="string">'CCG'</span>:<span class="number">0</span>,<span class="string">'AAU'</span>:<span class="number">0</span>,<span class="string">'AAC'</span>:<span class="number">0</span>,<span class="string">'GAU'</span>:<span class="number">0</span>,<span class="string">'GAC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UGU'</span>:<span class="number">0</span>,<span class="string">'UGC'</span>:<span class="number">0</span>,<span class="string">'CAA'</span>:<span class="number">0</span>,<span class="string">'CAG'</span>:<span class="number">0</span>,<span class="string">'GAA'</span>:<span class="number">0</span>,<span class="string">'GAG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'CAU'</span>:<span class="number">0</span>,<span class="string">'CAC'</span>:<span class="number">0</span>,<span class="string">'AAA'</span>:<span class="number">0</span>,<span class="string">'AAG'</span>:<span class="number">0</span>,<span class="string">'UUU'</span>:<span class="number">0</span>,<span class="string">'UUC'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UAU'</span>:<span class="number">0</span>,<span class="string">'UAC'</span>:<span class="number">0</span>,<span class="string">'AUG'</span>:<span class="number">0</span>,<span class="string">'UGG'</span>:<span class="number">0</span>,<span class="string">'UAG'</span>:<span class="number">0</span>,</span><br><span class="line"><span class="string">'UGA'</span>:<span class="number">0</span>,<span class="string">'UAA'</span>:<span class="number">0</span>}</span><br><span class="line"></span><br><span class="line"><span class="comment"># Writes the frequency of each codon to a file</span></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">calc_freq</span>(<span class="params">codon_count, out_file</span>):</span><br><span class="line">    count_tot = {}</span><br><span class="line">    <span class="keyword">for</span> aa <span class="keyword">in</span> aa_codon.keys():</span><br><span class="line">        n = <span class="number">0</span></span><br><span class="line">        <span class="keyword">for</span> codon <span class="keyword">in</span> aa_codon[aa]:</span><br><span class="line">            n = n + codon_count[codon]</span><br><span class="line">        count_tot[aa] = <span class="built_in">float</span>(n)</span><br><span class="line">    <span class="keyword">for</span> aa <span class="keyword">in</span> aa_codon.keys():</span><br><span class="line">        <span class="keyword">for</span> codon <span class="keyword">in</span> aa_codon[aa]:</span><br><span class="line">            <span class="keyword">if</span> count_tot[aa] != <span class="number">0.0</span>:</span><br><span class="line">                freq = codon_count[codon] / count_tot[aa]</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                freq = <span class="number">0.0</span></span><br><span class="line">            out_file.write(<span class="string">'%4s\t%5s\t%4d\t%9.3f\n'</span>% \</span><br><span class="line">                (aa,codon,codon_count[codon], freq))</span><br><span class="line">in_file = <span class="built_in">open</span>(<span class="string">'A06662.1.fasta'</span>)</span><br><span class="line">out_file = <span class="built_in">open</span>(<span class="string">'CodonFrequency.txt'</span>, <span class="string">'w'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Reads the RNA sequence into a single string</span></span><br><span class="line">rna = <span class="string">''</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> line[<span class="number">0</span>] == <span class="string">'&gt;'</span>:</span><br><span class="line">        rna = rna + line.strip()</span><br><span class="line">        </span><br><span class="line"><span class="comment"># Scans the sequence frame by frame,</span></span><br><span class="line"><span class="comment"># counts the number of occurrences</span></span><br><span class="line"><span class="comment"># of each codon, and stores it in codon_count dictionary.</span></span><br><span class="line"><span class="comment"># Then calls calc_freq()</span></span><br><span class="line"><span class="keyword">for</span> j <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">3</span>):</span><br><span class="line">    out_file.write(<span class="string">'!!Codon frequency in frame %d\n'</span> %(j+<span class="number">1</span>))</span><br><span class="line">    out_file.write(<span class="string">' AA\tcodon\thits\tfrequency\n'</span>)</span><br><span class="line">    prot = <span class="string">''</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(j, <span class="built_in">len</span>(rna), <span class="number">3</span>):</span><br><span class="line">        codon = rna[i:i + <span class="number">3</span>]</span><br><span class="line">        <span class="keyword">if</span> codon <span class="keyword">in</span> codon_count:</span><br><span class="line">            codon_count[codon] = codon_count[codon] + <span class="number">1</span></span><br><span class="line">    calc_freq(codon_count, out_file)</span><br><span class="line">    </span><br><span class="line">out_file.close()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈08: 解析Vienna格式的RNA二级结构</li>
</ul>
<p>RNA序列最重要的性质之一是其折叠和配对的能力. 与DNA不同, RNA的碱基配对不限于双螺旋, 而是能够形成复杂的结构. 预测这些RNA的碱基配对和二级结构是RNA生物信息学的常见任务.</p>
<p><code>Vienna</code>包: <a target="_blank" rel="noopener" href="http://rna.tbi.univie.ac.at/">ViennaRNA Web Services (univie.ac.at)</a></p>
<p>一个命令行和网络的工具集, 可以进行如从RNA序列预测碱基配对等基础任务</p>
<p>这些结果通常被保存为Vienna格式</p>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt; two hairpin loops</span><br><span class="line">AAACCCCGUUUCGGGGAACCACCA</span><br><span class="line">((((...)))).((((..)).)).</span><br></pre></td></tr></tbody></table></figure>
<p>第一行包含了RNA序列名称, 第二行是序列本身</p>
<p>第三行的点-括号结构表达RNA二级结构</p>
<p>对应的括号指示一个碱基对, 而不配对的碱基对用点表示</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">RNAStructure</span>:</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">__init__</span>(<span class="params">self, vienna</span>):</span><br><span class="line">        lines = vienna.split(<span class="string">'\n'</span>)</span><br><span class="line">        self.name = lines[<span class="number">0</span>].strip()</span><br><span class="line">        self.sequence = lines[<span class="number">1</span>].strip()</span><br><span class="line">        self.basepairs = \</span><br><span class="line">        <span class="built_in">sorted</span>(self.parse_basepairs(lines[<span class="number">2</span>].strip()))</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">parse_basepairs</span>(<span class="params">self, dotbracket</span>):</span><br><span class="line">        stack = []</span><br><span class="line">        <span class="keyword">for</span> i, char <span class="keyword">in</span> <span class="built_in">enumerate</span>(dotbracket):</span><br><span class="line">            <span class="keyword">if</span> char == <span class="string">'('</span>:</span><br><span class="line">                stack.append(i)</span><br><span class="line">            <span class="keyword">elif</span> char == <span class="string">')'</span>:</span><br><span class="line">                j = stack.pop()</span><br><span class="line">        <span class="keyword">yield</span> j, i</span><br><span class="line">        </span><br><span class="line">vienna = <span class="string">'''&gt; two hairpin loops</span></span><br><span class="line"><span class="string">AAACCCCGUUUCGGGGAACCACCA</span></span><br><span class="line"><span class="string">((((...)))).((((..)).)).</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line">rna = RNAStructure(vienna)</span><br><span class="line"><span class="built_in">print</span> rna.name</span><br><span class="line"><span class="built_in">print</span> rna.sequence</span><br><span class="line"><span class="keyword">for</span> base1, base2 <span class="keyword">in</span> rna.basepairs:</span><br><span class="line">    <span class="built_in">print</span> <span class="string">'(%i, %i)'</span>%(base1, base2)</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈09: 解析BLAST的XML输出</li>
</ul>
<p>Biopython BLAST XML解析器读取XML BLAST输出文件</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec125">7.3 Parsing BLAST output (biopython.org)</a></p>
<p>NCBIXML.parse()函数返回一个blast_out实例, 它包含一到多个记录对象(BLAST结果在一个记录里, 而如PSI-BLAST则是对每个查询有一个纪录). 每个查询都包含一到多个匹配(hit)或联配(alignment), 依次包含一到多个局部联配(HSP).</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>blastp -query P05480.fasta -db nr<span class="number">.00</span> -out blast_output.xml</span><br><span class="line">-outfmt <span class="number">5</span></span><br><span class="line"></span><br><span class="line">xml_file = <span class="built_in">open</span>(<span class="string">"blast_output.xml"</span>)</span><br><span class="line">blast_out = NCBIXML.parse(xml_file)</span><br><span class="line"><span class="keyword">for</span> record <span class="keyword">in</span> blast_out:</span><br><span class="line">    <span class="keyword">for</span> alignment <span class="keyword">in</span> record.alignments:</span><br><span class="line">        <span class="built_in">print</span> alignment.title</span><br><span class="line">        <span class="keyword">for</span> hsp <span class="keyword">in</span> alignment.hsps:</span><br><span class="line">            <span class="comment"># filter by e-value</span></span><br><span class="line">            <span class="keyword">if</span> hsp.expect &lt; <span class="number">0.0001</span>:</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"score:"</span>, hsp.score</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"query:"</span>, hsp.query</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"match:"</span>, hsp.<span class="keyword">match</span></span><br><span class="line">                <span class="built_in">print</span> <span class="string">"sbjct:"</span>, hsp.sbjct</span><br><span class="line">                <span class="built_in">print</span> <span class="string">'#'</span> * <span class="number">70</span></span><br><span class="line">                <span class="built_in">print</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈10: 解析SBML文件</li>
</ul>
<p><strong>Systems Biology Markup Language</strong> (<strong>SBML</strong>), a free and open interchange format for computer models of biological processes. SBML is useful for models of metabolism, cell signaling, and more. It <strong>continues to be evolved and expanded</strong> by an international community.</p>
<p><a target="_blank" rel="noopener" href="http://sbml.org/Main_Page">Main Page - SBML.caltech.edu</a></p>
<p>SBML格式用来储存通路,反应和调控网络信息的标准格式</p>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/xml/xml-tutorial.html">XML 教程 | 菜鸟教程 (runoob.com)</a></p>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/xml.html?highlight=xml#module-xml">19.5. XML处理模块 — Python 2.7.18 文档</a></p>
<figure class="highlight xml"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&lt;?xml version=<span class="string">"1.0"</span> encoding=<span class="string">"UTF-8"</span>?&gt;</span></span><br><span class="line"><span class="tag">&lt;<span class="name">sbml</span> <span class="attr">xmlns</span>=<span class="string">"http://www.sbml.org/sbml/level2"</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;<span class="name">model</span> <span class="attr">name</span>=<span class="string">"SBML file with three metabolites"</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;<span class="name">listOfSpecies</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m78"</span> <span class="attr">name</span>=<span class="string">"Inosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O5<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m79"</span> <span class="attr">name</span>=<span class="string">"Xanthosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O6<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;<span class="name">species</span> <span class="attr">id</span>=<span class="string">"M_m80"</span> <span class="attr">name</span>=<span class="string">"Xanthosine"</span>&gt;</span></span><br><span class="line">            <span class="tag">&lt;<span class="name">p</span>&gt;</span>FORMULA: C10H12N4O6<span class="tag">&lt;/<span class="name">p</span>&gt;</span></span><br><span class="line">        <span class="tag">&lt;/<span class="name">species</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;/<span class="name">listOfSpecies</span>&gt;</span></span><br><span class="line">    <span class="tag">&lt;/<span class="name">model</span>&gt;</span></span><br><span class="line"><span class="tag">&lt;/<span class="name">sbml</span>&gt;</span></span><br></pre></td></tr></tbody></table></figure>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> xml.dom.minidom <span class="keyword">import</span> parse</span><br><span class="line"></span><br><span class="line">document = parse(<span class="string">'sbml_example.xml'</span>)</span><br><span class="line">species_list = document.getElementsByTagName(<span class="string">"species"</span>)</span><br><span class="line"><span class="keyword">for</span> species <span class="keyword">in</span> species_list:</span><br><span class="line">    species_id = species.getAttribute(<span class="string">'id'</span>)</span><br><span class="line">    name = species.getAttribute(<span class="string">'name'</span>)</span><br><span class="line">    p_list = species.getElementsByTagName(<span class="string">"p"</span>)</span><br><span class="line">    p = p_list[<span class="number">0</span>]</span><br><span class="line">    text = p.childNodes[<span class="number">0</span>]</span><br><span class="line">    formula = text.nodeValue</span><br><span class="line">    <span class="built_in">print</span> <span class="string">"%-20s\t%5s\t%s"</span>%(name, species_id, formula)</span><br><span class="line">    </span><br><span class="line"><span class="comment"># Inosine      M_m78 FORMULA: C10H12N4O5</span></span><br><span class="line"><span class="comment"># Xanthosine   M_m79 FORMULA: C10H12N4O6</span></span><br><span class="line"><span class="comment"># Xanthosine   M_m80 FORMULA: C10H12N4O6</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈11: 运行BLAST</li>
</ul>
<blockquote>
<p>The NCBI provides <strong>a suite of command-line tools to run BLAST called BLAST+</strong>. This allows users to perform BLAST searches on their own server without size, volume and database restrictions. BLAST+ can be used with a command line so it can be integrated directly into your workflow.</p>
</blockquote>
<p><a target="_blank" rel="noopener" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">Download BLAST Software and Databases Documentation (nih.gov)</a></p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec121">7.2 Running BLAST locally (Biopython)</a></p>
<ul>
<li>编程秘笈12: 访问、 下载和读取网页</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/urllib.html?highlight=urllib">20.5. urllib — Open arbitrary resources by URL — Python 2.7.18 文档</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> urllib <span class="keyword">import</span> urlopen</span><br><span class="line">url = urlopen(<span class="string">'http://www.uniprot.org/uniprot/P01308.fasta'</span>)</span><br><span class="line">doc = url.read()</span><br><span class="line"><span class="built_in">print</span> doc</span><br></pre></td></tr></tbody></table></figure>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/urllib2.html">20.6. urllib2 — extensible library for opening URLs — Python 2.7.18 文档</a></p>
<ul>
<li>编程秘笈13: 解析HTML文件</li>
</ul>
<p><a target="_blank" rel="noopener" href="https://docs.python.org/zh-cn/2.7/library/htmlparser.html?highlight=htmlparser#module-HTMLParser">19.1. HTMLParser — Simple HTML and XHTML parser — Python 2.7.18 文档</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> HTMLParser <span class="keyword">import</span> HTMLParser</span><br><span class="line"><span class="keyword">import</span> urllib</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">MyHTMLParser</span>(<span class="title class_ inherited__">HTMLParser</span>):</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_starttag</span>(<span class="params">self, tag, attrs</span>):</span><br><span class="line">        self.start_tag = tag</span><br><span class="line">        <span class="built_in">print</span> <span class="string">"Start tag:"</span>, self.start_tag</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_endtag</span>(<span class="params">self, tag</span>):</span><br><span class="line">        self.end_tag = tag</span><br><span class="line">        <span class="built_in">print</span> <span class="string">"End tag :"</span>, self.end_tag</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">handle_data</span>(<span class="params">self, data</span>):</span><br><span class="line">        self.data = data.strip()</span><br><span class="line">        <span class="keyword">if</span> self.data:</span><br><span class="line">            <span class="built_in">print</span> <span class="string">"Data :"</span>, self.data</span><br><span class="line">            </span><br><span class="line">parser = MyHTMLParser()</span><br><span class="line">url = <span class="string">'http://www.ncbi.nlm.nih.gov/pubmed/21998472'</span></span><br><span class="line">page = urllib.urlopen(url)</span><br><span class="line">data = page.read()</span><br><span class="line">parser.feed(data)</span><br><span class="line"></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈14: 将PDB文件分割成PDB链文件</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> struct <span class="keyword">import</span> unpack</span><br><span class="line"><span class="keyword">import</span> os.path</span><br><span class="line"></span><br><span class="line">filename = <span class="string">'2H8L.pdb'</span></span><br><span class="line">in_file = <span class="built_in">open</span>(filename)</span><br><span class="line">pdb_id = filename.split(<span class="string">'.'</span>)[<span class="number">0</span>]</span><br><span class="line">pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s6s4s2s3s'</span></span><br><span class="line">chain_old = <span class="string">'@'</span></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">    <span class="keyword">if</span> line[<span class="number">0</span>:<span class="number">4</span>] == <span class="string">"ATOM"</span>:</span><br><span class="line">        col = unpack(pdb_format, line)</span><br><span class="line">        chain = col[<span class="number">7</span>].strip()</span><br><span class="line">        <span class="keyword">if</span> chain != chain_old:</span><br><span class="line">            <span class="keyword">if</span> os.path.exists(pdb_id+chain_old+<span class="string">'.pdb'</span>):</span><br><span class="line">                chain_file.close()</span><br><span class="line">                <span class="built_in">print</span> <span class="string">"closed:"</span>, pdb_id+chain_old+<span class="string">'.pdb'</span></span><br><span class="line">            chain_file = <span class="built_in">open</span>(pdb_id+chain+<span class="string">'.pdb'</span>,<span class="string">'w'</span>)</span><br><span class="line">            chain_file.write(line)</span><br><span class="line">            chain_old = chain</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            chain_file.write(line)</span><br><span class="line">chain_file.close()</span><br><span class="line"><span class="built_in">print</span> <span class="string">"closed:"</span>, pdb_id+chain_old+<span class="string">'.pdb'</span></span><br></pre></td></tr></tbody></table></figure>
<p>也可以使用终端的<code>&gt;</code>将结果导出到文件</p>
<ul>
<li>编程秘笈15: 在PDB结构上找到两个最靠近的Cα原子</li>
</ul>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> math <span class="keyword">import</span> sqrt</span><br><span class="line"><span class="keyword">from</span> struct <span class="keyword">import</span> unpack</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">calc_dist</span>(<span class="params">p1, p2</span>):</span><br><span class="line">    <span class="string">'''returns the distance between two 3D points'''</span></span><br><span class="line">    tmp = <span class="built_in">pow</span>(p1[<span class="number">0</span>] - p2[<span class="number">0</span>], <span class="number">2</span>) + \</span><br><span class="line">        <span class="built_in">pow</span>(p1[<span class="number">1</span>] - p2[<span class="number">1</span>], <span class="number">2</span>) + \</span><br><span class="line">        <span class="built_in">pow</span>(p1[<span class="number">2</span>] - p2[<span class="number">2</span>], <span class="number">2</span>)</span><br><span class="line">    tmp = sqrt(tmp)</span><br><span class="line">    <span class="keyword">return</span> tmp</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">min_dist</span>(<span class="params">arglist</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    returns the closest residue pair and their</span></span><br><span class="line"><span class="string">    CA_CA distance</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    <span class="comment"># initialize variables</span></span><br><span class="line">    maxval = <span class="number">10000</span></span><br><span class="line">    residue_pair = ()</span><br><span class="line">    <span class="comment"># read arglist starting from the 1st position</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(<span class="built_in">len</span>(arglist)):</span><br><span class="line">        <span class="comment"># save x,y,z coordinates from the arglist</span></span><br><span class="line">        <span class="comment"># i-element into the atom1 variable</span></span><br><span class="line">        atom1 = arglist[i][<span class="number">1</span>:]</span><br><span class="line">        <span class="comment"># run over all other elements</span></span><br><span class="line">        <span class="keyword">for</span> j <span class="keyword">in</span> <span class="built_in">range</span>(i + <span class="number">1</span>, <span class="built_in">len</span>(arglist)):</span><br><span class="line">            atom2 = arglist[j][<span class="number">1</span>:]</span><br><span class="line">            <span class="comment"># calculate the distance</span></span><br><span class="line">            tmp = calc_dist(atom1, atom2)</span><br><span class="line">            <span class="comment"># check if the distance is lower than</span></span><br><span class="line">            <span class="comment"># the previously recorded lowest value</span></span><br><span class="line">            <span class="keyword">if</span> tmp &lt; maxval :</span><br><span class="line">                <span class="comment"># save the new data</span></span><br><span class="line">                residue_pair = (arglist[i][<span class="number">0</span>], \</span><br><span class="line">                                arglist[j][<span class="number">0</span>])</span><br><span class="line">                maxval = tmp</span><br><span class="line">    <span class="keyword">return</span> residue_pair, maxval</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_list_ca_atoms</span>(<span class="params">pdb_file, chain</span>):</span><br><span class="line">    <span class="string">'''</span></span><br><span class="line"><span class="string">    returns a list of CA atoms, the residues</span></span><br><span class="line"><span class="string">    they belong to, and their x,y,z coordinates</span></span><br><span class="line"><span class="string">    from the input PDB file</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    in_file = <span class="built_in">open</span>(pdb_file)</span><br><span class="line">    CA_list = []</span><br><span class="line">    pdb_format = <span class="string">'6s5s1s4s1s3s1s1s4s1s3s8s8s8s6s6s6s4s2s3s'</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> in_file:</span><br><span class="line">        tmp = unpack(pdb_format, line)</span><br><span class="line">        tmp = [i.strip() <span class="keyword">for</span> i <span class="keyword">in</span> tmp]</span><br><span class="line">        <span class="comment"># only save CA coords belonging to input chain</span></span><br><span class="line">        <span class="keyword">if</span> tmp[<span class="number">0</span>] ==<span class="string">"ATOM"</span> <span class="keyword">and</span> tmp[<span class="number">7</span>] == chain <span class="keyword">and</span> \</span><br><span class="line">            tmp[<span class="number">3</span>] == <span class="string">"CA"</span>:</span><br><span class="line">            <span class="comment"># create a tuple (aa_number, x, y, x)</span></span><br><span class="line">            tmp = (tmp[<span class="number">5</span>]+tmp[<span class="number">8</span>], <span class="built_in">float</span>(tmp[<span class="number">11</span>]), \</span><br><span class="line">            <span class="built_in">float</span>(tmp[<span class="number">12</span>]), <span class="built_in">float</span>(tmp[<span class="number">13</span>]))</span><br><span class="line">            <span class="comment"># add the tuple to the list</span></span><br><span class="line">            CA_list.append(tmp)</span><br><span class="line">    in_file.close()</span><br><span class="line">    <span class="keyword">return</span> CA_list</span><br><span class="line"></span><br><span class="line"><span class="comment"># obtain the list of CA atoms of Chain A</span></span><br><span class="line">CA_list = get_list_ca_atoms(<span class="string">"2H8L.pdb"</span>, <span class="string">"A"</span>)</span><br><span class="line"><span class="comment"># identify the closest atoms</span></span><br><span class="line">res_pair, dist = min_dist(CA_list)</span><br><span class="line"><span class="built_in">print</span> <span class="string">'The distance between'</span>, res_pair, <span class="string">'is:'</span>, dist</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈16: 提取两个PDB链间的界面</li>
</ul>
<p>在一个三维结构中, 两条多肽链之间的界面被定义为满足这两个条件的所有残基对:1.两个残基分别属于不同的链;2.它们之间的距离小于某个阈值</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec189">11.2 Structure representation (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> PDB</span><br><span class="line">parser = PDB.PDBParser()</span><br><span class="line">s = parser.get_structure(<span class="string">"2H8L"</span>,<span class="string">"2H8L.pdb"</span>)</span><br><span class="line">first_model = s[<span class="number">0</span>]</span><br><span class="line">chain_A = first_model[<span class="string">"A"</span>]</span><br><span class="line">chain_B = first_model[<span class="string">"B"</span>]</span><br><span class="line"><span class="keyword">for</span> res1 <span class="keyword">in</span> chain_A:</span><br><span class="line">    <span class="keyword">for</span> res2 <span class="keyword">in</span> chain_B:</span><br><span class="line">        d = res1[<span class="string">"CA"</span>]-res2[<span class="string">"CA"</span>]</span><br><span class="line">        <span class="keyword">if</span> d &lt;= <span class="number">6.0</span>:</span><br><span class="line">            <span class="built_in">print</span> res1.resname,res1.get_id()[<span class="number">1</span>], res2.resname,\</span><br><span class="line">                res2.get_id()[<span class="number">1</span>], d</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈17: 用Modeller建立同源模型</li>
</ul>
<p><code>Modeller</code> 是一个为建立蛋白质三维结构的同源模型设计的软件包</p>
<p><a target="_blank" rel="noopener" href="https://salilab.org/modeller/">About MODELLER (salilab.org)</a></p>
<blockquote>
<p>MODELLER is used for homology or comparative modeling of protein three-dimensional structures (1,2). The user provides an alignment of a sequence to be modeled with known related structures and MODELLER automatically calculates a model containing all non-hydrogen atoms. MODELLER implements comparative protein structure modeling by satisfaction of spatial restraints (3,4), and can perform many additional tasks, including de novo modeling of loops in protein structures, optimization of various models of protein structure with respect to a flexibly defined objective function, multiple alignment of protein sequences and/or structures, clustering, searching of sequence databases, comparison of protein structures, etc.</p>
</blockquote>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> modeller <span class="keyword">import</span> *</span><br><span class="line"><span class="keyword">from</span> modeller.automodel <span class="keyword">import</span> *</span><br><span class="line">log.verbose()</span><br><span class="line">env = environ()</span><br><span class="line">env.io.atom_files_directory = [<span class="string">'.'</span>, <span class="string">'../atom_files'</span>]</span><br><span class="line">a = automodel(env,</span><br><span class="line">            alnfile = <span class="string">'alignment.ali'</span>,</span><br><span class="line">            knowns = <span class="string">'1eq9A'</span>,</span><br><span class="line">            sequence = <span class="string">'MyTarget_Seq'</span>)</span><br><span class="line">a.starting_model = <span class="number">1</span></span><br><span class="line">a.ending_model = <span class="number">1</span></span><br><span class="line">a.make()</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈18: 用ModeRNA分析RNA三维同源模型</li>
</ul>
<p><code>ModeRNA</code> 库: <a target="_blank" rel="noopener" href="https://pypi.org/project/moderna/">moderna · PyPI</a> 服务维护停止</p>
<p>分析/操纵/建模RNA三维结构</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> moderna <span class="keyword">import</span> *</span><br><span class="line"></span><br><span class="line">ehz = load_model(<span class="string">'1ehz.ent'</span>, <span class="string">'A'</span>)</span><br><span class="line">clean_structure(ehz)</span><br><span class="line"><span class="built_in">print</span> get_sequence(ehz)</span><br><span class="line"><span class="built_in">print</span> get_secstruc(ehz)</span><br><span class="line"></span><br><span class="line">m = create_model()</span><br><span class="line">copy_some_residues(ehz[<span class="string">'1'</span>:<span class="string">'15'</span>], m)</span><br><span class="line">write_model(m, <span class="string">'1ehz_15r.ent'</span>)</span><br><span class="line">temp = load_template(<span class="string">'1ehz_15r.ent'</span>)</span><br><span class="line">ali = load_alignment(<span class="string">'''&gt; model sequence</span></span><br><span class="line"><span class="string">ACUGUGAYUA[UACCU#PG</span></span><br><span class="line"><span class="string">&gt; template: first 15 bases from 1ehz</span></span><br><span class="line"><span class="string">GCGGA--UUUALCUCAG'''</span>)</span><br><span class="line">model = create_model(temp, ali)</span><br><span class="line"><span class="built_in">print</span> get_sequence(model)</span><br><span class="line"></span><br><span class="line">GCGGAUUUALCUCAGDDGGGAGAGCRCCAGABU<span class="comment">#AAYAP?UGGAG7UC?UGUGTPCG"UCC</span></span><br><span class="line">  ACAGAAUUCGCACCA</span><br><span class="line">(((((((..((((........)))).((((.........)))).....</span><br><span class="line">   (((((.......))))))))))))....</span><br><span class="line">ACUGUGAYUA[UACCU<span class="comment">#PG</span></span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈19: 从三级结构计算RNA碱基配对</li>
</ul>
<p><code>RNAView</code>: 计算一个RNA分子PDB结构中的碱基配对</p>
<p><a target="_blank" rel="noopener" href="http://ndbserver.rutgers.edu/ndbmodule/services/download/rnaview.html">NDB | Tools | Downloads (rutgers.edu)</a></p>
<blockquote>
<p><strong><a target="_blank" rel="noopener" href="http://ndbserver.rutgers.edu/ndbmodule/services/rna_viewer/rnaViewerIndex.html">RNA Viewer</a></strong>, a tool for viewing RNA 2-dimensional structure, base pairs and RNA patterns. 2D structure is fully annotated according to Leontis and Westhof (2001) <em>RNA</em> <strong>7</strong>:499-512.</p>
<p>Yang, H., Jossinet, F., Leontis, N., Chen, L., Westbrook, J., Berman, H.M. and Westhof, E. (2003) Tools for the automatic identification and classification of RNA base pairs. Nucleic Acids Res, 31, 3450-3460.</p>
</blockquote>
<p>PyCogent库</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;&gt;&gt; </span>rnaview lehz.pdb &gt; basepair.out</span><br><span class="line"><span class="number">1_72</span>, A:  <span class="number">1</span> G-C <span class="number">72</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">2_71</span>, A:  <span class="number">2</span> C-G <span class="number">71</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">3_70</span>, A:  <span class="number">3</span> G-C <span class="number">70</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">4_69</span>, A:  <span class="number">4</span> G-C <span class="number">69</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">5_68</span>, A:  <span class="number">5</span> G-C <span class="number">68</span>  A: +/+ cis  XIX</span><br><span class="line"><span class="number">6_67</span>, A:  <span class="number">6</span> A-U <span class="number">67</span>  A: W/W cis  n/a</span><br><span class="line"><span class="number">7_66</span>, A:  <span class="number">7</span> A-U <span class="number">66</span>  A: -/- cis  XX</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> cogent.app.rnaview <span class="keyword">import</span> RnaView</span><br><span class="line"><span class="keyword">from</span> cogent.parse.rnaview <span class="keyword">import</span> RnaviewParser</span><br><span class="line">rna_prog = RnaView()</span><br><span class="line">result = rna_prog(<span class="string">'1ehz.pdb'</span>)</span><br><span class="line">bpairs = result[<span class="string">'base_pairs'</span>]</span><br><span class="line">errors = result[<span class="string">'StdErr'</span>].read()</span><br><span class="line">stdout = result[<span class="string">'StdOut'</span>].read()</span><br><span class="line"></span><br><span class="line">bp_dict = RnaviewParser(bpairs)</span><br><span class="line"><span class="built_in">print</span> <span class="string">'INFORMATION:'</span></span><br><span class="line">sys.stderr.write(errors)</span><br><span class="line"><span class="built_in">print</span> stdout</span><br><span class="line"><span class="built_in">print</span> <span class="string">'BASE PAIRS:'</span></span><br><span class="line"><span class="keyword">for</span> key <span class="keyword">in</span> bp_dict:</span><br><span class="line">    <span class="built_in">print</span> key, bp_dict[key]</span><br></pre></td></tr></tbody></table></figure>
<ul>
<li>编程秘笈20: 结构重叠的真实实例: 丝氨酸蛋白酶催化三分子</li>
</ul>
<p>结构重叠: 在重叠的过程中, 两个结构之一的(靶标)被放置在一个固定的位置上,另一个(探针)被平移和旋转, 直到连个结构的均方差达到最小. 一个重叠必须要求两个集合的原子数相同才能工作, 所以在重叠两个结构之前, 需要决定每个结构上哪些原子需要重叠, 这甚至可以识整个结构的所有原子.</p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec238">11.8 Accessing the Protein Data Bank (biopython.org)</a></p>
<p><a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec223">11.6.6 Superimposing two structures (biopython.org)</a></p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Superimpose the catalytic triads of two different serine</span></span><br><span class="line"><span class="comment"># proteases(on CA and N atoms of res H57, D102, and S195 of chain A)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> PDB</span><br><span class="line"></span><br><span class="line"><span class="comment"># Retrieve PDB files   网络获取三个结构文件</span></span><br><span class="line">pdbl = PDB.PDBList()</span><br><span class="line">pdbl.retrieve_pdb_file(<span class="string">"1EQ9"</span>)</span><br><span class="line">pdbl.retrieve_pdb_file(<span class="string">"1FXY"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Parse the two structures  到对应目录读取文件结构信息</span></span><br><span class="line"><span class="keyword">from</span> Bio.PDB <span class="keyword">import</span> PDBParser, Superimposer, PDBIO</span><br><span class="line">parser = PDB.PDBParser()</span><br><span class="line">struct_1 = parser.get_structure(<span class="string">"1EQ9"</span>, <span class="string">"eq/pdb1eq9.ent"</span>)</span><br><span class="line">struct_2 = parser.get_structure(<span class="string">"1FXY"</span>, <span class="string">"fx/pdb1fxy.ent"</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># get the catalytic triads  提取三个催化位点氨基酸信息</span></span><br><span class="line">res_57_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">57</span>]</span><br><span class="line">res_102_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">102</span>]</span><br><span class="line">res_195_struct_1 = struct_1[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">195</span>]</span><br><span class="line">res_57_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">57</span>]</span><br><span class="line">res_102_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">102</span>]</span><br><span class="line">res_195_struct_2 = struct_2[<span class="number">0</span>][<span class="string">'A'</span>][<span class="number">195</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># Build 2 lists of atoms for calculating a rot.-trans. matrix</span></span><br><span class="line"><span class="comment"># (target and probe). 每个位点氨基酸的α碳以及氮原子放入目标和探针列表</span></span><br><span class="line">target = []</span><br><span class="line">backbone_names = [<span class="string">'CA'</span>, <span class="string">'N'</span>]</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> backbone_names:</span><br><span class="line">    target.append(res_57_struct_1[name])</span><br><span class="line">    target.append(res_102_struct_1[name])</span><br><span class="line">    target.append(res_195_struct_1[name])</span><br><span class="line">    </span><br><span class="line">probe = []</span><br><span class="line"><span class="keyword">for</span> name <span class="keyword">in</span> backbone_names:</span><br><span class="line">    probe.append(res_57_struct_2[name])</span><br><span class="line">    probe.append(res_102_struct_2[name])</span><br><span class="line">    probe.append(res_195_struct_2[name])</span><br><span class="line">    </span><br><span class="line"><span class="comment"># Check whether target and probe lists are equal in size.</span></span><br><span class="line"><span class="comment"># This is needed for calculating a rot.-trans. matrix</span></span><br><span class="line"><span class="keyword">assert</span> <span class="built_in">len</span>(target) == <span class="built_in">len</span>(probe)  <span class="comment"># 确定原子数量相等</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Calculate the rotation-translation matrix.</span></span><br><span class="line">sup = Superimposer()   <span class="comment"># 旋转平移矩阵计算</span></span><br><span class="line">sup.set_atoms(target, probe)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Apply the matrix. Remember that it can be applied only on</span></span><br><span class="line"><span class="comment"># lists of atoms.  应用矩阵, 注意只能按照列表提交</span></span><br><span class="line">struct_2_atoms = [at <span class="keyword">for</span> at <span class="keyword">in</span> struct_2.get_atoms()]</span><br><span class="line">sup.apply(struct_2_atoms)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Write the rotation-translated structure 保存旋转平移结构</span></span><br><span class="line">out = PDBIO()</span><br><span class="line">out.set_structure(struct_2)</span><br><span class="line">out.save(<span class="string">'1FXY-superimposed.pdb'</span>)</span><br></pre></td></tr></tbody></table></figure>
<h2><a href="#fu-lu" class="header-anchor"> ⚓️</a><span id="fu-lu">附录</span></h2>
<h3><a href="#fu-lu-a-ming-ling-gai-lan" class="header-anchor"> ⚓️</a><span id="fu-lu-a-ming-ling-gai-lan">附录A 命令概览</span></h3>
<p>Linux命令见: <a target="_blank" rel="noopener" href="https://wangchujiang.com/linux-command/">Linux命令搜索引擎 命令，Linux Linux命令搜索引擎 命令详解：最专业的Linux命令大全，内容包含Linux命令手册、详解、学习，值得收藏的Linux命令速查手册。 - Linux 命令搜索引擎</a></p>
<p>Python命令见: <a target="_blank" rel="noopener" href="https://www.runoob.com/python3/python3-tutorial.html">Python3 教程 | 菜鸟教程 (runoob.com)</a></p>
<h3><a href="#fu-lu-b-python-zi-yuan" class="header-anchor"> ⚓️</a><span id="fu-lu-b-python-zi-yuan">附录B Python资源</span></h3>
<p>Python官网: <a target="_blank" rel="noopener" href="https://www.python.org/">Welcome to Python.org</a></p>
<p>Python官方文档集合: <a target="_blank" rel="noopener" href="https://docs.python.org/3/">3.9.0 Documentation (python.org)</a></p>
<p>Python标准库: <a target="_blank" rel="noopener" href="https://docs.python.org/3/library/index.html#library-index">The Python Standard Library</a></p>
<p>Biopython教程: <a target="_blank" rel="noopener" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec249">Biopython Tutorial and Cookbook</a></p>
<h3><a href="#fu-lu-c-ji-lu-yang-ban" class="header-anchor"> ⚓️</a><span id="fu-lu-c-ji-lu-yang-ban">附录C 记录样板</span></h3>
<figure class="highlight cmd"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br></pre></td><td class="code"><pre><span class="line"># C.<span class="number">1</span> A SINGLE PROTEIN SEQUENCE FILE <span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;sp |P03372|ESR1_HUMAN Estrogen receptor OS = Homo sapiens GN =</span><br><span class="line">ESR1 PE = <span class="number">1</span> SV = <span class="number">2</span></span><br><span class="line">MTM TLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY</span><br><span class="line">EFN AAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF</span><br><span class="line">LQP HGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK</span><br><span class="line">ETR YCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC</span><br><span class="line">RLR KCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR</span><br><span class="line">SKK NSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW</span><br><span class="line">AKR VPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG</span><br><span class="line">MVE IFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD</span><br><span class="line">KIT DTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL</span><br><span class="line">LEM LDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</span><br><span class="line"></span><br><span class="line"># C.<span class="number">2</span> A SINGLE NUCLEOTIDE SEQUENCE FILE <span class="keyword">IN</span> FASTA</span><br><span class="line"><span class="built_in">FORMAT</span></span><br><span class="line">&gt;ENSG00000188536|hemoglobin alpha <span class="number">2</span></span><br><span class="line">ATGGTGCTGTCTCCTGCCGACAAGACCAACGTCAAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCT</span><br><span class="line">GGCGAGTATGGTGCGGAGGCCCTGGAGAGGATGTTCCTGTCCTTCCCCACCACCAAGACCTACTTC</span><br><span class="line">CCGCACTTCGACCTGAGCCACGGCTCTGCCCAGGTTAAGGGCCACGGCAAGAAGGTGGCCGACGCG</span><br><span class="line">CTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCAC</span><br><span class="line">GCGCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGCTCCTAAGCCACTGCCTGCTGGTGACCCTG</span><br><span class="line">GCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCCTCCCTGGACAAGTTCCTGGCTTCT</span><br><span class="line">GTGAGCACCGTGCTGACCTCCAAATACCGTTAA</span><br><span class="line"></span><br><span class="line"># C.<span class="number">3</span> AN EXAMPLE OF AN RNA SEQUENCE</span><br><span class="line"><span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;ENSG00000188536|hemoglobin alpha <span class="number">2</span></span><br><span class="line">ATGGTGCTGTCTCCTGCCGACAAGACCAACGTCAAGGCCGCCTGGGGTAAGGTCGGCGCGCACGCT</span><br><span class="line">GGCGAGTATGGTGCGGAGGCCCTGGAGAGGATGTTCCTGTCCTTCCCCACCACCAAGACCTACTTC</span><br><span class="line">CCGCACTTCGACCTGAGCCACGGCTCTGCCCAGGTTAAGGGCCACGGCAAGAAGGTGGCCGACGCG</span><br><span class="line">CTGACCAACGCCGTGGCGCACGTGGACGACATGCCCAACGCGCTGTCCGCCCTGAGCGACCTGCAC</span><br><span class="line">GCGCACAAGCTTCGGGTGGACCCGGTCAACTTCAAGCTCCTAAGCCACTGCCTGCTGGTGACCCTG</span><br><span class="line">GCCGCCCACCTCCCCGCCGAGTTCACCCCTGCGGTGCACGCCTCCCTGGACAAGTTCCTGGCTTCT</span><br><span class="line">GTGAGCACCGTGCTGACCTCCAAATACCGTTAA</span><br><span class="line"></span><br><span class="line"># C.<span class="number">4</span> A MULTIPLE SEQUENCE FILE <span class="keyword">IN</span> FASTA <span class="built_in">FORMAT</span></span><br><span class="line">&gt;sp |P03372|ESR1_HUMAN Estrogen receptor OS = Homo sapiens GN =</span><br><span class="line">ESR1 PE = <span class="number">1</span> SV = <span class="number">2</span></span><br><span class="line">MTM TLHTKASGMALLHQIQGNELEPLNRPQLKIPLERPLGEVYLDSSKPAVYNYPEGAAY</span><br><span class="line">EFN AAAAANAQVYGQTGLPYGPGSEAAAFGSNGLGGFPPLNSVSPSPLMLLHPPPQLSPF</span><br><span class="line">LQP HGQQVPYYLENEPSGYTVREAGPPAFYRPNSDNRRQGGRERLASTNDKGSMAMESAK</span><br><span class="line">ETR YCAVCNDYASGYHYGVWSCEGCKAFFKRSIQGHNDYMCPATNQCTIDKNRRKSCQAC</span><br><span class="line">RLR KCYEVGMMKGGIRKDRRGGRMLKHKRQRDDGEGRGEVGSAGDMRAANLWPSPLMIKR</span><br><span class="line">SKK NSLALSLTADQMVSALLDAEPPILYSEYDPTRPFSEASMMGLLTNLADRELVHMINW</span><br><span class="line">AKR VPGFVDLTLHDQVHLLECAWLEILMIGLVWRSMEHPGKLLFAPNLLLDRNQGKCVEG</span><br><span class="line">MVE IFDMLLATSSRFRMMNLQGEEFVCLKSIILLNSGVYTFLSSTLKSLEEKDHIHRVLD</span><br><span class="line">KIT DTLIHLMAKAGLTLQQQHQRLAQLLLILSHIRHMSNKGMEHLYSMKCKNVVPLYDLL</span><br><span class="line">LEM LDAHRLHAPTSRGGASVEETDQSHLATAGSTSSHSLQKYYITGEAEGFPATV</span><br><span class="line">&gt;sp |P62333|PRS10_HUMAN <span class="number">26</span>S protease regulatory subunit</span><br><span class="line"><span class="number">10</span>B OS = Homo sapiens GN = PSMC6 PE = <span class="number">1</span> SV = <span class="number">1</span></span><br><span class="line">MAD PRDKALQDYRKKLLEHKEIDGRLKELREQLKELTKQYEKSENDLKALQSVGQIVGEV</span><br><span class="line">LKQ LTEEKFIVKATNGPRYVVGCRRQLDKSKLKPGTRVALDMTTLTIMRYLPREVDPLVY</span><br><span class="line">NMS HEDPGNVSYSEIGGLSEQIRELREVIELPLTNPELFQRVGIIPPKGCLLYGPPGTGK</span><br><span class="line">TLL ARAVASQLDCNFLKVVSSSIVDKYIGESARLIREMFNYARDHQPCIIFMDEIDAIGG</span><br><span class="line">RRF SEGTSADREIQRTLMELLNQMDGFDTLHRVKMIMATNRPDTLDPALLRPGRLDRKIH</span><br><span class="line">IDL PNEQARLDILKIHAGPITKHGEIDYEAIVKLSDGFNGADLRNVCTEAGMFAIRADHD</span><br><span class="line">FVV QEDFMKAVRKVADSKKLESKLDYKPV</span><br><span class="line">&gt;sp |P62509|ERR3_MOUSE Estrogen-related receptor gamma OS = Mus</span><br><span class="line">musculus GN = Esrrg PE = <span class="number">1</span> SV = <span class="number">1</span></span><br><span class="line">MDS VELCLPESFSLHYEEELLCRMSNKDRHIDSSCSSFIKTEPSSPASLTDSVNHHSPGG</span><br><span class="line">SSD ASGSYSSTMNGHQNGLDSPPLYPSAPILGGSGPVRKLYDDCSSTIVEDPQTKCEYML</span><br><span class="line">NSM PKRLCLVCGDIASGYHYGVASCEACKAFFKRTIQGNIEYSCPATNECEITKRRRKSC</span><br><span class="line">QACR FMKCLKVGMLKEGVRLDRVRGGRQKYKRRIDAENSPYLNPQLVQPAKKPYNKIVSH</span><br><span class="line">LLV AEPEKIYAMPDPTVPDSDIKALTTLCDLADRELVVIIGWAKHIPGFSTLSLADQMSL</span><br><span class="line">LQS AWMEILILGVVYRSLSFEDELVYADDYIMDEDQSKLAGLLDLNNAILQLVKKYKSMK</span><br><span class="line">LEK EEFVTLKAIALANSDSMHIEDVEAVQKLQDVLHEALQDYEAGQHMEDPRRAGKMLMT</span><br><span class="line">LPLLRQTSTKAVQHFYNIKLEGKVPMHKLFLEMLEAKV</span><br><span class="line"></span><br><span class="line"># C.<span class="number">8</span> AN EXAMPLE OF THE SEQRES LINES OF</span><br><span class="line">A PDB FILE (FROM FILE <span class="number">1</span>TDL)</span><br><span class="line">SEQ RES <span class="number">1</span> A <span class="number">223</span> ILE VAL GLY GLY TYR THR CYS GLY ALA ASN THR VAL PRO</span><br><span class="line">SEQ RES <span class="number">2</span> A <span class="number">223</span> TYR GLN VAL SER LEU ASN SER GLY TYR HIS PHE CYS GLY</span><br><span class="line">SEQ RES <span class="number">3</span> A <span class="number">223</span> GLY SER LEU ILE ASN SER GLN TRP VAL VAL SER ALA ALA</span><br><span class="line">SEQ RES <span class="number">4</span> A <span class="number">223</span> HIS CYS TYR LYS SER GLY ILE GLN VAL ARG LEU GLY GLU</span><br><span class="line">SEQ RES <span class="number">5</span> A <span class="number">223</span> ASP ASN ILE ASN VAL VAL GLU GLY ASN GLU GLN PHE ILE</span><br><span class="line">SEQ RES <span class="number">6</span> A <span class="number">223</span> SER ALA SER LYS SER ILE VAL HIS PRO SER TYR ASN SER</span><br><span class="line">SEQ RES <span class="number">7</span> A <span class="number">223</span> ASN THR LEU ASN ASN ASP ILE MET LEU ILE LYS LEU LYS</span><br><span class="line">SEQ RES <span class="number">8</span> A <span class="number">223</span> SER ALA ALA SER LEU ASN SER ARG VAL ALA SER ILE SER</span><br><span class="line">SEQ RES <span class="number">9</span> A <span class="number">223</span> LEU PRO THR SER CYS ALA SER ALA GLY THR GLN CYS LEU</span><br><span class="line">SEQ RES <span class="number">10</span> A <span class="number">223</span> ILE SER GLY TRP GLY ASN THR LYS SER SER GLY THR SER</span><br><span class="line">SEQ RES <span class="number">11</span> A <span class="number">223</span> TYR PRO ASP VAL LEU LYS CYS LEU LYS ALA PRO ILE LEU</span><br><span class="line">SEQ RES <span class="number">12</span> A <span class="number">223</span> SER ASP SER SER CYS LYS SER ALA TYR PRO GLY GLN ILE</span><br><span class="line">SEQ RES <span class="number">13</span> A <span class="number">223</span> THR SER ASN MET PHE CYS ALA GLY TYR LEU GLU GLY GLY</span><br><span class="line">SEQ RES <span class="number">14</span> A <span class="number">223</span> LYS ASP SER CYS GLN GLY ASP SER GLY GLY PRO VAL VAL</span><br><span class="line">SEQ RES <span class="number">15</span> A <span class="number">223</span> CYS SER GLY LYS LEU GLN GLY ILE VAL SER TRP GLY SER</span><br><span class="line">SEQ RES <span class="number">16</span> A <span class="number">223</span> GLY CYS ALA GLN LYS ASN LYS PRO GLY VAL TYR THR LYS</span><br><span class="line">SEQ RES <span class="number">17</span> A <span class="number">223</span> VAL CYS ASN TYR VAL SER TRP ILE LYS GLN THR ILE ALA</span><br><span class="line">SEQRES <span class="number">18</span> A <span class="number">223</span> SER ASN</span><br><span class="line"></span><br><span class="line"># C.<span class="number">9</span> AN EXAMPLE OF THE CUFFCOMPARE OUTPUT <span class="keyword">FOR</span> THREE SAMPLES (Q1, Q2, AND Q3)</span><br><span class="line"># Because of the length of each line, the first field of each new line is <span class="keyword">in</span> bold,and different lines are highlighted using indentation.</span><br><span class="line">Med ullo-Diff_00000001 XLOC_000001 Lypla1|uc007afh.<span class="number">1</span></span><br><span class="line"><span class="function">    q1:<span class="title">NSC.P419</span>.228|<span class="title">uc007afh</span>.1 |100|35.109496|34.188903</span></span><br><span class="line"><span class="function">    |36.030089|397.404732|2433 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.18|<span class="title">uc007afh</span>.1|100 |15.885823|15.240240</span></span><br><span class="line"><span class="function">    |16.531407|171.011325 |2433 <span class="title">q3:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P437</span>.15|<span class="title">uc007afh</span>.1|100 |18.338541|17.704857|18.97222</span></span><br><span class="line"><span class="function">    4|181.643949|2433</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000002</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.228|<span class="title">uc007afi</span>.2|</span></span><br><span class="line"><span class="function">    18|1.653393|1.409591|1.897195 |18.587029|2671</span></span><br><span class="line"><span class="function">    - <span class="title">q3:NSC</span>.<span class="title">P437</span>.108|<span class="title">uc007afi</span>.2|100 |4.624079|4.258801|4</span></span><br><span class="line"><span class="function">    .989356|45.379750|2671</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000003</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc011wht</span>.1</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.228|<span class="title">uc011wht</span>.1|100 |9.011253|8.560848</span></span><br><span class="line"><span class="function">    |9.461657|101.302266|2668 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.116|<span class="title">uc011wht</span>.1|100 |6.889020|6.503460|7.27458</span></span><br><span class="line"><span class="function">    0|73.238938 |2668<span class="title">q3:NSC</span>.<span class="title">P437</span>.108 |<span class="title">uc011wht</span>.1|90</span></span><br><span class="line"><span class="function">    |4.170527 |3.817430|4.523625|40.928694|2668</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000004</span> <span class="title">XLOC_000003</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.231|<span class="title">NSC.P419</span>.231.1</span></span><br><span class="line"><span class="function">    |100|31.891396|30.892103 |32.890690|379.023601</span></span><br><span class="line"><span class="function">    |1568<span class="title">q2:NSC</span>.<span class="title">P429</span>.121|<span class="title">NSC.P429</span>.121.1 |100|27.991543</span></span><br><span class="line"><span class="function">    |27.007869|28.975218 |313.481210|1532 -</span></span><br><span class="line"><span class="function"><span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000005</span> <span class="title">XLOC_000002</span> <span class="title">Tcea1</span>|<span class="title">uc007afi</span>.2</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.236 |<span class="title">NSC.P419</span>.236.1</span></span><br><span class="line"><span class="function">    |100|1.164739|0.868895 |1.460583|13.879881|- - -</span></span><br><span class="line"><span class="function">    <span class="title">Med</span> <span class="title">ullo</span>-<span class="title">Diff_00000006</span> <span class="title">XLOC_000004</span> <span class="title">Atp6v1h</span>|<span class="title">uc007afn</span>.1</span></span><br><span class="line"><span class="function">    <span class="title">q1:NSC</span>.<span class="title">P419</span>.55|<span class="title">uc007afn</span>.1 |100|39.526818 |38.58510</span></span><br><span class="line"><span class="function">    2|40.468533|455.599775|1976 <span class="title">q2:NSC</span>.</span></span><br><span class="line"><span class="function">    <span class="title">P429</span>.43|<span class="title">uc007afn</span>.1 |100|25.034945 |24.182398|25.887</span></span><br><span class="line"><span class="function">    493|271.738343|1976 <span class="title">q3:NSC</span>.<span class="title">P437</span>.37|<span class="title">uc007afn</span>.1</span></span><br><span class="line"><span class="function">    |100|20.848047 |20.043989|21.652104|205.866771|1976</span></span><br><span class="line"><span class="function">    </span></span><br><span class="line"><span class="function"></span></span><br></pre></td></tr></tbody></table></figure>
<h3><a href="#fu-lu-d-chu-li-mu-lu-he-yong-unix-bian-cheng" class="header-anchor"> ⚓️</a><span id="fu-lu-d-chu-li-mu-lu-he-yong-unix-bian-cheng">附录D 处理目录和用UNIX编程</span></h3>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/linux/linux-shell.html">Shell 教程 | 菜鸟教程 (runoob.com)</a></p>
<p><a target="_blank" rel="noopener" href="https://www.runoob.com/linux/linux-tutorial.html">Linux 教程 | 菜鸟教程 (runoob.com)</a></p>
<br>
<br>
        </div>
        <footer class="article-footer">
            



    <a data-url="https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/" data-id="clkva0d1f0066owuu4gnm2qrw" class="article-share-link"><i class="fa fa-share"></i>分享到</a>
<script>
    (function ($) {
        $('body').on('click', function() {
            $('.article-share-box.on').removeClass('on');
        }).on('click', '.article-share-link', function(e) {
            e.stopPropagation();

            var $this = $(this),
                url = $this.attr('data-url'),
                encodedUrl = encodeURIComponent(url),
                id = 'article-share-box-' + $this.attr('data-id'),
                offset = $this.offset(),
                box;

            if ($('#' + id).length) {
                box = $('#' + id);

                if (box.hasClass('on')){
                    box.removeClass('on');
                    return;
                }
            } else {
                var html = [
                    '<div id="' + id + '" class="article-share-box">',
                        '<input class="article-share-input" value="' + url + '">',
                        '<div class="article-share-links">',
                            '<a href="https://twitter.com/intent/tweet?url=' + encodedUrl + '" class="article-share-twitter" target="_blank" title="Twitter"></a>',
                            '<a href="https://www.facebook.com/sharer.php?u=' + encodedUrl + '" class="article-share-facebook" target="_blank" title="Facebook"></a>',
                            '<a href="http://pinterest.com/pin/create/button/?url=' + encodedUrl + '" class="article-share-pinterest" target="_blank" title="Pinterest"></a>',
                            '<a href="https://plus.google.com/share?url=' + encodedUrl + '" class="article-share-google" target="_blank" title="Google+"></a>',
                        '</div>',
                    '</div>'
                ].join('');

              box = $(html);

              $('body').append(box);
            }

            $('.article-share-box.on').hide();

            box.css({
                top: offset.top + 25,
                left: offset.left
            }).addClass('on');

        }).on('click', '.article-share-box', function (e) {
            e.stopPropagation();
        }).on('click', '.article-share-box-input', function () {
            $(this).select();
        }).on('click', '.article-share-box-link', function (e) {
            e.preventDefault();
            e.stopPropagation();

            window.open(this.href, 'article-share-box-window-' + Date.now(), 'width=500,height=450');
        });
    })(jQuery);
</script>

        </footer>
    </div>
    <script type="application/ld+json">
    {
        "@context": "https://schema.org",
        "@type": "BlogPosting",
        "author": {
            "@type": "Person",
            "name": "苑长"
        },
        "headline": "读书笔记-Python生物信息学数据管理",
        "image": "https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/Python&amp;DNA.png",
        "keywords": "Python 数据处理",
        "genre": "科技 生物",
        "datePublished": "2020-10-04",
        "dateCreated": "2020-10-04",
        "dateModified": "2021-06-21",
        "url": "https://yuanz1949.github.io/du-shu-bi-ji-python-sheng-wu-xin-xi-xue-shu-ju-guan-li/",
        "description": "
当生物与Python交织之时
故事将就此展开
",
        "wordCount": 17848
    }
</script>

</article>

    <section id="comments">
    
        
<div id="comment-container"></div>


    
    </section>



                        </div>
                    </section>
                    <aside id="sidebar">
    <a class="sidebar-toggle" title="Expand Sidebar"><i class="toggle icon"></i></a>
    <div class="sidebar-top">
        <p>关注我 :</p>
        <ul class="social-links">
            
                
                <li>
                    <a class="social-tooltip" title="zhihu" href="https://www.zhihu.com/people/WoodYuan" target="_blank" rel="noopener">
                        <i class="fab fa-zhihu"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="github" href="https://github.com/YuanZ1949" target="_blank" rel="noopener">
                        <i class="fab fa-github"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="bilibili" href="https://space.bilibili.com/12549559" target="_blank" rel="noopener">
                        <i class="fab fa-bimobject"></i>
                    </a>
                </li>
                
            
                
                <li>
                    <a class="social-tooltip" title="rss" href="/atom.xml" target="_blank" rel="noopener">
                        <i class="fas fa-rss"></i>
                    </a>
                </li>
                
            
        </ul>
    </div>
    
        
<nav id="article-nav">
    
        <a href="/ri-chang-za-ji/" id="article-nav-newer" class="article-nav-link-wrap">
        <strong class="article-nav-caption">下一篇</strong>
        <p class="article-nav-title">
        
            日常杂记
        
        </p>
        <i class="icon fa fa-chevron-right" id="icon-chevron-right"></i>
    </a>
    
    
        <a href="/justscoresummary-ge-ren-cheng-ji-gai-lan/" id="article-nav-older" class="article-nav-link-wrap">
        <strong class="article-nav-caption">上一篇</strong>
        <p class="article-nav-title">JUSTScoreSummary-个人成绩概览</p>
        <i class="icon fa fa-chevron-left" id="icon-chevron-left"></i>
        </a>
    
</nav>

    
    <div class="widgets-container">
        
            
                

            
                
    <div class="widget-wrap">
        <h3 class="widget-title">最新文章</h3>
        <div class="widget">
            <ul id="recent-post" class="">
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/di-gui-wei-xing-xing-zuo-xiang-guan-xin-xi-hui-zong/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">空天</a></p>
                            <p class="item-title"><a href="/di-gui-wei-xing-xing-zuo-xiang-guan-xin-xi-hui-zong/" class="title">低轨卫星星座相关信息汇总</a></p>
                            <p class="item-date"><time datetime="2023-07-25T14:38:28.000Z" itemprop="datePublished">2023-07-25</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/ubuntu-wei-hu-bei-wang-lu/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">信息技术</a></p>
                            <p class="item-title"><a href="/ubuntu-wei-hu-bei-wang-lu/" class="title">Ubuntu维护备忘录</a></p>
                            <p class="item-date"><time datetime="2023-07-16T10:13:23.000Z" itemprop="datePublished">2023-07-16</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/qbittorrent-xiang-guan/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"></p>
                            <p class="item-title"><a href="/qbittorrent-xiang-guan/" class="title">Qbittorrent相关</a></p>
                            <p class="item-date"><time datetime="2023-05-04T12:45:37.000Z" itemprop="datePublished">2023-05-04</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/wang-luo-fu-shu-cun-chu-nas-pei-zhi-zhi-nan/" class="thumbnail">
    
    
        <span class="thumbnail-image thumbnail-none"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">信息技术</a></p>
                            <p class="item-title"><a href="/wang-luo-fu-shu-cun-chu-nas-pei-zhi-zhi-nan/" class="title">网络附属存储(NAS)配置指南</a></p>
                            <p class="item-date"><time datetime="2023-02-23T13:05:48.000Z" itemprop="datePublished">2023-02-23</time></p>
                        </div>
                    </li>
                
                    <li>
                        
                        <div class="item-thumbnail">
                            <a href="/shu-qian-suo-yin/" class="thumbnail">
    
    
        <span style="background-image:url(/shu-qian-suo-yin/logocat.JPG)" alt="书签索引" class="thumbnail-image"></span>
    
    
</a>

                        </div>
                        
                        <div class="item-inner">
                            <p class="item-category"><a class="article-category-link" href="/categories/%E7%B4%A2%E5%BC%95/">索引</a><i class="icon fa fa-angle-right"></i><a class="article-category-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">资源索引</a></p>
                            <p class="item-title"><a href="/shu-qian-suo-yin/" class="title">书签索引</a></p>
                            <p class="item-date"><time datetime="2023-02-19T07:47:44.399Z" itemprop="datePublished">2023-02-19</time></p>
                        </div>
                    </li>
                
            </ul>
        </div>
    </div>

            
                
    <div class="widget-wrap widget-list">
        <h3 class="widget-title">分类</h3>
        <div class="widget">
            <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E6%B4%BB/">生活</a><span class="category-list-count">7</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%94%9F%E6%B4%BB/%E6%90%9E%E4%BA%8B%E6%83%85%E5%90%96/">搞事情吖</a><span class="category-list-count">6</span></li></ul></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/">科技</a><span class="category-list-count">22</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E4%BF%A1%E6%81%AF%E6%8A%80%E6%9C%AF/">信息技术</a><span class="category-list-count">12</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E5%A4%9A%E5%AA%92%E4%BD%93/">多媒体</a><span class="category-list-count">4</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%94%9F%E7%89%A9/">生物</a><span class="category-list-count">4</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%A7%91%E6%8A%80/%E7%A9%BA%E5%A4%A9/">空天</a><span class="category-list-count">2</span></li></ul></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/">索引</a><span class="category-list-count">5</span><ul class="category-list-child"><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%AE%9E%E6%97%B6%E4%BF%A1%E6%81%AF/">实时信息</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E5%B7%A5%E4%BD%9C%E6%B5%81%E7%B4%A2%E5%BC%95/">工作流索引</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E6%94%BF%E7%AD%96%E7%B4%A2%E5%BC%95/">政策索引</a><span class="category-list-count">1</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/%E7%B4%A2%E5%BC%95/%E8%B5%84%E6%BA%90%E7%B4%A2%E5%BC%95/">资源索引</a><span class="category-list-count">2</span></li></ul></li></ul>
        </div>
    </div>


            
                
    <div class="widget-wrap widget-float">
        <h3 class="widget-title">标签云</h3>
        <div class="widget tagcloud">
            <a href="/tags/3D/" style="font-size: 10px;">3D</a> <a href="/tags/Blender/" style="font-size: 10px;">Blender</a> <a href="/tags/Blog/" style="font-size: 10px;">Blog</a> <a href="/tags/Chagas/" style="font-size: 10px;">Chagas</a> <a href="/tags/DIY/" style="font-size: 10px;">DIY</a> <a href="/tags/Excel/" style="font-size: 10px;">Excel</a> <a href="/tags/Github/" style="font-size: 10px;">Github</a> <a href="/tags/Hexo/" style="font-size: 10px;">Hexo</a> <a href="/tags/IT/" style="font-size: 10px;">IT</a> <a href="/tags/NAS/" style="font-size: 10px;">NAS</a> <a href="/tags/Python/" style="font-size: 20px;">Python</a> <a href="/tags/Ubuntu/" style="font-size: 10px;">Ubuntu</a> <a href="/tags/VBA/" style="font-size: 10px;">VBA</a> <a href="/tags/VPN/" style="font-size: 10px;">VPN</a> <a href="/tags/python/" style="font-size: 10px;">python</a> <a href="/tags/%E4%B9%A6%E7%AD%BE/" style="font-size: 10px;">书签</a> <a href="/tags/%E4%BB%AA%E5%99%A8/" style="font-size: 10px;">仪器</a> <a href="/tags/%E4%BD%8E%E8%BD%A8%E5%8D%AB%E6%98%9F%E6%98%9F%E5%BA%A7/" style="font-size: 10px;">低轨卫星星座</a> <a href="/tags/%E5%85%B3%E4%BA%8E/" style="font-size: 10px;">关于</a> <a href="/tags/%E5%88%86%E4%BA%AB/" style="font-size: 20px;">分享</a> <a href="/tags/%E5%8C%96%E5%AD%A6/" style="font-size: 10px;">化学</a> <a href="/tags/%E5%8C%96%E5%AD%A6%E5%8F%91%E5%85%89/" style="font-size: 10px;">化学发光</a> <a href="/tags/%E5%8D%95%E5%8F%8D/" style="font-size: 10px;">单反</a> <a href="/tags/%E5%91%BD%E4%BB%A4%E8%A1%8C/" style="font-size: 10px;">命令行</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 10px;">图像处理</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 10px;">天文</a> <a href="/tags/%E5%AE%9E%E9%AA%8C/" style="font-size: 10px;">实验</a> <a href="/tags/%E5%B7%A5%E4%BD%9C/" style="font-size: 10px;">工作</a> <a href="/tags/%E5%B7%A5%E4%BD%9C%E6%B5%81/" style="font-size: 10px;">工作流</a> <a href="/tags/%E5%B7%A5%E5%85%B7/" style="font-size: 13.33px;">工具</a> <a href="/tags/%E5%BC%80%E5%8F%91/" style="font-size: 10px;">开发</a> <a href="/tags/%E6%91%84%E5%BD%B1/" style="font-size: 10px;">摄影</a> <a href="/tags/%E6%94%BF%E7%AD%96/" style="font-size: 10px;">政策</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86/" style="font-size: 10px;">数据处理</a> <a href="/tags/%E6%98%9F%E7%BD%91/" style="font-size: 10px;">星网</a> <a href="/tags/%E6%98%9F%E9%93%BE/" style="font-size: 10px;">星链</a> <a href="/tags/%E6%9C%8D%E5%8A%A1%E5%99%A8/" style="font-size: 10px;">服务器</a> <a href="/tags/%E6%B1%9F%E7%A7%91%E5%A4%A7/" style="font-size: 10px;">江科大</a> <a href="/tags/%E6%B5%8B%E8%AF%95/" style="font-size: 10px;">测试</a> <a href="/tags/%E7%83%AD%E7%82%B9/" style="font-size: 10px;">热点</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 10px;">版本控制</a> <a href="/tags/%E7%94%9F%E6%B4%BB/" style="font-size: 10px;">生活</a> <a href="/tags/%E7%96%BE%E7%97%85/" style="font-size: 10px;">疾病</a> <a href="/tags/%E7%9E%8E%E9%97%B2%E9%80%9B/" style="font-size: 10px;">瞎闲逛</a> <a href="/tags/%E7%AC%94%E8%AE%B0/" style="font-size: 16.67px;">笔记</a> <a href="/tags/%E7%BB%88%E7%AB%AF/" style="font-size: 10px;">终端</a> <a href="/tags/%E7%BB%A9%E7%82%B9/" style="font-size: 10px;">绩点</a> <a href="/tags/%E7%BD%91%E7%AB%99/" style="font-size: 10px;">网站</a> <a href="/tags/%E7%BD%91%E7%BB%9C%E9%99%84%E5%B1%9E%E5%AD%98%E5%82%A8/" style="font-size: 10px;">网络附属存储</a> <a href="/tags/%E8%87%AA%E5%8A%A8%E5%8C%96/" style="font-size: 13.33px;">自动化</a> <a href="/tags/%E8%A7%86%E9%A2%91/" style="font-size: 10px;">视频</a> <a href="/tags/%E8%AE%A1%E5%88%92/" style="font-size: 10px;">计划</a> <a href="/tags/%E8%AE%BE%E5%A4%87/" style="font-size: 13.33px;">设备</a> <a href="/tags/%E9%80%89%E8%B4%AD/" style="font-size: 10px;">选购</a>
        </div>
    </div>


            
                
    <div class="widget-wrap widget-list">
        <h3 class="widget-title">链接</h3>
        <div class="widget">
            <ul>
                
                    <li>
                        <a href="/sheng-huo/gao-shi-qing-a/zhe-shi-yi-pian-testblog/">网站测试页</a>
                    </li>
                
                    <li>
                        <a target="_blank" rel="noopener" href="https://lib.just.edu.cn/">江苏科技大学图书馆</a>
                    </li>
                
                    <li>
                        <a target="_blank" rel="noopener" href="http://hexo.io">Hexo官网</a>
                    </li>
                
            </ul>
        </div>
    </div>


            
        
    </div>
</aside>

                </div>
            </div>
        </div>
        <footer id="footer">
    <div class="container">
        <div class="container-inner">
            <a id="back-to-top" href="javascript:;"><i class="icon fa fa-angle-up"></i></a>
            <div class="credit">
                <h1 class="logo-wrap">
                    <a href="/" class="logo"></a>
                </h1>
                <p>&copy; 2023 苑长</p>
                
                <p>Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>. Theme by <a href="https://github.com/ppoffice" target="_blank">PPOffice</a></p>
                
            </div>
            <div class="footer-plugins">
              
    


            </div>
        </div>
    </div>
</footer>

        
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css">
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>
<script>
    var gitalk = new Gitalk({
        clientID: 'ffffe2a96b4c373c1d4d',
        clientSecret: 'fc0ec8dd07cfb61e09dd06962678251d943bef37',
        id: '05acbd50fa835f8f9bf44ef4b390abd3',
        repo: 'yuanz1949.github.io',
        owner: 'YuanZ1949',
        admin: "YuanZ1949"
    })
    gitalk.render('comment-container')
</script>





    
        
<script src="/libs/lightgallery/js/lightgallery.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-thumbnail.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-pager.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-autoplay.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-fullscreen.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-zoom.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-hash.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-share.min.js"></script>

        
<script src="/libs/lightgallery/js/lg-video.min.js"></script>

    
    
        
<script src="/libs/justified-gallery/jquery.justifiedGallery.min.js"></script>

    
    
         <script type="text/x-mathjax-config">
            MathJax.Hub.Config({
                showProcessingMessages: false,
                messageStyle: "none",
                extensions: ["tex2jax.js"],
                jax: ["input/TeX", "output/HTML-CSS"],
                tex2jax: {
                    inlineMath:  [ ["$", "$"] ],
                    displayMath: [ ["$$","$$"] ],
                    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre','code','a'],
                    ignoreClass:"comment-content"
                },
                "HTML-CSS": {
                    availableFonts: ["STIX","TeX"],
                    showMathMenu: true
                }
            });
            MathJax.Hub.Queue(["Typeset",MathJax.Hub]);
         </script>
         <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
         <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    

    



<!-- Custom Scripts -->

<script src="/js/main.js"></script>


    </div>
</body>
</html>
